Kongoh <- function(){
  
  #############################
  # Install required packages #
  #############################
  
  #tcltk
  if(require(tcltk)){
    print("tcltk is loaded correctly")
  }else{
    print("trying to install tcltk...")
    install.packages("tcltk")
    if(require("tcltk")){
      print("tcltk installed and loaded")
      tclRequire("Tktable")
    }else{
      stop("could not install tcltk")
    }
  }
  
  #tcltk2
  if(require(tcltk2)){
    print("tcltk2 is loaded correctly")
  }else{
    print("trying to install tcltk2...")
    install.packages("tcltk2")
    if(require("tcltk2")){
      print("tcltk2 installed and loaded")
    }else{
      stop("could not install tcltk2")
    }
  }
  
  #gtools
  if(require(gtools)){
    print("gtools is loaded correctly")
  }else{
    print("trying to install gtools...")
    install.packages("gtools")
    if(require("gtools")){
      print("gtools installed and loaded")
    }else{
      stop("could not install gtools")
    }
  }
  
  #parallel
  if(require(parallel)){
    print("parallel is loaded correctly")
  }else{
    print("trying to install parallel...")
    install.packages("parallel")
    if(require("parallel")){
      print("parallel installed and loaded")
    }else{
      stop("could not install parallel")
    }
  }
  
  #truncnorm
  if(require(truncnorm)){
    print("truncnorm is loaded correctly")
  }else{
    print("trying to install truncnorm...")
    install.packages("truncnorm")
    if(require("truncnorm")){
      print("truncnorm installed and loaded")
    }else{
      stop("could not install truncnorm")
    }
  }
  
  #GenSA
  if(require(GenSA)){
    print("GenSA is loaded correctly")
  }else{
    print("trying to install GenSA...")
    install.packages("GenSA")
    if(require("GenSA")){
      print("GenSA installed and loaded")
    }else{
      stop("could not install GenSA")
    }
  }
  
  #Tktable
  addTclPath('/usr/local/lib/Tktable2.9')
  tclRequire("Tktable")
  
  
  #############
  # Parameter #
  #############
  
  #software version
  softVer <- "Kongoh v3.0.1"
  
  #Minimum or maximum of each parameter for Monte Carlo simulation
  aeMin <- 0.058
  hbMin <- 0.046
  srB1Max <- 0.7
  srF1Max <- 0.35
  srB2Max <- 0.13
  srM2Max <- 0.12
  
  #Default of mixture ratios (MRn) of one contributor (1 to n-1)
  mrDefault <- c(0.0025, 0.005, 0.0075, 0.01, 0.025, 0.05, 0.075, 0.1, 0.2, 0.3, 0.4, 0.5)
  mrOne <<- mrDefault
  
  #Default of degradation parameter (d)
  degDefault <- round(seq(-0.05, 0, 0.0025), 4)
  degOne <<- degDefault
  
  #Threshold of likelihoods for selecting MR and d
  mrDegCutVar <- tclVar("0.0001")
  
  #Sex chromosomal marker
  sexChrMar <- c("AMEL", "Amelogenin", "Yindel", "YIndel", "YInDel", "DYS391")
  
  
  ###################################################################################################
  # Size of each allele (used to determine sizes of drop-out alleles in the case of locus drop-out) #
  ###################################################################################################
  
  #Identifiler
  lociID <- c("D8S1179", "D21S11", "D7S820", "CSF1PO", "D3S1358", "TH01", "D13S317", "D16S539", "D2S1338", "D19S433", "vWA", "TPOX", "D18S51", "D5S818", "FGA")
  alleleID <- list()
  alleleID[[1]] <- c(7, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18)
  alleleID[[2]] <- c(27, 28, 28.2, 29, 30, 30.2, 30.3, 31, 31.2, 32, 32.2, 33, 33.1, 33.2, 34, 34.2)
  alleleID[[3]] <- c(7, 8, 9, 9.1, 10, 10.1, 10.3, 11, 12, 13, 14, 15)
  alleleID[[4]] <- c(7, 8, 9, 10, 11, 12, 13, 14, 15, 16)
  alleleID[[5]] <- c(12, 13, 14, 15, 16, 17, 18, 19, 21)
  alleleID[[6]] <- c(5, 6, 7, 8, 9, 9.3, 10)
  alleleID[[7]] <- c(7, 8, 9, 10, 11, 12, 13, 14, 15)
  alleleID[[8]] <- c(7, 8, 9, 10, 11, 12, 13, 14, 15)
  alleleID[[9]] <- c(16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28)
  alleleID[[10]] <- c(9.2, 10.2, 11, 11.2, 12, 12.2, 13, 13.2, 14, 14.2, 15, 15.2, 16, 16.2, 17.2, 18)
  alleleID[[11]] <- c(13, 14, 15, 16, 17, 18, 19, 20, 21, 22)
  alleleID[[12]] <- c(7, 8, 9, 10, 11, 12, 13, 14)
  alleleID[[13]] <- c(10, 11, 12, 13, 14, 15, 16, 17, 17.1, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27)
  alleleID[[14]] <- c(7, 8, 9, 10, 11, 12, 13, 14, 15)
  alleleID[[15]] <- c(17, 18, 19, 20, 21, 22, 22.2, 23, 23.2, 24, 24.2, 25, 25.2, 26, 27, 28)
  names(alleleID) <- lociID
  sizeID <- list()
  sizeID[[1]] <- c(123, 131, 135, 139, 143, 147, 151, 155, 159, 163, 167)
  sizeID[[2]] <- c(198, 202, 204, 206, 210, 212, 213, 214, 216, 218, 220, 222, 223, 224, 226, 228)
  sizeID[[3]] <- c(261, 265, 269, 270, 273, 274, 276, 277, 281, 285, 289, 293)
  sizeID[[4]] <- c(309, 313, 317, 321, 325, 329, 333, 337, 341, 345)
  sizeID[[5]] <- c(113, 117, 121, 125, 129, 133, 137, 141, 149)
  sizeID[[6]] <- c(168, 172, 176, 180, 184, 187, 188)
  sizeID[[7]] <- c(201, 205, 209, 213, 217, 221, 225, 229, 233)
  sizeID[[8]] <- c(260, 264, 268, 272, 276, 280, 284, 288, 292)
  sizeID[[9]] <- c(311, 315, 319, 323, 327, 331, 335, 339, 343, 347, 351, 355, 359)
  sizeID[[10]] <- c(108, 112, 114, 116, 118, 120, 122, 124, 126, 128, 130, 132, 134, 136, 140, 142)
  sizeID[[11]] <- c(164, 168, 172, 176, 180, 184, 188, 192, 196, 200)
  sizeID[[12]] <- c(221, 225, 229, 233, 237, 241, 245, 249)
  sizeID[[13]] <- c(276, 280, 284, 288, 292, 296, 300, 304, 305, 308, 312, 316, 320, 324, 328, 332, 336, 340, 344)
  sizeID[[14]] <- c(134, 138, 142, 146, 150, 154, 158, 162, 166)
  sizeID[[15]] <- c(214, 218, 222, 226, 230, 234, 236, 238, 240, 242, 244, 246, 248, 250, 254, 258)
  names(sizeID) <- lociID
  
  #GlobalFiler
  lociGF <- c("D3S1358", "vWA", "D16S539", "CSF1PO", "TPOX", "D8S1179", "D21S11", "D18S51", "D2S441", "D19S433", "TH01", "FGA", "D22S1045", "D5S818", "D13S317", "D7S820", "SE33", "D10S1248", "D1S1656", "D12S391", "D2S1338")
  alleleGF <- list()
  alleleGF[[1]] <- c(12, 13, 14, 15, 16, 17, 18, 19, 20, 21)
  alleleGF[[2]] <- c(13, 14, 15, 16, 17, 18, 19, 20, 21, 22)
  alleleGF[[3]] <- c(7, 8, 9, 10, 11, 12, 13, 14, 15)
  alleleGF[[4]] <- c(7, 8, 9, 10, 11, 12, 13, 14, 15, 16)
  alleleGF[[5]] <- c(8, 9, 10, 11, 12, 13, 14)
  alleleGF[[6]] <- c(7, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18)
  alleleGF[[7]] <- c(27, 28, 28.2, 29, 30, 30.2, 30.3, 31, 31.2, 32, 32.2, 33, 33.1, 33.2, 34, 34.2)
  alleleGF[[8]] <- c(10, 11, 12, 13, 14, 15, 16, 17, 17.1, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27)
  alleleGF[[9]] <- c(8, 8.1, 9.1, 10, 10.1, 11, 11.1, 11.3, 12, 13, 14, 15, 16)
  alleleGF[[10]] <- c(7, 9.2, 10.2, 11, 11.2, 12, 12.2, 13, 13.2, 14, 14.2, 15, 15.2, 16, 16.2, 17.2, 18)
  alleleGF[[11]] <- c(5, 6, 7, 8, 9, 9.3, 10)
  alleleGF[[12]] <- c(16, 17, 18, 19, 20, 21, 22, 22.2, 23, 23.2, 24, 24.2, 25, 25.2, 26, 27, 28, 29)
  alleleGF[[13]] <- c(11, 12, 13, 14, 15, 16, 17, 18, 19, 20)
  alleleGF[[14]] <- c(7, 8, 9, 10, 11, 12, 13, 14, 15)
  alleleGF[[15]] <- c(7, 8, 9, 10, 11, 12, 13, 14, 15)
  alleleGF[[16]] <- c(7, 8, 9, 9.1, 10, 10.1, 10.3, 11, 12, 13, 14, 15)
  alleleGF[[17]] <- c(13, 14, 14.1, 14.2, 15, 16, 16.2, 17, 17.2, 18, 18.2, 19, 19.2, 20, 20.2, 21, 21.2, 22, 22.2, 23, 23.2, 24, 24.2, 25, 25.2, 26.2, 27, 27.1, 27.2, 27.3, 28, 28.1, 28.2, 28.3, 29.2, 29.3, 30.2, 30.3, 31.1, 31.2, 32.1, 32.2, 33.2, 34.2)
  alleleGF[[18]] <- c(10, 11, 12, 13, 14, 15, 16, 17, 18)
  alleleGF[[19]] <- c(11, 12, 13, 14, 14.2, 15, 15.3, 16, 16.3, 17, 17.3, 18, 18.3, 19.3, 20.3)
  alleleGF[[20]] <- c(12, 14, 15, 16, 17, 18, 19, 19.3, 20, 20.3, 21, 21.3, 22, 23, 24, 25, 26, 27)
  alleleGF[[21]] <- c(16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28)
  names(alleleGF) <- lociGF
  sizeGF <- list()
  sizeGF[[1]] <- c(109, 113, 117, 121, 125, 129, 133, 137, 141, 145)
  sizeGF[[2]] <- c(165, 169, 173, 177, 181, 185, 189, 193, 197, 201)
  sizeGF[[3]] <- c(236, 240, 244, 248, 252, 256, 260, 264, 268)
  sizeGF[[4]] <- c(286, 290, 294, 298, 302, 306, 310, 314, 318, 322)
  sizeGF[[5]] <- c(351, 355, 359, 363, 367, 371, 375)
  sizeGF[[6]] <- c(122, 130, 134, 138, 142, 146, 150, 154, 158, 162, 166)
  sizeGF[[7]] <- c(195, 199, 201, 203, 207, 209, 210, 211, 213, 215, 217, 219, 220, 221, 223, 225)
  sizeGF[[8]] <- c(273, 277, 281, 285, 289, 293, 297, 301, 302, 305, 309, 313, 317, 321, 325, 329, 333, 337, 341)
  sizeGF[[9]] <- c(77, 78, 82, 85, 86, 89, 90, 92, 93, 97, 101, 105, 109)
  sizeGF[[10]] <- c(122, 131, 135, 137, 139, 141, 143, 145, 147, 149, 151, 153, 155, 157, 159, 163, 165)
  sizeGF[[11]] <- c(183, 187, 191, 195, 199, 202, 203)
  sizeGF[[12]] <- c(236, 240, 244, 248, 252, 256, 260, 262, 264, 266, 268, 270, 272, 274, 276, 280, 284, 288)
  sizeGF[[13]] <- c(97, 100, 103, 106, 109, 112, 115, 118, 121, 124)
  sizeGF[[14]] <- c(139, 143, 147, 151, 155, 159, 163, 167, 171)
  sizeGF[[15]] <- c(207, 211, 215, 219, 223, 227, 231, 235, 239)
  sizeGF[[16]] <- c(266, 270, 274, 275, 278, 279, 281, 282, 286, 290, 294, 298)
  sizeGF[[17]] <- c(342, 346, 347, 348, 350, 354, 356, 358, 360, 362, 364, 366, 368, 370, 372, 374, 376, 378, 380, 382, 384, 386, 388, 390, 392, 396, 398, 399, 400, 401, 402, 403, 404, 405, 408, 409, 412, 413, 415, 416, 419, 420, 424, 428)
  sizeGF[[18]] <- c(94, 98, 102, 106, 110, 114, 118, 122, 126)
  sizeGF[[19]] <- c(168, 172, 176, 180, 182, 184, 187, 188, 191, 192, 195, 196, 199, 203, 207)
  sizeGF[[20]] <- c(208, 216, 220, 224, 228, 232, 236, 239, 240, 243, 244, 247, 248, 252, 256, 260, 264, 268)
  sizeGF[[21]] <- c(302, 306, 310, 314, 318, 322, 326, 330, 334, 338, 342, 346, 350)
  names(sizeGF) <- lociGF
  
  #NGM
  lociNGM <- c("D10S1248", "vWA", "D16S539", "D2S1338", "D8S1179", "D21S11", "D18S51", "D22S1045", "D19S433", "TH01", "FGA", "D2S441", "D3S1358", "D1S1656", "D12S391")
  alleleNGM <- list()
  alleleNGM[[1]] <- c(8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18)
  alleleNGM[[2]] <- c(11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24)
  alleleNGM[[3]] <- c(5, 8, 9, 10, 11, 12, 13, 14, 15)
  alleleNGM[[4]] <- c(15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28)
  alleleNGM[[5]] <- c(8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19)
  alleleNGM[[6]] <- c(24, 24.2, 25, 26, 27, 28, 28.2, 29, 29.2, 30, 30.2, 31, 31.2, 32, 32.2, 33, 33.2, 34, 34.2, 35, 35.2, 36, 37, 38)
  alleleNGM[[7]] <- c(7, 9, 10, 10.2, 11, 12, 13, 13.2, 14, 14.2, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27)
  alleleNGM[[8]] <- c(8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19)
  alleleNGM[[9]] <- c(9, 10, 11, 12, 12.2, 13, 13.2, 14, 14.2, 15, 15.2, 16, 16.2, 17, 17.2)
  alleleNGM[[10]] <- c(4, 5, 6, 7, 8, 9, 9.3, 10, 11, 13.3)
  alleleNGM[[11]] <- c(17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 26.2, 27, 28, 29, 30, 30.2, 31.2, 32.2, 33.2)
  alleleNGM[[12]] <- c(9, 10, 11, 11.3, 12, 13, 14, 15, 16)
  alleleNGM[[13]] <- c(12, 13, 14, 15, 16, 17, 18, 19)
  alleleNGM[[14]] <- c(9, 10, 11, 12, 13, 14, 14.3, 15, 15.3, 16, 16.3, 17, 17.3, 18.3, 19.3, 20.3)
  alleleNGM[[15]] <- c(14, 15, 16, 17, 18, 19, 19.3, 20, 21, 22, 23, 24, 25, 26, 27)
  names(alleleNGM) <- lociNGM
  sizeNGM <- list()
  sizeNGM[[1]] <- c(77, 81, 85, 89, 93, 97, 101, 105, 109, 113, 117)
  sizeNGM[[2]] <- c(154, 158, 162, 166, 170, 174, 178, 182, 186, 190, 194, 198, 202, 206)
  sizeNGM[[3]] <- c(229, 241, 245, 249, 253, 257, 261, 265, 269)
  sizeNGM[[4]] <- c(290, 294, 298, 302, 306, 310, 314, 318, 322, 326, 330, 334, 338, 342)
  sizeNGM[[5]] <- c(124, 128, 132, 136, 140, 144, 148, 152, 156, 160, 164, 168)
  sizeNGM[[6]] <- c(185, 187, 189, 193, 197, 201, 203, 205, 207, 209, 211, 213, 215, 217, 219, 221, 223, 225, 227, 229, 231, 233, 237, 241)
  sizeNGM[[7]] <- c(262, 270, 274, 276, 278, 282, 286, 288, 290, 292, 294, 298, 302, 306, 310, 314, 318, 322, 326, 330, 334, 338, 342)
  sizeNGM[[8]] <- c(80, 83, 86, 89, 92, 95, 98, 101, 104, 107, 110, 113)
  sizeNGM[[9]] <- c(127, 131, 135, 139, 141, 143, 145, 147, 149, 151, 153, 155, 157, 159, 161)
  sizeNGM[[10]] <- c(182, 186, 190, 194, 198, 202, 205, 206, 210, 221)
  sizeNGM[[11]] <- c(233, 237, 241, 245, 249, 253, 257, 261, 265, 269, 271, 273, 277, 281, 285, 287, 291, 295, 299)
  sizeNGM[[12]] <- c(80, 84, 88, 91, 92, 96, 100, 104, 108)
  sizeNGM[[13]] <- c(135, 139, 143, 147, 151, 155, 159, 163)
  sizeNGM[[14]] <- c(176, 180, 184, 188, 192, 196, 199, 200, 203, 204, 207, 208, 211, 215, 219, 223)
  sizeNGM[[15]] <- c(230, 234, 238, 242, 246, 250, 253, 254, 258, 262, 266, 270, 274, 278, 282)
  names(sizeNGM) <- lociNGM
  
  #NGM SElect
  lociNGMSE <- c("D10S1248", "vWA", "D16S539", "D2S1338", "D8S1179", "D21S11", "D18S51", "D22S1045", "D19S433", "TH01", "FGA", "D2S441", "D3S1358", "D1S1656", "D12S391", "SE33")
  alleleNGMSE <- list()
  alleleNGMSE[[1]] <- c(8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18)
  alleleNGMSE[[2]] <- c(11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24)
  alleleNGMSE[[3]] <- c(5, 8, 9, 10, 11, 12, 13, 14, 15)
  alleleNGMSE[[4]] <- c(15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28)
  alleleNGMSE[[5]] <- c(8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19)
  alleleNGMSE[[6]] <- c(24, 24.2, 25, 26, 27, 28, 28.2, 29, 29.2, 30, 30.2, 31, 31.2, 32, 32.2, 33, 33.2, 34, 34.2, 35, 35.2, 36, 37, 38)
  alleleNGMSE[[7]] <- c(7, 9, 10, 10.2, 11, 12, 13, 13.2, 14, 14.2, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27)
  alleleNGMSE[[8]] <- c(8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19)
  alleleNGMSE[[9]] <- c(9, 10, 11, 12, 12.2, 13, 13.2, 14, 14.2, 15, 15.2, 16, 16.2, 17, 17.2)
  alleleNGMSE[[10]] <- c(4, 5, 6, 7, 8, 9, 9.3, 10, 11, 13.3)
  alleleNGMSE[[11]] <- c(17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 26.2, 27, 28, 29, 30, 30.2, 31.2, 32.2, 33.2)
  alleleNGMSE[[12]] <- c(9, 10, 11, 11.3, 12, 13, 14, 15, 16)
  alleleNGMSE[[13]] <- c(12, 13, 14, 15, 16, 17, 18, 19)
  alleleNGMSE[[14]] <- c(9, 10, 11, 12, 13, 14, 14.3, 15, 15.3, 16, 16.3, 17, 17.3, 18.3, 19.3, 20.3)
  alleleNGMSE[[15]] <- c(14, 15, 16, 17, 18, 19, 19.3, 20, 21, 22, 23, 24, 25, 26, 27)
  alleleNGMSE[[16]] <- c(4.2, 6.3, 8, 9, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 20.2, 21, 21.2, 22.2, 23.2, 24.2, 25.2, 26.2, 27.2, 28.2, 29.2, 30.2, 31.2, 32.2, 33.2, 34.2, 35, 35.2, 36, 37)
  names(alleleNGMSE) <- lociNGMSE
  sizeNGMSE <- list()
  sizeNGMSE[[1]] <- c(75, 79, 83, 87, 91, 95, 99, 103, 107, 111, 115)
  sizeNGMSE[[2]] <- c(151, 155, 159, 163, 167, 171, 175, 179, 183, 187, 191, 195, 199, 203)
  sizeNGMSE[[3]] <- c(227, 239, 243, 247, 251, 255, 259, 263, 267)
  sizeNGMSE[[4]] <- c(288, 292, 296, 300, 304, 308, 312, 316, 320, 324, 328, 332, 336, 340)
  sizeNGMSE[[5]] <- c(123, 127, 131, 135, 139, 143, 147, 151, 155, 159, 163, 167)
  sizeNGMSE[[6]] <- c(183, 185, 187, 191, 195, 199, 201, 203, 205, 207, 209, 211, 213, 215, 217, 219, 221, 223, 225, 227, 229, 231, 235, 239)
  sizeNGMSE[[7]] <- c(261, 269, 273, 275, 277, 281, 285, 287, 289, 291, 293, 297, 301, 305, 309, 313, 317, 321, 325, 329, 333, 337, 341)
  sizeNGMSE[[8]] <- c(78, 81, 84, 87, 90, 93, 96, 99, 102, 105, 108, 111)
  sizeNGMSE[[9]] <- c(126, 130, 134, 138, 140, 142, 144, 146, 148, 150, 152, 154, 156, 158, 160)
  sizeNGMSE[[10]] <- c(179, 183, 187, 191, 195, 199, 202, 203, 207, 218)
  sizeNGMSE[[11]] <- c(232, 236, 240, 244, 248, 252, 256, 260, 264, 268, 270, 272, 276, 280, 284, 286, 290, 294, 298)
  sizeNGMSE[[12]] <- c(78, 82, 86, 89, 90, 94, 98, 102, 106)
  sizeNGMSE[[13]] <- c(134, 138, 142, 146, 150, 154, 158, 162)
  sizeNGMSE[[14]] <- c(173, 177, 181, 185, 189, 193, 196, 197, 200, 201, 204, 205, 208, 212, 216, 220)
  sizeNGMSE[[15]] <- c(229, 233, 237, 241, 245, 249, 252, 253, 257, 261, 265, 269, 273, 277, 281)
  sizeNGMSE[[16]] <- c(311, 320, 325, 329, 337, 341, 345, 349, 353, 357, 361, 365, 369, 373, 375, 377, 379, 383, 387, 391, 395, 399, 403, 407, 411, 415, 419, 423, 427, 431, 433, 435, 437, 441)
  names(sizeNGMSE) <- lociNGMSE
  
  #Minifiler
  lociMF <- c("D13S317", "D7S820", "D2S1338", "D21S11", "D16S539", "D18S51", "CSF1PO", "FGA")
  alleleMF <- list()
  alleleMF[[1]] <- c(8, 9, 10, 11, 12, 13, 14, 15)
  alleleMF[[2]] <- c(6, 7, 8, 9, 10, 11, 12, 13, 14, 15)
  alleleMF[[3]] <- c(15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28)
  alleleMF[[4]] <- c(24, 24.2, 25, 26, 27, 28, 28.2, 29, 29.2, 30, 30.2, 31, 31.2, 32, 32.2, 33, 33.2, 34, 34.2, 35, 35.2, 36, 37, 38)
  alleleMF[[5]] <- c(5, 8, 9, 10, 11, 12, 13, 14, 15)
  alleleMF[[6]] <- c(7, 9, 10, 10.2, 11, 12, 13, 13.2, 14, 14.2, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27)
  alleleMF[[7]] <- c(6, 7, 8, 9, 10, 11, 12, 13, 14, 15)
  alleleMF[[8]] <- c(17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 26.2, 27, 28, 29, 30, 30.2, 31.2, 32.2, 33.2)
  names(alleleMF) <- lociMF
  sizeMF <- list()
  sizeMF[[1]] <- c(103, 107, 111, 115, 119, 123, 127, 131)
  sizeMF[[2]] <- c(150, 154, 158, 162, 166, 170, 174, 178, 182, 186)
  sizeMF[[3]] <- c(120, 124, 128, 132, 136, 140, 144, 148, 152, 156, 160, 164, 168, 172)
  sizeMF[[4]] <- c(187, 189, 191, 195, 199, 203, 205, 207, 209, 211, 213, 215, 217, 219, 221, 223, 225, 227, 229, 231, 233, 235, 239, 243)
  sizeMF[[5]] <- c(76, 88, 92, 96, 100, 104, 108, 112, 116)
  sizeMF[[6]] <- c(125, 133, 137, 139, 141, 145, 149, 151, 153, 155, 157, 161, 165, 169, 173, 177, 181, 185, 189, 193, 197, 201, 205)
  sizeMF[[7]] <- c(87, 91, 95, 99, 103, 107, 111, 115, 119, 123)
  sizeMF[[8]] <- c(150, 154, 158, 162, 166, 170, 174, 178, 182, 186, 188, 190, 194, 198, 202, 204, 208, 212, 216)
  names(sizeMF) <- lociMF
  
  #SGM Plus
  lociSGMP <- c("D3S1358", "vWA", "D16S539", "D2S1338", "D8S1179", "D21S11", "D18S51", "D19S433", "TH01", "FGA")
  alleleSGMP <- list()
  alleleSGMP[[1]] <- c(12, 13, 14, 15, 16, 17, 18, 19)
  alleleSGMP[[2]] <- c(11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24)
  alleleSGMP[[3]] <- c(5, 8, 9, 10, 11, 12, 13, 14, 15)
  alleleSGMP[[4]] <- c(15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28)
  alleleSGMP[[5]] <- c(8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19)
  alleleSGMP[[6]] <- c(24, 24.2, 25, 26, 27, 28, 28.2, 29, 29.2, 30, 30.2, 31, 31.2, 32, 32.2, 33, 33.2, 34, 34.2, 35, 35.2, 36, 37, 38)
  alleleSGMP[[7]] <- c(7, 9, 10, 10.2, 11, 12, 13, 13.2, 14, 14.2, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27)
  alleleSGMP[[8]] <- c(9, 10, 11, 12, 12.2, 13, 13.2, 14, 14.2, 15, 15.2, 16, 16.2, 17, 17.2)
  alleleSGMP[[9]] <- c(4, 5, 6, 7, 8, 9, 9.3, 10, 11, 13.3)
  alleleSGMP[[10]] <- c(17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 26.2, 27, 28, 29, 30, 30.2, 31.2, 32.2, 33.2)
  names(alleleSGMP) <- lociSGMP
  sizeSGMP <- list()
  sizeSGMP[[1]] <- c(111, 115, 119, 123, 127, 131, 135, 139)
  sizeSGMP[[2]] <- c(155, 159, 163, 167, 171, 175, 179, 183, 187, 191, 195, 199, 203, 207)
  sizeSGMP[[3]] <- c(229, 241, 245, 249, 253, 257, 261, 265, 269)
  sizeSGMP[[4]] <- c(290, 294, 298, 302, 306, 310, 314, 318, 322, 326, 330, 334, 338, 342)
  sizeSGMP[[5]] <- c(124, 128, 132, 136, 140, 144, 148, 152, 156, 160, 164, 168)
  sizeSGMP[[6]] <- c(185, 187, 189, 193, 197, 201, 203, 205, 207, 209, 211, 213, 215, 217, 219, 221, 223, 225, 227, 229, 231, 233, 237, 241)
  sizeSGMP[[7]] <- c(263, 271, 275, 277, 279, 283, 287, 289, 291, 293, 295, 299, 303, 307, 311, 315, 319, 323, 327, 331, 335, 339, 343)
  sizeSGMP[[8]] <- c(102, 106, 110, 114, 116, 118, 120, 122, 124, 126, 128, 130, 132, 134, 136)
  sizeSGMP[[9]] <- c(163, 167, 171, 175, 179, 183, 186, 187, 191, 202)
  sizeSGMP[[10]] <- c(213, 217, 221, 225, 229, 233, 237, 241, 245, 249, 251, 253, 257, 261, 265, 267, 271, 275, 279)
  names(sizeSGMP) <- lociSGMP
  
  #PowerPlex Fusion 6C
  lociPPF6C <- c("D3S1358", "D1S1656", "D2S441", "D10S1248", "D13S317", "Penta E", "D16S539", "D18S51", "D2S1338", "CSF1PO", "Penta D", "TH01", "vWA", "D21S11", "D7S820", "D5S818", "TPOX", "D8S1179", "D12S391", "D19S433", "SE33", "D22S1045", "FGA")
  allelePPF6C <- list()
  allelePPF6C[[1]] <- c(9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20)
  allelePPF6C[[2]] <- c(9, 10, 11, 12, 13, 14, 14.3, 15, 15.3, 16, 16.3, 17, 17.3, 18, 18.3, 19, 19.3, 20.3)
  allelePPF6C[[3]] <- c(8, 9, 10, 11, 11.3, 12, 13, 14, 15, 16, 17)
  allelePPF6C[[4]] <- c(8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19)
  allelePPF6C[[5]] <- c(5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17)
  allelePPF6C[[6]] <- c(5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25)
  allelePPF6C[[7]] <- c(4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16)
  allelePPF6C[[8]] <- c(7, 8, 9, 10, 10.2, 11, 12, 13, 13.2, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27)
  allelePPF6C[[9]] <- c(10, 12, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28)
  allelePPF6C[[10]] <- c(5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16)
  allelePPF6C[[11]] <- c(2.2, 3.2, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17)
  allelePPF6C[[12]] <- c(3, 4, 5, 6, 7, 8, 9, 9.3, 10, 11, 13.3)
  allelePPF6C[[13]] <- c(10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24)
  allelePPF6C[[14]] <- c(24, 24.2, 25, 25.2, 26, 27, 28, 28.2, 29, 29.2, 30, 30.2, 31, 31.2, 32, 32.2, 33, 33.2, 34, 34.2, 35, 35.2, 36, 37, 38)
  allelePPF6C[[15]] <- c(5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16)
  allelePPF6C[[16]] <- c(6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18)
  allelePPF6C[[17]] <- c(4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16)
  allelePPF6C[[18]] <- c(7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19)
  allelePPF6C[[19]] <- c(14, 15, 16, 17, 17.3, 18, 18.3, 19, 20, 21, 22, 23, 24, 25, 26, 27)
  allelePPF6C[[20]] <- c(5.2, 6.2, 8, 9, 10, 11, 12, 12.2, 13, 13.2, 14, 14.2, 15, 15.2, 16, 16.2, 17, 17.2, 18, 18.2)
  allelePPF6C[[21]] <- c(4.2, 6.3, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 20.2, 21, 21.2, 22, 22.2, 23.2, 24.2, 25.2, 26.2, 27.2, 28.2, 29.2, 30.2, 31.2, 32.2, 33.2, 34.2, 35, 36, 37, 39)
  allelePPF6C[[22]] <- c(7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20)
  allelePPF6C[[23]] <- c(14, 15, 16, 17, 18, 18.2, 19, 19.2, 20, 20.2, 21, 21.2, 22, 22.2, 23, 23.2, 24, 24.2, 25, 25.2, 26, 27, 28, 29, 30, 31.2, 32.2, 33.2)
  names(allelePPF6C) <- lociPPF6C
  sizePPF6C <- list()
  sizePPF6C[[1]] <- c(103, 107, 111, 115, 119, 123, 127, 131, 135, 139, 143, 147)
  sizePPF6C[[2]] <- c(161, 165, 169, 173, 177, 181, 184, 185,  188, 189, 192, 193, 196, 197, 200, 201, 204, 208)
  sizePPF6C[[3]] <- c(216, 220, 224, 228, 231, 232, 236, 240, 244, 248, 252)
  sizePPF6C[[4]] <- c(259, 263, 267, 271, 275, 279, 283, 287, 291, 295, 299, 303)
  sizePPF6C[[5]] <- c(308, 312, 316, 320, 324, 328, 332, 336, 340, 344, 348, 352, 356)
  sizePPF6C[[6]] <- c(371, 376, 381, 386, 391, 396, 401, 406, 411, 416, 421, 426, 431, 436, 441, 446, 451, 456, 461, 466, 471)
  sizePPF6C[[7]] <- c(84, 88, 92, 96, 100, 104, 108, 112, 116, 120, 124, 128, 132)
  sizePPF6C[[8]] <- c(134, 138, 142, 146, 148, 150, 154, 158, 160, 162, 166, 170, 174, 178, 182, 186, 190, 194, 198, 202, 206, 210, 214)
  sizePPF6C[[9]] <- c(224, 232, 240, 244, 248, 252, 256, 260, 264, 268, 272, 276, 280, 284, 288, 292, 296)
  sizePPF6C[[10]] <- c(318, 322, 326, 330, 334, 338, 342, 346, 350, 354, 358, 362)
  sizePPF6C[[11]] <- c(377, 382, 390, 395, 400, 405, 410, 415, 420, 425, 430, 435, 440, 445, 450)
  sizePPF6C[[12]] <- c(72, 76, 80, 84, 88, 92, 96, 99, 100, 104, 115)
  sizePPF6C[[13]] <- c(127, 131, 135, 139, 143, 147, 151, 155, 159, 163, 167, 171, 175, 179, 183)
  sizePPF6C[[14]] <- c(203, 205, 207, 209, 211, 215, 219, 221, 223, 225, 227, 229, 231, 233, 235, 237, 239, 241, 243, 245, 247, 249, 251, 255, 259)
  sizePPF6C[[15]] <- c(269, 273, 277, 281, 285, 289, 293, 297, 301, 305, 309, 313)
  sizePPF6C[[16]] <- c(321, 325, 329, 333, 337, 341, 345, 349, 353, 357, 361, 365, 369)
  sizePPF6C[[17]] <- c(393, 397, 401, 405, 409, 413, 417, 421, 425, 429, 433, 437, 441)
  sizePPF6C[[18]] <- c(76, 80, 84, 88, 92, 96, 100, 104, 108, 112, 116, 120, 124)
  sizePPF6C[[19]] <- c(133, 137, 141, 145, 148, 149, 152, 153, 157, 161, 165, 169, 173, 177, 181, 185)
  sizePPF6C[[20]] <- c(193, 197, 203, 207, 211, 215, 219, 221, 223, 225, 227, 229, 231, 233, 235, 237, 239, 241, 243, 245)
  sizePPF6C[[21]] <- c(270, 279, 284, 288, 292, 296, 300, 304, 308, 312, 316, 320, 324, 328, 332, 334, 336, 338, 340, 342, 346, 350, 354, 358, 362, 366, 370, 374, 378, 382, 386, 390, 392, 396, 400, 408)
  sizePPF6C[[22]] <- c(431, 434, 437, 440, 443, 446, 449, 452, 455, 458, 461, 464, 467, 470)
  sizePPF6C[[23]] <- c(143, 147, 151, 155, 159, 161, 163, 165, 167, 169, 171, 173, 175, 177, 179, 181, 183, 185, 187, 189, 191, 195, 199, 203, 207, 213, 217, 221)
  names(sizePPF6C) <- lociPPF6C
  
  #PowerPlex Fusion
  lociPPF <- c("D3S1358", "D1S1656", "D2S441", "D10S1248", "D13S317", "Penta E", "D16S539", "D18S51", "D2S1338", "CSF1PO", "Penta D", "TH01", "vWA", "D21S11", "D7S820", "D5S818", "TPOX", "D8S1179", "D12S391", "D19S433", "FGA", "D22S1045")
  allelePPF <- list()
  allelePPF[[1]] <- c(9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20)
  allelePPF[[2]] <- c(9, 10, 11, 12, 13, 14, 14.3, 15, 15.3, 16, 16.3, 17, 17.3, 18, 18.3, 19, 19.3, 20.3)
  allelePPF[[3]] <- c(8, 9, 10, 11, 11.3, 12, 13, 14, 15, 16, 17)
  allelePPF[[4]] <- c(8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19)
  allelePPF[[5]] <- c(5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17)
  allelePPF[[6]] <- c(5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24)
  allelePPF[[7]] <- c(4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16)
  allelePPF[[8]] <- c(7, 8, 9, 10, 10.2, 11, 12, 13, 13.2, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27)
  allelePPF[[9]] <- c(10, 12, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28)
  allelePPF[[10]] <- c(5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16)
  allelePPF[[11]] <- c(2.2, 3.2, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17)
  allelePPF[[12]] <- c(3, 4, 5, 6, 7, 8, 9, 9.3, 10, 11, 13.3)
  allelePPF[[13]] <- c(10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24)
  allelePPF[[14]] <- c(24, 24.2, 25, 25.2, 26, 27, 28, 28.2, 29, 29.2, 30, 30.2, 31, 31.2, 32, 32.2, 33, 33.2, 34, 34.2, 35, 35.2, 36, 37, 38)
  allelePPF[[15]] <- c(5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16)
  allelePPF[[16]] <- c(6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18)
  allelePPF[[17]] <- c(4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16)
  allelePPF[[18]] <- c(7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19)
  allelePPF[[19]] <- c(14, 15, 16, 17, 17.3, 18, 18.3, 19, 20, 21, 22, 23, 24, 25, 26, 27)
  allelePPF[[20]] <- c(5.2, 6.2, 8, 9, 10, 11, 12, 12.2, 13, 13.2, 14, 14.2, 15, 15.2, 16, 16.2, 17, 17.2, 18, 18.2)
  allelePPF[[21]] <- c(14, 15, 16, 17, 18, 18.2, 19, 19.2, 20, 20.2, 21, 21.2, 22, 22.2, 23, 23.2, 24, 24.2, 25, 25.2, 26, 27, 28, 29, 30, 31.2, 32.2, 33.2)
  allelePPF[[22]] <- c(7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20)
  names(allelePPF) <- lociPPF
  sizePPF <- list()
  sizePPF[[1]] <- c(103, 107, 111, 115, 119, 123, 127, 131, 135, 139, 143, 147)
  sizePPF[[2]] <- c(161, 165, 169, 173, 177, 181, 184, 185,  188, 189, 192, 193, 196, 197, 200, 201, 204, 208)
  sizePPF[[3]] <- c(214, 218, 222, 226, 229, 230, 234, 238, 242, 246, 250)
  sizePPF[[4]] <- c(255, 259, 263, 267, 271, 275, 279, 283, 287, 291, 295, 299)
  sizePPF[[5]] <- c(302, 306, 310, 314, 318, 322, 326, 330, 334, 338, 342, 346, 350)
  sizePPF[[6]] <- c(371, 376, 381, 386, 391, 396, 401, 406, 411, 416, 421, 426, 431, 436, 441, 446, 451, 456, 461, 466)
  sizePPF[[7]] <- c(84, 88, 92, 96, 100, 104, 108, 112, 116, 120, 124, 128, 132)
  sizePPF[[8]] <- c(134, 138, 142, 146, 148, 150, 154, 158, 160, 162, 166, 170, 174, 178, 182, 186, 190, 194, 198, 202, 206, 210, 214)
  sizePPF[[9]] <- c(224, 232, 240, 244, 248, 252, 256, 260, 264, 268, 272, 276, 280, 284, 288, 292, 296)
  sizePPF[[10]] <- c(318, 322, 326, 330, 334, 338, 342, 346, 350, 354, 358, 362)
  sizePPF[[11]] <- c(377, 382, 390, 395, 400, 405, 410, 415, 420, 425, 430, 435, 440, 445, 450)
  sizePPF[[12]] <- c(72, 76, 80, 84, 88, 92, 96, 99, 100, 104, 115)
  sizePPF[[13]] <- c(127, 131, 135, 139, 143, 147, 151, 155, 159, 163, 167, 171, 175, 179, 183)
  sizePPF[[14]] <- c(203, 205, 207, 209, 211, 215, 219, 221, 223, 225, 227, 229, 231, 233, 235, 237, 239, 241, 243, 245, 247, 249, 251, 255, 259)
  sizePPF[[15]] <- c(269, 273, 277, 281, 285, 289, 293, 297, 301, 305, 309, 313)
  sizePPF[[16]] <- c(321, 325, 329, 333, 337, 341, 345, 349, 353, 357, 361, 365, 369)
  sizePPF[[17]] <- c(393, 397, 401, 405, 409, 413, 417, 421, 425, 429, 433, 437, 441)
  sizePPF[[18]] <- c(76, 80, 84, 88, 92, 96, 100, 104, 108, 112, 116, 120, 124)
  sizePPF[[19]] <- c(133, 137, 141, 145, 148, 149, 152, 153, 157, 161, 165, 169, 173, 177, 181, 185)
  sizePPF[[20]] <- c(193, 197, 203, 207, 211, 215, 219, 221, 223, 225, 227, 229, 231, 233, 235, 237, 239, 241, 243, 245)
  sizePPF[[21]] <- c(265, 269, 273, 277, 281, 283, 285, 287, 289, 291, 293, 295, 297, 299, 301, 303, 305, 307, 309, 311, 313, 317, 321, 325, 329, 335, 339, 343)
  sizePPF[[22]] <- c(425, 428, 431, 434, 437, 440, 443, 446, 449, 452, 455, 458, 461, 464)
  names(sizePPF) <- lociPPF
  
  #PowerPlex 21
  lociPP21 <- c("D3S1358", "D1S1656", "D6S1043", "D13S317", "Penta E", "D16S539", "D18S51", "D2S1338", "CSF1PO", "Penta D", "TH01", "vWA", "D21S11", "D7S820", "D5S818", "TPOX", "D8S1179", "D12S391", "D19S433", "FGA")
  allelePP21 <- list()
  allelePP21[[1]] <- c(9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20)
  allelePP21[[2]] <- c(9, 10, 11, 12, 13, 14, 14.3, 15, 15.3, 16, 16.3, 17, 17.3, 18, 18.3, 19, 19.3, 20.3)
  allelePP21[[3]] <- c(7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25)
  allelePP21[[4]] <- c(5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17)
  allelePP21[[5]] <- c(5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24)
  allelePP21[[6]] <- c(4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16)
  allelePP21[[7]] <- c(7, 8, 9, 10, 10.2, 11, 12, 13, 13.2, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27)
  allelePP21[[8]] <- c(10, 12, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28)
  allelePP21[[9]] <- c(5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16)
  allelePP21[[10]] <- c(2.2, 3.2, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17)
  allelePP21[[11]] <- c(3, 4, 5, 6, 7, 8, 9, 9.3, 10, 11, 13.3)
  allelePP21[[12]] <- c(10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24)
  allelePP21[[13]] <- c(24, 24.2, 25, 25.2, 26, 27, 28, 28.2, 29, 29.2, 30, 30.2, 31, 31.2, 32, 32.2, 33, 33.2, 34, 34.2, 35, 35.2, 36, 37, 38)
  allelePP21[[14]] <- c(5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16)
  allelePP21[[15]] <- c(6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18)
  allelePP21[[16]] <- c(4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16)
  allelePP21[[17]] <- c(7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19)
  allelePP21[[18]] <- c(14, 15, 16, 17, 17.3, 18, 18.3, 19, 20, 21, 22, 23, 24, 25, 26, 27)
  allelePP21[[19]] <- c(5.2, 6.2, 8, 9, 10, 11, 12, 12.2, 13, 13.2, 14, 14.2, 15, 15.2, 16, 16.2, 17, 17.2, 18, 18.2)
  allelePP21[[20]] <- c(14, 15, 16, 17, 18, 18.2, 19, 19.2, 20, 20.2, 21, 21.2, 22, 22.2, 23, 23.2, 24, 24.2, 25, 25.2, 26, 27, 28, 29, 30, 31.2, 32.2, 33.2)
  names(allelePP21) <- lociPP21
  sizePP21 <- list()
  sizePP21[[1]] <- c(103, 107, 111, 115, 119, 123, 127, 131, 135, 139, 143, 147)
  sizePP21[[2]] <- c(161, 165, 169, 173, 177, 181, 184, 185,  188, 189, 192, 193, 196, 197, 200, 201, 204, 208)
  sizePP21[[3]] <- c(215, 219, 223, 227, 231, 235, 239, 243, 247, 251, 255, 259, 263, 267, 271, 275, 279, 283, 287)
  sizePP21[[4]] <- c(302, 306, 310, 314, 318, 322, 326, 330, 334, 338, 342, 346, 350)
  sizePP21[[5]] <- c(371, 376, 381, 386, 391, 396, 401, 406, 411, 416, 421, 426, 431, 436, 441, 446, 451, 456, 461, 466)
  sizePP21[[6]] <- c(84, 88, 92, 96, 100, 104, 108, 112, 116, 120, 124, 128, 132)
  sizePP21[[7]] <- c(134, 138, 142, 146, 148, 150, 154, 158, 160, 162, 166, 170, 174, 178, 182, 186, 190, 194, 198, 202, 206, 210, 214)
  sizePP21[[8]] <- c(224, 232, 240, 244, 248, 252, 256, 260, 264, 268, 272, 276, 280, 284, 288, 292, 296)
  sizePP21[[9]] <- c(318, 322, 326, 330, 334, 338, 342, 346, 350, 354, 358, 362)
  sizePP21[[10]] <- c(377, 382, 390, 395, 400, 405, 410, 415, 420, 425, 430, 435, 440, 445, 450)
  sizePP21[[11]] <- c(72, 76, 80, 84, 88, 92, 96, 99, 100, 104, 115)
  sizePP21[[12]] <- c(127, 131, 135, 139, 143, 147, 151, 155, 159, 163, 167, 171, 175, 179, 183)
  sizePP21[[13]] <- c(203, 205, 207, 209, 211, 215, 219, 221, 223, 225, 227, 229, 231, 233, 235, 237, 239, 241, 243, 245, 247, 249, 251, 255, 259)
  sizePP21[[14]] <- c(269, 273, 277, 281, 285, 289, 293, 297, 301, 305, 309, 313)
  sizePP21[[15]] <- c(321, 325, 329, 333, 337, 341, 345, 349, 353, 357, 361, 365, 369)
  sizePP21[[16]] <- c(393, 397, 401, 405, 409, 413, 417, 421, 425, 429, 433, 437, 441)
  sizePP21[[17]] <- c(76, 80, 84, 88, 92, 96, 100, 104, 108, 112, 116, 120, 124)
  sizePP21[[18]] <- c(133, 137, 141, 145, 148, 149, 152, 153, 157, 161, 165, 169, 173, 177, 181, 185)
  sizePP21[[19]] <- c(193, 197, 203, 207, 211, 215, 219, 221, 223, 225, 227, 229, 231, 233, 235, 237, 239, 241, 243, 245)
  sizePP21[[20]] <- c(265, 269, 273, 277, 281, 283, 285, 287, 289, 291, 293, 295, 297, 299, 301, 303, 305, 307, 309, 311, 313, 317, 321, 325, 329, 335, 339, 343)
  names(sizePP21) <- lociPP21
  
  #PowerPlex 16
  lociPP16 <- c("D3S1358", "TH01", "D21S11", "D18S51", "Penta E", "D5S818", "D13S317", "D7S820", "D16S539", "CSF1PO", "Penta D", "vWA", "D8S1179", "TPOX", "FGA")
  allelePP16 <- list()
  allelePP16[[1]] <- c(12, 13, 14, 15, 16, 17, 18, 19, 20)
  allelePP16[[2]] <- c(4, 5, 6, 7, 8, 9, 9.3, 10, 11, 13.3)
  allelePP16[[3]] <- c(24, 24.2, 25, 25.2, 26, 27, 28, 28.2, 29, 29.2, 30, 30.2, 31, 31.2, 32, 32.2, 33, 33.2, 34, 34.2, 35, 35.2, 36, 37, 38)
  allelePP16[[4]] <- c(8, 9, 10, 10.2, 11, 12, 13, 13.2, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27)
  allelePP16[[5]] <- c(5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24)
  allelePP16[[6]] <- c(7, 8, 9, 10, 11, 12, 13, 14, 15, 16)
  allelePP16[[7]] <- c(7, 8, 9, 10, 11, 12, 13, 14, 15)
  allelePP16[[8]] <- c(6, 7, 8, 9, 10, 11, 12, 13, 14)
  allelePP16[[9]] <- c(5, 8, 9, 10, 11, 12, 13, 14, 15)
  allelePP16[[10]] <- c(6, 7, 8, 9, 10, 11, 12, 13, 14, 15)
  allelePP16[[11]] <- c(2.2, 3.2, 5, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17)
  allelePP16[[12]] <- c(10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22)
  allelePP16[[13]] <- c(7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18)
  allelePP16[[14]] <- c(6, 7, 8, 9, 10, 11, 12, 13)
  allelePP16[[15]] <- c(16, 17, 18, 18.2, 19, 19.2, 20, 20.2, 21, 21.2, 22, 22.2, 23, 23.2, 24, 24.2, 25, 25.2, 26, 27, 28, 29, 30, 31.2)
  names(allelePP16) <- lociPP16
  sizePP16 <- list()
  sizePP16[[1]] <- c(115, 119, 123, 127, 131, 135, 139, 143, 147)
  sizePP16[[2]] <- c(156, 160, 164, 168, 172, 176, 179, 180, 184, 195)
  sizePP16[[3]] <- c(203, 205, 207, 209, 211, 215, 219, 221, 223, 225, 227, 229, 231, 233, 235, 237, 239, 241, 243, 245, 247, 249, 251, 255, 259)
  sizePP16[[4]] <- c(290, 294, 298, 300, 302, 306, 310, 312, 314, 318, 322, 326, 330, 334, 338, 342, 346, 350, 354, 358, 362, 366)
  sizePP16[[5]] <- c(379, 384, 389, 394, 399, 404, 409, 414, 419, 424, 429, 434, 439, 444, 449, 454, 459, 464, 469, 474)
  sizePP16[[6]] <- c(119, 123, 127, 131, 135, 139, 143, 147, 151, 155)
  sizePP16[[7]] <- c(176, 180, 184, 188, 192, 196, 200, 204, 208)
  sizePP16[[8]] <- c(215, 219, 223, 227, 231, 235, 239, 243, 247)
  sizePP16[[9]] <- c(264, 276, 280, 284, 288, 292, 296, 300, 304)
  sizePP16[[10]] <- c(321, 325, 329, 333, 337, 341, 345, 349, 353, 357)
  sizePP16[[11]] <- c(376, 381, 389, 399, 404, 409, 414, 419, 424, 429, 434, 439, 444, 449)
  sizePP16[[12]] <- c(123, 127, 131, 135, 139, 143, 147, 151, 155, 159, 163, 167, 171)
  sizePP16[[13]] <- c(203, 207, 211, 215, 219, 223, 227, 231, 235, 239, 243, 247)
  sizePP16[[14]] <- c(262, 266, 270, 274, 278, 282, 286, 290)
  sizePP16[[15]] <- c(322, 326, 330, 332, 334, 336, 338, 340, 342, 344, 346, 348, 350, 352, 354, 356, 358, 360, 362, 366, 370, 374, 378, 384)
  names(sizePP16) <- lociPP16
  
  
  #############################
  # Default of GUI parameters #
  #############################
  
  cspName <- tclVar("")
  cspFp <- tclVar("")
  cspButtS <- tclVar("disabled")
  cspButtC <- tclVar("arrow")
  refName <- tclVar("")
  refFp <- tclVar("")
  refButtS <- tclVar("disabled")
  refButtC <- tclVar("arrow")
  afName <- tclVar("")
  afFp <- tclVar("")
  afButtS <- tclVar("disabled")
  afButtC <- tclVar("arrow")
  mcName <- tclVar("")
  mcFp <- tclVar("")
  mcButtS <- tclVar("disabled")
  mcButtC <- tclVar("")
  alCorName <- tclVar("")
  alCorFp <- tclVar("")
  alCorButtS <- tclVar("disabled")
  alCorButtC <- tclVar("")
  
  hpVars <- hdVars <- atVars <- list()
  
  hncFromVar <- tclVar("1")
  hncToVar <- tclVar("4")
  mafVar <- tclVar("0.001")
  thetaVar <- tclVar("0")
  numMcVar <- tclVar("1000")
  
  stVar <- tclVar("500")
  hbFltrVar <- tclVar("0.25")
  stB1FltrVar <- tclVar("0.7")
  stF1FltrVar <- tclVar("0.35")
  stB2FltrVar <- tclVar("0.13")
  stM2FltrVar <- tclVar("0.12")
  afMethVar <- tclVar("0")
  
  fileCkFin <- tclVar("0")
  calcFin <- tclVar("0")
  
  hncFromSpinVar <- tclVar("normal")
  hncToSpinVar <- tclVar("normal")
  atEntryVar <- tclVar("normal")
  thetaEntryVar <- tclVar("normal")
  otherArrowVar <- tclVar("hand2")
  otherButtVar <- tclVar("normal")
  resetArrowVar <- tclVar("arrow")
  resetButtVar <- tclVar("disable")
  
  
  ###############
  # 'Files' tab #
  ###############
  
  #Make the 'Files' tab
  tab1Make <- function(){
    
    #Input required files
    openFile <- function(fp, var, top, fileState, cursor){
      imputOk <- "ok"
      if(tclvalue(calcFin) == "1"){
        imputOk <- tclvalue(tkmessageBox(message = "Calculation results will be deleted. Do you want to continue?", type = "okcancel", icon = "warning"))
      }
      if(imputOk == "ok"){
        fileName <- tclvalue(tkgetOpenFile(parent = top, initialdir = tclvalue(fp), multiple = "true", filetypes = "{{CSV Files} {.csv}}"))
        if(!nchar(fileName)){
          tkmessageBox(message = "No file was selected!", icon = "error", type = "ok")
        }else{
          tmp <- sub("\\}", fileName, replacement = "")
          tmp2 <- sub("\\{", tmp, replacement = "")
          tclvalue(fp) <- tmp2
          foo3 <- strsplit(tmp2, "/")[[1]]
          tclvalue(var) <- strsplit(foo3[length(foo3)], "\\.csv")[[1]][1]
          tclvalue(fileState) <- "normal"
          tclvalue(cursor) <- "hand2"
        }
        tclvalue(fileCkFin) <- 0
        tclvalue(calcFin) <- 0
        tab1Make()
        tab2Make()
        tab3Make()
        tab4Make()
      }
    }
    
    #View the inputted CSP profile
    cspView <- function(){
      if(paste(tclvalue(cspName)) == ""){
        tkmessageBox(message = "Load Crime Stain Profile!", icon = "error", type = "ok")
      }else{
        csp <- read.csv(tclvalue(cspFp), header = FALSE)
        csp <- as.matrix(csp)
        csp <- gsub(" ", "", csp, fixed = TRUE)
        csp <- csp[, - grep("Sample", csp[1, ])]
        cspLoci <- csp[, grep("Marker", csp[1, ])]
        csp <- csp[!is.element(cspLoci, sexChrMar), , drop = FALSE]
        cspTclArray <- tclArray()
        for(i in 1:nrow(csp)){
          for(j in 1:ncol(csp)){
            if(is.na(csp[i, j])){
              cspTclArray[[i - 1, j - 1]] <- "-"
            }else if(csp[i, j] == ""){
              cspTclArray[[i - 1, j - 1]] <- "-"
            }else{
              cspTclArray[[i - 1, j - 1]] <- csp[i, j]
            }
          }
        }
        cspTf <- tktoplevel()
        tkwm.title(cspTf, "Crime Stain Profile")
        cspFrame <- tkframe(cspTf)
        cspTable <- tkwidget(cspFrame, "table", variable = cspTclArray, rows = nrow(csp), cols = ncol(csp), titlerows = "1", titlecols = "1", selectmode = "extended", colwidth = "15", xscrollcommand = function(...) tkset(xscr, ...), yscrollcommand = function(...) tkset(yscr, ...))
        xscr <- tkscrollbar(cspFrame, orient = "horizontal", command = function(...) tkxview(cspTable, ...))
        yscr <- tkscrollbar(cspFrame, command = function(...) tkyview(cspTable, ...))
        tkgrid(cspTable, yscr)
        tkgrid.configure(yscr, sticky = "nsw")
        tkgrid(xscr, sticky = "new")
        tkconfigure(cspTable, variable = cspTclArray, background = "white", selectmode = "extended", state = "disable")
        tkgrid(cspFrame, padx = 50, pady = 50)
      }
    }
    
    #View the inputted reference profile
    refView <- function(){
      if(paste(tclvalue(refName)) == ""){
        tkmessageBox(message = "Load Reference Profiles!", icon = "error", type = "ok")
      }else{
        ref <- read.csv(tclvalue(refFp), header = FALSE)
        ref <- as.matrix(ref)
        ref <- gsub(" ", "", ref, fixed = TRUE)
        refLoci <- ref[, grep("Marker", ref[1, ])]
        ref <- ref[!is.element(refLoci, sexChrMar), , drop = FALSE]
        refTclArray <- tclArray()
        for(i in 1:nrow(ref)){
          for(j in 1:ncol(ref)){
            refTclArray[[i - 1, j - 1]] <- ref[i, j]
          }
        }
        refTf <- tktoplevel()
        tkwm.title(refTf, "Reference Profiles")
        refTable <- tkwidget(refTf, "table", variable = refTclArray, rows = nrow(ref), cols = ncol(ref), titlerows = "1", titlecols = "1", selectmode = "extended", colwidth = "15", background = "white", state = "disable")
        tkpack(refTable)
      }
    }
    
    #View the inputted allele frequencies
    afView <- function(){
      if(paste(tclvalue(afName)) == ""){
        tkmessageBox(message = "Load Allele Frequencies!", icon = "error", type = "ok")
      }else{
        af <- read.csv(tclvalue(afFp), header = FALSE, as.is = TRUE, na.strings = "")
        af <- as.matrix(af)
        af <- af[, !is.element(af[1, ], sexChrMar), drop = FALSE]
        afVal <- as.numeric(af[-1, -1])
        afVal <- afVal[!is.na(afVal)]
        if(length(which(afVal %% 1 != 0)) == 0){
          afName <- "Dirichlet distribution"
          for(i in 2:ncol(af)){
            afOneL <- as.numeric(af[-1, i])
            alPos <- which(is.na(afOneL) == FALSE)
            afOneL <- afOneL[alPos]
            afOneL <- (afOneL + 1) / (sum(afOneL) + length(afOneL))
            af[alPos + 1, i] <- afOneL
          }
        }else{
          afName <- "Observed frequencies"
        }
        afTclArray <- tclArray()
        for(j in 1:ncol(af)){
          afTclArray[[0, j - 1]] <- af[1, j]
        }
        for(i in 2:nrow(af)){
          for(j in 1:ncol(af)){
            if(is.na(af[i, j])){
              afTclArray[[i - 1, j - 1]] <- "-"
            }else{
              afTclArray[[i - 1, j - 1]] <- round(as.numeric(af[i, j]), 5)
            }
          }
        }
        afTf <- tktoplevel()
        tkwm.title(afTf, "Allele Frequencies")
        tkgrid(tklabel(afTf, text = afName), padx = 20, pady = 10, sticky = "w")
        afFrame <- tkframe(afTf)
        afTable <- tkwidget(afFrame, "table", rows = nrow(af), cols = ncol(af), titlerows = 1, titlecols = 1, height = 30, colwidth = "10", xscrollcommand = function(...) tkset(xscr, ...), yscrollcommand = function(...) tkset(yscr, ...))
        xscr <- tkscrollbar(afFrame, orient = "horizontal", command = function(...) tkxview(afTable, ...))
        yscr <- tkscrollbar(afFrame, command = function(...) tkyview(afTable, ...))
        tkgrid(afTable, yscr)
        tkgrid.configure(yscr, sticky = "nsw")
        tkgrid(xscr, sticky = "new")
        tkconfigure(afTable, variable = afTclArray, background = "white", selectmode = "extended", state = "disable")
        tkgrid(afFrame, padx = 50, pady = 10)
      }
    }
    
    #View the inputted parameters for Monte Carlo simulation
    mcView <- function(){
      if(paste(tclvalue(mcName)) == ""){
        tkmessageBox(message = "Load parameters for Monte Carlo simulation!", icon = "error", type = "ok")
      }else{
        mc <- read.csv(tclvalue(mcFp), header = FALSE)
        mc <- as.matrix(mc)
        mcLoci <- mc[, grep("Marker", mc[1, ])]
        mc <- mc[!is.element(mcLoci, sexChrMar), , drop = FALSE]
        mcTclArray <- tclArray()
        for(i in 1:nrow(mc)){
          for(j in 1:ncol(mc)){
            if(mc[i, j] == ""){
              mcTclArray[[i - 1, j - 1]] <- "-"
            }else{
              mcTclArray[[i - 1, j - 1]] <- mc[i, j]
            }
          }
        }
        mcTf <- tktoplevel()
        tkwm.title(mcTf, "Parameters for Monte Carlo simulation")
        mcFrame <- tkframe(mcTf)
        mcTable <- tkwidget(mcFrame, "table", variable = mcTclArray, rows = nrow(mc), cols = ncol(mc), titlerows = "1", titlecols = "1", selectmode = "extended", colwidth = "20", xscrollcommand = function(...) tkset(xscr, ...), yscrollcommand = function(...) tkset(yscr, ...))
        xscr <- tkscrollbar(mcFrame, orient = "horizontal", command = function(...) tkxview(mcTable, ...))
        yscr <- tkscrollbar(mcFrame, command = function(...) tkyview(mcTable, ...))
        tkgrid(mcTable, yscr)
        tkgrid.configure(yscr, sticky = "nsw")
        tkgrid(xscr, sticky = "new")
        tkconfigure(mcTable, variable = mcTclArray, background = "white", selectmode = "extended", state = "disable")
        tkgrid(mcFrame, padx = 50, pady = 50)
      }
    }
    
    #View the inputted information of allele repeat correction
    alCorView <- function(){
      if(paste(tclvalue(alCorName)) == ""){
        tkmessageBox(message = "Load Crime Stain Profile!", icon = "error", type = "ok")
      }else{
        alCor <- read.csv(tclvalue(alCorFp), header = FALSE)
        alCor <- as.matrix(alCor)
        alCorTclArray <- tclArray()
        for(i in 1:nrow(alCor)){
          for(j in 1:ncol(alCor)){
            if(alCor[i, j] == ""){
              alCorTclArray[[i - 1, j - 1]] <- "-"
            }else{
              alCorTclArray[[i - 1, j - 1]] <- alCor[i, j]
            }
          }
        }
        alCorTf <- tktoplevel()
        tkwm.title(alCorTf, "Allele Repeat Correction")
        alCorFrame <- tkframe(alCorTf)
        alCorTable <- tkwidget(alCorFrame, "table", variable = alCorTclArray, rows = nrow(alCor), cols = ncol(alCor), titlerows = "1", titlecols = "1", selectmode = "extended", colwidth = "20", xscrollcommand = function(...) tkset(xscr, ...), yscrollcommand = function(...) tkset(yscr, ...))
        xscr <- tkscrollbar(alCorFrame, orient = "horizontal", command = function(...) tkxview(alCorTable, ...))
        yscr <- tkscrollbar(alCorFrame, command = function(...) tkyview(alCorTable, ...))
        tkgrid(alCorTable, yscr)
        tkgrid.configure(yscr, sticky = "nsw")
        tkgrid(xscr, sticky = "new")
        tkconfigure(alCorTable, variable = alCorTclArray, background = "white", selectmode = "extended", state = "disable")
        tkgrid(alCorFrame, padx = 50, pady = 50)
      }
    }
    
    tkdestroy(frameTab1)
    frameTab1 <<- tkframe(tab1)
    
    frameTab1_1 <- tkframe(frameTab1)
    cspLabel <- tklabel(frameTab1_1, text = "Crime Stain Profile")
    cspFileLabel <- tklabel(frameTab1_1, textvariable = cspName, width = 30, highlightthickness = 1, relief = "groove", justify = "center", background = "white")
    cspButt <- tkbutton(frameTab1_1, text = "    Load    ", cursor = "hand2", command = function() openFile(cspFp, cspName, tf, cspButtS, cspButtC))
    cspEpgButt <- tkbutton(frameTab1_1, text = "        View Crime Stain Profile        ", cursor = tclvalue(cspButtC), state = tclvalue(cspButtS), command = function() cspView())
    
    refLabel <- tklabel(frameTab1_1, text = "Reference Profiles")
    refFileLabel <- tklabel(frameTab1_1, textvariable = refName, width = 30, highlightthickness = 1, relief = "groove", justify = "center", background = "white")
    refButt <- tkbutton(frameTab1_1, text = "    Load    ", cursor = "hand2", command = function() openFile(refFp, refName, tf, refButtS, refButtC))
    refGtButt <- tkbutton(frameTab1_1, text = "     View Reference Profiles     ", cursor = tclvalue(refButtC), state = tclvalue(refButtS), command = function() refView())
    
    afLabel <- tklabel(frameTab1_1, text = "Allele Frequencies")
    afFileLabel <- tklabel(frameTab1_1, textvariable = afName, width = 30, highlightthickness = 1, relief = "groove", justify = "center", background = "white")
    afButt <- tkbutton(frameTab1_1, text = "    Load    ", cursor = "hand2", command = function() openFile(afFp, afName, tf, afButtS, afButtC))
    afShowButt <- tkbutton(frameTab1_1, text = "  View Frequencies  ", cursor = tclvalue(afButtC), state = tclvalue(afButtS), command = function() afView())
    
    mcLabel <- tklabel(frameTab1_1, text = "Parameters for Monte Carlo simulation")
    mcFileLabel <- tklabel(frameTab1_1, textvariable = mcName, width = 30, highlightthickness = 1, relief = "groove", justify = "center", background = "white")
    mcButt <- tkbutton(frameTab1_1, text = "    Load    ", cursor = "hand2", command = function() openFile(mcFp, mcName, tf, mcButtS, mcButtC))
    mcShowButt <- tkbutton(frameTab1_1, text = "  View parameters for Monte Carlo simulation  ", cursor = tclvalue(mcButtC), state = tclvalue(mcButtS), command = function() mcView())
    
    alCorLabel <- tklabel(frameTab1_1, text = "Allele Repeat Correction")
    alCorFileLabel <- tklabel(frameTab1_1, textvariable = alCorName, width = 30, highlightthickness = 1, relief = "groove", justify = "center", background = "white")
    alCorButt <- tkbutton(frameTab1_1, text = "    Load    ", cursor = "hand2", command = function() openFile(alCorFp, alCorName, tf, alCorButtS, alCorButtC))
    alCorShowButt <- tkbutton(frameTab1_1, text = "  View Allele Repeat Correction  ", cursor = tclvalue(alCorButtC), state = tclvalue(alCorButtS), command = function() alCorView())
    
    frameTab1_2 <- tkframe(frameTab1)
    nextButt <- tkbutton(frameTab1_2, text = "    Next    ", cursor = "hand2", command = function() fileCk())
    
    tkgrid(cspLabel, cspFileLabel, cspButt, cspEpgButt, padx = 20, pady = 20, sticky = "w")
    tkgrid(refLabel, refFileLabel, refButt, refGtButt, padx = 20, pady = 20, sticky = "w")
    tkgrid(afLabel, afFileLabel, afButt, afShowButt, padx = 20, pady = 20, sticky = "w")
    tkgrid(mcLabel, mcFileLabel, mcButt, mcShowButt, padx = 20, pady = 20, sticky = "w")
    tkgrid(alCorLabel, alCorFileLabel, alCorButt, alCorShowButt, padx = 20, pady = 20, sticky = "w")
    tkgrid(nextButt, padx = 40, pady = 20, sticky = "w")
    tkgrid(frameTab1_1)
    tkgrid(frameTab1_2)
    tkgrid(frameTab1)
  }
  
  
  #####################
  # Check input files #
  #####################
  
  #Check input files
  fileCk <- function(){
    if(any(c(tclvalue(cspFp) == "", tclvalue(refFp) == "", tclvalue(afFp) == "", tclvalue(mcFp) == ""))){
      tkmessageBox(message = "Load required file(s)!", icon = "error", type = "ok")
    }else{
      csp <- read.csv(tclvalue(cspFp), header = TRUE)
      csp <- as.matrix(csp)
      csp <- gsub(" ", "", csp, fixed = TRUE)
      cspLoci <- csp[, grep("Marker", colnames(csp))]
      cspPos <- !is.element(cspLoci, sexChrMar)
      csp <- csp[cspPos, , drop = FALSE]
      cspLoci <- cspLoci[cspPos]
      
      ref <- read.csv(tclvalue(refFp), header = TRUE)
      ref <- as.matrix(ref)
      ref <- gsub(" ", "", ref, fixed = TRUE)
      refLoci <- ref[, grep("Marker", colnames(ref))]
      refPos <- !is.element(refLoci, sexChrMar)
      ref <- ref[refPos, , drop = FALSE]
      refLoci <- refLoci[refPos]
      
      af <- read.csv(tclvalue(afFp), header = TRUE)
      af <- as.matrix(af)
      afPos <- !is.element(colnames(af), sexChrMar)
      af <- af[, afPos, drop = FALSE]
      afLoci <- colnames(af)[afPos]
      afLoci <- afLoci[- grep("Allele", colnames(af))]
      
      mcParam <- read.csv(tclvalue(mcFp), header = TRUE)
      mcParam <- as.matrix(mcParam)
      aeParam <- mcParam[mcParam[, grep("Factor", colnames(mcParam))] == "AE", , drop = FALSE]
      hbParam <- mcParam[mcParam[, grep("Factor", colnames(mcParam))] == "Hb", , drop = FALSE]
      srB1Param <- mcParam[mcParam[, grep("Factor", colnames(mcParam))] == "BSR", , drop = FALSE]
      srF1Param <- mcParam[mcParam[, grep("Factor", colnames(mcParam))] == "FSR", , drop = FALSE]
      srB2Param <- mcParam[mcParam[, grep("Factor", colnames(mcParam))] == "DSR", , drop = FALSE]
      srM2Param <- mcParam[mcParam[, grep("Factor", colnames(mcParam))] == "M2SR", , drop = FALSE]
      
      aeLoci <- aeParam[, grep("Marker", colnames(aeParam))]
      aePos <- !is.element(aeLoci, sexChrMar)
      aeParam <- aeParam[aePos, , drop = FALSE]
      aeLoci <- aeLoci[aePos]
      
      hbLoci <- hbParam[, grep("Marker", colnames(hbParam))]
      hbPos <- !is.element(hbLoci, sexChrMar)
      hbParam <- hbParam[hbPos, , drop = FALSE]
      hbLoci <- hbLoci[hbPos]
      
      srB1Loci <- srB1Param[, grep("Marker", colnames(srB1Param))]
      srB1Pos <- !is.element(srB1Loci, sexChrMar)
      srB1Param <- srB1Param[srB1Pos, , drop = FALSE]
      srB1Loci <- srB1Loci[srB1Pos]
      
      srF1Loci <- srF1Param[, grep("Marker", colnames(srF1Param))]
      srF1Pos <- !is.element(srF1Loci, sexChrMar)
      srF1Param <- srF1Param[srF1Pos, , drop = FALSE]
      srF1Loci <- srF1Loci[srF1Pos]
      
      srB2Loci <- srB2Param[, grep("Marker", colnames(srB2Param))]
      srB2Pos <- !is.element(srB2Loci, sexChrMar)
      srB2Param <- srB2Param[srB2Pos, , drop = FALSE]
      srB2Loci <- srB2Loci[srB2Pos]
      
      srM2Loci <- srM2Param[, grep("Marker", colnames(srM2Param))]
      srM2Pos <- !is.element(srM2Loci, sexChrMar)
      srM2Param <- srM2Param[srM2Pos, , drop = FALSE]
      srM2Loci <- srM2Loci[srM2Pos]
      
      if(all(c(setequal(cspLoci, refLoci), setequal(cspLoci, afLoci), setequal(cspLoci, aeLoci), setequal(cspLoci, hbLoci), setequal(cspLoci, srB1Loci), setequal(cspLoci, srF1Loci), setequal(cspLoci, srB2Loci), setequal(cspLoci, srM2Loci)))){
        alCorJudge <- TRUE
        if(any(is.element(unique(mcParam[, grep("Model", colnames(mcParam))]), c("LUS", "Multi-seq")))){
          if(tclvalue(alCorFp) == ""){
            alCorJudge <- FALSE
            tkmessageBox(message = "Load a file of 'allele repeat correction'!", icon = "error", type = "ok")
          }
        }
        if(alCorJudge){
          refPos <- match(cspLoci, refLoci)
          ref <- ref[refPos, , drop = FALSE]
          
          afPos <- match(cspLoci, afLoci)
          af <- af[, c(1, afPos + 1), drop = FALSE]
          afVal <- as.numeric(af[, -1])
          afVal <- afVal[!is.na(afVal)]
          if(length(which(afVal %% 1 != 0)) == 0){
            tclvalue(afMethVar) <- 1
          }else{
            tclvalue(afMethVar) <- 0
          }
          
          aePos <- match(cspLoci, aeLoci)
          aeParam <- aeParam[aePos, ]
          aeParamVal <- aeParam[, c(grep("Parameter1", colnames(mcParam)), grep("Parameter2", colnames(mcParam))), drop = FALSE]
          aeParamVal <- matrix(as.numeric(aeParamVal), nrow = nrow(aeParamVal))
          
          hbPos <- match(cspLoci, hbLoci)
          hbParam <- hbParam[hbPos, ]
          hbParamVal <- hbParam[, c(grep("Parameter1", colnames(mcParam)), grep("Parameter2", colnames(mcParam))), drop = FALSE]
          hbParamVal <- matrix(as.numeric(hbParamVal), nrow = nrow(hbParamVal))
          
          srB1Pos <- match(cspLoci, srB1Loci)
          srB1Param <- srB1Param[srB1Pos, ]
          srB1Consider <- !is.element(srB1Param[, grep("Model", colnames(srB1Param))], "No")
          srB1Model <- srB1Param[, grep("Model", colnames(srB1Param))]
          srB1ParamVal <- srB1Param[, c(grep("Parameter1", colnames(mcParam)), grep("Parameter2", colnames(mcParam)), grep("Parameter3", colnames(mcParam))), drop = FALSE]
          srB1ParamVal <- matrix(as.numeric(srB1ParamVal), nrow = nrow(srB1ParamVal))
          srB1ParamVal[which(is.na(srB1ParamVal) == TRUE, arr.ind = TRUE)] <- 0
          
          srF1Pos <- match(cspLoci, srF1Loci)
          srF1Param <- srF1Param[srF1Pos, ]
          srF1Consider <- !is.element(srF1Param[, grep("Model", colnames(srF1Param))], "No")
          srF1Model <- srF1Param[, grep("Model", colnames(srF1Param))]
          srF1ParamVal <- srF1Param[, c(grep("Parameter1", colnames(mcParam)), grep("Parameter2", colnames(mcParam)), grep("Parameter3", colnames(mcParam))), drop = FALSE]
          srF1ParamVal <- matrix(as.numeric(srF1ParamVal), nrow = nrow(srF1ParamVal))
          srF1ParamVal[which(is.na(srF1ParamVal) == TRUE, arr.ind = TRUE)] <- 0
          
          srB2Pos <- match(cspLoci, srB2Loci)
          srB2Param <- srB2Param[srB2Pos, ]
          srB2Consider <- !is.element(srB2Param[, grep("Model", colnames(srB2Param))], "No")
          srB2Model <- srB2Param[, grep("Model", colnames(srB2Param))]
          srB2ParamVal <- srB2Param[, c(grep("Parameter1", colnames(mcParam)), grep("Parameter2", colnames(mcParam)), grep("Parameter3", colnames(mcParam))), drop = FALSE]
          srB2ParamVal <- matrix(as.numeric(srB2ParamVal), nrow = nrow(srB2ParamVal))
          srB2ParamVal[which(is.na(srB2ParamVal) == TRUE, arr.ind = TRUE)] <- 0
          
          srM2Pos <- match(cspLoci, srM2Loci)
          srM2Param <- srM2Param[srM2Pos, ]
          srM2Consider <- !is.element(srM2Param[, grep("Model", colnames(srM2Param))], "No")
          srM2Model <- srM2Param[, grep("Model", colnames(srM2Param))]
          srM2ParamVal <- srM2Param[, c(grep("Parameter1", colnames(mcParam)), grep("Parameter2", colnames(mcParam))), drop = FALSE]
          srM2ParamVal <- matrix(as.numeric(srM2ParamVal), nrow = nrow(srM2ParamVal))
          srM2ParamVal[which(is.na(srM2ParamVal) == TRUE, arr.ind = TRUE)] <- 0
          
          colnames(aeParamVal) <- colnames(hbParamVal) <- c("meanlog", "variance")
          rownames(aeParamVal) <- rownames(hbParamVal) <- rownames(srB1ParamVal) <- rownames(srF1ParamVal) <- rownames(srB2ParamVal) <- rownames(srM2ParamVal) <- cspLoci
          colnames(srB1ParamVal) <- colnames(srF1ParamVal) <- colnames(srB2ParamVal) <- c("slope", "intercept", "variance")
          colnames(srM2ParamVal) <- c("mean", "variance")
          
          srConsider <- cbind(srB1Consider, srF1Consider, srB2Consider, srM2Consider)
          srModel <- cbind(srB1Model, srF1Model, srB2Model, srM2Model)
          colnames(srConsider) <- c("srB1Consider", "srF1Consider", "srB2Consider", "srM2Consider")
          colnames(srModel) <- c("srB1Model", "srF1Model", "srB2Model", "srM2Model")
          rownames(srConsider) <- rownames(srModel) <- cspLoci
          
          repLengthAll <- as.numeric(aeParam[, grep("Repeat", colnames(aeParam))])
          names(repLengthAll) <- cspLoci
          
          numKnown <- (ncol(ref) - 1) / 2
          for(i in 1:numKnown){
            hpVars[[i]] <- tclVar("0")
            hdVars[[i]] <- tclVar("0")
          }
          
          dyeAllL <- aeParam[, grep("Dye", colnames(aeParam))]
          names(dyeAllL) <- cspLoci
          dyeNames <- unique(dyeAllL)
          for(i in 1:length(dyeNames)){
            atVars[[i]] <- tclVar("100")
          }
          names(atVars) <- dyeNames
          
          tclvalue(fileCkFin) <- 1
          tclvalue(hncFromSpinVar) <- "normal"
          tclvalue(hncToSpinVar) <- "normal"
          tclvalue(atEntryVar) <- "normal"
          tclvalue(thetaEntryVar) <- "normal"
          tclvalue(otherArrowVar) <- "hand2"
          tclvalue(otherButtVar) <- "normal"
          tclvalue(resetArrowVar) <- "arrow"
          tclvalue(resetButtVar) <- "disable"
          tab2Make(csp, ref, af, aeParamVal, hbParamVal, srB1ParamVal, srF1ParamVal, srB2ParamVal, srM2ParamVal, srConsider, srModel, repLengthAll, dyeAllL, hpVars, hdVars, atVars)
          tk2notetab.select(tabs, "Calculation")
        }
      }else{
        tkmessageBox(message = "Format of some files are inappropriate!", icon = "error", type = "ok")
      }
    }
  }
  
  #####################
  # 'Calculation' tab #
  #####################
  
  #Correct repeat number of variant alleles
  variantCor <- function(peakOne, repLength){
    corVal <- round(peakOne %% 1 * 10, 0)
    peakCor <- peakOne %/% 1 + round(corVal / repLength, 2)
    return(peakCor)
  }
  
  #Calculate template amount
  tempAmountCalc <- function(heightAllL){
    heightAllL[which(is.na(heightAllL) == TRUE, arr.ind = TRUE)] <- 0
    return(apply(heightAllL, 1, sum))
  }
  
  #Make population frequencies
  popFreqMake <- function(afOneL, afAlOneL, afMeth, maf = NULL, unobsAl = NULL){
    nUnobs <- length(unobsAl)
    if(nUnobs > 0){
      afAlOneL <- c(afAlOneL, unobsAl)
      if(afMeth == 1){
        afOneL <- c(afOneL, rep(0, nUnobs))
        afOneL <- (afOneL + 1) / (sum(afOneL) + length(afOneL))
      }else{
        afOneL <- c(afOneL, rep(maf, nUnobs))
      }
      afAlOneL2 <- afAlOneL[order(afAlOneL)]
      afOneL <- afOneL[order(afAlOneL)]
    }else{
      afAlOneL2 <- afAlOneL
      if(afMeth == 1){
        afOneL <- (afOneL + 1) / (sum(afOneL) + length(afOneL))
      }
    }
    return(list(afOneL, afAlOneL2))
  }
  
  #Extract drop-out alleles from reference profiles
  dropExt <- function(peakOneL, refOneL){
    refAl <- sort(unique(refOneL))
    return(refAl[!is.element(refAl, peakOneL)])
  }
  
  #Estimate size of a drop-out allele
  sizeDropEst <- function(dropOneAl, peakOneL, sizeOneL, repLength){
    peakOneLCor <- sapply(peakOneL, variantCor, repLength = repLength)
    dropOneAlCor <- variantCor(dropOneAl, repLength)
    distance <- abs(peakOneLCor - dropOneAlCor)
    basePeakPos <- which(distance == min(distance))
    basePeak <- peakOneLCor[basePeakPos[1]]
    baseSize <- sizeOneL[basePeakPos[1]]
    if(basePeak < dropOneAlCor){
      sizeDrop <- baseSize + repLength * min(distance) 
    }else{
      sizeDrop <- baseSize - repLength * min(distance)
    }
    return(sizeDrop)
  }
  
  #Estimate size of a drop-out allele in the case of locus drop-out
  sizeDropEst2 <- function(dropOneAl, kitAl, kitSize, repLength){
    dropAlPos <- which(kitAl == dropOneAl)
    if(length(dropAlPos) == 1){
      return(kitSize[dropAlPos])
    }else{
      kitAlCor <- sapply(kitAl, variantCor, repLength = repLength)
      dropOneAlCor <- variantCor(dropOneAl, repLength)
      distance <- abs(kitAlCor - dropOneAlCor)
      basePeakPos <- which(distance == min(distance))
      basePeak <- kitAlCor[basePeakPos[1]]
      baseSize <- kitSize[basePeakPos[1]]
      if(basePeak < dropOneAlCor){
        sizeDrop <- baseSize + repLength * min(distance)
      }else{
        sizeDrop <- baseSize - repLength * min(distance)
      }
      return(sizeDrop)
    }
  }
  
  #Arrange input data for calculation
  dataArrange <- function(csp, ref, af, atVals, dyeAllL, repLengthAll, afMeth, maf, alCor = NULL, srModel = NULL){
    srXmake <- function(srId){
      modelName <- srModelOneL[srId]
      if(modelName == "Allele"){
        srX <- peakOneL
      }else if(modelName == "LUS"){
        peakOneL2 <- peakOneL
        peakOneL2[peakOneL == 99] <- alQOneL
        srX <- as.numeric(alCorOneL[which(as.numeric(alCorOneL[, 2]) %in% peakOneL2), 3])
      }else if(modelName == "Multi-seq"){
        peakOneL2 <- peakOneL
        peakOneL2[peakOneL == 99] <- alQOneL
        srX <- as.numeric(alCorOneL[which(as.numeric(alCorOneL[, 2]) %in% peakOneL2), 3 + srId])
      }else{
        srX <- numeric(0)
      }
      return(srX)
    }
    
    colCsp <- colnames(csp)
    colRef <- colnames(ref)
    peakAllL <- csp[, grep("Allele", colCsp)]
    peakAllL <- matrix(as.numeric(peakAllL), nrow = nrow(peakAllL))
    sizeAllL <- csp[, grep("Size", colCsp)]
    sizeAllL <- matrix(as.numeric(sizeAllL), nrow = nrow(sizeAllL))
    sizeVec <- as.numeric(sizeAllL)
    sizeVec <- sizeVec[!is.na(sizeVec)]
    heightAllL <- csp[, grep("Height", colCsp)]
    heightAllL <- matrix(as.numeric(heightAllL), nrow = nrow(heightAllL))
    heightVec <- as.numeric(heightAllL)
    heightVec <- heightVec[!is.na(heightVec)]
    tempPerL <- tempAmountCalc(heightAllL)
    lociName <- csp[, grep("Marker", colCsp)]
    refAllL <- ref[, - grep("Marker", colRef)]
    refAllL <- matrix(as.numeric(refAllL), nrow = nrow(refAllL))
    colnames(refAllL) <- colRef[ -grep("Marker", colRef)]
    rownames(refAllL) <- lociName
    afAl <- af[, grep("Allele", colnames(af))]
    
    cspPeak <- cspSize <- cspHeight <- list()
    nL <- length(lociName)
    atAllL <- alQ <- QFreqAll <- rep(0, nL)
    popAlList <- popFreqList <- list()
    srB1Peak <- srF1Peak <- srB2Peak <- list()
    noSeqInfo <- rep("", nL)
    
    for(i in 1:nL){
      lociNameOne <- lociName[i]
      peakOneL <- peakAllL[i, !is.na(peakAllL[i, ])]
      sizeOneL <- sizeAllL[i, !is.na(sizeAllL[i, ])]
      heightOneL <- heightAllL[i, !is.na(heightAllL[i, ])]
      dyePos <- names(atVals) == dyeAllL[i]
      atAllL[i] <- atVals[dyePos]
      peakPos <- heightOneL >= atAllL[i]
      peakOneL <- peakOneL[peakPos]
      sizeOneL <- sizeOneL[peakPos]
      heightOneL <- heightOneL[peakPos]
      refOneL <- refAllL[i, ]
      
      afOneL <- af[, i + 1]
      afAlPos <- !is.na(afOneL)
      afAlOneL <- afAl[afAlPos]
      afOneL <- afOneL[afAlPos]
      unobsAl <- setdiff(unique(c(peakOneL, refOneL)), afAlOneL)
      popFreqData <- popFreqMake(afOneL, afAlOneL, afMeth, maf, unobsAl)
      popFreq <- popFreqData[[1]]
      popAl <- popFreqData[[2]]
      repLength <- repLengthAll[i]
      srModelOneL <- srModel[i, ]
      
      if(length(peakOneL) > 0){
        dropAl <- dropExt(peakOneL, refOneL)
        nD <- length(dropAl)
        if(nD > 0){
          sizeDrop <- sapply(dropAl, sizeDropEst, peakOneL = peakOneL, sizeOneL = sizeOneL, repLength = repLength)
          sizeOneL <- c(sizeOneL, sizeDrop)
          peakOneL <- c(peakOneL, dropAl)
          heightOneL <- c(heightOneL, rep(0, nD))
        }
      }else{
        dropAl <- sort(unique(refOneL))
        if(setequal(lociName, lociID)){
          posLID <- which(lociID == lociNameOne)
          kitAl <- alleleID[[posLID]]
          kitSize <- sizeID[[posLID]]
        }else if(setequal(lociName, lociGF)){
          posLGF <- which(lociGF == lociNameOne)
          kitAl <- alleleGF[[posLGF]]
          kitSize <- sizeGF[[posLGF]]
        }else if(setequal(lociName, lociNGM)){
          posLNGM <- which(lociNGM == lociNameOne)
          kitAl <- alleleNGM[[posLNGM]]
          kitSize <- sizeNGM[[posLNGM]]
        }else if(setequal(lociName, lociNGMSE)){
          posLNGMSE <- which(lociNGMSE == lociNameOne)
          kitAl <- alleleNGMSE[[posLNGMSE]]
          kitSize <- sizeNGMSE[[posLNGMSE]]
        }else if(setequal(lociName, lociMF)){
          posLMF <- which(lociMF == lociNameOne)
          kitAl <- alleleMF[[posLMF]]
          kitSize <- sizeMF[[posLMF]]
        }else if(setequal(lociName, lociSGMP)){
          posLSGMP <- which(lociSGMP == lociNameOne)
          kitAl <- alleleSGMP[[posLSGMP]]
          kitSize <- sizeSGMP[[posLSGMP]]
        }else if(setequal(lociName, lociPPF6C)){
          posLPPF6C <- which(lociPPF6C == lociNameOne)
          kitAl <- allelePPF6C[[posLPPF6C]]
          kitSize <- sizePPF6C[[posLPPF6C]]
        }else if(setequal(lociName, lociPPF)){
          posLPPF <- which(lociPPF == lociNameOne)
          kitAl <- allelePPF[[posLPPF]]
          kitSize <- sizePPF[[posLPPF]]
        }else if(setequal(lociName, lociPP21)){
          posLPP21 <- which(lociPP21 == lociNameOne)
          kitAl <- allelePP21[[posLPP21]]
          kitSize <- sizePP21[[posLPP21]]
        }else if(setequal(lociName, lociPP16)){
          posLPP16 <- which(lociPP16 == lociNameOne)
          kitAl <- allelePP16[[posLPP16]]
          kitSize <- sizePP16[[posLPP16]]
        }else{
          stop("ERROR: Sizes of drop-out alleles cannot be determined because of locus drop-out!")
        }
        sizeOneL <- sapply(dropAl, sizeDropEst2, kitAl, kitSize, repLength)
        peakOneL <- dropAl
        heightOneL <- rep(0, length(dropAl))
      }
      
      alQCandPos <- !is.element(popAl, peakOneL)
      alQCand <- popAl[alQCandPos]
      nQ <- length(alQCand)
      if(nQ > 0){
        alQ[i] <- alQOneL <- alQCand[which(popFreq[alQCandPos] == max(popFreq[alQCandPos]))[1]]
        sizeQ <- sizeDropEst(alQOneL, peakOneL, sizeOneL, repLength)
        QFreqAll[i] <- sum(popFreq[alQCandPos])
        peakOneL <- c(peakOneL, alQOneL)
        sizeOneL <- c(sizeOneL, sizeQ)
        heightOneL <- c(heightOneL, 0)
      }
      
      if(length(alCor) > 0){
        alCorOneL <- alCor[alCor[, grep("Marker", colnames(alCor))] == lociNameOne, ]
        seqAlOneL <- as.numeric(alCorOneL[, grep("Allele", colnames(alCorOneL))])
        noSeqAl <- setdiff(unique(c(peakOneL, refOneL)), seqAlOneL)
        noSeqJudge <- TRUE
        nNoSeq <- length(noSeqAl)
        if(nNoSeq > 0){
          srB1SeqJudge <- is.element(srModelOneL[1], c("LUS", "Multi-seq"))
          srF1SeqJudge <- is.element(srModelOneL[2], c("LUS", "Multi-seq"))
          srB2SeqJudge <- is.element(srModelOneL[3], c("LUS", "Multi-seq"))
          if(any(c(srB1SeqJudge, srF1SeqJudge, srB2SeqJudge))){
            noSeqJudge <- FALSE
            noSeqInfoOneL <- paste(lociNameOne, ": ", sep = "")
            for(j in 1:nNoSeq){
              noSeqInfoOneL <- paste(noSeqInfoOneL, noSeqAl[j], " ", sep = "")
            }
            noSeqInfo[i] <- noSeqInfoOneL
          }
        }
      }
      
      if(nQ > 0){
        peakOneL[length(peakOneL)] <- 99
      }
      
      if(noSeqJudge){
        srB1Peak[[i]] <- srXmake(1)
        srF1Peak[[i]] <- srXmake(2)
        srB2Peak[[i]] <- srXmake(3)
      }else{
        srB1Peak[[i]] <- srF1Peak[[i]] <- srB2Peak[[i]] <- numeric(0)
      }
      cspPeak[[i]] <- peakOneL
      cspSize[[i]] <- sizeOneL
      cspHeight[[i]] <- heightOneL
      popAlList[[i]] <- popAl
      popFreqList[[i]] <- popFreq
    }
    names(cspPeak) <- names(cspSize) <- names(cspHeight) <- names(popAlList) <- names(popFreqList) <- names(alQ) <- names(QFreqAll) <- lociName
    return(list(cspPeak, cspSize, cspHeight, refAllL, popAlList, popFreqList, alQ, QFreqAll, atAllL, tempPerL, srB1Peak, srF1Peak, srB2Peak, noSeqInfo))
  }
  
  #Extract calculation results of Hp and Hd from 'resultAllHnc' (including results of all hypotheses)
  resultExt <- function(resultAllHnc, nKnown, HpKnownID, HdKnownID, hncFrom, hncTo, hypIdAllList){
    resultHp <- resultHd <- list()
    countHnc <- 0
    for(i in hncFrom:hncTo){
      countHnc <- countHnc + 1
      hypIdUnique <- sapply(sapply(hypIdAllList[[countHnc]], as.numeric), unique)
      if(i - length(HpKnownID) >= 0){
        resultHp[[countHnc]] <- resultAllHnc[[countHnc]][[which(sapply(hypIdUnique, setequal, c(HpKnownID, rep(0, i - length(HpKnownID)))) == TRUE)]]
      }else{
        resultHp[[countHnc]] <- NULL
      }
      if(i - length(HdKnownID) >= 0){
        resultHd[[countHnc]] <- resultAllHnc[[countHnc]][[which(sapply(hypIdUnique, setequal, c(HdKnownID, rep(0, i - length(HdKnownID)))) == TRUE)]]
      }else{
        resultHd[[countHnc]] <- NULL
      }
    }
    return(list(resultHp, resultHd))
  }
  
  #Make a spinbox (GUI)
  tkspinbox <- function(parent, ...){
    tkwidget(parent, "tk::spinbox", ...)
  }
  
  #Make the 'Calculation' tab
  tab2Make <- function(csp = NULL, ref = NULL, af = NULL, aeParamVal = NULL, hbParamVal = NULL, srB1ParamVal = NULL, srF1ParamVal = NULL, srB2ParamVal = NULL, srM2ParamVal = NULL, srConsider = NULL, srModel = NULL, repLengthAll = NULL, dyeAllL = NULL, hpVars = NULL, hdVars = NULL, atVars = NULL){
    
    #Customize mixture ratios of one contributor
    mrCustomize <- function(){
      mrSort <- function(){
        mrList <- tclvalue(mrListVar)
        mrList <- unlist(strsplit(mrList, " "))
        mrList <- sort(as.numeric(mrList))
        nMrOne <- length(mrList)
        tkdelete(mrListbox, 0, size(mrListbox) - 1)
        for(i in 1:nMrOne){
          tkinsert(mrListbox, "end", mrList[i])
        }
      }
      
      mrRestore <- function(){
        tkdelete(mrListbox, 0, size(mrListbox) - 1)
        nMrOne <- length(mrDefault)
        for(i in 1:nMrOne){
          tkinsert(mrListbox, "end", mrDefault[i])
        }
      }
      
      mrAdd <- function(){
        mrNew <- tclvalue(mrAddVar)
        if(mrNew != ""){
          tkinsert(mrListbox, "end", tclvalue(mrAddVar))
        }
      }
      
      mrDelete <- function(){
        selectPos <- tkcurselection(mrListbox)
        selectPos <- tclvalue(selectPos)
        if(selectPos != ""){
          tkdelete(mrListbox, selectPos)
        }
      }
      
      mrSave <- function(){
        mrList <- tclvalue(mrListVar)
        mrList <- unlist(strsplit(mrList, " "))
        mrOne <<- sort(as.numeric(mrList))
        tkdestroy(mr.tf)
      }
      
      mrListVar <- tclVar("")
      mrAddVar <- tclVar("")
      mr.tf <- tktoplevel()
      tkwm.title(mr.tf, "Customize mixture ratios")
      mrListFrame <- tkframe(mr.tf)
      mrListbox <- tk2listbox(mrListFrame, listvariable = mrListVar, height = 10, justify = "center", selectmode = "single")
      tkgrid(tklabel(mrListFrame, text = "    mixture ratios per contributor"), padx = 20, sticky = "w")
      tkgrid(mrListbox, padx = 20, sticky = "w")
      nMrOne <- length(mrOne)
      for(i in 1:nMrOne){
        tkinsert(mrListbox, "end", mrOne[i])
      }
      mrButtFrame <- tkframe(mrListFrame)
      mrSortButt <- tkbutton(mrButtFrame, text = " Sort ", command = function() mrSort())
      mrRestoreButt <- tkbutton(mrButtFrame, text = " Restore Defaults ", command = function() mrRestore())
      tkgrid(mrSortButt, mrRestoreButt, padx = 10)
      tkgrid(mrButtFrame, pady = 10)
      
      mrSetFrame <- tkframe(mr.tf)
      mrAddFrame <- tkframe(mrSetFrame)
      mrAddEntry <- tkentry(mrAddFrame, textvariable = mrAddVar, width = 10, highlightthickness = 1, relief = "solid", justify = "center", background = "white")
      mrAddButt <- tkbutton(mrAddFrame, text = " Add ", command = function() mrAdd())
      mrDeleteFrame <- tkframe(mrSetFrame)
      mrDeleteButt <- tkbutton(mrDeleteFrame, text = " Delete ", command = function() mrDelete())
      mrSaveFrame <- tkframe(mrSetFrame)
      mrSaveButt <- tkbutton(mrSaveFrame, text = "Save", command = function() mrSave())
      tkgrid(tklabel(mrSetFrame, text = "Add an entered value to the list of mixture ratios"), padx = 20, sticky = "w")
      tkgrid(mrAddEntry, mrAddButt, padx = 20)
      tkgrid(mrAddFrame, padx = 20, sticky = "w")
      tkgrid(tklabel(mrSetFrame, text = ""), padx = 20, sticky = "w")
      tkgrid(tklabel(mrSetFrame, text = "Delete a selected mixture ratio"), padx = 20, sticky = "w")
      tkgrid(tklabel(mrDeleteFrame, text = "                    "), mrDeleteButt, padx = 20)
      tkgrid(mrDeleteFrame, padx = 20, sticky = "w")
      tkgrid(tklabel(mrSetFrame, text = ""), padx = 20, sticky = "w")
      tkgrid(tklabel(mrSetFrame, text = "Save the customized mixture ratios"), padx = 20, sticky = "w")
      tkgrid(tklabel(mrSaveFrame, text = "                    "), mrSaveButt, padx = 20)
      tkgrid(mrSaveFrame, padx = 20, sticky = "w")
      tkgrid(mrListFrame, mrSetFrame, pady = 10)
    }
    
    #Customize degradation parameters
    degCustomize <- function(){
      degSort <- function(){
        degList <- tclvalue(degListVar)
        degList <- unlist(strsplit(degList, " "))
        degList <- sort(as.numeric(degList))
        ndegOne <- length(degList)
        tkdelete(degListbox, 0, size(degListbox) - 1)
        for(i in 1:ndegOne){
          tkinsert(degListbox, "end", degList[i])
        }
      }
      
      degRestore <- function(){
        tkdelete(degListbox, 0, size(degListbox) - 1)
        ndegOne <- length(degDefault)
        for(i in 1:ndegOne){
          tkinsert(degListbox, "end", degDefault[i])
        }
      }
      
      degAdd <- function(){
        degNew <- tclvalue(degAddVar)
        if(degNew != ""){
          tkinsert(degListbox, "end", tclvalue(degAddVar))
        }
      }
      
      degDelete <- function(){
        selectPos <- tkcurselection(degListbox)
        selectPos <- tclvalue(selectPos)
        if(selectPos != ""){
          tkdelete(degListbox, selectPos)
        }
      }
      
      degSave <- function(){
        degList <- tclvalue(degListVar)
        degList <- unlist(strsplit(degList, " "))
        degOne <<- sort(as.numeric(degList))
        tkdestroy(deg.tf)
      }
      
      degListVar <- tclVar("")
      degAddVar <- tclVar("")
      deg.tf <- tktoplevel()
      tkwm.title(deg.tf, "Customize degradation parameters")
      degListFrame <- tkframe(deg.tf)
      degListbox <- tk2listbox(degListFrame, listvariable = degListVar, height = 10, justify = "center", selectmode = "single")
      tkgrid(tklabel(degListFrame, text = "    degradation parameters"), padx = 20, sticky = "w")
      tkgrid(degListbox, padx = 20, sticky = "w")
      ndegOne <- length(degOne)
      for(i in 1:ndegOne){
        tkinsert(degListbox, "end", degOne[i])
      }
      degButtFrame <- tkframe(degListFrame)
      degSortButt <- tkbutton(degButtFrame, text = " Sort ", command = function() degSort())
      degRestoreButt <- tkbutton(degButtFrame, text = " Restore Defaults ", command = function() degRestore())
      tkgrid(degSortButt, degRestoreButt, padx = 10)
      tkgrid(degButtFrame, pady = 10)
      
      degSetFrame <- tkframe(deg.tf)
      degAddFrame <- tkframe(degSetFrame)
      degAddEntry <- tkentry(degAddFrame, textvariable = degAddVar, width = 10, highlightthickness = 1, relief = "solid", justify = "center", background = "white")
      degAddButt <- tkbutton(degAddFrame, text = " Add ", command = function() degAdd())
      degDeleteFrame <- tkframe(degSetFrame)
      degDeleteButt <- tkbutton(degDeleteFrame, text = " Delete ", command = function() degDelete())
      degSaveFrame <- tkframe(degSetFrame)
      degSaveButt <- tkbutton(degSaveFrame, text = "Save", command = function() degSave())
      tkgrid(tklabel(degSetFrame, text = "Add an entered value to the list of degradation parameters"), padx = 20, sticky = "w")
      tkgrid(degAddEntry, degAddButt, padx = 20)
      tkgrid(degAddFrame, padx = 20, sticky = "w")
      tkgrid(tklabel(degSetFrame, text = ""), padx = 20, sticky = "w")
      tkgrid(tklabel(degSetFrame, text = "Delete a selected degradation parameter"), padx = 20, sticky = "w")
      tkgrid(tklabel(degDeleteFrame, text = "                    "), degDeleteButt, padx = 20)
      tkgrid(degDeleteFrame, padx = 20, sticky = "w")
      tkgrid(tklabel(degSetFrame, text = ""), padx = 20, sticky = "w")
      tkgrid(tklabel(degSetFrame, text = "Save the customized degradation parameters"), padx = 20, sticky = "w")
      tkgrid(tklabel(degSaveFrame, text = "                    "), degSaveButt, padx = 20)
      tkgrid(degSaveFrame, padx = 20, sticky = "w")
      tkgrid(degListFrame, degSetFrame, pady = 10)
    }
    
    #Make a window for setting other conditions
    paramSetting <- function(){
      tfParamSet <- tktoplevel()
      tkwm.title(tfParamSet, "Setting other conditions")
      frameParamSet <- tkframe(tfParamSet)
      
      mcLabel <- tklabel(frameParamSet, text = "Number of Monte Carlo simulations")
      mcEntry <- tkentry(frameParamSet, textvariable = numMcVar, width = 10, highlightthickness = 1, relief = "solid", justify = "center", background = "white")
      
      afLabel <- tklabel(frameParamSet, text = "Allele frequencies")
      mafLabel <- tklabel(frameParamSet, text = "                - Minimum allele frequency")
      if(tclvalue(afMethVar) == 1){
        dirRadioButt <- tkradiobutton(frameParamSet, anchor = "w", width = 40, state = "normal", text = "Dirichlet distribution", variable = afMethVar, value = 1)
        obsRadioButt <- tkradiobutton(frameParamSet, anchor = "w", width = 40, state = "disable", text = "Observed frequencies", variable = afMethVar, value = 0)
        mafEntry <- tkentry(frameParamSet, textvariable = mafVar, width = 10, state = "disable", highlightthickness = 1, relief = "solid", justify = "center", background = "white")
      }else{
        dirRadioButt <- tkradiobutton(frameParamSet, anchor = "w", width = 40, state = "disable", text = "Dirichlet distribution", variable = afMethVar, value = 1)
        obsRadioButt <- tkradiobutton(frameParamSet, anchor = "w", width = 40, state = "normal", text = "Observed frequencies", variable = afMethVar, value = 0)
        mafEntry <- tkentry(frameParamSet, textvariable = mafVar, width = 10, state = "normal", highlightthickness = 1, relief = "solid", justify = "center", background = "white")
      }
      
      mrLabel <- tklabel(frameParamSet, text = "Mixture ratio (MR)")
      mrCusButt <- tkbutton(frameParamSet, text = "    Customize    ", cursor = "hand2", state = "normal", command = function() mrCustomize())
      
      degLabel <- tklabel(frameParamSet, text = "Degradation (d)")
      degCusButt <- tkbutton(frameParamSet, text = "    Customize    ", cursor = "hand2", state = "normal", command = function() degCustomize())
      
      gtCombExcludeLab <- tklabel(frameParamSet, text = "Thresholds to exclude unrealistic genotype combinations")
      stLabel <- tklabel(frameParamSet, text = "        - Stochastic threshold")
      stEntry <- tkentry(frameParamSet, textvariable = stVar, width = 10, highlightthickness = 1, relief = "solid", justify = "center", background = "white")
      hbFltrLabel <- tklabel(frameParamSet, text = "        - Heterozygote balance filter")
      hbFltrEntry <- tkentry(frameParamSet, textvariable = hbFltrVar, width = 10, highlightthickness = 1, relief = "solid", justify = "center", background = "white")
      stB1FltrLabel <- tklabel(frameParamSet, text = "        - Back stutter filter")
      stB1FltrEntry <- tkentry(frameParamSet, textvariable = stB1FltrVar, width = 10, highlightthickness = 1, relief = "solid", justify = "center", background = "white")
      stF1FltrLabel <- tklabel(frameParamSet, text = "        - Forward stutter filter")
      stF1FltrEntry <- tkentry(frameParamSet, textvariable = stF1FltrVar, width = 10, highlightthickness = 1, relief = "solid", justify = "center", background = "white")
      stB2FltrLabel <- tklabel(frameParamSet, text = "        - Double-back stutter filter")
      stB2FltrEntry <- tkentry(frameParamSet, textvariable = stB2FltrVar, width = 10, highlightthickness = 1, relief = "solid", justify = "center", background = "white")
      stM2FltrLabel <- tklabel(frameParamSet, text = "        - Minus 2-nt stutter filter")
      stM2FltrEntry <- tkentry(frameParamSet, textvariable = stM2FltrVar, width = 10, highlightthickness = 1, relief = "solid", justify = "center", background = "white")
      
      tkgrid(mcLabel, mcEntry, padx = 20, sticky = "w")
      tkgrid(afLabel, padx = 20, sticky = "w")
      tkgrid(dirRadioButt, padx = 40, sticky = "w")
      tkgrid(obsRadioButt, padx = 40, sticky = "w")
      tkgrid(mafLabel, mafEntry, padx = 20, sticky = "w")
      tkgrid(mrLabel, mrCusButt, padx = 20, sticky = "w")
      tkgrid(degLabel, degCusButt, padx = 20, sticky = "w")
      tkgrid(gtCombExcludeLab, padx = 20, sticky = "w")
      tkgrid(stLabel, stEntry, padx = 20, sticky = "w")
      tkgrid(hbFltrLabel, hbFltrEntry, padx = 20, sticky = "w")
      tkgrid(stB1FltrLabel, stB1FltrEntry, padx = 20, sticky = "w")
      tkgrid(stF1FltrLabel, stF1FltrEntry, padx = 20, sticky = "w")
      tkgrid(stB2FltrLabel, stB2FltrEntry, padx = 20, sticky = "w")
      tkgrid(stM2FltrLabel, stM2FltrEntry, padx = 20, sticky = "w")
      tkgrid(frameParamSet, pady = 10)
    }
    
    #Users can change calculational conditions
    resetCondition <- function(){
      imputOk <- tclvalue(tkmessageBox(message = "Calculation results will be deleted. Do you want to continue?", type = "okcancel", icon = "warning"))
      if(imputOk == "ok"){
        tclvalue(calcFin) <- 0
        tkconfigure(hncFromSpin, state = "normal")
        tkconfigure(hncToSpin, state = "normal")
        for(i in 1:length(atEntry)){
          tkconfigure(atEntry[[i]], state = "normal")
        }
        tkconfigure(thetaEntry, state = "normal")
        tkconfigure(otherButt, cursor = "hand2", state = "normal")
        tkconfigure(resetButt, cursor = "arrow", state = "disable")
        tab3Make()
        tab4Make()
      }
    }
    
    #Prepare and finish calculation
    calcFunc <- function(){
      hncFrom <- as.numeric(tclvalue(hncFromVar))
      hncTo <- as.numeric(tclvalue(hncToVar))
      atVals <- as.numeric(sapply(atVars, tclvalue))
      names(atVals) <- names(atVars)
      afMeth <- as.numeric(tclvalue(afMethVar))
      if(afMeth == 1){
        maf <- 1
      }else{
        maf <- as.numeric(tclvalue(mafVar))
      }
      theta <- as.numeric(tclvalue(thetaVar))
      numMc <- as.numeric(tclvalue(numMcVar))
      mrDegCut <- as.numeric(tclvalue(mrDegCutVar))
      st <- as.numeric(tclvalue(stVar))
      hbFltr <- as.numeric(tclvalue(hbFltrVar))
      stFltr <- c(as.numeric(tclvalue(stB1FltrVar)), as.numeric(tclvalue(stF1FltrVar)), as.numeric(tclvalue(stB2FltrVar)), as.numeric(tclvalue(stM2FltrVar)))
      names(stFltr) <- c("stB1Fltr", "stF1Fltr", "stB2Fltr", "stM2Fltr")
      calcCond <- c(hncFrom, hncTo, atVals, afMeth, maf, theta, numMc, mrDegCut, st, hbFltr, stFltr)
      names(calcCond) <- c("hncFrom", "hncTo", names(atVals), "afMeth", "maf", "theta", "numMc", "mrDegCut", "st", "hbFltr", names(stFltr))
      alCorJudge <- TRUE
      
      if(tclvalue(alCorFp) != ""){
        alCor <- read.csv(tclvalue(alCorFp), header = TRUE)
        alCor <- as.matrix(alCor)
      }else{
        alCor <- NULL
      }
      dataAll <- dataArrange(csp, ref, af, atVals, dyeAllL, repLengthAll, afMeth, maf, alCor, srModel)
      cspPeak <- dataAll[[1]]
      cspSize <- dataAll[[2]]
      cspHeight <- dataAll[[3]]
      refAllL <- dataAll[[4]]
      popAlList <- dataAll[[5]]
      popFreqList <- dataAll[[6]]
      alQ <- dataAll[[7]]
      QFreqAll <- dataAll[[8]]
      atAllL <- dataAll[[9]]
      tempPerL <- dataAll[[10]]
      srB1Peak <- dataAll[[11]]
      srF1Peak <- dataAll[[12]]
      srB2Peak <- dataAll[[13]]
      noSeqInfo <- dataAll[[14]]
      
      if(tclvalue(calcFin) == "0"){
        if(!all(is.element(noSeqInfo, ""))){
          alCorJudge <- FALSE
          noSeqInfo <- noSeqInfo[noSeqInfo != ""]
          noSeqMessage <- paste(noSeqInfo[1])
          if(length(noSeqInfo) >= 2){
            for(i in 2:length(noSeqInfo)){
              noSeqMessage <- paste(noSeqMessage, "\n", noSeqInfo[i], sep = "")
            }
          }
          tkmessageBox(message = paste("There is a lack of information about allele repeat correction of the following allele(s): ", "\n", noSeqMessage, sep = ""), icon = "error", type = "ok")
        }
        
        if(alCorJudge){
          resultData <- cspInterpret(cspPeak, cspSize, cspHeight, refAllL, popAlList, popFreqList, QFreqAll, atAllL, tempPerL, srB1Peak, srF1Peak, srB2Peak, aeParamVal, hbParamVal, srB1ParamVal, srF1ParamVal, srB2ParamVal, srM2ParamVal, srConsider, repLengthAll, hncFrom, hncTo, mrOne, degOne, theta, numMc, mrDegCut, st, hbFltr, stFltr)
          resultAllHnc <<- resultData[[1]]
          gammaAllList <<- resultData[[2]]
          hypIdAllList <<- resultData[[3]]
          calcTime <<- resultData[[4]]
          mrOneCList <<- resultData[[5]]
          degOneC <<- resultData[[6]]
        }
      }
      if(alCorJudge){
        if(length(resultAllHnc) == 0){
          tkmessageBox(message = "Your crime stain profile cannot be explained by your setting hypotheses. Please recalculate after setting appropriate hypotheses.", icon = "error", type = "ok")
        }else{
          HpKnownID <- setdiff(which(as.numeric(sapply(hpVars, tclvalue)) == 1), 0)
          HdKnownID <- setdiff(which(as.numeric(sapply(hdVars, tclvalue)) == 1), 0)
          resultHpHd <- resultExt(resultAllHnc, numKnown, HpKnownID, HdKnownID, hncFrom, hncTo, hypIdAllList)
          resultHp <- resultHpHd[[1]]
          resultHd <- resultHpHd[[2]]
          tclvalue(calcFin) <- 1
          tclvalue(hncFromSpinVar) <- "disable"
          tclvalue(hncToSpinVar) <- "disable"
          tclvalue(atEntryVar) <- "disable"
          tclvalue(thetaEntryVar) <- "disable"
          tclvalue(otherArrowVar) <- "arrow"
          tclvalue(otherButtVar) <- "disable"
          tclvalue(resetArrowVar) <- "hand2"
          tclvalue(resetButtVar) <- "normal"
          try(dev.off(), silent = TRUE)
          tab2Make(csp, ref, af, aeParamVal, hbParamVal, srB1ParamVal, srF1ParamVal, srB2ParamVal, srM2ParamVal, srConsider, srModel, repLengthAll, dyeAllL, hpVars, hdVars, atVars)
          tab3Make(resultHp, resultHd, cspPeak, cspHeight, gammaAllList, dyeAllL, calcCond, mrOneCList, degOneC, repLengthAll)
          tab4Make(resultHp, resultHd, calcCond, mrOne, unique(dyeAllL), calcTime)
          tk2notetab.select(tabs, "Likelihood ratio")
        }
      }
    }
    
    tkdestroy(frameTab2)
    frameTab2 <<- tkframe(tab2)
    if(tclvalue(cspName) == ""){
      tkgrid(tklabel(frameTab2, text = "        Load crime stain profile!        "), pady = 10, sticky = "w")
    }
    if(tclvalue(refName) == ""){
      tkgrid(tklabel(frameTab2, text = "        Load reference profile!        "), pady = 10, sticky = "w")
    }
    if(tclvalue(afName) == ""){
      tkgrid(tklabel(frameTab2, text = "        Load allele frequencies!        "), pady = 10, sticky = "w")
    }
    if(tclvalue(mcName) == ""){
      tkgrid(tklabel(frameTab2, text = "        Load parameters for Monte Carlo simulation!        "), pady = 10, sticky = "w")
    }
    if(all(c(tclvalue(cspName), tclvalue(refName), tclvalue(afName), tclvalue(mcName)) != "")){
      if(tclvalue(fileCkFin) == "0"){
        tkgrid(tklabel(frameTab2, text = "        Push 'Next' button in the Files tab.        "), pady = 10, sticky = "w")
      }else{
        frameTab2_hypotheses <- tkframe(frameTab2)
        frameTab2_Hp <- tkframe(frameTab2_hypotheses, relief = "groove", borderwidth = 2)
        frameTab2_Hd <- tkframe(frameTab2_hypotheses, relief = "groove", borderwidth = 2)
        labelHp <- tklabel(frameTab2_Hp, text = "Prosecutor hypothesis", font = "Helvetica 10 bold")
        labelHd <- tklabel(frameTab2_Hd, text = "Defense hypothesis", font = "Helvetica 10 bold")
        frameHpKnown <- tkframe(frameTab2_Hp)
        frameHdKnown <- tkframe(frameTab2_Hd)
        
        numKnown <- (ncol(ref) - 1) / 2
        nameKnown <- colnames(ref[, 2:(numKnown * 2 + 1)])[2:(numKnown * 2 + 1) %% 2 == 0]
        for(i in 1:length(hpVars)){
          tkgrid(tkcheckbutton(frameHpKnown, text = paste(nameKnown[i]), variable = hpVars[[i]]), sticky = "w")
          tkgrid(tkcheckbutton(frameHdKnown, text = paste(nameKnown[i]), variable = hdVars[[i]]), sticky = "w")
        }
        tkgrid(labelHp, padx = 20)
        tkgrid(labelHd, padx = 20)
        tkgrid(frameHpKnown, padx = 20)
        tkgrid(frameHdKnown, padx = 20)
        tkgrid(tklabel(frameTab2_Hp, text = "(+ unknown contributor(s))"))
        tkgrid(tklabel(frameTab2_Hd, text = "(+ unknown contributor(s))"))
        tkgrid(frameTab2_Hp, frameTab2_Hd, padx = 20, pady = 10)
        tkgrid(frameTab2_hypotheses, padx = 20, pady = 10)
        
        frameTab2_param <- tkframe(frameTab2, relief = "groove", borderwidth = 2)
        tkgrid(tklabel(frameTab2_param, text = "Calculational conditions", font = "Helvetica 10 bold"), sticky = "w")
        frameTab2_param1 <- tkframe(frameTab2_param)
        hncFromSpin <- tkspinbox(frameTab2_param1, from = 1, to = 4, increment = 1, textvariable = hncFromVar, width = 6, highlightthickness = 1, justify = "center", background = "white", state = tclvalue(hncFromSpinVar))
        hncToSpin <- tkspinbox(frameTab2_param1, from = 1, to = 4, increment = 1, textvariable = hncToVar, width = 6, highlightthickness = 1, justify = "center", background = "white", state = tclvalue(hncToSpinVar))
        tkgrid(tklabel(frameTab2_param1, text = "Number of contributors"), sticky = "w")
        tkgrid(tklabel(frameTab2_param1, text = "    From"), hncFromSpin, sticky = "w")
        tkgrid(tklabel(frameTab2_param1, text = "    To"),  hncToSpin, sticky = "w")
        thetaEntry <- tkentry(frameTab2_param1, textvariable = thetaVar, width = 6, highlightthickness = 1, relief = "solid", justify = "center", background = "white", state = tclvalue(thetaEntryVar))
        tkgrid(tklabel(frameTab2_param1, text = "Theta"), thetaEntry, sticky = "w")
        
        frameTab2_param2 <- tkframe(frameTab2_param)
        tkgrid(tklabel(frameTab2_param2, text = "Analytical threshold"), sticky = "w")
        dyeNames <- unique(dyeAllL)
        atLabel <- atEntry <- list()
        for(i in 1:length(atVars)){
          atLabel[[i]] <- tklabel(frameTab2_param2, text = dyeNames[i])
          atEntry[[i]] <- tkentry(frameTab2_param2, textvariable = atVars[[i]], width = 6, highlightthickness = 1, relief = "solid", justify = "center", background = "white", state = tclvalue(atEntryVar))
          tkgrid(atLabel[[i]], atEntry[[i]], padx = 20, sticky = "w")
        }
        tkgrid(frameTab2_param1, frameTab2_param2, padx = 20, pady = 10, sticky = "nw")
        
        otherButt <- tkbutton(frameTab2_param, text = "    Others    ", cursor = tclvalue(otherArrowVar), state = tclvalue(otherButtVar), command = function() paramSetting())
        resetButt <- tkbutton(frameTab2_param, text = "    Change conditions    ", cursor = tclvalue(resetArrowVar), state = tclvalue(resetButtVar), command = function() resetCondition())
        tkgrid(otherButt, resetButt, padx = 20, pady = 10, sticky = "w")
        tkgrid(frameTab2_param, padx = 20, pady = 10)
        tkgrid(tkbutton(frameTab2, text = "    Calculate    ", cursor = "hand2", command = function() calcFunc()), padx = 20, pady = 10)
      }
    }
    tkgrid(frameTab2)
  }
  
  
  ####################
  # Main calculation #
  ####################
  
  #Set candidates of mixture ratios
  mrSetting <- function(hnc, mrOne){
    if(hnc == 1){
      return(matrix(1, 1, 1))
    }else{
      mrDigit <- 5
      mrOne <- round(mrOne, mrDigit)
      mrOne <- mrOne[mrOne > 0]
      mrOne <- sort(unique(mrOne[mrOne < 0.5]))
      nMrOne <- length(mrOne)
      mrOutN <- combinations(n = nMrOne, r = hnc - 1, v = mrOne, repeats.allowed = TRUE)
      mrOutN <- mrOutN[which(apply(mrOutN, 1, sum) < 1), , drop = FALSE]
      mrN <- round(1 - apply(mrOutN, 1, sum), mrDigit)
      mr <- cbind(mrOutN, mrN)
      mrEqualOutN <- rep(round(1 / hnc, mrDigit), hnc - 1)
      mr <- rbind(mr, c(mrEqualOutN, 1 - sum(mrEqualOutN)))
      mr <- t(apply(mr, 1, sort))
      mrCutCount <- 1
      repeat{
        mrCutPos <- setdiff(which(apply(mr, 1, setequal, mr[mrCutCount, ]) == TRUE), mrCutCount)
        if(length(mrCutPos) != 0){
          mr <- mr[-mrCutPos, ]
        }
        mrCutCount <- mrCutCount + 1
        if(nrow(mr) < mrCutCount){
          break
        }
      }
      for(i in 1:(hnc - 1)){
        mr <- mr[is.element(mr[, i], c(mrOne, mrEqualOutN)), ]
      }
      return(mr)
    }
  }
  
  #Set candidates of the degradation parameter of one contributor
  degOneCMake <- function(cspSize, cspHeight, tempPerL, degOne){
    sizeMeanCalc <- function(sizeVec, heightVec){
      sum(sizeVec * heightVec) / sum(heightVec)
    }
    
    lsDeg <- function(d, sizeVec, sizeMean, heightVec, tempMean){
      sum((exp(d * (sizeVec - sizeMean)) * tempMean / 2 - heightVec)^2)
    }
    
    sizeVec <- unlist(cspSize)
    heightVec <- unlist(cspHeight)
    sizeMean <- sizeMeanCalc(sizeVec, heightVec)
    tempMean <- mean(tempPerL)
    degError <- sapply(degOne, lsDeg, sizeVec = sizeVec, sizeMean = sizeMean, heightVec = heightVec, tempMean = tempMean)
    degOneContrib <- degOne[is.element(rank(degError), 1:3)]
    return(list(degOneContrib, sizeMean))
  }
  
  #Make genotype combinations
  gtCombMake <- function(peakOneL, heightOneL, hnc, srConsiderOneL, hbFltr, stFltr, st){
    
    #Make possible peak of observation in CSP under a hypothesized genotype combination
    possiblePeakMake <- function(gtCombOne, srConsiderOneL){
      possiblePeak <- gtCombOne
      if(srConsiderOneL[1]){
        possiblePeak <- c(possiblePeak, round(gtCombOne - 1, 1))
      }
      if(srConsiderOneL[2]){
        possiblePeak <- c(possiblePeak, round(gtCombOne + 1, 1))
      }
      if(srConsiderOneL[3]){
        possiblePeak <- c(possiblePeak, round(gtCombOne - 2, 1))
      }
      if(srConsiderOneL[4]){
        possiblePeak <- c(possiblePeak, round(gtCombOne[gtCombOne %% 1 < 0.15] - 0.8, 1), round(gtCombOne[gtCombOne %% 1 >= 0.15] - 0.2, 1))
      }
      return(sort(unique(possiblePeak)))
    }
    
    #Judge a genotype combination
    gtCombJudge <- function(gtCombOne, srConsiderOneL, peakOneL, heightOneL, hbFltr, stFltr, st){
      possiblePeak <- possiblePeakMake(gtCombOne, srConsiderOneL)
      peakObs <- peakOneL[heightOneL != 0]
      expJudge <- all(is.element(peakObs, possiblePeak))
      hbJudge <- TRUE
      stJudge <- TRUE
      
      #Memo: whether a genotype combination can explain all observed peaks in CSP
      if(expJudge){
        
        #Memo: exclude unrealistic genotype combination based on stutter filters
        stutters <- setdiff(peakObs, gtCombOne)
        nSt <- length(stutters)
        if(nSt > 0){
          stHeight <- heightOneL[which(peakOneL %in% stutters)]
          stHeight <- sapply(stHeight, rep, 4)
          stFltrRep <- matrix(rep(stFltr, nSt), nrow = 4)
          alNum <- matrix(0, 4, nSt)
          alNum[1, ] <- round(stutters + 1, 1)
          alNum[2, ] <- round(stutters - 1, 1)
          alNum[3, ] <- round(stutters + 2, 1)
          alNum[4, stutters %% 1 < 0.15] <- round(stutters[stutters %% 1 < 0.15] + 0.2, 1)
          alNum[4, stutters %% 1 >= 0.15] <- round(stutters[stutters %% 1 >= 0.15] + 0.8, 1)
          alNum <- as.numeric(alNum)
          alHeight <- rep(0.01, 4 * nSt)
          alObsPos <- is.element(alNum, peakOneL)
          alHeight[alObsPos] <- heightOneL[match(alNum[alObsPos], peakOneL)]
          alHeight <- matrix(alHeight, nrow = 4)
          stJudgeMat <- stHeight / alHeight <= stFltrRep
          stJudgeMat <- stJudgeMat[srConsiderOneL, , drop = FALSE]
          if(!all(apply(stJudgeMat, 2, any))){
            stJudge <- FALSE
          }
        }
        
        #Memo: exclude unrealistic genotype combination based on stochastic threshold and heterozygote balance
        if(stJudge){
          notShareHeightAll <- numeric(0)
          hnc <- (length(gtCombOne) / 2)
          for(i in 1:hnc){
            gtOne <- gtCombOne[c(2 * i - 1, 2 * i)]
            if((!any(is.element(gtOne, gtCombOne[- c(2 * i - 1, 2 * i)]))) && (gtOne[1] != gtOne[2])){
              heights <- c(heightOneL[peakOneL == gtOne[1]], heightOneL[peakOneL == gtOne[2]])
              #Memo: judge of Hb for one contributor
              if(all(heights >= st)){
                if((heights[1] / heights[2] < hbFltr) || (heights[1] / heights[2] > 1 / hbFltr)){
                  hbJudge <- FALSE
                  break
                }
              }else if(any(heights >= st) && any(heights == 0)){
                hbJudge <- FALSE
                break
              }
            }
            notShare <- setdiff(gtOne, gtCombOne[- c(2 * i - 1, 2 * i)])
            if(length(notShare) > 0){
              notShareHeight <- heightOneL[which(peakOneL %in% notShare)]
              notShareHeightAll <- c(notShareHeightAll, sum(notShareHeight) / 2)
            }
          }
          nNotShare <- length(notShareHeightAll)
          if(hbJudge && (nNotShare >= 2)){
            for(i in 1:(nNotShare - 1)){
              #Memo: judge of Hb between each contributor
              notShareHeight2 <- notShareHeightAll[i:length(notShareHeightAll)]
              hbFalsePos <- which(notShareHeight2[1] / notShareHeight2[-1] >= 1 / hbFltr)
              if((length(hbFalsePos) > 0) && (notShareHeight2[1] >= st)){
                hbJudge <- FALSE
                break
              }
            }
          }
        }
      }
      return(all(c(expJudge, hbJudge, stJudge)))
    }
    
    nAl <- length(peakOneL)
    gt <- combinations(n = nAl, r = 2, v = peakOneL, repeats.allowed = TRUE)
    nGt <- nrow(gt)
    gtCombID <- permutations(n = nGt, r = hnc, repeats.allowed = TRUE)
    nGtComb <- nrow(gtCombID)
    gtComb <- matrix(0, nGtComb, 2 * hnc)
    for(i in 1:hnc){
      gtComb[, c(2 * (i - 1) + 1, 2 * i)] <- gt[gtCombID[, i], ]
    }
    gtCombPassPos <- rep(0, nGtComb)
    for(i in 1:nGtComb){
      gtCombPassPos[i] <- gtCombJudge(gtComb[i, ], srConsiderOneL, peakOneL, heightOneL, hbFltr, stFltr, st)
    }
    gtCombPassPos <- which(gtCombPassPos == 1)
    if(length(gtCombPassPos) != 0){
      return(gtComb[gtCombPassPos, , drop = FALSE])
    }else{
      return(NULL)
    }
  }
  
  #Estimate expected peak heights
  ephEstimate <- function(peakOneL, sizeOneL, numMc, tempMean, mrOneC, degOneC, sizeMean, srB1PeakOneL, srF1PeakOneL, srB2PeakOneL, aeParamOneL, hbParamOneL, srB1ParamOneL, srF1ParamOneL, srB2ParamOneL, srM2ParamOneL){
    
    #Make conditions of mr, peak name, and d
    paramCondMake <- function(mrOneC, peakOneL, degOneC){
      numMr <- length(mrOneC)
      numPeak <- length(peakOneL)
      numDeg <- length(degOneC)
      numCond <- numMr * numPeak * numDeg
      paramCond <- matrix(0, 3, numCond)
      rownames(paramCond) <- c("mixture ratio", "allele", "degradation")
      paramCond[1, ] <- rep(mrOneC, numCond / numMr)
      paramCond[2, ] <- rep(as.vector(sapply(peakOneL, rep, numMr)), numDeg)
      paramCond[3, ] <- as.vector(sapply(degOneC, rep, numMr * numPeak))
      return(paramCond)
    }
    
    #Cauculate Dal values
    dalCalc <- function(d, sizeOneL, sizeMean){
      exp(d * (sizeOneL - sizeMean))
    }
    
    paramCond <- paramCondMake(mrOneC, peakOneL, degOneC)
    
    step2Height <- tempMean * mrOneC
    dalVal <- sapply(degOneC, dalCalc, sizeOneL = sizeOneL, sizeMean = sizeMean)
    step4Height <- as.vector(matrix(step2Height / 2, ncol = 1) %*% matrix(dalVal, nrow = 1))
    
    nStep4 <- length(step4Height)
    step5Height <- matrix(0, numMc, nStep4)
    aeMean <- as.numeric(aeParamOneL[1])
    aeVar <- as.numeric(aeParamOneL[2])
    for(i in 1:nStep4){
      aeVal <- exp(rtruncnorm(numMc, a = log(aeMin), b = log(1 / aeMin), mean = aeMean, sd = sqrt(aeVar / step4Height[i])))
      step5Height[, i] <- aeVal * step4Height[i]
    }
    
    nStep5 <- ncol(step5Height)
    step6Height <- matrix(0, numMc, nStep5)
    hbMean <- as.numeric(hbParamOneL[1])
    hbVar <- as.numeric(hbParamOneL[2])
    for(i in 1:nStep5){
      hbVal <- exp(rtruncnorm(numMc, a = log(hbMin), b = log(1 / hbMin), mean = hbMean, sd = sqrt(hbVar / step5Height[, i])))
      step6Height[, i] <- 2 / (1 + hbVal) * step5Height[, i]
    }
    
    nMrOne <- length(mrOneC)
    nDegOne <- length(degOneC)
    
    nStep6 <- ncol(step6Height)
    srB1Val <- matrix(0, numMc, nStep6)
    if(length(srB1PeakOneL) != 0){
      srB1Mean <- srB1ParamOneL[1] * srB1PeakOneL + srB1ParamOneL[2]      
      srB1Mean <- rep(as.vector(matrix(rep(srB1Mean, nMrOne), nrow = nMrOne, byrow = TRUE)), nDegOne)
      srB1Var <- srB1ParamOneL[3]
      for(i in 1:nStep6){
        srB1Val[, i] <- exp(rtruncnorm(numMc, b = log(srB1Max), mean = log(srB1Mean[i]), sd = sqrt(srB1Var / step6Height[, i])))
      }
    }
    
    srF1Val <- srB1Val * 0
    if(length(srF1PeakOneL) != 0){
      srF1Mean <- srF1ParamOneL[1] * srF1PeakOneL + srF1ParamOneL[2]
      srF1Mean <- rep(as.vector(matrix(rep(srF1Mean, nMrOne), nrow = nMrOne, byrow = TRUE)), nDegOne)
      srF1Var <- srF1ParamOneL[3]
      for(i in 1:nStep6){
        srF1Val[, i] <- exp(rtruncnorm(numMc, b = log(srF1Max), mean = log(srF1Mean[i]), sd = sqrt(srF1Var / step6Height[, i])))
      }
    }
    
    srB2Val <- srB1Val * 0
    if(length(srB2PeakOneL) != 0){
      srB2Mean <- srB2ParamOneL[1] * srB2PeakOneL + srB2ParamOneL[2]
      srB2Mean <- rep(as.vector(matrix(rep(srB2Mean, nMrOne), nrow = nMrOne, byrow = TRUE)), nDegOne)
      srB2Var <- srB2ParamOneL[3]
      for(i in 1:nStep6){
        srB2Val[, i] <- exp(rtruncnorm(numMc, b = log(srB2Max), mean = log(srB2Mean[i]), sd = sqrt(srB2Var / step6Height[, i])))
      }
    }
    
    srM2Val <- srB1Val * 0
    srM2Mean <- as.numeric(srM2ParamOneL[1])
    srM2Var <- as.numeric(srM2ParamOneL[2])
    if((srM2Mean != 0) && (srM2Var != 0)){
      for(i in 1:nStep6){
        srM2Val[, i] <- exp(rtruncnorm(numMc, b = log(srM2Max), mean = log(srM2Mean), sd = sqrt(srM2Var / step6Height[, i])))
      }
    }
    
    srSum <- 1 + srB1Val + srF1Val + srB2Val + srM2Val
    ephAl <- 1 / srSum * step6Height
    ephStB1 <- srB1Val / srSum * step6Height
    ephStF1 <- srF1Val / srSum * step6Height
    ephStB2 <- srB2Val / srSum * step6Height
    ephStM2 <- srM2Val / srSum * step6Height
    
    return(list(ephAl, ephStB1, ephStF1, ephStB2, ephStM2, paramCond))
  }
  
  #Estimate parameters of gamma distribution
  gammaEstimate <- function(eph){
    nCond <- ncol(eph)
    gammaParam <- matrix(0, 2, nCond)
    if(any(eph != 0)){
      m <- apply(eph, 2, mean)
      v <- apply(eph, 2, var)
      gammaParam[1, ] <- m^2 / v
      gammaParam[2, ] <- v / m
    }
    return(gammaParam)
  }
  
  #Obtain density values of gamma distribution
  gammaDens <- function(gtCombOne, mrDegAll, peakOneL, heightOneL, gammaAl, gammaStB1, gammaStF1, gammaStB2, gammaStM2, mrOneC, degOneC, at){
    
    #Sum gamma distributions of same allele repeat number
    gammaSumCalc <- function(gtCombOne, mr, deg, peakOneL, gammaAl, gammaStB1, gammaStF1, gammaStB2, gammaStM2, mrOneC, degOneC){
      nAl <- length(gtCombOne)
      nPeak <- length(peakOneL)
      nMrOneC <- length(mrOneC)
      nDegOneC <- length(degOneC)
      
      gammaSumPre1Al <- matrix(0, nAl, nPeak)
      gammaSumPre1StB1 <- gammaSumPre1StF1 <- gammaSumPre1StB2 <- gammaSumPre1StM2 <- gammaSumPre2Al <- gammaSumPre2StB1 <- gammaSumPre2StF1 <- gammaSumPre2StB2 <- gammaSumPre2StM2 <- gammaSumPre1Al
      
      peakAlPos <- match(gtCombOne, peakOneL)
      mrPos <- rep(match(mr, mrOneC), times = 1, each = 2)
      degPos <- rep(match(deg, degOneC), times = 1, each = 2)
      condPos <- nMrOneC * nPeak * (degPos - 1) + nMrOneC * (peakAlPos - 1) + mrPos
      
      gammaAlPos <- cbind(1:nAl, peakAlPos)
      p1 <- gammaAl[1, condPos]
      p2 <- gammaAl[2, condPos]
      gammaSumPre1Al[gammaAlPos] <- p1 * p2
      gammaSumPre2Al[gammaAlPos] <- p1 * (p2^2)
      
      gammaStPos <- matrix(0, nAl, 2)
      gammaStPos[, 1] <- 1:nAl
      
      peakStB1Pos <- match(round(gtCombOne - 1, 1), peakOneL)
      gammaStPos[, 2] <- peakStB1Pos
      peakStB1ObsPos <- !is.na(peakStB1Pos)
      gammaStB1Pos <- gammaStPos[peakStB1ObsPos, , drop = FALSE]
      p1 <- gammaStB1[1, condPos]
      p2 <- gammaStB1[2, condPos]
      gammaSumPre1StB1[gammaStB1Pos] <- (p1 * p2)[peakStB1ObsPos]
      gammaSumPre2StB1[gammaStB1Pos] <- (p1 * (p2^2))[peakStB1ObsPos]
      
      peakStF1Pos <- match(round(gtCombOne + 1, 1), peakOneL)
      gammaStPos[, 2] <- peakStF1Pos
      peakStF1ObsPos <- !is.na(peakStF1Pos)
      gammaStF1Pos <- gammaStPos[peakStF1ObsPos, , drop = FALSE]
      p1 <- gammaStF1[1, condPos]
      p2 <- gammaStF1[2, condPos]
      gammaSumPre1StF1[gammaStF1Pos] <- (p1 * p2)[peakStF1ObsPos]
      gammaSumPre2StF1[gammaStF1Pos] <- (p1 * (p2^2))[peakStF1ObsPos]
      
      peakStB2Pos <- match(round(gtCombOne - 2, 1), peakOneL)
      gammaStPos[, 2] <- peakStB2Pos
      peakStB2ObsPos <- !is.na(peakStB2Pos)
      gammaStB2Pos <- gammaStPos[peakStB2ObsPos, , drop = FALSE]
      p1 <- gammaStB2[1, condPos]
      p2 <- gammaStB2[2, condPos]
      gammaSumPre1StB2[gammaStB2Pos] <- (p1 * p2)[peakStB2ObsPos]
      gammaSumPre2StB2[gammaStB2Pos] <- (p1 * (p2^2))[peakStB2ObsPos]
      
      peakStM2Pos <- rep(0, nAl)
      m2AlPos1 <- gtCombOne %% 1 < 0.15
      m2AlPos2 <- gtCombOne %% 1 >= 0.15
      peakStM2Pos[m2AlPos1] <- match(round(gtCombOne[m2AlPos1] - 0.8, 1), peakOneL)
      peakStM2Pos[m2AlPos2] <- match(round(gtCombOne[m2AlPos2] - 0.2, 1), peakOneL)
      gammaStPos[, 2] <- peakStM2Pos
      peakStM2ObsPos <- !is.na(peakStM2Pos)
      gammaStM2Pos <- gammaStPos[peakStM2ObsPos, , drop = FALSE]
      p1 <- gammaStM2[1, condPos]
      p2 <- gammaStM2[2, condPos]
      gammaSumPre1StM2[gammaStM2Pos] <- (p1 * p2)[peakStM2ObsPos]
      gammaSumPre2StM2[gammaStM2Pos] <- (p1 * (p2^2))[peakStM2ObsPos]
      
      gammaSumPre1 <- apply(gammaSumPre1Al + gammaSumPre1StB1 + gammaSumPre1StF1 + gammaSumPre1StB2 + gammaSumPre1StM2, 2, sum)
      gammaSumPre2 <- apply(gammaSumPre2Al + gammaSumPre2StB1 + gammaSumPre2StF1 + gammaSumPre2StB2 + gammaSumPre2StM2, 2, sum)
      
      gammaSum <- matrix(0, 2, nPeak)
      gammaSum[1, ] <- gammaSumPre1^2 / gammaSumPre2
      gammaSum[2, ] <- gammaSumPre2 / gammaSumPre1
      return(gammaSum)
    }
    
    #Compare observed peak heights with gamma distribution (expected peak heights)
    cfOphEph <- function(heightOneL, gammaSum, at){
      numH <- length(heightOneL)
      dens <- rep(0, numH)
      for(i in 1:numH){
        if(heightOneL[i] == 0){
          dens[i] <- sum(dgamma(1:(at - 1), shape = gammaSum[1, i], scale = gammaSum[2, i]))
        }else{
          dens[i] <- dgamma(heightOneL[i], shape = gammaSum[1, i], scale = gammaSum[2, i])
        }
      }
      return(prod(dens))
    }
    
    hnc <- length(gtCombOne) / 2
    prodGtCombOne <- rep(0, nrow(mrDegAll))
    for(i in 1:nrow(mrDegAll)){
      mrDeg <- mrDegAll[i, ]
      mr <- mrDeg[1:hnc]
      deg <- mrDeg[(hnc + 1):(2 * hnc)]
      gammaSum <- gammaSumCalc(gtCombOne, mr, deg, peakOneL, gammaAl, gammaStB1, gammaStF1, gammaStB2, gammaStM2, mrOneC, degOneC)
      densPos <- !is.nan(gammaSum[1, ])
      dens <- cfOphEph(heightOneL[densPos], gammaSum[, densPos, drop = FALSE], at)
      prodGtCombOne[i] <- dens
    }
    return(prodGtCombOne)
  }
  
  #Calculate probabilities of a genotype combination
  gtProbCalc <- function(gtCombOne, hypIdOne, popAl, popFreq, theta, knownGt = NULL, QFreq = NULL){
    hnc <- length(gtCombOne) / 2
    gtUnique <- unique(gtCombOne)
    alCount <- rep(0, length(gtUnique))
    names(alCount) <- sort(gtUnique)
    nKnownAl <- length(knownGt)
    if(nKnownAl > 0){
      for(i in 1:nKnownAl){
        alCountPos <- which(as.numeric(names(alCount)) == knownGt[i])
        if(length(alCountPos) != 0){
          alCount[alCountPos] <- alCount[alCountPos] + 1
        }
      }
    }
    gtOneProb <- rep(0, hnc)
    for(i in 1:hnc){
      Id <- hypIdOne[i]
      gtOne <- gtCombOne[c(2 * i - 1, 2 * i)]
      if(Id != 0){
        if(setequal(gtOne, knownGt[c(2 * Id - 1, 2 * Id)])){
          gtOneProb[i] <- 1
        }
      }else{
        alProb <- rep(0, 2)
        for(j in 1:2){
          alOne <- gtOne[j]
          if(alOne == 99){
            alOneFreq <- QFreq
          }else{
            alOneFreq <- popFreq[popAl == alOne]
          }
          alOnePos <- as.numeric(names(alCount)) == alOne
          alProb[j] <- (alCount[alOnePos] * theta + (1 - theta) * alOneFreq) / ((nKnownAl + 2 * (i - 1) + j - 1) * theta + 1 - theta)
          alCount[alOnePos] <- alCount[alOnePos] + 1
        }
        if(gtOne[1] == gtOne[2]){
          gtOneProb[i] <- prod(alProb)
        }else{
          gtOneProb[i] <- 2 * prod(alProb)
        }
      }
    }
    return(prod(gtOneProb))
  }
  
  #Make all hypotheses
  hypIdMake <- function(nRef, hnc){
    personId <- 0:nRef
    hypIdComb <- combinations(n = length(personId), r = hnc, v = personId, repeats.allowed = TRUE)
    n1 <- nrow(hypIdComb)
    adoptPos <- numeric(0)
    for(i in 1:n1){
      hypIdOneCount <- table(hypIdComb[i, ])
      hypIdOneCount <- hypIdOneCount[names(hypIdOneCount) != 0]
      if((length(hypIdOneCount) == 0) || (all(is.element(hypIdOneCount, 1)))){
        adoptPos <- c(adoptPos, i)
      }
    }
    hypIdComb <- hypIdComb[adoptPos, , drop = FALSE]
    hypIdAll <- list()
    n1 <- nrow(hypIdComb)
    for(i in 1:n1){
      hypIdCombOne <- hypIdComb[i, hypIdComb[i, ] != 0]
      n2 <- length(hypIdCombOne)
      if(n2 > 0){
        kcOrder <- permutations(n = hnc, r = n2, repeats.allowed = FALSE)
        n3 <- nrow(kcOrder)
        hypId <- matrix(0, n3, hnc)
        for(j in 1:n3){
          hypId[j, kcOrder[j, ]] <- hypIdCombOne
        }
        hypIdAll[[i]] <- hypId
      }else{
        hypIdAll[[i]] <- matrix(0, 1, hnc)
      }
    }
    return(hypIdAll)
  }
  
  #Make the best hypothesis
  hypBestMake <- function(hypBestId, nameKnown){
    n1 <- length(hypBestId)
    hypBest <- rep(0, n1)
    for(i in 1:n1){
      if(hypBestId[i] == 0){
        hypBest[i] <- "Unknown"
      }else{
        hypBest[i] <- nameKnown[hypBestId[i]]
      }
    }
    return(hypBest)
  }
  
  #Make a list of results of Hp and Hd
  resultMake <- function(hypIdAll, nameKnown, mrDegAll, gtCombList, productsList, gtProbList, log10LikeList, overallLike, lociName){
    resultList <- list()
    nL <- length(lociName)
    countHyp <- 0
    for(i in 1:length(hypIdAll)){
      hypIdOneAll <- hypIdAll[[i]]
      maxLikeCand <- apply(overallLike[countHyp + 1:nrow(hypIdOneAll), , drop = FALSE], 1, max)
      maxLike <- max(maxLikeCand)
      hypBestIdPos <- which(maxLikeCand == maxLike)[1]
      hypBestId <- hypIdOneAll[hypBestIdPos, ]
      hypBest <- hypBestMake(hypBestId, nameKnown)
      mrDegBestId <- which(overallLike[countHyp + hypBestIdPos, ] == maxLike)[1]
      mrDegBest <- mrDegAll[mrDegBestId, ]
      mrBest <- mrDegBest[1:(length(mrDegBest) / 2)]
      degBest <- mrDegBest[(length(mrDegBest) / 2 + 1):(length(mrDegBest))]
      names(mrBest) <- names(degBest) <- hypBest
      pgResultList <- list()
      likeBestPerLocus <- rep(0, nL)
      for(j in 1:nL){
        gtComb <- gtCombList[[j]]
        productsBest <- productsList[[j]][, mrDegBestId]
        gtProbBest <- gtProbList[[j]][countHyp + hypBestIdPos, ]
        likeEachComb <- productsBest * gtProbBest
        possibleCombPos <- likeEachComb != 0
        pgResult <- cbind(gtComb[possibleCombPos, , drop = FALSE], productsBest[possibleCombPos] / sum(productsBest[possibleCombPos]), productsBest[possibleCombPos], gtProbBest[possibleCombPos])
        colnames(pgResult) <- c(as.vector(sapply(hypBest, rep, 2)), "weight", "product", "genotype freq")
        pgResultList[[j]] <- pgResult[order(pgResult[, ncol(pgResult) - 2], decreasing = TRUE), , drop = FALSE]
        likeBestPerLocus[j] <- log10LikeList[[j]][countHyp + hypBestIdPos, mrDegBestId]
      }
      names(likeBestPerLocus) <- lociName
      names(pgResultList) <- lociName
      resultOne <- list()
      resultOne[[1]] <- likeBestPerLocus
      resultOne[[2]] <- maxLike
      resultOne[[3]] <- pgResultList
      resultOne[[4]] <- mrBest
      resultOne[[5]] <- degBest
      resultList[[i]] <- resultOne
      names(resultList[[i]]) <- c("likelihood per locus", "overall likelihood", "probabilistic genotyping", "mixture ratio", "degradation")
      countHyp <- countHyp + nrow(hypIdOneAll)
    }
    return(resultList)
  }
  
  #Cut gtComb
  gtCombCut <- function(gtComb, peakOneL, heightOneL){
    hnc <- ncol(gtComb) / 2
    sortH <- sort(heightOneL, decreasing = TRUE)
    judgeMat <- matrix(TRUE, nrow(gtComb), hnc)
    for(i in 1:hnc){
      gt1p <- gtComb[, 2 * hnc - c(2 * (i - 1) + 1, 2 * (i - 1))]
      candAl <- peakOneL[which(heightOneL %in% sortH[1:(2 * i)])]
      judge1 <- apply(gt1p, 1, is.element, candAl)
      judgeMat[, hnc - i + 1] <- apply(judge1, 2, all)
    }
    gtCombRough <- gtComb[apply(judgeMat, 1, all), , drop = FALSE]
    return(gtCombRough)
  }
  
  #Interpret a CSP
  cspInterpret <- function(cspPeak, cspSize, cspHeight, refAllL, popAlList, popFreqList, QFreqAll, atAllL, tempPerL, srB1Peak, srF1Peak, srB2Peak, aeParamVal, hbParamVal, srB1ParamVal, srF1ParamVal, srB2ParamVal, srM2ParamVal, srConsider, repLengthAll, hncFrom, hncTo, mrOne, degOne, theta, numMc, mrDegCut, st, hbFltr, stFltr){
    lociName <- names(cspPeak)
    nL <- length(lociName)
    nK <- ncol(refAllL) / 2
    nameKnown <- colnames(refAllL)[1:(nK * 2) %% 2 == 1]
    tempMean <- mean(tempPerL)
    degData <- degOneCMake(cspSize, cspHeight, tempPerL, degOne)
    degOneC <- degData[[1]]
    sizeMean <- degData[[2]]
    
    plot(c(0, 100), c(0, 2 * (hncTo - hncFrom + 1) + 2), type = "n", xlab = "", ylab = "", axes = FALSE, mar = c(3, 3, 3, 3))
    for(i in hncFrom:hncTo){
      nBerPos <- 2 * (hncTo - i) + 1
      rect(0, nBerPos, 100, nBerPos + 1, col = "white")
      text(25, nBerPos + 1.5, paste("Hypothesis : ", i, "-person contribution", sep = ""))
    }
    par(xpd = TRUE)
    text(50, 2 * (hncTo - hncFrom + 1) + 2, "Progress Bar", cex = 1.25)
    
    calcTime <- matrix(0, nrow = 1, ncol = hncTo - hncFrom + 1)
    calcTimeName <- rep(0, hncTo - hncFrom + 1)
    
    resultAllHnc <- hypIdAllList <- gammaAllList <- mrOneCList <- list()
    countHnc <- 0
    cl <- makeCluster(4, type = "PSOCK")
    for(i in hncFrom:hncTo){
      t <- proc.time()
      countHnc <- countHnc + 1
      hnc <- i
      cat(paste("CSP Interpretation (", hnc, "-person contribution) was started.", sep = ""), "\n")
      nBerPos <- 2 * (hncTo - hnc) + 1
      mrAll <- mrSetting(hnc, mrOne)
      degAll <- matrix(rep(degOneC, hnc), nrow = length(degOneC))
      mrDegAll <- cbind(matrix(as.numeric(apply(mrAll, 1, rep, nrow(degAll))), ncol = hnc, byrow = TRUE), matrix(rep(as.numeric(t(degAll)), nrow(mrAll)), ncol = hnc, byrow = TRUE))
      nMD <- nrow(mrDegAll)
      mrDegID <- 1:nMD
      mrOneC <- sort(unique(as.vector(mrAll)))
      mrOneCList[[countHnc]] <- mrOneC
      hypIdAllList[[countHnc]] <- hypIdAll <- hypIdMake(ncol(refAllL) / 2, hnc)
      nH <- sum(sapply(hypIdAll, nrow))
      gammaList <- gtCombList <- productsList <- log10LikeList <- gtProbList <- list()
      overallLike <- matrix(0, nH, nMD)
      impossible <- FALSE
      for(j in 1:nL){
        peakOneL <- cspPeak[[j]]
        sizeOneL <- cspSize[[j]]
        heightOneL <- cspHeight[[j]]
        srConsiderOneL <- srConsider[j, ]
        refOneL <- refAllL[j, ]
        srB1PeakOneL <- srB1Peak[[j]]
        srF1PeakOneL <- srF1Peak[[j]]
        srB2PeakOneL <- srB2Peak[[j]]
        aeParamOneL <- aeParamVal[j, ]
        hbParamOneL <- hbParamVal[j, ]
        srB1ParamOneL <- srB1ParamVal[j, ]
        srF1ParamOneL <- srF1ParamVal[j, ]
        srB2ParamOneL <- srB2ParamVal[j, ]
        srM2ParamOneL <- srM2ParamVal[j, ]
        repLength <- repLengthAll[names(repLengthAll) == lociName[j]]
        popAl <- popAlList[[j]]
        popFreq <- popFreqList[[j]]
        QFreq <- QFreqAll[j]
        at <- atAllL[j]
        
        gtComb <- gtCombMake(peakOneL, heightOneL, hnc, srConsiderOneL, hbFltr, stFltr, st)
        gtCombList[[j]] <- gtComb
        if(length(gtComb) == 0){
          impossible <- TRUE
          break
        }else{
          gtCombRough <- gtCombCut(gtComb, peakOneL, heightOneL)
          ephData <- ephEstimate(peakOneL, sizeOneL, numMc, tempMean, mrOneC, degOneC, sizeMean, srB1PeakOneL, srF1PeakOneL, srB2PeakOneL, aeParamOneL, hbParamOneL, srB1ParamOneL, srF1ParamOneL, srB2ParamOneL, srM2ParamOneL)
          gammaAl <- gammaEstimate(ephData[[1]])
          gammaStB1 <- gammaEstimate(ephData[[2]])
          gammaStF1 <- gammaEstimate(ephData[[3]])
          gammaStB2 <- gammaEstimate(ephData[[4]])
          gammaStM2 <- gammaEstimate(ephData[[5]])
          gammaList[[j]] <- list(gammaAl, gammaStB1, gammaStF1, gammaStB2, gammaStM2)
          
          nGC <- nrow(gtCombRough)
          products <- matrix(0, nGC, nMD)
          if(hnc >= 3){
            productsOneL <- parRapply(cl, gtCombRough, gammaDens, mrDegAll[mrDegID, , drop = FALSE], peakOneL, heightOneL, gammaAl, gammaStB1, gammaStF1, gammaStB2, gammaStM2, mrOneC, degOneC, at) 
            products[, mrDegID] <- matrix(productsOneL, nrow = nGC, byrow = TRUE)
          }else{
            for(k in 1:nGC){
              products[k, mrDegID] <- gammaDens(gtCombRough[k, ], mrDegAll[mrDegID, , drop = FALSE], peakOneL, heightOneL, gammaAl, gammaStB1, gammaStF1, gammaStB2, gammaStM2, mrOneC, degOneC, at) 
            }
          }
          productsList[[j]] <- products
          log10Like <- mrDegAdopt <- matrix(0, nH, nMD)
          gtProb <- matrix(0, nH, nGC)
          countHyp <- 0
          nHI1 <- length(hypIdAll)
          for(k in 1:nHI1){
            hypIdOneAll <- hypIdAll[[k]]
            nHI2 <- nrow(hypIdOneAll)
            for(l in 1:nHI2){
              countHyp <- countHyp + 1
              hypIdOne <- hypIdOneAll[l, ]
              gtProb[countHyp, ] <- apply(gtCombRough, 1, gtProbCalc, hypIdOne, popAl, popFreq, theta, refOneL, QFreq)
              log10LikeOne <- log10(matrix(gtProb[countHyp, ], nrow = 1) %*% products)
              log10Like[countHyp, ] <- log10LikeOne
            }
            likePos <- (countHyp - nHI2 + 1):countHyp
            log10LikeOne2 <- as.numeric(log10Like[likePos, ])
            mrDegAdoptPos <- intersect(which(log10LikeOne2 != -Inf), which(log10LikeOne2 >= max(log10LikeOne2) + log10(mrDegCut)))
            if(length(mrDegAdoptPos) > 0){
              mrDegAdopt2 <- rep(0, length(log10LikeOne2))
              mrDegAdopt2[mrDegAdoptPos] <- 1
              mrDegAdopt[likePos, ] <- matrix(mrDegAdopt2, ncol = nMD)
            }
          }
          mrDegID <- which(apply(mrDegAdopt, 2, sum) != 0)
          if(length(mrDegID) == 0){
            impossible <- TRUE
            break
          }
          log10LikeList[[j]] <- log10Like
          overallLike <- overallLike + log10Like
          gtProbList[[j]] <- gtProb
          rect(0, nBerPos, 100 * j / nL, nBerPos + 1, col = "greenyellow")
        }
      }
      if(impossible){
        resultAllHnc[[countHnc]] <- NULL
        rect(0, nBerPos, 100, nBerPos + 1, col = "greenyellow")
      }else{
        mrDegFinal <- mrDegAll[mrDegID, , drop = FALSE]
        nMD <- nrow(mrDegFinal)
        
        for(j in 1:nL){
          peakOneL <- cspPeak[[j]]
          sizeOneL <- cspSize[[j]]
          heightOneL <- cspHeight[[j]]
          srConsiderOneL <- srConsider[j, ]
          refOneL <- refAllL[j, ]
          srB1PeakOneL <- srB1Peak[[j]]
          srF1PeakOneL <- srF1Peak[[j]]
          srB2PeakOneL <- srB2Peak[[j]]
          aeParamOneL <- aeParamVal[j, ]
          hbParamOneL <- hbParamVal[j, ]
          srB1ParamOneL <- srB1ParamVal[j, ]
          srF1ParamOneL <- srF1ParamVal[j, ]
          srB2ParamOneL <- srB2ParamVal[j, ]
          srM2ParamOneL <- srM2ParamVal[j, ]
          repLength <- repLengthAll[names(repLengthAll) == lociName[j]]
          popAl <- popAlList[[j]]
          popFreq <- popFreqList[[j]]
          QFreq <- QFreqAll[j]
          at <- atAllL[j]
          
          gtComb <- gtCombList[[j]]
          ephData <- ephEstimate(peakOneL, sizeOneL, numMc, tempMean, mrOneC, degOneC, sizeMean, srB1PeakOneL, srF1PeakOneL, srB2PeakOneL, aeParamOneL, hbParamOneL, srB1ParamOneL, srF1ParamOneL, srB2ParamOneL, srM2ParamOneL)
          gammaAl <- gammaEstimate(ephData[[1]])
          gammaStB1 <- gammaEstimate(ephData[[2]])
          gammaStF1 <- gammaEstimate(ephData[[3]])
          gammaStB2 <- gammaEstimate(ephData[[4]])
          gammaStM2 <- gammaEstimate(ephData[[5]])
          gammaList[[j]] <- list(gammaAl, gammaStB1, gammaStF1, gammaStB2, gammaStM2)
          
          nGC <- nrow(gtComb)
          if(hnc >= 3){
            productsOneL <- parRapply(cl, gtComb, gammaDens, mrDegFinal, peakOneL, heightOneL, gammaAl, gammaStB1, gammaStF1, gammaStB2, gammaStM2, mrOneC, degOneC, at) 
            products <- matrix(productsOneL, nrow = nGC, byrow = TRUE)
          }else{
            products <- matrix(0, nGC, nMD)
            for(k in 1:nGC){
              products[k, ] <- gammaDens(gtComb[k, ], mrDegFinal, peakOneL, heightOneL, gammaAl, gammaStB1, gammaStF1, gammaStB2, gammaStM2, mrOneC, degOneC, at) 
            }
          }
          productsList[[j]] <- products
          log10Like <- mrDegAdopt <- matrix(0, nH, nMD)
          gtProb <- matrix(0, nH, nGC)
          countHyp <- 0
          nHI1 <- length(hypIdAll)
          
          for(k in 1:nHI1){
            hypIdOneAll <- hypIdAll[[k]]
            nHI2 <- nrow(hypIdOneAll)
            for(l in 1:nHI2){
              countHyp <- countHyp + 1
              hypIdOne <- hypIdOneAll[l, ]
              gtProb[countHyp, ] <- apply(gtComb, 1, gtProbCalc, hypIdOne, popAl, popFreq, theta, refOneL, QFreq)
              log10LikeOne <- log10(matrix(gtProb[countHyp, ], nrow = 1) %*% products)
              log10Like[countHyp, ] <- log10LikeOne
            }
            likePos <- (countHyp - nHI2 + 1):countHyp
            log10LikeOne2 <- as.numeric(log10Like[likePos, ])
          }
          log10LikeList[[j]] <- log10Like
          overallLike <- overallLike + log10Like
          gtProbList[[j]] <- gtProb
        }
        resultAllHnc[[countHnc]] <- resultMake(hypIdAll, nameKnown, mrDegFinal, gtCombList, productsList, gtProbList, log10LikeList, overallLike, lociName)
      }
      gammaAllList[[countHnc]] <- gammaList
      timeOneHnc <- proc.time() - t
      calcTime[1, countHnc] <- timeOneHnc[3]
      calcTimeName[countHnc] <- paste(hncFrom + countHnc - 1, "-person", sep = "")
      cat(paste("Time of CSP interpretation (", hnc, "-person contribution) : ", round(timeOneHnc[3], 3), " sec.", sep = ""), "\n")
    }
    stopCluster(cl)
    colnames(calcTime) <- calcTimeName
    return(list(resultAllHnc, gammaAllList, hypIdAllList, calcTime, mrOneCList, degOneC))
  }
  
  
  ##################################
  # 'Probabilistic genotyping' tab #
  ##################################
  
  #Make the 'Probabilistic genotyping' tab
  tab3Make <- function(resultHp = NULL, resultHd = NULL, cspPeak = NULL, cspHeight = NULL, gammaList = NULL, dyeAllL = NULL, calcCond = NULL, mrOneCList = NULL, degOneC = NULL, repLengthAll = NULL){
    tkdestroy(frameTab3)
    frameTab3 <<- tkframe(tab3)
    if(tclvalue(cspName) == ""){
      tkgrid(tklabel(frameTab3, text = "        Load crime stain profile!        "), pady = 10, sticky = "w")
    }
    if(tclvalue(refName) == ""){
      tkgrid(tklabel(frameTab3, text = "        Load reference profile!        "), pady = 10, sticky = "w")
    }
    if(tclvalue(afName) == ""){
      tkgrid(tklabel(frameTab3, text = "        Load allele frequencies!        "), pady = 10, sticky = "w")
    }
    if(tclvalue(mcName) == ""){
      tkgrid(tklabel(frameTab3, text = "        Load parameters for Monte Carlo simulation!        "), pady = 10, sticky = "w")
    }
    if(all(c(tclvalue(cspName), tclvalue(refName), tclvalue(afName)) != "")){
      if(tclvalue(calcFin) == "0"){
        tkgrid(tklabel(frameTab3, text = "        Perform calculation!        "), pady = 10, sticky = "w")
      }else{
        #Make it easy to see the result of probabilistic genotyping
        pgRefine <- function(pg){
          n1 <- ncol(pg)
          gt <- pg[, 1:(n1 - 3), drop = FALSE]
          nC <- ncol(gt) / 2
          gtRefine <- matrix(0, nrow(gt), nC)
          nameC <- rep("", nC)
          for(i in 1:nC){
            for(j in 1:nrow(gt)){
              al_1 <- gt[j, 2 * i - 1]
              if(al_1 == 99){
                al_1 <- "Q"
              }
              al_2 <- gt[j, 2 * i]
              if(al_2 == 99){
                al_2 <- "Q"
              }
              gtRefine[j, i] <- paste("(", al_1, ", ", al_2, ")", sep = "")
            }
            nameC[i] <- colnames(gt)[2 * i - 1]
          }
          pgNew <- cbind(gtRefine, sprintf("%.2e", pg[, n1 - 2]))
          colnames(pgNew) <- c(nameC, colnames(pg)[n1 - 2])
          return(pgNew)
        }
        
        #Sum gamma distributions of the same allele repeat number
        gammaSumCalc <- function(gtCombOne, mr, deg, peakOneL, gammaAl, gammaStB1, gammaStF1, gammaStB2, gammaStM2, mrOneC, degOneC){
          nAl <- length(gtCombOne)
          nPeak <- length(peakOneL)
          nMrOneC <- length(mrOneC)
          nDegOneC <- length(degOneC)
          
          gammaSumPre1Al <- matrix(0, nAl, nPeak)
          gammaSumPre1StB1 <- gammaSumPre1StF1 <- gammaSumPre1StB2 <- gammaSumPre1StM2 <- gammaSumPre2Al <- gammaSumPre2StB1 <- gammaSumPre2StF1 <- gammaSumPre2StB2 <- gammaSumPre2StM2 <- gammaSumPre1Al
          
          peakAlPos <- match(gtCombOne, peakOneL)
          mrPos <- rep(match(mr, mrOneC), times = 1, each = 2)
          degPos <- rep(match(deg, degOneC), times = 1, each = 2)
          condPos <- nMrOneC * nPeak * (degPos - 1) + nMrOneC * (peakAlPos - 1) + mrPos
          
          gammaAlPos <- cbind(1:nAl, peakAlPos)
          p1 <- gammaAl[1, condPos]
          p2 <- gammaAl[2, condPos]
          gammaSumPre1Al[gammaAlPos] <- p1 * p2
          gammaSumPre2Al[gammaAlPos] <- p1 * (p2^2)
          
          gammaStPos <- matrix(0, nAl, 2)
          gammaStPos[, 1] <- 1:nAl
          
          peakStB1Pos <- match(round(gtCombOne - 1, 1), peakOneL)
          gammaStPos[, 2] <- peakStB1Pos
          peakStB1ObsPos <- !is.na(peakStB1Pos)
          gammaStB1Pos <- gammaStPos[peakStB1ObsPos, , drop = FALSE]
          p1 <- gammaStB1[1, condPos]
          p2 <- gammaStB1[2, condPos]
          gammaSumPre1StB1[gammaStB1Pos] <- (p1 * p2)[peakStB1ObsPos]
          gammaSumPre2StB1[gammaStB1Pos] <- (p1 * (p2^2))[peakStB1ObsPos]
          
          peakStF1Pos <- match(round(gtCombOne + 1, 1), peakOneL)
          gammaStPos[, 2] <- peakStF1Pos
          peakStF1ObsPos <- !is.na(peakStF1Pos)
          gammaStF1Pos <- gammaStPos[peakStF1ObsPos, , drop = FALSE]
          p1 <- gammaStF1[1, condPos]
          p2 <- gammaStF1[2, condPos]
          gammaSumPre1StF1[gammaStF1Pos] <- (p1 * p2)[peakStF1ObsPos]
          gammaSumPre2StF1[gammaStF1Pos] <- (p1 * (p2^2))[peakStF1ObsPos]
          
          peakStB2Pos <- match(round(gtCombOne - 2, 1), peakOneL)
          gammaStPos[, 2] <- peakStB2Pos
          peakStB2ObsPos <- !is.na(peakStB2Pos)
          gammaStB2Pos <- gammaStPos[peakStB2ObsPos, , drop = FALSE]
          p1 <- gammaStB2[1, condPos]
          p2 <- gammaStB2[2, condPos]
          gammaSumPre1StB2[gammaStB2Pos] <- (p1 * p2)[peakStB2ObsPos]
          gammaSumPre2StB2[gammaStB2Pos] <- (p1 * (p2^2))[peakStB2ObsPos]
          
          peakStM2Pos <- rep(0, nAl)
          m2AlPos1 <- gtCombOne %% 1 < 0.15
          m2AlPos2 <- gtCombOne %% 1 >= 0.15
          peakStM2Pos[m2AlPos1] <- match(round(gtCombOne[m2AlPos1] - 0.8, 1), peakOneL)
          peakStM2Pos[m2AlPos2] <- match(round(gtCombOne[m2AlPos2] - 0.2, 1), peakOneL)
          gammaStPos[, 2] <- peakStM2Pos
          peakStM2ObsPos <- !is.na(peakStM2Pos)
          gammaStM2Pos <- gammaStPos[peakStM2ObsPos, , drop = FALSE]
          p1 <- gammaStM2[1, condPos]
          p2 <- gammaStM2[2, condPos]
          gammaSumPre1StM2[gammaStM2Pos] <- (p1 * p2)[peakStM2ObsPos]
          gammaSumPre2StM2[gammaStM2Pos] <- (p1 * (p2^2))[peakStM2ObsPos]
          
          gammaSumPre1 <- apply(gammaSumPre1Al + gammaSumPre1StB1 + gammaSumPre1StF1 + gammaSumPre1StB2 + gammaSumPre1StM2, 2, sum)
          gammaSumPre2 <- apply(gammaSumPre2Al + gammaSumPre2StB1 + gammaSumPre2StF1 + gammaSumPre2StB2 + gammaSumPre2StM2, 2, sum)
          
          gammaSum <- matrix(0, 2, nPeak)
          gammaSum[1, ] <- gammaSumPre1^2 / gammaSumPre2
          gammaSum[2, ] <- gammaSumPre2 / gammaSumPre1
          return(gammaSum)
        }
        
        #Make a graph of observed peaks vs estimated gamma distributions
        pgGraphMake <- function(gtComb, at, mrEstimate, degEstimate, gammaParamOneL, mrOneC, degOneC, repLength){
          if(tclvalue(tkcurselection(pgMlb)) == ""){
            tkmessageBox(message = "Select one genotype combination!", icon = "error", type = "ok")
          }else{
            gtCombOne <- gtComb[as.numeric(tclvalue(tkcurselection(pgMlb))) + 1, ]
            
            lociName <- names(cspPeak)
            lPos <- which(lociName == tclvalue(selectL))
            peakOneL <- cspPeak[[lPos]]
            heightOneL <- cspHeight[[lPos]]
            if(max(heightOneL) >= at){
              yMax <- 2 * max(heightOneL)
            }else{
              yMax <- 2 * at
            }
            gamHeightCand <- 1:yMax
            
            gammaSum <- gammaSumCalc(gtCombOne, mrEstimate, degEstimate, peakOneL, gammaParamOneL[[1]], gammaParamOneL[[2]], gammaParamOneL[[3]], gammaParamOneL[[4]], gammaParamOneL[[5]], mrOneC, degOneC)
            densPos <- !is.nan(gammaSum[1, ])
            gammaSumGraph <- gammaSum[, densPos, drop = FALSE]
            peakOneLGraph <- peakOneLGraph2 <- peakOneL[densPos]
            if(any(is.element(peakOneLGraph, 99))){
              peakOneLGraph[which(peakOneLGraph == 99)] <- max(peakOneLGraph[which(peakOneLGraph != 99)]) + 2
            }
            heightOneLGraph <- heightOneL[densPos]
            
            plot(-1, -1, xlim = c(min(peakOneLGraph) - 1, max(peakOneLGraph) + 1), ylim = c(0, yMax), xaxt = "n", xlab = "Allele repeat number", ylab = "Peak height (RFU)", las = 1)
            segments(min(peakOneLGraph) - 1, 0, max(peakOneLGraph) + 1, 0)
            for(i in 1:length(peakOneLGraph)){
              peakOne <- variantCor(peakOneLGraph[i], repLength)
              segments(peakOne - 0.1, 0, peakOne, heightOneLGraph[i], col = "blue")
              segments(peakOne + 0.1, 0, peakOne, heightOneLGraph[i], col = "blue")
              dens <- dgamma(gamHeightCand, shape = gammaSumGraph[1, i], scale = gammaSumGraph[2, i])
              densPos <- dens > 0.005 * max(dens)
              dens <- dens[densPos]
              par(new = TRUE)
              plot(peakOne - 0.1 - 0.3 / max(dens) * dens, gamHeightCand[densPos], type = "l", col = "red", xlim = c(min(peakOneLGraph) - 1, max(peakOneLGraph) + 1), ylim = c(0, yMax), xlab = "", ylab = "", axes = FALSE)
              if(peakOneLGraph2[i] == 99){
                xLabelOne <- "Q"
              }else{
                xLabelOne <- peakOneLGraph2[i]
              }
              par(xpd = TRUE)
              text(peakOne, - yMax / 12, xLabelOne)
              par(xpd = FALSE)
            }
          }
        }
        
        #Show results of probabilistic genotyping
        pgShow <- function(hyp, numContributor, locusId, hncFrom){
          tkdestroy(pgMlb)
          tkdestroy(pgGraphButt)
          if(hyp == "Hp"){
            resultList <- resultHp
          }else{
            resultList <- resultHd
          }
          resultOneHnc <- resultList[[numContributor - hncFrom + 1]]
          if(length(resultOneHnc) == 0){
            pgMlb <<- tklabel(frameTab3_2, text = "Inappropriate hypothesis")
            tkgrid(pgMlb, padx = 20, pady = 20, sticky = "w")
            pgGraphButt <<- tklabel(frameTab3_2, text = "")
            tkgrid(pgGraphButt, padx = 20, pady = 20, sticky = "w")
          }else if(length(resultOneHnc[[3]][[locusId]]) == 0){
            pgMlb <<- tklabel(frameTab3_2, text = "Inappropriate hypothesis")
            tkgrid(pgMlb, padx = 20, pady = 20, sticky = "w")
            pgGraphButt <<- tklabel(frameTab3_2, text = "")
            tkgrid(pgGraphButt, padx = 20, pady = 20, sticky = "w")
          }else{
            pgResultOneL <- pgRefine(resultOneHnc[[3]][[locusId]])
            pgMlb <<- tk2mclistbox(frameTab3_2, width = 15 * ncol(pgResultOneL), height = 20, resizablecolumns = TRUE, selectmode = "single")
            for(i in 1:ncol(pgResultOneL)){
              tk2column(pgMlb, "add", label = colnames(pgResultOneL)[i], width = 15)
            }
            tkgrid(pgMlb, padx = 20, pady = 20, sticky = "w")
            tk2insert.multi(pgMlb, "end", pgResultOneL)
            pgGraphButt <<- tkbutton(frameTab3_2, text = "    Graph    ", cursor = "hand2", command = function() pgGraphMake(resultOneHnc[[3]][[locusId]][, 1:(2 * numContributor), drop = FALSE], calcCond[names(calcCond) == dyeAllL[locusId]], resultOneHnc[[4]], resultOneHnc[[5]], gammaAllList[[numContributor - hncFrom + 1]][[locusId]], mrOneCList[[numContributor - hncFrom + 1]], degOneC, repLengthAll[[locusId]]))
            tkgrid(pgGraphButt, padx = 20)
          }
        }
        
        #Export all results of probabilistic genotyping
        exportPg <- function(hncFrom, hncTo){
          savePg <- function(){
            if(tclvalue(pgFolderName) == ""){
              tkmessageBox(message = "Enter the folder name for saving files!", icon = "error", type = "ok")
            }else{
              folderPath <- choose.dir()
              if(!is.na(folderPath)){
                folderPath <- paste(folderPath, "/", tclvalue(pgFolderName), sep = "")
                dir.create(path = folderPath)
                countHnc <- 0
                for(i in hncFrom:hncTo){
                  countHnc <- countHnc + 1
                  resultOneHnc <- resultHp[[countHnc]]
                  if(length(resultOneHnc) != 0){
                    lociName <- names(resultOneHnc[[3]])
                    for(j in 1:length(resultOneHnc[[3]])){
                      if(length(resultOneHnc[[3]][[j]]) != 0){
                        pgResultOneL <- pgRefine(resultOneHnc[[3]][[j]])
                        write.csv(pgResultOneL, paste(folderPath, "/pgResult_Hp_", i, "-contributors_", lociName[j], ".csv", sep = ""), row.names = FALSE)
                      }
                    }
                  }
                  resultOneHnc <- resultHd[[countHnc]]
                  if(length(resultOneHnc) != 0){
                    lociName <- names(resultOneHnc[[3]])
                    for(j in 1:length(resultOneHnc[[3]])){
                      if(length(resultOneHnc[[3]][[j]]) != 0){
                        pgResultOneL <- pgRefine(resultOneHnc[[3]][[j]])
                        write.csv(pgResultOneL, paste(folderPath, "/pgResult_Hd_", i, "-contributors_", lociName[j], ".csv", sep = ""), row.names = FALSE)
                      }
                    }
                  }
                }
                tkmessageBox(message = paste("Results of probabilistic genotyping have been saved successfully in the folder of ", tclvalue(pgFolderName), ".", sep = ""), icon = "info", type = "ok")
              }
            }
          }
          exportPg.tf <- tktoplevel()
          tkwm.title(exportPg.tf, "Export results of probabilistic genotyping")
          tkgrid(tklabel(exportPg.tf, text = "Enter the folder name for saving files."), padx = 10, pady = 5, sticky = "w")
          pgFolderName <- tclVar("")
          tkgrid(tkentry(exportPg.tf, textvariable = pgFolderName, width = 30, highlightthickness = 1, relief = "solid", justify = "center", background = "white"), padx = 20, pady = 5, sticky = "w")
          tkgrid(tkbutton(exportPg.tf, text = "    Save    ", cursor = "hand2", command = function() savePg()), padx = 20, pady = 20, sticky = "w")
        }
        
        frameTab3_1 <- frameTab3_2 <- tkframe(frameTab3)
        frameTab3_1_1 <- tkframe(frameTab3_1)
        hypotheses <- c("Hp", "Hd")
        selectHyp <- tclVar("Hp")
        selectHypComboBox <- ttkcombobox(frameTab3_1_1, values = hypotheses, textvariable = selectHyp, state = "readonly")
        tkgrid(tklabel(frameTab3_1_1, text = "Hypothesis : "), selectHypComboBox, sticky = "w")
        hncFrom <- calcCond[names(calcCond) == "hncFrom"]
        hncTo <- calcCond[names(calcCond) == "hncTo"]
        nC <- hncFrom:hncTo
        selectNc <- tclVar(nC[1])
        selectNcComboBox <- ttkcombobox(frameTab3_1, values = nC, textvariable = selectNc, state = "readonly")
        tkgrid(tklabel(frameTab3_1_1, text = "Number of contributors : "), selectNcComboBox, sticky = "w")
        lociName <- names(resultHp[[which(sapply(resultHp, length) != 0)[1]]][[1]])
        selectL <- tclVar(lociName[1])
        selectLComboBox <- ttkcombobox(frameTab3_1_1, values = lociName, textvariable = selectL, state = "readonly")
        tkgrid(tklabel(frameTab3_1_1, text = "Locus : "), selectLComboBox, sticky = "w")
        pgShowButt <- tkbutton(frameTab3_1_1, text = "    Show    ", cursor = "hand2", command = function() pgShow(tclvalue(selectHyp), as.numeric(tclvalue(selectNc)), which(lociName == tclvalue(selectL)), hncFrom))
        pgExportButt <- tkbutton(frameTab3_1_1, text = "    Export    ", cursor = "hand2", command = function() exportPg(hncFrom, hncTo))
        tkgrid(pgShowButt, pgExportButt, sticky = "w")
        tkgrid(frameTab3_1_1, sticky = "w")
        tkgrid(frameTab3_1, padx = 20, pady = 20, sticky = "w")
        resultOneHnc <- resultHp[[as.numeric(tclvalue(selectNc)) - hncFrom + 1]]
        if(length(resultOneHnc) == 0){
          pgMlb <- tklabel(frameTab3_2, text = "Inappropriate hypothesis")
          tkgrid(pgMlb, padx = 20, pady = 20, sticky = "w")
          pgGraphButt <- tklabel(frameTab3_2, text = "")
          tkgrid(pgGraphButt, padx = 20, pady = 20, sticky = "w")
        }else if(length(resultOneHnc[[3]][[1]]) == 0){
          pgMlb <- tklabel(frameTab3_2, text = "Inappropriate hypothesis")
          tkgrid(pgMlb, padx = 20, pady = 20, sticky = "w")
          pgGraphButt <- tklabel(frameTab3_2, text = "")
          tkgrid(pgGraphButt, padx = 20, pady = 20, sticky = "w")
        }else{
          pgResultOneL <- pgRefine(resultOneHnc[[3]][[1]])
          pgMlb <- tk2mclistbox(frameTab3_2, width = 15 * ncol(pgResultOneL), height = 20, resizablecolumns = TRUE, selectmode = "single")
          for(i in 1:ncol(pgResultOneL)){
            tk2column(pgMlb, "add", label = colnames(pgResultOneL)[i], width = 15)
          }
          tkgrid(pgMlb, padx = 20, pady = 20, sticky = "w")
          tk2insert.multi(pgMlb, "end", pgResultOneL)
          pgGraphButt <- tkbutton(frameTab3_2, text = "    Graph    ", cursor = "hand2", command = function() pgGraphMake(resultOneHnc[[3]][[1]][, 1:(2 * as.numeric(tclvalue(selectNc))), drop = FALSE], calcCond[names(calcCond) == dyeAllL[1]], resultOneHnc[[4]], resultOneHnc[[5]], gammaAllList[[1]][[1]], mrOneCList[[1]], degOneC, repLengthAll[[1]]))
          tkgrid(pgGraphButt, padx = 20)
        }
        tkgrid(frameTab3_2, padx = 20, pady = 20, sticky = "w")
      }
    }
    tkgrid(frameTab3)
  }
  
  
  ##########################
  # 'Likelihood Ratio' tab #
  ##########################
  
  #Make a report
  reportMake <- function(HpKnownName, HdKnownName, log10LikeHp, log10LikeHd, lrAll, likeHpMax, likeHdMax, lrMax, mrEstimateHpAll, mrEstimateHdAll, dEstimateHpAll, dEstimateHdAll, calcCond, mrOne, dyeNames, calcTime){
    saveAs <- tkgetSaveFile(filetypes = "{{CSV Files} {.csv}}")
    if(tclvalue(saveAs) != ""){
      if(substr(tclvalue(saveAs), nchar(tclvalue(saveAs)) - 3, nchar(tclvalue(saveAs))) == ".csv"){
        reportName <- tclvalue(saveAs)
      }else{
        reportName <- paste(tclvalue(saveAs), ".csv", sep = "")
      }
      write.table("======== Software version ========", file = reportName, sep = ",", row.names = FALSE, col.names = FALSE, append = FALSE)
      write.table(softVer, file = reportName, sep = ",", row.names = FALSE, col.names = FALSE, append = TRUE)
      write.table("", file = reportName, sep = ",", row.names = FALSE, col.names = FALSE, append = TRUE)
      
      write.table("======== Imput files ========", file = reportName, sep = ",", row.names = FALSE, col.names = FALSE, append = TRUE)
      write.table(paste("Crime stain profile : ", tclvalue(cspName), sep = ""), file = reportName, sep = ",", row.names = FALSE, col.names = FALSE, append = TRUE)
      write.table(paste("Reference profile : ", tclvalue(refName), sep = ""), file = reportName, sep = ",", row.names = FALSE, col.names = FALSE, append = TRUE)
      write.table(paste("Allele frequencies : ", tclvalue(afName), sep = ""), file = reportName, sep = ",", row.names = FALSE, col.names = FALSE, append = TRUE)
      write.table(paste("Parameters for Monte Carlo simulation : ", tclvalue(mcName), sep = ""), file = reportName, sep = ",", row.names = FALSE, col.names = FALSE, append = TRUE)
      write.table(paste("Allele repeat correction : ", tclvalue(alCorName), sep = ""), file = reportName, sep = ",", row.names = FALSE, col.names = FALSE, append = TRUE)
      write.table("", file = reportName, sep = ",", row.names = FALSE, col.names = FALSE, append = TRUE)
      
      write.table("======== Calculational conditions ========", file = reportName, sep = ",", row.names = FALSE, col.names = FALSE, append = TRUE)
      write.table(paste("Number of contributors : from ", calcCond[names(calcCond) == "hncFrom"], " to ", calcCond[names(calcCond) == "hncTo"], sep = ""), file = reportName, sep = ",", row.names = FALSE, col.names = FALSE, append = TRUE)
      write.table("Analytical Threshold", file = reportName, sep = ",", row.names = FALSE, col.names = FALSE, append = TRUE)
      for(i in 1:length(dyeNames)){
        write.table(paste("  - ", dyeNames[i], ": ", calcCond[names(calcCond) == dyeNames[i]], sep = ""), file = reportName, sep = ",", row.names = FALSE, col.names = FALSE, append = TRUE)
      }
      write.table(paste("Theta value : ", calcCond[names(calcCond) == "theta"], sep = ""), file = reportName, sep = ",", row.names = FALSE, col.names = FALSE, append = TRUE)
      write.table(paste("Number of Monte Carlo simulation : ", calcCond[names(calcCond) == "numMc"], sep = ""), file = reportName, sep = ",", row.names = FALSE, col.names = FALSE, append = TRUE)
      write.table("Mixture ratio", file = reportName, sep = ",", row.names = FALSE, col.names = FALSE, append = TRUE)
      write.table(matrix(mrOne, nrow = 1), file = reportName, sep = ",", row.names = FALSE, col.names = FALSE, append = TRUE)
      write.table("Degradation", file = reportName, sep = ",", row.names = FALSE, col.names = FALSE, append = TRUE)
      write.table(matrix(degOne, nrow = 1), file = reportName, sep = ",", row.names = FALSE, col.names = FALSE, append = TRUE)
      
      write.table("Allele frequencies", file = reportName, sep = ",", row.names = FALSE, col.names = FALSE, append = TRUE)
      afMeth <- calcCond[names(calcCond) == "afMeth"]
      if(afMeth == 1){
        afMethName <- "Dirichlet distribution"
        write.table(paste("  - Method : ", afMethName, sep = ""), file = reportName, sep = ",", row.names = FALSE, col.names = FALSE, append = TRUE)
      }else{
        afMethName <- "Observed frequencies"
        write.table(paste("  - Method : ", afMethName, sep = ""), file = reportName, sep = ",", row.names = FALSE, col.names = FALSE, append = TRUE)
        write.table(paste("  - Minimum allele frequency : ", calcCond[names(calcCond) == "maf"], sep = ""), file = reportName, sep = ",", row.names = FALSE, col.names = FALSE, append = TRUE)
      }
      write.table("Thresholds to exclude unrealistic genotype combinations", file = reportName, sep = ",", row.names = FALSE, col.names = FALSE, append = TRUE)
      write.table(paste("  - Stochastic threshold: ", calcCond[names(calcCond) == "st"], sep = ""), file = reportName, sep = ",", row.names = FALSE, col.names = FALSE, append = TRUE)
      write.table(paste("  - Heterozygote balance filter: ", calcCond[names(calcCond) == "hbFltr"], sep = ""), file = reportName, sep = ",", row.names = FALSE, col.names = FALSE, append = TRUE)
      write.table(paste("  - Back stutter filter : ", calcCond[names(calcCond) == "stB1Fltr"], sep = ""), file = reportName, sep = ",", row.names = FALSE, col.names = FALSE, append = TRUE)
      write.table(paste("  - Forward stutter filter : ", calcCond[names(calcCond) == "stF1Fltr"], sep = ""), file = reportName, sep = ",", row.names = FALSE, col.names = FALSE, append = TRUE)
      write.table(paste("  - Double-back stutter filter : ", calcCond[names(calcCond) == "stB2Fltr"], sep = ""), file = reportName, sep = ",", row.names = FALSE, col.names = FALSE, append = TRUE)
      write.table(paste("  - Minus 2-nt stutter filter : ", calcCond[names(calcCond) == "stM2Fltr"], sep = ""), file = reportName, sep = ",", row.names = FALSE, col.names = FALSE, append = TRUE)
      write.table("", file = reportName, sep = ",", row.names = FALSE, col.names = FALSE, append = TRUE)
      
      write.table("======== Hypothesis ========", file = reportName, sep = ",", row.names = FALSE, col.names = FALSE, append = TRUE)
      write.table(paste("Prosecutor hypothesis (Hp) : ", paste(HpKnownName, collapse = " + "), " (+ unknown contributors)", sep = ""), file = reportName, sep = ",", row.names = FALSE, col.names = FALSE, append = TRUE)
      write.table(paste("Defense hypothesis (Hd) : ", paste(HdKnownName, collapse = " + "), " (+ unknown contributors)", sep = ""), file = reportName, sep = ",", row.names = FALSE, col.names = FALSE, append = TRUE)
      write.table("", file = reportName, sep = ",", row.names = FALSE, col.names = FALSE, append = TRUE)
      
      write.table("======== Log10 (likelihood) in Hp ========", file = reportName, sep = ",", row.names = FALSE, col.names = FALSE, append = TRUE)
      write.table(log10LikeHp, file = reportName, sep = ",", row.names = TRUE, col.names = NA, append = TRUE)
      
      write.table("---- Estimated Mixture Ratio ----", file = reportName, sep = ",", row.names = FALSE, col.names = FALSE, append = TRUE)
      write.table(mrEstimateHpAll, file = reportName, sep = ",", row.names = FALSE, col.names = TRUE, append = TRUE)
      
      write.table("---- Estimated Degradation Parameters ----", file = reportName, sep = ",", row.names = FALSE, col.names = FALSE, append = TRUE)
      write.table(dEstimateHpAll, file = reportName, sep = ",", row.names = FALSE, col.names = TRUE, append = TRUE)
      write.table("", file = reportName, sep = ",", row.names = FALSE, col.names = FALSE, append = TRUE)
      
      write.table("======== Log10 (likelihood) in Hd ========", file = reportName, sep = ",", row.names = FALSE, col.names = FALSE, append = TRUE)
      write.table(log10LikeHd, file = reportName, sep = ",", row.names = TRUE, col.names = NA, append = TRUE)
      
      write.table("---- Estimated Mixture Ratio ----", file = reportName, sep = ",", row.names = FALSE, col.names = FALSE, append = TRUE)
      write.table(mrEstimateHdAll, file = reportName, sep = ",", row.names = FALSE, col.names = TRUE, append = TRUE)
      
      write.table("---- Estimated Degradation Parameters ----", file = reportName, sep = ",", row.names = FALSE, col.names = FALSE, append = TRUE)
      write.table(dEstimateHdAll, file = reportName, sep = ",", row.names = FALSE, col.names = TRUE, append = TRUE)
      write.table("", file = reportName, sep = ",", row.names = FALSE, col.names = FALSE, append = TRUE)
      
      write.table("======== Likelihood ratio (Hp/Hd) ========", file = reportName, sep = ",", row.names = FALSE, col.names = FALSE, append = TRUE)
      write.table(lrAll, file = reportName, sep = ",", row.names = TRUE, col.names = NA, append = TRUE)
      write.table(paste("---- Maximum Log10 (likelihood) in Hp (", colnames(log10LikeHp)[log10LikeHp[nrow(log10LikeHp), ] == likeHpMax], " contribution) ----", sep = ""), file = reportName, sep = ",", row.names = FALSE, col.names = FALSE, append = TRUE)
      write.table(likeHpMax, file = reportName, sep = ",", row.names = FALSE, col.names = FALSE, append = TRUE)
      write.table(paste("---- Maximum Log10 (likelihood) in Hd (", colnames(log10LikeHd)[log10LikeHd[nrow(log10LikeHd), ] == likeHdMax], " contribution) ----", sep = ""), file = reportName, sep = ",", row.names = FALSE, col.names = FALSE, append = TRUE)
      write.table(likeHdMax, file = reportName, sep = ",", row.names = FALSE, col.names = FALSE, append = TRUE)
      write.table("---- Ratio of Maximum Likelihoods ----", file = reportName, sep = ",", row.names = FALSE, col.names = FALSE, append = TRUE)
      write.table(lrMax, file = reportName, sep = ",", row.names = FALSE, col.names = FALSE, append = TRUE)
      
      write.table("======== Calculation time (sec) ========", file = reportName, sep = ",", row.names = FALSE, col.names = FALSE, append = TRUE)
      write.table(calcTime, file = reportName, sep = ",", row.names = FALSE, col.names = TRUE, append = TRUE)
    }
  }
  
  #Make the 'Likelihood Ratio' tab
  tab4Make <- function(resultHp = NULL, resultHd = NULL, calcCond = NULL, mrOne = NULL, dyeNames = NULL, calcTime = NULL){
    tkdestroy(frameTab4)
    frameTab4 <<- tkframe(tab4)
    if(tclvalue(cspName) == ""){
      tkgrid(tklabel(frameTab4, text = "        Load crime stain profile!        "), pady = 10, sticky = "w")
    }
    if(tclvalue(refName) == ""){
      tkgrid(tklabel(frameTab4, text = "        Load reference profile!        "), pady = 10, sticky = "w")
    }
    if(tclvalue(afName) == ""){
      tkgrid(tklabel(frameTab4, text = "        Load allele frequencies!        "), pady = 10, sticky = "w")
    }
    if(tclvalue(mcName) == ""){
      tkgrid(tklabel(frameTab4, text = "        Load parameters for Monte Carlo simulation!        "), pady = 10, sticky = "w")
    }
    if(all(c(tclvalue(cspName), tclvalue(refName), tclvalue(afName)) != "")){
      if(tclvalue(calcFin) == "0"){
        tkgrid(tklabel(frameTab4, text = "        Perform calculation!        "), pady = 10, sticky = "w")
      }else{
        frameTab4_1 <- tkframe(frameTab4, relief = "groove", borderwidth = 2)
        frameTab4_1_1 <- tkframe(frameTab4_1)
        frameTab4_1_2 <- tkframe(frameTab4_1)
        frameTab4_1_2_1 <- tkframe(frameTab4_1_2)
        frameTab4_1_2_2 <- tkframe(frameTab4_1_2)
        frameTab4_1_2_3 <- tkframe(frameTab4_1_2)
        frameTab4_1_2_4 <- tkframe(frameTab4_1_2)
        tkgrid(tklabel(frameTab4_1_1, text = "Prosecutor hypothesis", font = "Helvetica 10 bold"))
        tkgrid(tklabel(frameTab4_1_2_1, text = ""))
        tkgrid(tklabel(frameTab4_1_2_2, text = "Log10 (likelihood)"))
        tkgrid(tklabel(frameTab4_1_2_3, text = "Estimated mixture ratio"))
        tkgrid(tklabel(frameTab4_1_2_4, text = "Estimated degradation"))
        
        frameTab4_2 <- tkframe(frameTab4, relief = "groove", borderwidth = 2)
        frameTab4_2_1 <- tkframe(frameTab4_2)
        frameTab4_2_2 <- tkframe(frameTab4_2)
        frameTab4_2_2_1 <- tkframe(frameTab4_2_2)
        frameTab4_2_2_2 <- tkframe(frameTab4_2_2)
        frameTab4_2_2_3 <- tkframe(frameTab4_2_2)
        frameTab4_2_2_4 <- tkframe(frameTab4_2_2)
        tkgrid(tklabel(frameTab4_2_1, text = "Defense hypothesis", font = "Helvetica 10 bold"))
        tkgrid(tklabel(frameTab4_2_2_1, text = ""))
        tkgrid(tklabel(frameTab4_2_2_2, text = "Log10 (likelihood)"))
        tkgrid(tklabel(frameTab4_2_2_3, text = "Estimated mixture ratio"))
        tkgrid(tklabel(frameTab4_2_2_4, text = "Estimated degradation"))
        
        lociName <- names(resultHp[[which(sapply(resultHp, length) != 0)[1]]][[1]])
        nL <- length(lociName)
        HpKnownName <- setdiff(names(resultHp[[which(sapply(resultHp, length) != 0)[1]]][[4]]), "Unknown")
        HdKnownName <- setdiff(names(resultHd[[which(sapply(resultHd, length) != 0)[1]]][[4]]), "Unknown")
        log10LikeHp <- matrix(-Inf, nL + 1, length(resultHp))
        rownames(log10LikeHp) <- c(lociName, "Total")
        colnames(log10LikeHp) <- rep("", length(resultHp))
        log10LikeHd <- matrix(-Inf, nL + 1, length(resultHd))
        rownames(log10LikeHd) <- c(lociName, "Total")
        colnames(log10LikeHd) <- rep("", length(resultHd))
        hncFrom <- calcCond[names(calcCond) == "hncFrom"]
        hncTo <- calcCond[names(calcCond) == "hncTo"]
        mrEstimateHpAll <- mrEstimateHdAll <- dEstimateHpAll <- dEstimateHdAll <- matrix("", hncTo, 3 * hncTo - 1)
        colnames(mrEstimateHpAll) <- colnames(mrEstimateHdAll) <- colnames(dEstimateHpAll) <- colnames(dEstimateHdAll) <- rep("", 3 * hncTo - 1)
        
        countHnc <- 0
        for(i in hncFrom:hncTo){
          countHnc <- countHnc + 1
          tkgrid(tklabel(frameTab4_1_2_1, text = paste("Hp", countHnc, ": ", i, "-person", sep = "")))
          colnames(log10LikeHp)[countHnc] <- colnames(log10LikeHd)[countHnc] <- paste(hncFrom + countHnc - 1, "-person", sep = "")
          colnames(mrEstimateHpAll)[3 * countHnc - 1] <- colnames(dEstimateHpAll)[3 * countHnc - 1] <- colnames(mrEstimateHdAll)[3 * countHnc - 1] <- colnames(dEstimateHdAll)[3 * countHnc - 1] <- paste(hncFrom + countHnc - 1, "-person", sep = "")
          if(length(resultHp[[countHnc]]) != 0){
            if(!is.element(0, sapply(resultHp[[countHnc]][[3]], length))){
              log10LikeHp[1:nL, countHnc] <- resultHp[[countHnc]][[1]]
              log10LikeHp[nL + 1, countHnc] <- resultHp[[countHnc]][[2]]
              tkgrid(tklabel(frameTab4_1_2_2, text = paste(signif(log10LikeHp[nL + 1, countHnc], 3))))
              mrEstimateHpAll[1:i, 3 * countHnc - 2] <- names(resultHp[[countHnc]][[4]])
              mrEstimateHpAll[1:i, 3 * countHnc - 1] <- mrEstimateHp <- resultHp[[countHnc]][[4]]
              dEstimateHpAll[1:i, 3 * countHnc - 2] <- names(resultHp[[countHnc]][[5]])
              dEstimateHpAll[1:i, 3 * countHnc - 1] <- dEstimateHp <- resultHp[[countHnc]][[5]]
              tkgrid(tklabel(frameTab4_1_2_3, text = round(mrEstimateHp, 4)))
              tkgrid(tklabel(frameTab4_1_2_4, text = round(dEstimateHp, 4)))
            }else{
              mrEstimateHpAll[1:i, 3 * countHnc - 1] <- dEstimateHpAll[1:i, 3 * countHnc - 1] <- "-"
              tkgrid(tklabel(frameTab4_1_2_2, text = "-Inf"))
              tkgrid(tklabel(frameTab4_1_2_3, text = " - "))
              tkgrid(tklabel(frameTab4_1_2_4, text = " - "))
            }
          }else{
            mrEstimateHpAll[1:i, 3 * countHnc - 1] <- dEstimateHpAll[1:i, 3 * countHnc - 1] <- "-"
            tkgrid(tklabel(frameTab4_1_2_2, text = "-Inf"))
            tkgrid(tklabel(frameTab4_1_2_3, text = " - "))
            tkgrid(tklabel(frameTab4_1_2_4, text = " - "))
          }
          tkgrid(tklabel(frameTab4_2_2_1, text = paste("Hd", i - hncFrom + 1, ": ", i, "-person", sep = "")))
          if(length(resultHd[[countHnc]]) != 0){ 
            if(!is.element(0, sapply(resultHd[[countHnc]][[3]], length))){
              log10LikeHd[1:nL, countHnc] <- resultHd[[countHnc]][[1]]
              log10LikeHd[nL + 1, countHnc] <- resultHd[[countHnc]][[2]]
              tkgrid(tklabel(frameTab4_2_2_2, text = paste(signif(log10LikeHd[nL + 1, countHnc], 3))))
              mrEstimateHdAll[1:i, 3 * countHnc - 2] <- names(resultHd[[countHnc]][[4]])
              mrEstimateHdAll[1:i, 3 * countHnc - 1] <- mrEstimateHd <- resultHd[[countHnc]][[4]]
              dEstimateHdAll[1:i, 3 * countHnc - 2] <- names(resultHd[[countHnc]][[5]])
              dEstimateHdAll[1:i, 3 * countHnc - 1] <- dEstimateHd <- resultHd[[countHnc]][[5]]
              tkgrid(tklabel(frameTab4_2_2_3, text = round(mrEstimateHd, 4)))
              tkgrid(tklabel(frameTab4_2_2_4, text = round(dEstimateHd, 4)))
            }else{
              mrEstimateHdAll[1:i, 3 * countHnc - 1] <- dEstimateHdAll[1:i, 3 * countHnc - 1] <- "-"
              tkgrid(tklabel(frameTab4_2_2_2, text = "-Inf"))
              tkgrid(tklabel(frameTab4_2_2_3, text = " - "))
              tkgrid(tklabel(frameTab4_2_2_4, text = " - "))
            }
          }else{
            mrEstimateHdAll[1:i, 3 * countHnc - 1] <- dEstimateHdAll[1:i, 3 * countHnc - 1] <- "-"
            tkgrid(tklabel(frameTab4_2_2_2, text = "-Inf"))
            tkgrid(tklabel(frameTab4_2_2_3, text = " - "))
            tkgrid(tklabel(frameTab4_2_2_4, text = " - "))
          }
        }
        likeHpMaxPos <- which(log10LikeHp[nL + 1, ] == max(log10LikeHp[nL + 1, ]))[1]
        likeHdMaxPos <- which(log10LikeHd[nL + 1, ] == max(log10LikeHd[nL + 1, ]))[1]
        likeHpMax <- log10LikeHp[nL + 1, likeHpMaxPos]
        likeHdMax <- log10LikeHd[nL + 1, likeHdMaxPos]
        lrMax <- 10^(likeHpMax - likeHdMax)
        tkgrid(frameTab4_1_2_1, frameTab4_1_2_2, frameTab4_1_2_3, frameTab4_1_2_4, padx = 10)
        tkgrid(frameTab4_1_1, pady = 5)
        tkgrid(frameTab4_1_2, pady = 5)
        tkgrid(frameTab4_1, padx = 20, pady = 5)
        tkgrid(frameTab4_2_2_1, frameTab4_2_2_2, frameTab4_2_2_3, frameTab4_2_2_4, padx = 10)
        tkgrid(frameTab4_2_1, pady = 5)
        tkgrid(frameTab4_2_2, pady = 5)
        tkgrid(frameTab4_2, padx = 20, pady = 5)
        
        frameTab4_3 <- tkframe(frameTab4)
        frameTab4_3_1 <- tkframe(frameTab4_3, relief = "groove", borderwidth = 2)
        frameTab4_3_1_1 <- tkframe(frameTab4_3_1)
        frameTab4_3_1_2 <- tkframe(frameTab4_3_1)
        frameTab4_3_1_2_1 <- tkframe(frameTab4_3_1_2)
        frameTab4_3_1_2_2 <- tkframe(frameTab4_3_1_2)
        tkgrid(tklabel(frameTab4_3_1_1, text = "Likelihood ratio (LR)", font = "Helvetica 10 bold"))
        lrAll <- matrix(0, nL + 1, ncol(log10LikeHp) * ncol(log10LikeHd))
        rownames(lrAll) <- c(lociName, "Total")
        colnames(lrAll) <- rep("", ncol(log10LikeHp) * ncol(log10LikeHd))
        for(i in 1:ncol(log10LikeHp)){
          for(j in 1:ncol(log10LikeHd)){
            lrAll[, ncol(log10LikeHd) * (i - 1) + j] <- 10^(log10LikeHp[, i] - log10LikeHd[, j])
            colnames(lrAll)[ncol(log10LikeHd) * (i - 1) + j] <- paste(colnames(log10LikeHp)[i], " vs ", colnames(log10LikeHd)[j], sep = "")
            tkgrid(tklabel(frameTab4_3_1_2_1, text = paste("Hp", i, "/Hd", j, sep = "")))
            tkgrid(tklabel(frameTab4_3_1_2_2, text = paste(signif(lrAll[nL + 1, ncol(log10LikeHd) * (i - 1) + j], 3))))
          }
        }
        tkgrid(frameTab4_3_1_2_1, frameTab4_3_1_2_2, padx = 10)
        tkgrid(frameTab4_3_1_1, pady = 5)
        tkgrid(frameTab4_3_1_2, pady = 5)
        frameTab4_3_2 <- tkframe(frameTab4_3, relief = "groove", borderwidth = 2)
        frameTab4_3_2_1 <- tkframe(frameTab4_3_2)
        frameTab4_3_2_2 <- tkframe(frameTab4_3_2)
        frameTab4_3_2_2_1 <- tkframe(frameTab4_3_2_2)
        frameTab4_3_2_2_2 <- tkframe(frameTab4_3_2_2)
        tkgrid(tklabel(frameTab4_3_2_1, text = "Ratio of maximum likelihoods", font = "Helvetica 10 bold"))
        tkgrid(tklabel(frameTab4_3_2_2_1, text = paste("Hp", likeHpMaxPos, " / Hd", likeHdMaxPos, sep = "")))
        tkgrid(tklabel(frameTab4_3_2_2_2, text = paste(signif(lrMax, 3))))
        tkgrid(frameTab4_3_2_2_1, frameTab4_3_2_2_2, padx = 10)
        tkgrid(frameTab4_3_2_1, pady = 5)
        tkgrid(frameTab4_3_2_2, pady = 5)
        tkgrid(frameTab4_3_1, frameTab4_3_2, padx = 10, pady = 5, sticky = "n")
        tkgrid(frameTab4_3, padx = 20, pady = 5)
        log10LikeHp[which(log10LikeHp == -Inf, arr.ind = TRUE)] <- "NA"
        log10LikeHd[which(log10LikeHd == -Inf, arr.ind = TRUE)] <- "NA"
        
        frameTab4_4 <- tkframe(frameTab4)
        reportButt <- tkbutton(frameTab4_4, text = "    Report   ", cursor = "hand2", command = function() reportMake(HpKnownName, HdKnownName, log10LikeHp, log10LikeHd, lrAll, likeHpMax, likeHdMax, lrMax, mrEstimateHpAll, mrEstimateHdAll, dEstimateHpAll, dEstimateHdAll, calcCond, mrOne, dyeNames, calcTime))
        tkgrid(reportButt, padx = 10, pady = 5)
        tkgrid(frameTab4_4, pady = 5)
      }
    }
    tkgrid(frameTab4)
  }
  
  
  ###############################
  # Parameter estimation window #
  ###############################
  
  #Open a new window for parameter estimation
  parEstWindow <- function(){
    #Input required files
    openFile2 <- function(fp, var, top){
      imputOk <- "ok"
      if(tclvalue(paramEstFin) == "1"){
        imputOk <- tclvalue(tkmessageBox(message = "Results of estimating parameters will be deleted. Do you want to continue?", type = "okcancel", icon = "warning"))
      }
      if(imputOk == "ok"){
        fileName <- tclvalue(tkgetOpenFile(parent = top, initialdir = tclvalue(fp), multiple = "true", filetypes = "{{CSV Files} {.csv}}"))
        if(!nchar(fileName)){
          tkmessageBox(message = "No file was selected!", icon = "error", type = "ok")
        }else{
          tmp <- sub("\\}", fileName, replacement = "")
          tmp2 <- sub("\\{", tmp, replacement = "")
          tclvalue(fp) <- tmp2
          foo3 <- strsplit(tmp2, "/")[[1]]
          tclvalue(var) <- strsplit(foo3[length(foo3)], "\\.csv")[[1]][1]
        }
        tclvalue(fileCk2Fin) <- "0"
        tclvalue(paramEstFin) <- "0"
        tab2ParMake()
        tab3ParMake()
      }
    }
    
    #Make the 'Files' tab
    tab1ParMake <- function(){
      tkdestroy(fTab1Par)
      fTab1Par <<- tkframe(tab1Par)
      
      calibLabel <- tklabel(fTab1Par, text = "Experimental Data")
      calibFileLabel <- tklabel(fTab1Par, textvariable = calibName, width = 30, highlightthickness = 1, relief = "groove", justify = "center", background = "white")
      calibButt <- tkbutton(fTab1Par, text = "    Load    ", cursor = "hand2", command = function() openFile2(calibFp, calibName, tfPar))
      
      gtAnsLabel <- tklabel(fTab1Par, text = "Genotypes")
      gtAnsFileLabel <- tklabel(fTab1Par, textvariable = gtAnsName, width = 30, highlightthickness = 1, relief = "groove", justify = "center", background = "white")
      gtAnsButt <- tkbutton(fTab1Par, text = "    Load    ", cursor = "hand2", command = function() openFile2(gtAnsFp, gtAnsName, tfPar))
      
      seqLabel <- tklabel(fTab1Par, text = "Sequence Data")
      seqFileLabel <- tklabel(fTab1Par, textvariable = seqName, width = 30, highlightthickness = 1, relief = "groove", justify = "center", background = "white")
      seqButt <- tkbutton(fTab1Par, text = "    Load    ", cursor = "hand2", command = function() openFile2(seqFp, seqName, tfPar))
      
      nextButt <- tkbutton(fTab1Par, text = "    Next    ", cursor = "hand2", command = function() fileCk2())
      
      tkgrid(calibLabel, calibFileLabel, calibButt, padx = 20, pady = 20, sticky = "w")
      tkgrid(gtAnsLabel, gtAnsFileLabel, gtAnsButt, padx = 20, pady = 20, sticky = "w")
      tkgrid(seqLabel, seqFileLabel, seqButt, padx = 20, pady = 20, sticky = "w")
      tkgrid(nextButt, padx = 20, pady = 20, sticky = "w")
      tkgrid(fTab1Par)
    }
    
    #Check input files
    fileCk2 <- function(){
      if(any(c(tclvalue(calibFp) == "", tclvalue(gtAnsFp) == "", tclvalue(seqFp) == ""))){
        tkmessageBox(message = "Load required file(s)!", icon = "error", type = "ok")
      }else{
        calData <- read.csv(tclvalue(calibFp), header = TRUE)
        calData <- as.matrix(calData)
        calData <- gsub(" ", "", calData, fixed = TRUE)
        allLoci <- unique(calData[, grep("Marker", colnames(calData))])
        calDataLoci <- setdiff(allLoci, sexChrMar)
        gtAns <- read.csv(tclvalue(gtAnsFp), header = TRUE)
        gtAns <- as.matrix(gtAns)
        gtAns <- gsub(" ", "", gtAns, fixed = TRUE)
        gtAnsLoci <- unique(gtAns[, grep("Marker", colnames(gtAns))])
        gtAnsLoci <- setdiff(gtAnsLoci, sexChrMar)
        seqFileOk <- TRUE
        
        if(setequal(calDataLoci, gtAnsLoci)){
          seqData <- read.csv(tclvalue(seqFp), header = TRUE)
          seqData <- as.matrix(seqData)
          seqDataLoci <- unique(seqData[, which(colnames(seqData) == "Marker")])
          if(!all(is.element(calDataLoci, seqDataLoci))){
            tkmessageBox(message = "Some loci of experimental data are not included in the file of sequence data!", icon = "error", type = "ok")
            seqFileOk <- FALSE
          }
          if(seqFileOk){
            calDataSampleName <- unique(calData[, intersect(grep("Sample", colnames(calData)), grep("Name", colnames(calData)))])
            gtAnsSampleName <- unique(gtAns[, intersect(grep("Sample", colnames(gtAns)), grep("Name", colnames(gtAns)))])
            if(setequal(calDataSampleName, gtAnsSampleName)){
              parDye <- calData[1:length(allLoci), grep("Dye", colnames(calData))]
              parDye <- parDye[is.element(allLoci, gtAnsLoci)]
              tclvalue(fileCk2Fin) <- "1"
              tab2ParMake(gtAnsLoci, parDye)
              tab3ParMake()
              tk2notetab.select(tabsPar, "Setting")
            }else{
              tkmessageBox(message = "Sample names are not equal in each file!", icon = "error", type = "ok")
            }
          }
        }else{
          tkmessageBox(message = "Names of loci are not equal in each file!", icon = "error", type = "ok")
        }
      }
    }
    
    #Make a scroll bar
    scrollMake <- function(top){
      frame <- tkframe(top, class = "Scrollform")
      xbar <- tkscrollbar(frame, orient = "horizontal", command = function(...) tkxview(vport, ...))
      tkpack(xbar, side = "bottom", fill = "x")
      ybar <- tkscrollbar(frame, command = function(...) tkyview(vport, ...))
      tkpack(ybar, side = "right", fill = "y")
      vport <- tkcanvas(frame, xscrollcommand = function(...) tkset(xbar, ...), yscrollcommand = function(...) tkset(ybar, ...))
      tkpack(vport, side = "left", fill = "both", expand = TRUE)
      form <- tkframe(vport)
      tkcreate(vport, "window", "0 0", anchor = "nw", window = form$ID)
      tkbind(form, "<Configure>", function() scrollResize(form))
      tkaddtag(vport, "iform", "all")
      
      iform.win <- tkitemconfigure(vport, tkfind(vport, "withtag", "iform"), "-window")
      return(frame)
    }
    
    #Make a scroll bar
    scrollResize <- function(form){
      stopifnot(tclvalue(tkwinfo("class", form)) == "Frame")
      vport <- tkwinfo("parent", form)
      stopifnot(tclvalue(tkwinfo("class", vport)) == "Canvas")
      bbox <- tkbbox(vport, "all")
      w <- tkwinfo("width", form)
      tkconfigure(vport, width = w, scrollregion = bbox, yscrollincrement = "0.1i")
    }
    
    #Make a scroll bar
    scrollGet <- function(sform){
      stopifnot(inherits(sform, "tkwin") && tclvalue(tkwinfo("class", sform)) == "Scrollform")
      children.ids <- unlist(strsplit(tclvalue(tkwinfo("children", sform)), " "))
      for(child.id in children.ids){
        if(tclvalue(tkwinfo("class", child.id)) == "Canvas"){
          vport <- .Tk.newwin(child.id)
          iform.win <- tkitemconfigure(vport, tkfind(vport, "withtag", "iform"), "-window")
          iform.id <- unlist(strsplit(tclvalue(iform.win), " "))[5]
          stopifnot(tclvalue(tkwinfo("class", iform.id)) == "Frame")
          form <- .Tk.newwin(iform.id)
          return(form)
        }
      }
    }
    
    #Make conditions of maximum likelihood estimation for AE
    aeMleCondMake <- function(aeCandFin){
      aeMleCond <- list()
      nL <- length(aeCandFin)
      for(i in 1:nL){
        aeMleCondOneL <- list()
        aeCandOneL <- aeCandFin[[i]]
        nM <- length(aeCandOneL)
        for(j in 1:nM){
          aeMleCondOneM <- list()
          aeCondInitial <- list()
          aeCondLower <- list()
          aeCondUpper <- list()
          m1 <- aeCandOneL[j]
          if(m1 == "Log-normal"){
            aeCondInitial[[1]] <- tclVar(0)
            aeCondInitial[[2]] <- tclVar(50)
            aeCondLower[[1]] <- tclVar(-0.7)
            aeCondLower[[2]] <- tclVar(1)
            aeCondUpper[[1]] <- tclVar(0.7)
            aeCondUpper[[2]] <- tclVar(500)
            names(aeCondInitial) <- names(aeCondLower) <- names(aeCondUpper) <- c("Mean", "Variance")
          }
          aeMleCondOneM[[1]] <- aeCondInitial
          aeMleCondOneM[[2]] <- aeCondLower
          aeMleCondOneM[[3]] <- aeCondUpper
          aeMleCondOneL[[j]] <- aeMleCondOneM
        }
        names(aeMleCondOneL) <- aeCandOneL
        aeMleCond[[i]] <- aeMleCondOneL
      }
      names(aeMleCond) <- names(aeCandFin)
      return(aeMleCond)
    }
    
    #Make conditions of maximum likelihood estimation for Hb
    hbMleCondMake <- function(hbCandFin){
      hbMleCond <- list()
      nL <- length(hbCandFin)
      for(i in 1:nL){
        hbMleCondOneL <- list()
        hbCandOneL <- hbCandFin[[i]]
        nM <- length(hbCandOneL)
        for(j in 1:nM){
          hbMleCondOneM <- list()
          hbCondInitial <- list()
          hbCondLower <- list()
          hbCondUpper <- list()
          m1 <- hbCandOneL[j]
          if(m1 == "Log-normal"){
            hbCondInitial[[1]] <- tclVar(0)
            hbCondInitial[[2]] <- tclVar(50)
            hbCondLower[[1]] <- tclVar(-0.1)
            hbCondLower[[2]] <- tclVar(1)
            hbCondUpper[[1]] <- tclVar(0.1)
            hbCondUpper[[2]] <- tclVar(500)
            names(hbCondInitial) <- names(hbCondLower) <- names(hbCondUpper) <- c("Mean", "Variance")
          }
          hbMleCondOneM[[1]] <- hbCondInitial
          hbMleCondOneM[[2]] <- hbCondLower
          hbMleCondOneM[[3]] <- hbCondUpper
          hbMleCondOneL[[j]] <- hbMleCondOneM
        }
        names(hbMleCondOneL) <- hbCandOneL
        hbMleCond[[i]] <- hbMleCondOneL
      }
      names(hbMleCond) <- names(hbCandFin)
      return(hbMleCond)
    }
    
    #Make conditions of maximum likelihood estimation for BSR
    bsrMleCondMake <- function(bsrCandFin){
      bsrMleCond <- list()
      nL <- length(bsrCandFin)
      for(i in 1:nL){
        bsrMleCondOneL <- list()
        bsrCandOneL <- bsrCandFin[[i]]
        bsrCandOneL <- bsrCandOneL[bsrCandOneL != "Best"]
        nM <- length(bsrCandOneL)
        for(j in 1:nM){
          bsrMleCondOneM <- list()
          bsrCondInitial <- list()
          bsrCondLower <- list()
          bsrCondUpper <- list()
          m1 <- bsrCandOneL[j]
          if(m1 == "Allele"){
            bsrCondInitial[[1]] <- tclVar(0.01)
            bsrCondInitial[[2]] <- tclVar(-0.05)
            bsrCondInitial[[3]] <- tclVar(50)
            bsrCondLower[[1]] <- tclVar(0)
            bsrCondLower[[2]] <- tclVar(-1)
            bsrCondLower[[3]] <- tclVar(1)
            bsrCondUpper[[1]] <- tclVar(0.1)
            bsrCondUpper[[2]] <- tclVar(1)
            bsrCondUpper[[3]] <- tclVar(1000)
            names(bsrCondInitial) <- names(bsrCondLower) <- names(bsrCondUpper) <- c("Slope", "Intercept", "Variance")
          }else if(m1 == "LUS"){
            bsrCondInitial[[1]] <- tclVar(0.01)
            bsrCondInitial[[2]] <- tclVar(-0.05)
            bsrCondInitial[[3]] <- tclVar(50)
            bsrCondLower[[1]] <- tclVar(0)
            bsrCondLower[[2]] <- tclVar(-1)
            bsrCondLower[[3]] <- tclVar(1)
            bsrCondUpper[[1]] <- tclVar(0.1)
            bsrCondUpper[[2]] <- tclVar(1)
            bsrCondUpper[[3]] <- tclVar(1000)
            names(bsrCondInitial) <- names(bsrCondLower) <- names(bsrCondUpper) <- c("Slope", "Intercept", "Variance")
          }else if(m1 == "Multi-seq"){
            bsrCondInitial[[1]] <- tclVar(0.01)
            bsrCondInitial[[2]] <- tclVar(-0.05)
            bsrCondInitial[[3]] <- tclVar(50)
            bsrCondInitial[[4]] <- tclVar(2)
            bsrCondLower[[1]] <- tclVar(0)
            bsrCondLower[[2]] <- tclVar(-1)
            bsrCondLower[[3]] <- tclVar(1)
            bsrCondLower[[4]] <- tclVar(2)
            bsrCondUpper[[1]] <- tclVar(0.1)
            bsrCondUpper[[2]] <- tclVar(1)
            bsrCondUpper[[3]] <- tclVar(1000)
            bsrCondUpper[[4]] <- tclVar(30)
            names(bsrCondInitial) <- names(bsrCondLower) <- names(bsrCondUpper) <- c("Slope", "Intercept", "Variance", "Parameter x")
          }
          bsrMleCondOneM[[1]] <- bsrCondInitial
          bsrMleCondOneM[[2]] <- bsrCondLower
          bsrMleCondOneM[[3]] <- bsrCondUpper
          bsrMleCondOneL[[j]] <- bsrMleCondOneM
        }
        names(bsrMleCondOneL) <- bsrCandOneL
        bsrMleCond[[i]] <- bsrMleCondOneL
      }
      names(bsrMleCond) <- names(bsrCandFin)
      return(bsrMleCond)
    }
    
    #Make conditions of maximum likelihood estimation for FSR
    fsrMleCondMake <- function(fsrCandFin){
      fsrMleCond <- list()
      nL <- length(fsrCandFin)
      for(i in 1:nL){
        fsrMleCondOneL <- list()
        fsrCandOneL <- fsrCandFin[[i]]
        fsrCandOneL <- fsrCandOneL[fsrCandOneL != "Best"]
        nM <- length(fsrCandOneL)
        for(j in 1:nM){
          fsrMleCondOneM <- list()
          fsrCondInitial <- list()
          fsrCondLower <- list()
          fsrCondUpper <- list()
          m1 <- fsrCandOneL[j]
          if(m1 == "Allele"){
            fsrCondInitial[[1]] <- tclVar(0.005)
            fsrCondInitial[[2]] <- tclVar(-0.01)
            fsrCondInitial[[3]] <- tclVar(500)
            fsrCondLower[[1]] <- tclVar(0)
            fsrCondLower[[2]] <- tclVar(-1)
            fsrCondLower[[3]] <- tclVar(1)
            fsrCondUpper[[1]] <- tclVar(0.1)
            fsrCondUpper[[2]] <- tclVar(1)
            fsrCondUpper[[3]] <- tclVar(5000)
            names(fsrCondInitial) <- names(fsrCondLower) <- names(fsrCondUpper) <- c("Slope", "Intercept", "Variance")
          }else if(m1 == "LUS"){
            fsrCondInitial[[1]] <- tclVar(0.005)
            fsrCondInitial[[2]] <- tclVar(-0.01)
            fsrCondInitial[[3]] <- tclVar(500)
            fsrCondLower[[1]] <- tclVar(0)
            fsrCondLower[[2]] <- tclVar(-1)
            fsrCondLower[[3]] <- tclVar(1)
            fsrCondUpper[[1]] <- tclVar(0.1)
            fsrCondUpper[[2]] <- tclVar(1)
            fsrCondUpper[[3]] <- tclVar(5000)
            names(fsrCondInitial) <- names(fsrCondLower) <- names(fsrCondUpper) <- c("Slope", "Intercept", "Variance")
          }else if(m1 == "Multi-seq"){
            fsrCondInitial[[1]] <- tclVar(0.005)
            fsrCondInitial[[2]] <- tclVar(-0.01)
            fsrCondInitial[[3]] <- tclVar(500)
            fsrCondInitial[[4]] <- tclVar(2)
            fsrCondLower[[1]] <- tclVar(0)
            fsrCondLower[[2]] <- tclVar(-1)
            fsrCondLower[[3]] <- tclVar(1)
            fsrCondLower[[4]] <- tclVar(2)
            fsrCondUpper[[1]] <- tclVar(0.1)
            fsrCondUpper[[2]] <- tclVar(1)
            fsrCondUpper[[3]] <- tclVar(5000)
            fsrCondUpper[[4]] <- tclVar(30)
            names(fsrCondInitial) <- names(fsrCondLower) <- names(fsrCondUpper) <- c("Slope", "Intercept", "Variance", "Parameter x")
          }else if(m1 == "Uniform"){
            fsrCondInitial[[1]] <- tclVar(0.01)
            fsrCondInitial[[2]] <- tclVar(500)
            fsrCondLower[[1]] <- tclVar(0)
            fsrCondLower[[2]] <- tclVar(1)
            fsrCondUpper[[1]] <- tclVar(1)
            fsrCondUpper[[2]] <- tclVar(5000)
            names(fsrCondInitial) <- names(fsrCondLower) <- names(fsrCondUpper) <- c("Mean", "Variance")
          }
          fsrMleCondOneM[[1]] <- fsrCondInitial
          fsrMleCondOneM[[2]] <- fsrCondLower
          fsrMleCondOneM[[3]] <- fsrCondUpper
          fsrMleCondOneL[[j]] <- fsrMleCondOneM
        }
        names(fsrMleCondOneL) <- fsrCandOneL
        fsrMleCond[[i]] <- fsrMleCondOneL
      }
      names(fsrMleCond) <- names(fsrCandFin)
      return(fsrMleCond)
    }
    
    #Make conditions of maximum likelihood estimation for DSR
    dsrMleCondMake <- function(dsrCandFin){
      dsrMleCond <- list()
      nL <- length(dsrCandFin)
      for(i in 1:nL){
        dsrMleCondOneL <- list()
        dsrCandOneL <- dsrCandFin[[i]]
        dsrCandOneL <- dsrCandOneL[dsrCandOneL != "Best"]
        nM <- length(dsrCandOneL)
        for(j in 1:nM){
          dsrMleCondOneM <- list()
          dsrCondInitial <- list()
          dsrCondLower <- list()
          dsrCondUpper <- list()
          m1 <- dsrCandOneL[j]
          if(m1 == "Allele"){
            dsrCondInitial[[1]] <- tclVar(0.005)
            dsrCondInitial[[2]] <- tclVar(-0.01)
            dsrCondInitial[[3]] <- tclVar(1000)
            dsrCondLower[[1]] <- tclVar(0)
            dsrCondLower[[2]] <- tclVar(-1)
            dsrCondLower[[3]] <- tclVar(1)
            dsrCondUpper[[1]] <- tclVar(0.1)
            dsrCondUpper[[2]] <- tclVar(1)
            dsrCondUpper[[3]] <- tclVar(5000)
            names(dsrCondInitial) <- names(dsrCondLower) <- names(dsrCondUpper) <- c("Slope", "Intercept", "Variance")
          }else if(m1 == "LUS"){
            dsrCondInitial[[1]] <- tclVar(0.005)
            dsrCondInitial[[2]] <- tclVar(-0.01)
            dsrCondInitial[[3]] <- tclVar(1000)
            dsrCondLower[[1]] <- tclVar(0)
            dsrCondLower[[2]] <- tclVar(-1)
            dsrCondLower[[3]] <- tclVar(1)
            dsrCondUpper[[1]] <- tclVar(0.1)
            dsrCondUpper[[2]] <- tclVar(1)
            dsrCondUpper[[3]] <- tclVar(5000)
            names(dsrCondInitial) <- names(dsrCondLower) <- names(dsrCondUpper) <- c("Slope", "Intercept", "Variance")
          }else if(m1 == "Multi-seq"){
            dsrCondInitial[[1]] <- tclVar(0.005)
            dsrCondInitial[[2]] <- tclVar(-0.01)
            dsrCondInitial[[3]] <- tclVar(1000)
            dsrCondInitial[[4]] <- tclVar(2)
            dsrCondLower[[1]] <- tclVar(0)
            dsrCondLower[[2]] <- tclVar(-1)
            dsrCondLower[[3]] <- tclVar(1)
            dsrCondLower[[4]] <- tclVar(2)
            dsrCondUpper[[1]] <- tclVar(0.1)
            dsrCondUpper[[2]] <- tclVar(1)
            dsrCondUpper[[3]] <- tclVar(5000)
            dsrCondUpper[[4]] <- tclVar(30)
            names(dsrCondInitial) <- names(dsrCondLower) <- names(dsrCondUpper) <- c("Slope", "Intercept", "Variance", "Parameter x")
          }else if(m1 == "Uniform"){
            dsrCondInitial[[1]] <- tclVar(0.01)
            dsrCondInitial[[2]] <- tclVar(500)
            dsrCondLower[[1]] <- tclVar(0)
            dsrCondLower[[2]] <- tclVar(1)
            dsrCondUpper[[1]] <- tclVar(1)
            dsrCondUpper[[2]] <- tclVar(5000)
            names(dsrCondInitial) <- names(dsrCondLower) <- names(dsrCondUpper) <- c("Mean", "Variance")
          }
          dsrMleCondOneM[[1]] <- dsrCondInitial
          dsrMleCondOneM[[2]] <- dsrCondLower
          dsrMleCondOneM[[3]] <- dsrCondUpper
          dsrMleCondOneL[[j]] <- dsrMleCondOneM
        }
        names(dsrMleCondOneL) <- dsrCandOneL
        dsrMleCond[[i]] <- dsrMleCondOneL
      }
      names(dsrMleCond) <- names(dsrCandFin)
      return(dsrMleCond)
    }
    
    #Make conditions of maximum likelihood estimation for M2SR
    m2srMleCondMake <- function(m2srCandFin){
      m2srMleCond <- list()
      nL <- length(m2srCandFin)
      for(i in 1:nL){
        m2srMleCondOneL <- list()
        m2srCandOneL <- m2srCandFin[[i]]
        m2srCandOneL <- m2srCandOneL[m2srCandOneL != "Best"]
        nM <- length(m2srCandOneL)
        for(j in 1:nM){
          m2srMleCondOneM <- list()
          m2srCondInitial <- list()
          m2srCondLower <- list()
          m2srCondUpper <- list()
          m1 <- m2srCandOneL[j]
          if(m1 == "Uniform"){
            m2srCondInitial[[1]] <- tclVar(0.01)
            m2srCondInitial[[2]] <- tclVar(500)
            m2srCondLower[[1]] <- tclVar(0)
            m2srCondLower[[2]] <- tclVar(1)
            m2srCondUpper[[1]] <- tclVar(1)
            m2srCondUpper[[2]] <- tclVar(5000)
            names(m2srCondInitial) <- names(m2srCondLower) <- names(m2srCondUpper) <- c("Mean", "Variance")
          }
          m2srMleCondOneM[[1]] <- m2srCondInitial
          m2srMleCondOneM[[2]] <- m2srCondLower
          m2srMleCondOneM[[3]] <- m2srCondUpper
          m2srMleCondOneL[[j]] <- m2srMleCondOneM
        }
        names(m2srMleCondOneL) <- m2srCandOneL
        m2srMleCond[[i]] <- m2srMleCondOneL
      }
      names(m2srMleCond) <- names(m2srCandFin)
      return(m2srMleCond)
    }
    
    #Re-make the 'Setting' tab to restore default setting
    restoreDefaults <- function(parLocus, parDye){
      tab2ParMake(parLocus, parDye)
    }
    
    #Make the 'Setting' tab
    tab2ParMake <- function(parLocus = NULL, parDye = NULL){
      tkdestroy(fTab2Par)
      fTab2Par <<- tkframe(tab2Par)
      if(tclvalue(calibName) == ""){
        tkgrid(tklabel(fTab2Par, text = "        Load experimental data!        "), pady = 10, sticky = "w")
        tkgrid(fTab2Par)
      }
      if(tclvalue(gtAnsName) == ""){
        tkgrid(tklabel(fTab2Par, text = "        Load actual genotypes!        "), pady = 10, sticky = "w")
        tkgrid(fTab2Par)
      }
      if(tclvalue(seqName) == ""){
        tkgrid(tklabel(fTab2Par, text = "        Load sequence data!        "), pady = 10, sticky = "w")
        tkgrid(fTab2Par)
      }
      if(all(c(tclvalue(calibName), tclvalue(gtAnsName), tclvalue(seqName)) != "")){
        if(tclvalue(fileCk2Fin) == "0"){
          tkgrid(tklabel(fTab2Par, text = "        Push 'Next' button in the Files tab.        "), pady = 10, sticky = "w")
          tkgrid(fTab2Par)
        }else{
          advancedSetting <- function(){
            #Make a window for advanced setting of AE
            tab1AdvMake <- function(){
              showCond <- function(){
                tkdestroy(fTab1Adv_2)
                fTab1Adv_2 <<- tkframe(fTab1Adv, relief = "groove", borderwidth = 2)
                tkgrid(tklabel(fTab1Adv_2, text = "Model"), row = 0, column = 0, padx = 5, pady = 5)
                tkgrid(tklabel(fTab1Adv_2, text = "Parameter"), row = 0, column = 1, padx = 5, pady = 5)
                tkgrid(tklabel(fTab1Adv_2, text = "Initial value"), row = 0, column = 2, padx = 5, pady = 5)
                tkgrid(tklabel(fTab1Adv_2, text = "Lower value"), row = 0, column = 3, padx = 5, pady = 5)
                tkgrid(tklabel(fTab1Adv_2, text = "Upper value"), row = 0, column = 4, padx = 5, pady = 5)
                
                posL <- which(parLocus == tclvalue(selectL))
                modelName <- aeCandFin[[posL]]
                nM <- length(modelName)
                aeMleCondOneL <- aeMleCond[[posL]]
                rowCount <- 1
                for(i in 1:nM){
                  mleC <- aeMleCondOneL[[i]]
                  mleInitial <- mleC[[1]]
                  mleLower <- mleC[[2]]
                  mleUpper <- mleC[[3]]
                  nPar <- length(mleInitial)
                  parName <- names(mleInitial)
                  tkgrid(tklabel(fTab1Adv_2, text = modelName[i]), row = rowCount, column = 0, padx = 5, pady = 5)
                  for(j in 1:nPar){
                    tkgrid(tklabel(fTab1Adv_2, text = parName[j]), row = rowCount, column = 1, padx = 5, pady = 5)
                    tkgrid(tkentry(fTab1Adv_2, textvariable = mleInitial[[j]], width = 10, highlightthickness = 1, relief = "solid", justify = "center", background = "white"), row = rowCount, column = 2, padx = 5, pady = 5)
                    tkgrid(tkentry(fTab1Adv_2, textvariable = mleLower[[j]], width = 10, highlightthickness = 1, relief = "solid", justify = "center", background = "white"), row = rowCount, column = 3, padx = 5, pady = 5)
                    tkgrid(tkentry(fTab1Adv_2, textvariable = mleUpper[[j]], width = 10, highlightthickness = 1, relief = "solid", justify = "center", background = "white"), row = rowCount, column = 4, padx = 5, pady = 5)
                    rowCount <- rowCount + 1
                  }
                }
                tkgrid(fTab1Adv_2, padx = 5, pady = 5, sticky = "w")
              }
              
              tkdestroy(fTab1Adv)
              fTab1Adv <<- tkframe(tab1Adv)
              
              fTab1Adv_1 <- tkframe(fTab1Adv)
              fTab1Adv_2 <- tkframe(fTab1Adv)
              
              tkgrid(tklabel(fTab1Adv_1, text = "Minimum of locus-specific amplification efficiency : "), row = 0, column = 0, padx = 5, pady = 5, sticky = "w")
              tkgrid(tkentry(fTab1Adv_1, textvariable = aeMinVar, width = 10, highlightthickness = 1, relief = "solid", justify = "center", background = "white"), row = 0, column = 1, padx = 5, pady = 5, sticky = "w")
              tkgrid(tklabel(fTab1Adv_1, text = "Conditions of maximum likelihood estimation : "), row = 1, column = 0, padx = 5, pady = 5, sticky = "w")
              tkgrid(tklabel(fTab1Adv_1, text = "    Select locus : "), row = 2, column = 0, padx = 5, pady = 5, sticky = "w")
              selectL <- tclVar(parLocus[1])
              tkgrid(ttkcombobox(fTab1Adv_1, values = parLocus, textvariable = selectL, state = "readonly"), row = 2, column = 1, padx = 5, pady = 5, sticky = "w")
              tkgrid(tkbutton(fTab1Adv_1, text = "    Show    ", cursor = "hand2", command = function() showCond()), row = 2, column = 2, padx = 5, pady = 5, sticky = "w")
              tkgrid(fTab1Adv_1, padx = 10, sticky = "w")
              tkgrid(fTab1Adv_2, padx = 10, sticky = "w")
              tkgrid(fTab1Adv)
            }
            
            #Make a window for advanced setting of Hb
            tab2AdvMake <- function(){
              showCond <- function(){
                tkdestroy(fTab2Adv_2)
                fTab2Adv_2 <<- tkframe(fTab2Adv, relief = "groove", borderwidth = 2)
                tkgrid(tklabel(fTab2Adv_2, text = "Model"), row = 0, column = 0, padx = 5, pady = 5)
                tkgrid(tklabel(fTab2Adv_2, text = "Parameter"), row = 0, column = 1, padx = 5, pady = 5)
                tkgrid(tklabel(fTab2Adv_2, text = "Initial value"), row = 0, column = 2, padx = 5, pady = 5)
                tkgrid(tklabel(fTab2Adv_2, text = "Lower value"), row = 0, column = 3, padx = 5, pady = 5)
                tkgrid(tklabel(fTab2Adv_2, text = "Upper value"), row = 0, column = 4, padx = 5, pady = 5)
                
                posL <- which(parLocus == tclvalue(selectL))
                modelName <- hbCandFin[[posL]]
                nM <- length(modelName)
                hbMleCondOneL <- hbMleCond[[posL]]
                rowCount <- 1
                for(i in 1:nM){
                  mleC <- hbMleCondOneL[[i]]
                  mleInitial <- mleC[[1]]
                  mleLower <- mleC[[2]]
                  mleUpper <- mleC[[3]]
                  nPar <- length(mleInitial)
                  parName <- names(mleInitial)
                  tkgrid(tklabel(fTab2Adv_2, text = modelName[i]), row = rowCount, column = 0, padx = 5, pady = 5)
                  for(j in 1:nPar){
                    tkgrid(tklabel(fTab2Adv_2, text = parName[j]), row = rowCount, column = 1, padx = 5, pady = 5)
                    tkgrid(tkentry(fTab2Adv_2, textvariable = mleInitial[[j]], width = 10, highlightthickness = 1, relief = "solid", justify = "center", background = "white"), row = rowCount, column = 2, padx = 5, pady = 5)
                    tkgrid(tkentry(fTab2Adv_2, textvariable = mleLower[[j]], width = 10, highlightthickness = 1, relief = "solid", justify = "center", background = "white"), row = rowCount, column = 3, padx = 5, pady = 5)
                    tkgrid(tkentry(fTab2Adv_2, textvariable = mleUpper[[j]], width = 10, highlightthickness = 1, relief = "solid", justify = "center", background = "white"), row = rowCount, column = 4, padx = 5, pady = 5)
                    rowCount <- rowCount + 1
                  }
                }
                tkgrid(fTab2Adv_2, padx = 5, pady = 5, sticky = "w")
              }
              
              tkdestroy(fTab2Adv)
              fTab2Adv <<- tkframe(tab2Adv)
              
              fTab2Adv_1 <- tkframe(fTab2Adv)
              fTab2Adv_2 <- tkframe(fTab2Adv)
              
              tkgrid(tklabel(fTab2Adv_1, text = "Minimum of heterozygote balance : "), row = 0, column = 0, padx = 5, pady = 5, sticky = "w")
              tkgrid(tkentry(fTab2Adv_1, textvariable = hbMinVar, width = 10, highlightthickness = 1, relief = "solid", justify = "center", background = "white"), row = 0, column = 1, padx = 5, pady = 5, sticky = "w")
              tkgrid(tklabel(fTab2Adv_1, text = "Conditions of maximum likelihood estimation : "), row = 1, column = 0, padx = 5, pady = 5, sticky = "w")
              tkgrid(tklabel(fTab2Adv_1, text = "    Select locus : "), row = 2, column = 0, padx = 5, pady = 5, sticky = "w")
              selectL <- tclVar(parLocus[1])
              tkgrid(ttkcombobox(fTab2Adv_1, values = parLocus, textvariable = selectL, state = "readonly"), row = 2, column = 1, padx = 5, pady = 5, sticky = "w")
              tkgrid(tkbutton(fTab2Adv_1, text = "    Show    ", cursor = "hand2", command = function() showCond()), row = 2, column = 2, padx = 5, pady = 5, sticky = "w")
              tkgrid(fTab2Adv_1, padx = 10, sticky = "w")
              tkgrid(fTab2Adv_2, padx = 10, sticky = "w")
              tkgrid(fTab2Adv)
            }
            
            #Make a window for advanced setting of BSR
            tab3AdvMake <- function(){
              showCond <- function(){
                tkdestroy(fTab3Adv_2)
                fTab3Adv_2 <<- tkframe(fTab3Adv, relief = "groove", borderwidth = 2)
                tkgrid(tklabel(fTab3Adv_2, text = "Model"), row = 0, column = 0, padx = 5, pady = 5)
                tkgrid(tklabel(fTab3Adv_2, text = "Parameter"), row = 0, column = 1, padx = 5, pady = 5)
                tkgrid(tklabel(fTab3Adv_2, text = "Initial value"), row = 0, column = 2, padx = 5, pady = 5)
                tkgrid(tklabel(fTab3Adv_2, text = "Lower value"), row = 0, column = 3, padx = 5, pady = 5)
                tkgrid(tklabel(fTab3Adv_2, text = "Upper value"), row = 0, column = 4, padx = 5, pady = 5)
                
                posL <- which(parLocus == tclvalue(selectL))
                bsrMleCondOneL <- bsrMleCond[[posL]]
                modelName <- names(bsrMleCondOneL)
                nM <- length(modelName)
                rowCount <- 1
                for(i in 1:nM){
                  mleC <- bsrMleCondOneL[[i]]
                  mleInitial <- mleC[[1]]
                  mleLower <- mleC[[2]]
                  mleUpper <- mleC[[3]]
                  nPar <- length(mleInitial)
                  parName <- names(mleInitial)
                  tkgrid(tklabel(fTab3Adv_2, text = modelName[i]), row = rowCount, column = 0, padx = 5, pady = 5)
                  for(j in 1:nPar){
                    tkgrid(tklabel(fTab3Adv_2, text = parName[j]), row = rowCount, column = 1, padx = 5, pady = 5)
                    tkgrid(tkentry(fTab3Adv_2, textvariable = mleInitial[[j]], width = 10, highlightthickness = 1, relief = "solid", justify = "center", background = "white"), row = rowCount, column = 2, padx = 5, pady = 5)
                    tkgrid(tkentry(fTab3Adv_2, textvariable = mleLower[[j]], width = 10, highlightthickness = 1, relief = "solid", justify = "center", background = "white"), row = rowCount, column = 3, padx = 5, pady = 5)
                    tkgrid(tkentry(fTab3Adv_2, textvariable = mleUpper[[j]], width = 10, highlightthickness = 1, relief = "solid", justify = "center", background = "white"), row = rowCount, column = 4, padx = 5, pady = 5)
                    rowCount <- rowCount + 1
                  }
                }
                tkgrid(fTab3Adv_2, padx = 5, pady = 5, sticky = "w")
              }
              
              tkdestroy(fTab3Adv)
              fTab3Adv <<- tkframe(tab3Adv)
              
              fTab3Adv_1 <- tkframe(fTab3Adv)
              fTab3Adv_2 <- tkframe(fTab3Adv)
              
              tkgrid(tklabel(fTab3Adv_1, text = "Maximum of back stutter ratio : "), row = 0, column = 0, padx = 5, pady = 5, sticky = "w")
              tkgrid(tkentry(fTab3Adv_1, textvariable = bsrMaxVar, width = 10, highlightthickness = 1, relief = "solid", justify = "center", background = "white"), row = 0, column = 1, padx = 5, pady = 5, sticky = "w")
              tkgrid(tklabel(fTab3Adv_1, text = "Conditions of maximum likelihood estimation : "), row = 1, column = 0, padx = 5, pady = 5, sticky = "w")
              tkgrid(tklabel(fTab3Adv_1, text = "    Select locus : "), row = 2, column = 0, padx = 5, pady = 5, sticky = "w")
              selectL <- tclVar(parLocus[1])
              tkgrid(ttkcombobox(fTab3Adv_1, values = parLocus, textvariable = selectL, state = "readonly"), row = 2, column = 1, padx = 5, pady = 5, sticky = "w")
              tkgrid(tkbutton(fTab3Adv_1, text = "    Show    ", cursor = "hand2", command = function() showCond()), row = 2, column = 2, padx = 5, pady = 5, sticky = "w")
              tkgrid(fTab3Adv_1, padx = 10, sticky = "w")
              tkgrid(fTab3Adv_2, padx = 10, sticky = "w")
              tkgrid(fTab3Adv)
            }
            
            #Make a window for advanced setting of FSR
            tab4AdvMake <- function(){
              showCond <- function(){
                tkdestroy(fTab4Adv_2)
                fTab4Adv_2 <<- tkframe(fTab4Adv, relief = "groove", borderwidth = 2)
                tkgrid(tklabel(fTab4Adv_2, text = "Model"), row = 0, column = 0, padx = 5, pady = 5)
                tkgrid(tklabel(fTab4Adv_2, text = "Parameter"), row = 0, column = 1, padx = 5, pady = 5)
                tkgrid(tklabel(fTab4Adv_2, text = "Initial value"), row = 0, column = 2, padx = 5, pady = 5)
                tkgrid(tklabel(fTab4Adv_2, text = "Lower value"), row = 0, column = 3, padx = 5, pady = 5)
                tkgrid(tklabel(fTab4Adv_2, text = "Upper value"), row = 0, column = 4, padx = 5, pady = 5)
                
                posL <- which(parLocus == tclvalue(selectL))
                fsrMleCondOneL <- fsrMleCond[[posL]]
                modelName <- names(fsrMleCondOneL)
                nM <- length(modelName)
                rowCount <- 1
                for(i in 1:3){#nM){
                  mleC <- fsrMleCondOneL[[i]]
                  mleInitial <- mleC[[1]]
                  mleLower <- mleC[[2]]
                  mleUpper <- mleC[[3]]
                  nPar <- length(mleInitial)
                  parName <- names(mleInitial)
                  tkgrid(tklabel(fTab4Adv_2, text = modelName[i]), row = rowCount, column = 0, padx = 5, pady = 5)
                  for(j in 1:nPar){
                    tkgrid(tklabel(fTab4Adv_2, text = parName[j]), row = rowCount, column = 1, padx = 5, pady = 5)
                    tkgrid(tkentry(fTab4Adv_2, textvariable = mleInitial[[j]], width = 10, highlightthickness = 1, relief = "solid", justify = "center", background = "white"), row = rowCount, column = 2, padx = 5, pady = 5)
                    tkgrid(tkentry(fTab4Adv_2, textvariable = mleLower[[j]], width = 10, highlightthickness = 1, relief = "solid", justify = "center", background = "white"), row = rowCount, column = 3, padx = 5, pady = 5)
                    tkgrid(tkentry(fTab4Adv_2, textvariable = mleUpper[[j]], width = 10, highlightthickness = 1, relief = "solid", justify = "center", background = "white"), row = rowCount, column = 4, padx = 5, pady = 5)
                    rowCount <- rowCount + 1
                  }
                }
                tkgrid(fTab4Adv_2, padx = 5, pady = 5, sticky = "w")
              }
              
              tkdestroy(fTab4Adv)
              fTab4Adv <<- tkframe(tab4Adv)
              
              fTab4Adv_1 <- tkframe(fTab4Adv)
              fTab4Adv_2 <- tkframe(fTab4Adv)
              
              tkgrid(tklabel(fTab4Adv_1, text = "Maximum of forward stutter ratio : "), row = 0, column = 0, padx = 5, pady = 5, sticky = "w")
              tkgrid(tkentry(fTab4Adv_1, textvariable = fsrMaxVar, width = 10, highlightthickness = 1, relief = "solid", justify = "center", background = "white"), row = 0, column = 1, padx = 5, pady = 5, sticky = "w")
              tkgrid(tklabel(fTab4Adv_1, text = "Conditions of maximum likelihood estimation : "), row = 1, column = 0, padx = 5, pady = 5, sticky = "w")
              tkgrid(tklabel(fTab4Adv_1, text = "    Select locus : "), row = 2, column = 0, padx = 5, pady = 5, sticky = "w")
              selectL <- tclVar(parLocus[1])
              tkgrid(ttkcombobox(fTab4Adv_1, values = parLocus, textvariable = selectL, state = "readonly"), row = 2, column = 1, padx = 5, pady = 5, sticky = "w")
              tkgrid(tkbutton(fTab4Adv_1, text = "    Show    ", cursor = "hand2", command = function() showCond()), row = 2, column = 2, padx = 5, pady = 5, sticky = "w")
              tkgrid(fTab4Adv_1, padx = 10, sticky = "w")
              tkgrid(fTab4Adv_2, padx = 10, sticky = "w")
              tkgrid(fTab4Adv)
            }
            
            #Make a window for advanced setting of DSR
            tab5AdvMake <- function(){
              showCond <- function(){
                tkdestroy(fTab5Adv_2)
                fTab5Adv_2 <<- tkframe(fTab5Adv, relief = "groove", borderwidth = 2)
                tkgrid(tklabel(fTab5Adv_2, text = "Model"), row = 0, column = 0, padx = 5, pady = 5)
                tkgrid(tklabel(fTab5Adv_2, text = "Parameter"), row = 0, column = 1, padx = 5, pady = 5)
                tkgrid(tklabel(fTab5Adv_2, text = "Initial value"), row = 0, column = 2, padx = 5, pady = 5)
                tkgrid(tklabel(fTab5Adv_2, text = "Lower value"), row = 0, column = 3, padx = 5, pady = 5)
                tkgrid(tklabel(fTab5Adv_2, text = "Upper value"), row = 0, column = 4, padx = 5, pady = 5)
                
                posL <- which(parLocus == tclvalue(selectL))
                dsrMleCondOneL <- dsrMleCond[[posL]]
                modelName <- names(dsrMleCondOneL)
                nM <- length(modelName)
                rowCount <- 1
                for(i in 1:nM){
                  mleC <- dsrMleCondOneL[[i]]
                  mleInitial <- mleC[[1]]
                  mleLower <- mleC[[2]]
                  mleUpper <- mleC[[3]]
                  nPar <- length(mleInitial)
                  parName <- names(mleInitial)
                  tkgrid(tklabel(fTab5Adv_2, text = modelName[i]), row = rowCount, column = 0, padx = 5, pady = 5)
                  for(j in 1:nPar){
                    tkgrid(tklabel(fTab5Adv_2, text = parName[j]), row = rowCount, column = 1, padx = 5, pady = 5)
                    tkgrid(tkentry(fTab5Adv_2, textvariable = mleInitial[[j]], width = 10, highlightthickness = 1, relief = "solid", justify = "center", background = "white"), row = rowCount, column = 2, padx = 5, pady = 5)
                    tkgrid(tkentry(fTab5Adv_2, textvariable = mleLower[[j]], width = 10, highlightthickness = 1, relief = "solid", justify = "center", background = "white"), row = rowCount, column = 3, padx = 5, pady = 5)
                    tkgrid(tkentry(fTab5Adv_2, textvariable = mleUpper[[j]], width = 10, highlightthickness = 1, relief = "solid", justify = "center", background = "white"), row = rowCount, column = 4, padx = 5, pady = 5)
                    rowCount <- rowCount + 1
                  }
                }
                tkgrid(fTab5Adv_2, padx = 5, pady = 5, sticky = "w")
              }
              
              tkdestroy(fTab5Adv)
              fTab5Adv <<- tkframe(tab5Adv)
              
              fTab5Adv_1 <- tkframe(fTab5Adv)
              fTab5Adv_2 <- tkframe(fTab5Adv)
              
              tkgrid(tklabel(fTab5Adv_1, text = "Maximum of double-back stutter ratio : "), row = 0, column = 0, padx = 5, pady = 5, sticky = "w")
              tkgrid(tkentry(fTab5Adv_1, textvariable = dsrMaxVar, width = 10, highlightthickness = 1, relief = "solid", justify = "center", background = "white"), row = 0, column = 1, padx = 5, pady = 5, sticky = "w")
              tkgrid(tklabel(fTab5Adv_1, text = "Conditions of maximum likelihood estimation : "), row = 1, column = 0, padx = 5, pady = 5, sticky = "w")
              tkgrid(tklabel(fTab5Adv_1, text = "    Select locus : "), row = 2, column = 0, padx = 5, pady = 5, sticky = "w")
              selectL <- tclVar(parLocus[1])
              tkgrid(ttkcombobox(fTab5Adv_1, values = parLocus, textvariable = selectL, state = "readonly"), row = 2, column = 1, padx = 5, pady = 5, sticky = "w")
              tkgrid(tkbutton(fTab5Adv_1, text = "    Show    ", cursor = "hand2", command = function() showCond()), row = 2, column = 2, padx = 5, pady = 5, sticky = "w")
              tkgrid(fTab5Adv_1, padx = 10, sticky = "w")
              tkgrid(fTab5Adv_2, padx = 10, sticky = "w")
              tkgrid(fTab5Adv)
            }
            
            #Make a window for advanced setting of M2SR
            tab6AdvMake <- function(){
              showCond <- function(){
                tkdestroy(fTab6Adv_2)
                fTab6Adv_2 <<- tkframe(fTab6Adv, relief = "groove", borderwidth = 2)
                tkgrid(tklabel(fTab6Adv_2, text = "Model"), row = 0, column = 0, padx = 5, pady = 5)
                tkgrid(tklabel(fTab6Adv_2, text = "Parameter"), row = 0, column = 1, padx = 5, pady = 5)
                tkgrid(tklabel(fTab6Adv_2, text = "Initial value"), row = 0, column = 2, padx = 5, pady = 5)
                tkgrid(tklabel(fTab6Adv_2, text = "Lower value"), row = 0, column = 3, padx = 5, pady = 5)
                tkgrid(tklabel(fTab6Adv_2, text = "Upper value"), row = 0, column = 4, padx = 5, pady = 5)
                
                posL <- which(parLocus == tclvalue(selectL))
                m2srMleCondOneL <- m2srMleCond[[posL]]
                modelName <- names(m2srMleCondOneL)
                nM <- length(modelName)
                rowCount <- 1
                for(i in 1:nM){
                  mleC <- m2srMleCondOneL[[i]]
                  mleInitial <- mleC[[1]]
                  mleLower <- mleC[[2]]
                  mleUpper <- mleC[[3]]
                  nPar <- length(mleInitial)
                  parName <- names(mleInitial)
                  tkgrid(tklabel(fTab6Adv_2, text = modelName[i]), row = rowCount, column = 0, padx = 5, pady = 5)
                  for(j in 1:nPar){
                    tkgrid(tklabel(fTab6Adv_2, text = parName[j]), row = rowCount, column = 1, padx = 5, pady = 5)
                    tkgrid(tkentry(fTab6Adv_2, textvariable = mleInitial[[j]], width = 10, highlightthickness = 1, relief = "solid", justify = "center", background = "white"), row = rowCount, column = 2, padx = 5, pady = 5)
                    tkgrid(tkentry(fTab6Adv_2, textvariable = mleLower[[j]], width = 10, highlightthickness = 1, relief = "solid", justify = "center", background = "white"), row = rowCount, column = 3, padx = 5, pady = 5)
                    tkgrid(tkentry(fTab6Adv_2, textvariable = mleUpper[[j]], width = 10, highlightthickness = 1, relief = "solid", justify = "center", background = "white"), row = rowCount, column = 4, padx = 5, pady = 5)
                    rowCount <- rowCount + 1
                  }
                }
                tkgrid(fTab6Adv_2, padx = 5, pady = 5, sticky = "w")
              }
              
              tkdestroy(fTab6Adv)
              fTab6Adv <<- tkframe(tab6Adv)
              
              fTab6Adv_1 <- tkframe(fTab6Adv)
              fTab6Adv_2 <- tkframe(fTab6Adv)
              
              tkgrid(tklabel(fTab6Adv_1, text = "Maximum of minus 2-nt stutter ratio : "), row = 0, column = 0, padx = 5, pady = 5, sticky = "w")
              tkgrid(tkentry(fTab6Adv_1, textvariable = m2srMaxVar, width = 10, highlightthickness = 1, relief = "solid", justify = "center", background = "white"), row = 0, column = 1, padx = 5, pady = 5, sticky = "w")
              tkgrid(tklabel(fTab6Adv_1, text = "Conditions of maximum likelihood estimation : "), row = 1, column = 0, padx = 5, pady = 5, sticky = "w")
              tkgrid(tklabel(fTab6Adv_1, text = "    Select locus : "), row = 2, column = 0, padx = 5, pady = 5, sticky = "w")
              selectL <- tclVar(parLocus[1])
              tkgrid(ttkcombobox(fTab6Adv_1, values = parLocus, textvariable = selectL, state = "readonly"), row = 2, column = 1, padx = 5, pady = 5, sticky = "w")
              tkgrid(tkbutton(fTab6Adv_1, text = "    Show    ", cursor = "hand2", command = function() showCond()), row = 2, column = 2, padx = 5, pady = 5, sticky = "w")
              tkgrid(fTab6Adv_1, padx = 10, sticky = "w")
              tkgrid(fTab6Adv_2, padx = 10, sticky = "w")
              tkgrid(fTab6Adv)
            }
            
            tfAdv <- tktoplevel()
            tkwm.title(tfAdv, "Advanced setting")
            tabsAdv <- tk2notebook(tfAdv, tabs = c("Locus-specific amplification efficiency", "Heterozygote balance", "Back stutter ratio", "Forward stutter ratio", "Double-back stutter ratio", "Minus 2-nt stutter ratio"))
            tkpack(tabsAdv, fill = "both", expand = 1)
            tab1Adv <- tk2notetab(tabsAdv, "Locus-specific amplification efficiency")
            tab2Adv <- tk2notetab(tabsAdv, "Heterozygote balance")
            tab3Adv <- tk2notetab(tabsAdv, "Back stutter ratio")
            tab4Adv <- tk2notetab(tabsAdv, "Forward stutter ratio")
            tab5Adv <- tk2notetab(tabsAdv, "Double-back stutter ratio")
            tab6Adv <- tk2notetab(tabsAdv, "Minus 2-nt stutter ratio")
            
            fTab1Adv <- tkframe(tab1Adv)
            tab1AdvMake()
            fTab2Adv <- tkframe(tab2Adv)
            tab2AdvMake()
            fTab3Adv <- tkframe(tab3Adv)
            tab3AdvMake()
            fTab4Adv <- tkframe(tab4Adv)
            tab4AdvMake()
            fTab5Adv <- tkframe(tab5Adv)
            tab5AdvMake()
            fTab6Adv <- tkframe(tab6Adv)
            tab6AdvMake()
          }
          
          #Link a combobox1 and a combobox2
          comboBind <- function(combo1, combo2, srDealVar){
            tkbind(combo1, "<<ComboboxSelected>>", function(){
              if(tclvalue(srDealVar) == "Locus specific"){
                tkconfigure(combo2, state = "readonly")
              }else{
                tkconfigure(combo2, state = "disable")
              }
            })
          }
          
          tkdestroy(fTab2Par)
          fTab2Par <<- scrollMake(tab2Par)
          tkpack(fTab2Par, expand = TRUE, fill = "both")
          fTab2Par_scr <- scrollGet(fTab2Par)
          
          tkgrid(tklabel(fTab2Par_scr, text = "Locus", font = "Helvetica 10 bold"), row = 0, column = 0, padx = 5)
          tkgrid(tklabel(fTab2Par_scr, text = "Repeat length", font = "Helvetica 10 bold"), row = 0, column = 1, padx = 5)
          tkgrid(tklabel(fTab2Par_scr, text = "Min. threshold", font = "Helvetica 10 bold"), row = 0, column = 2, padx = 5)
          tkgrid(tklabel(fTab2Par_scr, text = "Locus-specific", font = "Helvetica 10 bold"), row = 0, column = 3, padx = 5)
          tkgrid(tklabel(fTab2Par_scr, text = "amplification efficiency", font = "Helvetica 10 bold"), row = 1, column = 3, padx = 5)
          tkgrid(tklabel(fTab2Par_scr, text = "Model"), row = 2, column = 3, padx = 5, pady = 5)
          tkgrid(tklabel(fTab2Par_scr, text = "Heterozygote balance", font = "Helvetica 10 bold"), row = 0, column = 4, padx = 5)
          tkgrid(tklabel(fTab2Par_scr, text = "Model"), row = 2, column = 4, padx = 5, pady = 5)
          tkgrid(tklabel(fTab2Par_scr, text = "Back stutter ratio", font = "Helvetica 10 bold"), row = 0, column = 5, columnspan = 2, padx = 5)
          tkgrid(tklabel(fTab2Par_scr, text = "Method"), row = 2, column = 5, padx = 5, pady = 5)
          tkgrid(tklabel(fTab2Par_scr, text = "Model"), row = 2, column = 6, padx = 5, pady = 5)
          tkgrid(tklabel(fTab2Par_scr, text = "Forward stutter ratio", font = "Helvetica 10 bold"), row = 0, column = 7, columnspan = 2, padx = 5)
          tkgrid(tklabel(fTab2Par_scr, text = "Method"), row = 2, column = 7, padx = 5, pady = 5)
          tkgrid(tklabel(fTab2Par_scr, text = "Model"), row = 2, column = 8, padx = 5, pady = 5)
          tkgrid(tklabel(fTab2Par_scr, text = "Double-back stutter ratio", font = "Helvetica 10 bold"), row = 0, column = 9, columnspan = 2, padx = 5)
          tkgrid(tklabel(fTab2Par_scr, text = "Method"), row = 2, column = 9, padx = 5, pady = 5)
          tkgrid(tklabel(fTab2Par_scr, text = "Model"), row = 2, column = 10, padx = 5, pady = 5)
          tkgrid(tklabel(fTab2Par_scr, text = "Minus 2-nt stutter ratio", font = "Helvetica 10 bold"), row = 0, column = 11, columnspan = 2, padx = 5)
          tkgrid(tklabel(fTab2Par_scr, text = "Method"), row = 2, column = 11, padx = 5, pady = 5)
          tkgrid(tklabel(fTab2Par_scr, text = "Model"), row = 2, column = 12, padx = 5, pady = 5)
          repLengthVar <- list()
          repLengthEntry <- list()
          mtVar <- list()
          mtEntry <- list()
          aeModelVar <- aeModelCombobox <- list()
          hbModelVar <- hbModelCombobox <- list()
          bsrModelVar <- bsrModelCombobox1 <- bsrModelCombobox2 <- list()
          fsrModelVar <- fsrModelCombobox1 <- fsrModelCombobox2 <- list()
          dsrModelVar <- dsrModelCombobox1 <- dsrModelCombobox2 <- list()
          m2srModelVar <- m2srModelCombobox1 <- m2srModelCombobox2 <- list()
          
          srDeal <- c("Locus specific", "Multiple loci together", "Not consider")
          bsrDealVar <- fsrDealVar <- dsrDealVar <- m2srDealVar <- list()
          
          aeCandFin <- hbCandFin <- bsrCandFin <- fsrCandFin <- dsrCandFin <- m2srCandFin <- list()
          for(i in 1:length(parLocus)){
            locusLabel <- tklabel(fTab2Par_scr, text = parLocus[i])
            locusPos <- which(parDefault[, 1] == parLocus[i])
            if(length(locusPos) == 1){
              parDefaultOneL <- parDefault[locusPos, ]
              aeCandFin[[i]] <- aeCand[[locusPos]]
              hbCandFin[[i]] <- hbCand[[locusPos]]
              bsrCandFin[[i]] <- bsrCand[[locusPos]]
              fsrCandFin[[i]] <- fsrCand[[locusPos]]
              dsrCandFin[[i]] <- dsrCand[[locusPos]]
              m2srCandFin[[i]] <- m2srCand[[locusPos]]
            }else{
              parDefaultOneL <- parGeneral
              aeCandFin[[i]] <- aeCandGen
              hbCandFin[[i]] <- hbCandGen
              bsrCandFin[[i]] <- bsrCandGen
              fsrCandFin[[i]] <- fsrCandGen
              dsrCandFin[[i]] <- dsrCandGen
              m2srCandFin[[i]] <- m2srCandGen
            }
            
            repLengthVar[[i]] <- tclVar(parDefaultOneL[2])
            repLengthEntry[[i]] <- tkentry(fTab2Par_scr, textvariable = repLengthVar[[i]], width = 10, highlightthickness = 1, relief = "solid", justify = "center", background = "white")
            mtVar[[i]] <- tclVar(parDefaultOneL[3])
            mtEntry[[i]] <- tkentry(fTab2Par_scr, textvariable = mtVar[[i]], width = 10, highlightthickness = 1, relief = "solid", justify = "center", background = "white")
            
            aeModelVar[[i]] <- tclVar(parDefaultOneL[4])
            aeModelCombobox[[i]] <- ttkcombobox(fTab2Par_scr, values = aeCandFin[[i]], textvariable = aeModelVar[[i]], width = 12, state = "readonly")
            
            hbModelVar[[i]] <- tclVar(parDefaultOneL[5])
            hbModelCombobox[[i]] <- ttkcombobox(fTab2Par_scr, values = hbCandFin[[i]], textvariable = hbModelVar[[i]], width = 12, state = "readonly")
            
            bsrDealVar[[i]] <- tclVar(parDefaultOneL[6])
            bsrModelCombobox1[[i]] <- ttkcombobox(fTab2Par_scr, values = srDeal, textvariable = bsrDealVar[[i]], state = "readonly")
            bsrModelVar[[i]] <- tclVar(parDefaultOneL[7])
            bsrModelCombobox2[[i]] <- ttkcombobox(fTab2Par_scr, values = bsrCandFin[[i]], textvariable = bsrModelVar[[i]], width = 10, state = "readonly")
            if(tclvalue(bsrDealVar[[i]]) != "Locus specific"){
              tkconfigure(bsrModelCombobox2[[i]], state = "disable")
            }
            
            fsrDealVar[[i]] <- tclVar(parDefaultOneL[8])
            fsrModelCombobox1[[i]] <- ttkcombobox(fTab2Par_scr, values = srDeal, textvariable = fsrDealVar[[i]], state = "readonly")
            fsrModelVar[[i]] <- tclVar(parDefaultOneL[9])
            fsrModelCombobox2[[i]] <- ttkcombobox(fTab2Par_scr, values = fsrCandFin[[i]], textvariable = fsrModelVar[[i]], width = 10, state = "readonly")
            if(tclvalue(fsrDealVar[[i]]) != "Locus specific"){
              tkconfigure(fsrModelCombobox2[[i]], state = "disable")
            }
            
            dsrDealVar[[i]] <- tclVar(parDefaultOneL[10])
            dsrModelCombobox1[[i]] <- ttkcombobox(fTab2Par_scr, values = srDeal, textvariable = dsrDealVar[[i]], state = "readonly")
            dsrModelVar[[i]] <- tclVar(parDefaultOneL[11])
            dsrModelCombobox2[[i]] <- ttkcombobox(fTab2Par_scr, values = dsrCandFin[[i]], textvariable = dsrModelVar[[i]], width = 10, state = "readonly")
            if(tclvalue(dsrDealVar[[i]]) != "Locus specific"){
              tkconfigure(dsrModelCombobox2[[i]], state = "disable")
            }
            
            m2srDealVar[[i]] <- tclVar(parDefaultOneL[12])
            m2srModelCombobox1[[i]] <- ttkcombobox(fTab2Par_scr, values = srDeal, textvariable = m2srDealVar[[i]], state = "readonly")
            m2srModelVar[[i]] <- tclVar(parDefaultOneL[13])
            m2srModelCombobox2[[i]] <- ttkcombobox(fTab2Par_scr, values = m2srCandFin[[i]], textvariable = m2srModelVar[[i]], width = 10, state = "readonly")
            if(tclvalue(m2srDealVar[[i]]) != "Locus specific"){
              tkconfigure(m2srModelCombobox2[[i]], state = "disable")
            }
            
            tkgrid(locusLabel, row = i + 2, column = 0, padx = 5, pady = 5)
            tkgrid(repLengthEntry[[i]], row = i + 2, column = 1, padx = 5, pady = 5)
            tkgrid(mtEntry[[i]], row = i + 2, column = 2, padx = 5, pady = 5)
            tkgrid(aeModelCombobox[[i]], row = i + 2, column = 3, padx = 5, pady = 5)
            tkgrid(hbModelCombobox[[i]], row = i + 2, column = 4, padx = 5, pady = 5)
            tkgrid(bsrModelCombobox1[[i]], row = i + 2, column = 5, padx = 5, pady = 5)
            tkgrid(bsrModelCombobox2[[i]], row = i + 2, column = 6, padx = 5, pady = 5)
            tkgrid(fsrModelCombobox1[[i]], row = i + 2, column = 7, padx = 5, pady = 5)
            tkgrid(fsrModelCombobox2[[i]], row = i + 2, column = 8, padx = 5, pady = 5)
            tkgrid(dsrModelCombobox1[[i]], row = i + 2, column = 9, padx = 5, pady = 5)
            tkgrid(dsrModelCombobox2[[i]], row = i + 2, column = 10, padx = 5, pady = 5)
            tkgrid(m2srModelCombobox1[[i]], row = i + 2, column = 11, padx = 5, pady = 5)
            tkgrid(m2srModelCombobox2[[i]], row = i + 2, column = 12, padx = 5, pady = 5)
          }
          names(aeCandFin) <- names(hbCandFin) <- names(bsrCandFin) <- names(fsrCandFin) <- names(dsrCandFin) <- names(m2srCandFin) <- parLocus
          
          aeMleCond <- aeMleCondMake(aeCandFin)
          hbMleCond <- hbMleCondMake(hbCandFin)
          bsrMleCond <- bsrMleCondMake(bsrCandFin)
          fsrMleCond <- fsrMleCondMake(fsrCandFin)
          dsrMleCond <- dsrMleCondMake(dsrCandFin)
          m2srMleCond <- m2srMleCondMake(m2srCandFin)
          
          bsrMultiModelVar <- tclVar(bsrCandGen[1])
          bsrMultiCombobox2 <- ttkcombobox(fTab2Par_scr, values = bsrCandGen, textvariable = bsrMultiModelVar, width = 10, state = "readonly")
          fsrMultiModelVar <- tclVar(fsrCandGen[1])
          fsrMultiCombobox2 <- ttkcombobox(fTab2Par_scr, values = fsrCandGen, textvariable = fsrMultiModelVar, width = 10, state = "readonly")
          dsrMultiModelVar <- tclVar(dsrCandGen[1])
          dsrMultiCombobox2 <- ttkcombobox(fTab2Par_scr, values = dsrCandGen, textvariable = dsrMultiModelVar, width = 10, state = "readonly")
          m2srMultiModelVar <- tclVar(m2srCandGen[1])
          m2srMultiCombobox2 <- ttkcombobox(fTab2Par_scr, values = m2srCandGen, textvariable = m2srMultiModelVar, width = 10, state = "readonly")
          tkgrid(tklabel(fTab2Par_scr, text = "Multiple loci"), row = i + 3, column = 0, padx = 5, pady = 5)
          tkgrid(tklabel(fTab2Par_scr, text = ""), row = i + 3, column = 1, columnspan = 5, padx = 5, pady = 5)
          tkgrid(bsrMultiCombobox2, row = i + 3, column = 6, padx = 5, pady = 5)
          tkgrid(tklabel(fTab2Par_scr, text = ""), row = i + 3, column = 7, padx = 5, pady = 5)
          tkgrid(fsrMultiCombobox2, row = i + 3, column = 8, padx = 5, pady = 5)
          tkgrid(tklabel(fTab2Par_scr, text = ""), row = i + 3, column = 9, padx = 5, pady = 5)
          tkgrid(dsrMultiCombobox2, row = i + 3, column = 10, padx = 5, pady = 5)
          tkgrid(tklabel(fTab2Par_scr, text = ""), row = i + 3, column = 11, padx = 5, pady = 5)
          tkgrid(m2srMultiCombobox2, row = i + 3, column = 12, padx = 5, pady = 5)
          mapply(comboBind, bsrModelCombobox1, bsrModelCombobox2, bsrDealVar)
          mapply(comboBind, fsrModelCombobox1, fsrModelCombobox2, fsrDealVar)
          mapply(comboBind, dsrModelCombobox1, dsrModelCombobox2, dsrDealVar)
          mapply(comboBind, m2srModelCombobox1, m2srModelCombobox2, m2srDealVar)
          
          tkgrid(tkbutton(fTab2Par_scr, text = "    Advanced setting    ", cursor = "hand2", command = function() advancedSetting()), row = i + 4, column = 1, columnspan = 2, padx = 5, pady = 10)
          tkgrid(tkbutton(fTab2Par_scr, text = "    Restore defaults    ", cursor = "hand2", command = function() restoreDefaults(parLocus, parDye)), row = i + 4, column = 3, columnspan = 2, padx = 5, pady = 10)
          tkgrid(tkbutton(fTab2Par_scr, text = "    Estimate parameters    ", cursor = "hand2", command = function() 
            parEstimate(parLocus, parDye, as.numeric(sapply(mtVar, tclvalue)), as.numeric(sapply(repLengthVar, tclvalue)), 
                        sapply(bsrDealVar, tclvalue), sapply(fsrDealVar, tclvalue), sapply(dsrDealVar, tclvalue), sapply(m2srDealVar, tclvalue), 
                        sapply(aeModelVar, tclvalue), sapply(hbModelVar, tclvalue), sapply(bsrModelVar, tclvalue), sapply(fsrModelVar, tclvalue), sapply(dsrModelVar, tclvalue), sapply(m2srModelVar, tclvalue), 
                        tclvalue(bsrMultiModelVar), tclvalue(fsrMultiModelVar), tclvalue(dsrMultiModelVar), tclvalue(m2srMultiModelVar), 
                        as.numeric(tclvalue(aeMinVar)), as.numeric(tclvalue(hbMinVar)), as.numeric(tclvalue(bsrMaxVar)), as.numeric(tclvalue(fsrMaxVar)), as.numeric(tclvalue(dsrMaxVar)), as.numeric(tclvalue(m2srMaxVar)), 
                        aeCandFin, hbCandFin, bsrCandFin, fsrCandFin, dsrCandFin, m2srCandFin, 
                        aeMleCond, hbMleCond, bsrMleCond, fsrMleCond, dsrMleCond, m2srMleCond
            )
          ), row = i + 4, column = 5, columnspan = 2, padx = 5, pady = 10)
        }
      }
    }
    
    
    ##################################################
    # Estimate parameters for Monte Carlo simulation #
    ##################################################
    
    #Make a list of model names in each parameter
    modelNameListMake <- function(modelName, modelCandFin, parDeal = NULL, multiModelName = NULL){
      modelNameList <- list()
      nL <- length(modelName)
      for(i in 1:nL){
        modelCandOneL <- modelCandFin[[i]]
        if(length(parDeal) != 0){
          if(parDeal[i] == "Multiple loci together"){
            if(multiModelName == "Best"){
              modelNameList[[i]] <- modelCandOneL[modelCandOneL != "Best"]
            }else{
              modelNameList[[i]] <- multiModelName
            }
          }else if(parDeal[i] == "Locus specific"){
            if(modelName[i] == "Best"){
              modelNameList[[i]] <- modelCandOneL[modelCandOneL != "Best"]
            }else{
              modelNameList[[i]] <- modelName[i]
            }
          }else{
            modelNameList[[i]] <- integer(0)
          }
        }else{
          if(modelName[i] == "Best"){
            modelNameList[[i]] <- modelCandOneL[modelCandOneL != "Best"]
          }else{
            modelNameList[[i]] <- modelName[i]
          }
        }
      }
      return(modelNameList)
    }
    
    #Determine conditions of MLE
    mleCondFinMake <- function(mleCond, modelNameList){
      mleCondFin <- list()
      nL <- length(mleCond)
      for(i in 1:nL){
        mleCondFinOneL <- list()
        mleCondOneL <- mleCond[[i]]
        modelName <- modelNameList[[i]]
        if(length(modelName) != 0){
          modelPos <- which(is.element(names(mleCondOneL), modelName) == TRUE)
          nM <- length(modelPos)
          for(j in 1:nM){
            mleCondFinOneM <- list()
            mleCondOneM <- mleCondOneL[[modelPos[j]]]
            mleCondFinOneM[[1]] <- as.numeric(sapply(mleCondOneM[[1]], tclvalue))
            mleCondFinOneM[[2]] <- as.numeric(sapply(mleCondOneM[[2]], tclvalue))
            mleCondFinOneM[[3]] <- as.numeric(sapply(mleCondOneM[[3]], tclvalue))
            names(mleCondFinOneM[[1]]) <- names(mleCondFinOneM[[2]]) <- names(mleCondFinOneM[[3]]) <- names(mleCondOneM[[1]])
            mleCondFinOneL[[j]] <- mleCondFinOneM
          }
          mleCondFin[[i]] <- mleCondFinOneL
        }else{
          mleCondFin[[i]] <- integer(0)
        }
      }
      return(mleCondFin)
    }
    
    #Arrange experimental data
    calDataArrange <- function(calData, gtAns, parLocus, mt, srConsider){
      
      #Pick up allele sizes
      pickAlSize <- function(sampleL, peakAllL, sizeAllL, dyeAllL, gtOne, gtMarPos, gtAlPos){
        alPeak <- alSize <- alDye <- numeric(0)
        nL <- length(sampleL)
        for(i in 1:nL){
          L1 <- sampleL[i]
          peakOneL <- peakAllL[i, !is.na(peakAllL[i, ])]
          peakPos <- which(peakOneL != "")
          peakOneL <- peakOneL[peakPos]
          sizeOneL <- sizeAllL[i, peakPos]
          sizeOneL <- as.numeric(sizeOneL)
          dyeOneL <- dyeAllL[i]
          gtOneL <- gtOne[gtOne[, gtMarPos] == L1, gtAlPos]
          alPos <- which(is.element(peakOneL, gtOneL) == TRUE)
          alPeak <- c(alPeak, peakOneL[alPos])
          alSize <- c(alSize, sizeOneL[alPos])
          alDye <- c(alDye, rep(dyeOneL, length(alPos)))
        }
        alSizeInfo <- rbind(alSize, alDye)
        colnames(alSizeInfo) <- alPeak
        return(alSizeInfo)
      }
      
      #Make possible peaks of observation in CSP under a hypothesized genotype combination
      possiblePeakMake <- function(gtCombOne, srConsiderOneL){
        possiblePeak <- gtCombOne
        if(srConsiderOneL[1]){
          possiblePeak <- c(possiblePeak, round(gtCombOne - 1, 1))
        }
        if(srConsiderOneL[2]){
          possiblePeak <- c(possiblePeak, round(gtCombOne + 1, 1))
        }
        if(srConsiderOneL[3]){
          possiblePeak <- c(possiblePeak, round(gtCombOne - 2, 1))
        }
        if(srConsiderOneL[4]){
          possiblePeak <- c(possiblePeak, round(gtCombOne[gtCombOne %% 1 == 0] - 0.8, 1), round(gtCombOne[gtCombOne %% 1 != 0] - 0.2, 1))
        }
        return(sort(unique(possiblePeak)))
      }
      
      #Remove two or more peaks which are located in the same bin.
      removePeak <- function(calOne){
        peakCount <- table(calOne[1, ])
        peakCutPos <- which(peakCount != 1)
        if(length(peakCutPos) != 0){
          calOne <- calOne[, -peakCutPos, drop = FALSE]
        }
        return(calOne)
      }
      
      #Remove pull-up peaks
      removePU <- function(calOne, dyeOneL, alSize, alDye){
        peakOneL2 <- calOne[1, ]
        sizeOneL2 <- calOne[2, ]
        heightOneL2 <- calOne[3, ]
        alPos2 <- which(is.element(peakOneL2, gtOneL) == TRUE)
        puDyePos <- which(alDye != dyeOneL)
        puSize <- alSize[puDyePos]
        puPos <- numeric(0)
        for(k in 1:length(sizeOneL2)){
          if(!is.element(k, alPos2)){
            size2One <- sizeOneL2[k]
            if(length(which(abs(size2One - puSize) <= 1)) != 0){
              puPos <- c(puPos, k)
            }
          }
        }
        if(length(puPos) != 0){
          calOne <- calOne[, - puPos, drop = FALSE]
        }
        return(calOne)
      }
      
      colCal <- colnames(calData)
      sampleFilePos <- intersect(grep("Sample", colCal), grep("File", colCal))
      sampleFiles <- unique(calData[, sampleFilePos])
      sampleNamePos <- intersect(grep("Sample", colCal), grep("Name", colCal))
      markerPos <- grep("Marker", colCal)
      dyePos <- grep("Dye", colCal)
      peakPos <- grep("Allele", colCal)
      sizePos <- grep("Size", colCal)
      heightPos <- grep("Height", colCal)
      
      colGtAns <- colnames(gtAns)
      gtSampleNamePos <- intersect(grep("Sample", colGtAns), grep("Name", colGtAns))
      gtMarPos <- grep("Marker", colGtAns)
      gtAlPos <- grep("Allele", colGtAns)
      
      sampleOne <- calData[calData[, sampleFilePos] == sampleFiles[1], ]
      sampleL <- sampleOne[, markerPos]
      dyeAllL <- sampleOne[, dyePos]
      
      calData2 <- gtAns2 <- list()
      for(i in 1:length(sampleFiles)){
        sampleOne <- calData[calData[, sampleFilePos] == sampleFiles[i], ]
        sampleL <- sampleOne[, markerPos]
        peakAllL <- sampleOne[, peakPos]
        sizeAllL <- sampleOne[, sizePos]
        heightAllL <- sampleOne[, heightPos]
        sampleName <- sampleOne[1, sampleNamePos]
        gtOne <- gtAns[gtAns[, gtSampleNamePos] == sampleName, ]
        
        alSizeInfo <- pickAlSize(sampleL, peakAllL, sizeAllL, dyeAllL, gtOne, gtMarPos, gtAlPos)
        alSize <- as.numeric(alSizeInfo[1, ])
        alDye <- alSizeInfo[2, ]
        
        calData2SampleOne <- list()
        nL2 <- length(parLocus)
        gtOne2 <- matrix(NA, nrow = nL2, ncol = 2)
        for(j in 1:nL2){
          L2 <- parLocus[j]
          L2Pos <- grep(L2, sampleOne[, markerPos])
          peakOneL <- peakAllL[L2Pos, !is.na(peakAllL[L2Pos, ])]
          fillPos <- peakOneL != ""
          options(warn = -1)
          peakOneL <- as.numeric(peakOneL[fillPos])
          options(warn = 0)
          sizeOneL <- as.numeric(sizeAllL[L2Pos, fillPos])
          heightOneL <- as.numeric(heightAllL[L2Pos, fillPos])
          dyeOneL <- dyeAllL[L2Pos]
          upMtPos <- heightOneL >= mt[j]
          peakOneL <- peakOneL[upMtPos]
          sizeOneL <- sizeOneL[upMtPos]
          heightOneL <- heightOneL[upMtPos]
          
          gtOneL <- gtOne[gtOne[, gtMarPos] == L2, gtAlPos]
          gtOneL <- as.numeric(gtOneL)
          gtOne2[j, ] <- gtOneL
          gtOneL <- gtOneL[!is.na(gtOneL)]
          
          possiblePeaks <- possiblePeakMake(gtOneL, c(srConsider[j, ]))
          usePos <- which(is.element(peakOneL, possiblePeaks) == TRUE)
          
          if(length(usePos) != 0){
            peakOneL2 <- peakOneL[usePos]
            sizeOneL2 <- as.numeric(sizeOneL[usePos])
            heightOneL2 <- as.numeric(heightOneL[usePos])
            calOne <- rbind(peakOneL2, sizeOneL2, heightOneL2)
            calOne <- removePeak(calOne)
            calData2SampleOne[[j]] <- removePU(calOne, dyeOneL, alSize, alDye)
          }else{
            calData2SampleOne[[j]] <- 0
          }
        }
        calData2[[i]] <- calData2SampleOne
        gtAns2[[i]] <- gtOne2
      }
      return(list(calData2, gtAns2))
    }
    
    #Get data of locus-specific amplification efficiency
    aeDataGet <- function(calData2){
      nS <- length(calData2)
      nL <- length(calData2[[1]])
      aeHeightSum <- matrix(0, nS, nL)
      useJudge <- rep(TRUE, nS)
      for(i in 1:nS){
        calData2One <- calData2[[i]]
        for(j in 1:nL){
          calData2OneL <- calData2One[[j]]
          if(is.matrix(calData2OneL)){
            aeHeightSum[i, j] <- sum(calData2OneL[3, ]) / 2
          }else{
            useJudge[i] <- FALSE
          }
        }
      }
      aeData <- matrix(0, nS, nL)
      for(i in 1:nS){
        aeData[i, ] <- aeHeightSum[i, ] / mean(aeHeightSum[i, ])
      }
      aeData <- aeData[useJudge, ]
      aeHeightData <- aeHeightSum[useJudge, ]
      return(list(aeData, aeHeightData))
    }
    
    #Get data of heterozygote balance
    hbDataGet <- function(calData2, gtAns2, srConsider){
      nS <- length(calData2)
      nL <- length(calData2[[1]])
      hbData <- hbHeightData <- matrix(0, nS, nL)
      for(i in 1:nS){
        calData2One <- calData2[[i]]
        gtAnsOne <- gtAns2[[i]]
        for(j in 1:nL){
          calData2OneL <- calData2One[[j]]
          if(is.matrix(calData2OneL)){
            peak <- calData2OneL[1, ]
            height <- calData2OneL[3, ]
            gtOneL <- gtAnsOne[j, ]
            gtOneL <- sort(unique(gtOneL[!is.na(gtOneL)]))
            srConsiderOneL <- srConsider[j, ]
            
            #Memo: heterozygote && other than one repeat difference
            if((length(gtOneL) == 2) && ((abs(round(gtOneL[1] - gtOneL[2], 1)) != 1))){
              hbHeight <- rep(0, 2)
              
              #Memo: no allelic drop-out
              if(all(is.element(gtOneL, peak))){
                for(k in 1:2){
                  alOne <- gtOneL[k]
                  alTwo <- gtOneL[3 - k]
                  shareAlSt <- alTwo
                  if(srConsiderOneL[1]){
                    shareAlSt <- c(shareAlSt, alTwo - 1)
                  }
                  if(srConsiderOneL[2]){
                    shareAlSt <- c(shareAlSt, alTwo + 1)
                  }
                  if(srConsiderOneL[3]){
                    shareAlSt <- c(shareAlSt, alTwo - 2)
                  }
                  if(srConsiderOneL[4]){
                    if(alTwo %% 1 == 0){
                      shareAlSt <- c(shareAlSt, alTwo - 0.8)
                    }else{
                      shareAlSt <- c(shareAlSt, alTwo - 0.2)
                    }
                  }
                  shareAlSt <- round(sort(unique(shareAlSt)), 1)
                  heightSum <- height[which(peak == alOne)]
                  
                  #Memo: back stutter
                  if(srConsiderOneL[1]){
                    stB1 <- round(alOne - 1, 1)
                    if(is.element(stB1, peak)){
                      heightSum <- heightSum + height[which(peak == stB1)]
                    }
                  }
                  
                  #Memo: forward stutter
                  if(srConsiderOneL[2]){
                    stF1 <- round(alOne + 1, 1)
                    if((is.element(stF1, peak)) && !is.element(stF1, shareAlSt)){
                      heightSum <- heightSum + height[which(peak == stF1)]
                    }
                  }
                  
                  #Memo: double-back stutter
                  if(srConsiderOneL[3]){
                    stB2 <- round(alOne - 2, 1)
                    if((is.element(stB2, peak)) && !is.element(stB2, shareAlSt)){
                      heightSum <- heightSum + height[which(peak == stB2)]
                    }
                  }
                  
                  #Memo: minus 2-nt stutter
                  if(srConsiderOneL[4]){
                    if(alOne %% 1 < 0.15){
                      stM2 <- round(alOne - 0.8, 1)
                    }else{
                      stM2 <- round(alOne - 0.2, 1)
                    }
                    if((is.element(stM2, peak)) && !is.element(stM2, shareAlSt)){
                      heightSum <- heightSum + height[which(peak == stM2)]
                    }
                  }
                  hbHeight[k] <- heightSum
                }
                hbData[i, j] <- hbHeight[2] / hbHeight[1]
                hbHeightData[i, j] <- sum(height) / 2
              }
            }
          }
        }
      }
      return(list(hbData, hbHeightData))
    }
    
    #Get data of stutter ratios
    srDataGet <- function(calData2, gtAns2, srConsider){
      numSample <- length(calData2)
      nL <- length(calData2[[1]])
      bsrData <- fsrData <- dsrData <- m2srData <- srAlData <- srHeightData <- matrix(0, 2 * numSample, nL)
      for(i in 1:numSample){
        calData2One <- calData2[[i]]
        gtAnsOne <- gtAns2[[i]]
        for(j in 1:nL){
          calData2OneL <- calData2One[[j]]
          if(is.matrix(calData2OneL)){
            peak <- calData2OneL[1, ]
            height <- calData2OneL[3, ]
            gtOneL <- gtAnsOne[j, ]
            gtOneL <- sort(unique(gtOneL[!is.na(gtOneL)]))
            srConsiderOneL <- srConsider[j, ]
            
            #Memo: homozygote
            if(length(gtOneL) == 1){
              srAlData[2 * (i - 1) + 1, j] <- gtOneL
              
              #Memo: no allelic drop-out
              if(is.element(gtOneL, peak)){
                srHeightData[2 * (i - 1) + 1, j] <- sum(height)
                heightAl <- height[which(peak == gtOneL)]
                
                if(srConsiderOneL[1]){
                  if(is.element(round(gtOneL - 1, 1), peak)){
                    bsrData[2 * (i - 1) + 1, j] <- height[which(peak == round(gtOneL - 1, 1))] / heightAl
                  }
                }
                if(srConsiderOneL[2]){
                  if(is.element(round(gtOneL + 1, 1), peak)){
                    fsrData[2 * (i - 1) + 1, j] <- height[which(peak == round(gtOneL + 1, 1))] / heightAl
                  }
                }
                if(srConsiderOneL[3]){
                  if(is.element(round(gtOneL - 2, 1), peak)){
                    dsrData[2 * (i - 1) + 1, j] <- height[which(peak == round(gtOneL - 2, 1))] / heightAl
                  }
                }
                if(srConsiderOneL[4]){
                  if(gtOneL %% 1 < 0.15){  #allele x or x.1
                    if(is.element(round(gtOneL - 0.8, 1), peak)){
                      m2srData[2 * (i - 1) + 1, j] <- height[which(peak == round(gtOneL - 0.8, 1))] / heightAl
                    }
                  }else{  #allele x.2, x.3, ...
                    if(is.element(round(gtOneL - 0.2, 1), peak)){
                      m2srData[2 * (i - 1) + 1, j] <- height[which(peak == round(gtOneL - 0.2, 1))] / heightAl
                    }
                  }
                }
              }
              
              #Memo: heterozygote && other than one repeat difference
            }else if(abs(round(gtOneL[1] - gtOneL[2], 1)) != 1){
              for(k in 1:2){
                alOne <- gtOneL[k]
                #Memo: no allelic drop-out
                if(length(which(peak == alOne)) != 0){
                  alTwo <- gtOneL[3 - k]
                  shareAlSt <- alTwo
                  if(srConsiderOneL[1]){
                    shareAlSt <- c(shareAlSt, alTwo - 1)
                  }
                  if(srConsiderOneL[2]){
                    shareAlSt <- c(shareAlSt, alTwo + 1)
                  }
                  if(srConsiderOneL[3]){
                    shareAlSt <- c(shareAlSt, alTwo - 2)
                  }
                  if(srConsiderOneL[4]){
                    if(alTwo %% 1 < 0.15){  #allele x or x.1
                      shareAlSt <- c(shareAlSt, alTwo - 0.8)
                    }else{  #allele x.2, x.3, ...
                      shareAlSt <- c(shareAlSt, alTwo - 0.2)
                    }
                  }
                  shareAlSt <- round(sort(unique(shareAlSt)), 1)
                  srAlData[2 * (i - 1) + k, j] <- alOne
                  heightSum <- heightAl <- height[which(peak == alOne)]
                  
                  stB2Pos <- stB1Pos <- stF1Pos <- stM2Pos <- numeric(0)
                  
                  #Memo: back stutter
                  if(srConsiderOneL[1]){
                    stB1 <- round(alOne - 1, 1)
                    if(is.element(stB1, peak)){ 
                      heightSum <- heightSum + height[which(peak == stB1)]
                      if(!is.element(stB1, shareAlSt)){
                        stB1Pos <- which(peak == stB1)
                        bsrData[2 * (i - 1) + k, j] <- height[stB1Pos] / heightAl
                      }
                    }
                  }
                  
                  #Memo: forward stutter
                  if(srConsiderOneL[2]){
                    stF1 <- round(alOne + 1, 1)
                    if((is.element(stF1, peak)) && !is.element(stF1, shareAlSt)){
                      stF1Pos <- which(peak == stF1)
                      fsrData[2 * (i - 1) + k, j] <- height[stF1Pos] / heightAl
                      heightSum <- heightSum + height[stF1Pos]
                    }
                  }
                  
                  #Memo: double-back stutter
                  if(srConsiderOneL[3]){
                    stB2 <- round(alOne - 2, 1)
                    if((is.element(stB2, peak)) && !is.element(stB2, shareAlSt)){
                      stB2Pos <- which(peak == stB2)
                      dsrData[2 * (i - 1) + k, j] <- height[stB2Pos] / heightAl
                      heightSum <- heightSum + height[stB2Pos]
                    }
                  }
                  
                  #Memo: minus 2-nt stutter
                  if(srConsiderOneL[4]){
                    if(alOne %% 1 < 0.15){  #allele x or x.1
                      stM2 <- round(alOne - 0.8, 1)
                    }else{  #allele x.2, x.3, ...
                      stM2 <- round(alOne - 0.2, 1)
                    }
                    if((is.element(stM2, peak)) && !is.element(stM2, shareAlSt)){
                      stM2Pos <- which(peak == stM2)
                      m2srData[2 * (i - 1) + k, j] <- height[stM2Pos] / heightAl
                      heightSum <- heightSum + height[stM2Pos]
                    }
                  }
                  srHeightData[2 * (i - 1) + k, j] <- heightSum
                }
              }
            }
          }
        }
      }
      return(list(bsrData, fsrData, dsrData, m2srData, srAlData, srHeightData))
    }
    
    #Make information of motif length (for LUS and multi-seq model)
    motifLengthMake <- function(repeatSeq){
      motifLength <- list()
      for(i in 1:length(repeatSeq)){
        seqOne <- strsplit(repeatSeq[i], ' ')[[1]]
        motifLengthOne <- numeric(0)
        for(j in 1:length(seqOne)){
          seqOne_oneMotif <- seqOne[j]
          bracketPos <- regexpr("]", seqOne_oneMotif)[1]
          if(bracketPos != -1){
            motifLengthOne <- c(motifLengthOne, as.numeric(strsplit(seqOne_oneMotif, ']')[[1]][2]))
          }
        }
        motifLength[[i]] <- motifLengthOne
      }
      return(motifLength)
    }
    
    #Analyze sequence data
    seqAnalysis <- function(seqData, parLocus){
      lusData <- list()
      motifLengthList <- list()
      seqAlList <- list()
      seqCountList <- list()
      for(i in 1:length(parLocus)){
        locusPos <- which(seqData[, which(colnames(seqData) == "Marker")] == parLocus[i])
        seqAlList[[i]] <- alleles <- as.numeric(seqData[locusPos, which(colnames(seqData) == "Allele")])
        alNames <- sort(unique(alleles))
        seqCountList[[i]] <- count <- as.numeric(seqData[locusPos, which(colnames(seqData) == "Count")])
        repeatSeq <- seqData[locusPos, intersect(grep("Repeat", colnames(seqData)), grep("Region", colnames(seqData)))]
        
        motifLengthList[[i]] <- motifLengthMake(repeatSeq)
        lus <- sapply(motifLengthList[[i]], max)
        lusOneL <- rep(0, length(alNames))
        
        for(j in 1:length(alNames)){
          alPos <- which(alleles == alNames[j])
          lusOneAl <- lus[alPos]
          lusOneL[j] <- sum(lusOneAl * count[alPos] / sum(count[alPos]))
        }
        names(lusOneL) <- alNames
        lusData[[i]] <- lusOneL
      }
      return(list(lusData, motifLengthList, seqAlList, seqCountList))
    }
    
    #Calculate AIC
    aicCalc <- function(logLike, nP){
      return(-2 * logLike + 2 * nP)
    }
    
    #Modeling locus-specific amplification efficiency
    aeModeling <- function(modelName, ae, aeHeight, aeMin, aeMleCondFinOneL){
      #likelihood function of Log-normal model
        #par(mean, variance)
      aeLognormLike <- function(par){
        dens <- dtruncnorm(log(ae), a = log(aeMin), b = log(1 / aeMin), mean = par[1], sd = sqrt(par[2] / aeHeight))
        return(-sum(log(dens)))
      }
      
      aeModel <- list()
      
      #Log-normal model
      modelPos <- which(modelName == "Log-normal")
      if(length(modelPos) == 1){
        aeMleCondFinOneM <- aeMleCondFinOneL[[modelPos]]
        initialVal <- aeMleCondFinOneM[[1]]
        lowerVal <- aeMleCondFinOneM[[2]]
        upperVal <- aeMleCondFinOneM[[3]]
        aeLognormFit <- GenSA(par = initialVal, fn = aeLognormLike, lower = lowerVal, upper = upperVal)
        logLike <- - aeLognormFit[[1]]
        aic <- aicCalc(logLike, length(initialVal))
        modelInfo <- c(logLike, aic)
        names(modelInfo) <- c("likelihood", "AIC")
        estPar <- aeLognormFit[[2]]
        names(estPar) <- c("mean", "variance")
        aeModel[[modelPos]] <- list(modelInfo, estPar, modelName[modelPos])
      }
      
      return(aeModel)
    }
    
    #Modeling heterozygote balance
    hbModeling <- function(modelName, hb, hbHeight, hbMin, hbMleCondFinOneL){
      #likelihood function of Log-normal model
        #par(mean, variance)
      hbLognormLike <- function(par){
        dens <- dtruncnorm(log(hb), a = log(hbMin), b = log(1 / hbMin), mean = par[1], sd = sqrt(par[2] / hbHeight))
        return(-sum(log(dens)))
      }
      
      hbModel <- list()
      
      #Log-normal model
      modelPos <- which(modelName == "Log-normal")
      if(length(modelPos) == 1){
        hbMleCondFinOneM <- hbMleCondFinOneL[[modelPos]]
        initialVal <- hbMleCondFinOneM[[1]]
        lowerVal <- hbMleCondFinOneM[[2]]
        upperVal <- hbMleCondFinOneM[[3]]
        hbLognormFit <- GenSA(par = initialVal, fn = hbLognormLike, lower = lowerVal, upper = upperVal)
        logLike <- - hbLognormFit[[1]]
        aic <- aicCalc(logLike, length(initialVal))
        modelInfo <- c(logLike, aic)
        names(modelInfo) <- c("likelihood", "AIC")
        estPar <- hbLognormFit[[2]]
        names(estPar) <- c("mean", "variance")
        hbModel[[modelPos]] <- list(modelInfo, estPar, modelName[modelPos])
      }
      
      return(hbModel)
    }
    
    #Make usable SR data when considering each locus separately
    srUseMake <- function(sr, srAl, srHeight, lusOneL, srMax){
      nonZeroPos <- which(sr != 0)
      sr <- sr[nonZeroPos]
      srAl <- srAl[nonZeroPos]
      srHeight <- srHeight[nonZeroPos]
      srAdoptPos <- which(sr < srMax)
      sr <- sr[srAdoptPos]
      srAl <- srAl[srAdoptPos]
      srHeight <- srHeight[srAdoptPos]
      
      alNames <- as.numeric(names(lusOneL))
      srLus <- rep(-100, length(sr))
      for(j in 1:length(lusOneL)){
        lusPos <- which(srAl == alNames[j])
        if(length(lusPos) != 0){
          srLus[lusPos] <- lusOneL[j]
        }
      }
      
      noSeqPos <- which(srLus == -100)
      if(length(noSeqPos) != 0){
        sr <- sr[-noSeqPos]
        srAl <- srAl[-noSeqPos]
        srHeight <- srHeight[-noSeqPos]
        srLus <- srLus[-noSeqPos]
      }
      srUseList <- list(sr, srHeight, srAl, srLus)
      names(srUseList) <- c("stutter ratio", "total allelic product", "parent allele", "LUS of parent allele")
      return(srUseList)
    }
    
    #Make SR data for considering multiple loci together
    srGenMake <- function(srGenPos, srData, srAlData, srHeightData, lusData, motifLengthList, seqAlList, seqCountList, srMax){
      srGen <- srAlGen <- srHeightGen <- srLusGen <- motifLengthGen <- seqAlGen <- seqCountGen <- list()
      for(i in 1:length(srGenPos)){
        srPos <- srGenPos[i]
        srUseData <- srUseMake(srData[, srPos], srAlData[, srPos], srHeightData[, srPos], lusData[[srPos]], srMax)
        srGen[[i]] <- srUseData[[1]]
        srHeightGen[[i]] <- srUseData[[2]]
        srAlGen[[i]] <- srUseData[[3]]
        srLusGen[[i]] <- srUseData[[4]]
        motifLengthGen[[i]] <- motifLengthList[[srPos]]
        seqAlGen[[i]] <- seqAlList[[srPos]]
        seqCountGen[[i]] <- seqCountList[[srPos]]
      }
      srGenList <- list(srGen, srHeightGen, srAlGen, srLusGen, motifLengthGen, seqAlGen, seqCountGen)
      names(srGenList) <- c("stutter ratio", "total allelic product", "parent allele", "LUS of parent allele", "motif length", "alleles of sequence data", "counts of each sequence")
      return(srGenList)
    }
    
    #Modeling stutter ratio
      #When modeling multiple loci together, each object is the list of each locus data
    srModeling <- function(modelName, sr, srHeight, srAl, srLus = NULL, motifLength = NULL, seqAl = NULL, seqCount = NULL, srMax, srMleCondFinOneL){
      
      #likelihood function of Allele model
        #par(slope, intercept, variance)
      srAlLike <- function(par, srMax){
        meanVal <- par[1] * srAl + par[2]
        dens <- dtruncnorm(log(sr), b = log(srMax), mean = log(meanVal), sd = sqrt(par[3] / srHeight))
        return(-sum(log(dens)))
      }
      
      #likelihood function of LUS model
        #par(slope, intercept, variance)
      srLusLike <- function(par, srMax){
        meanVal <- par[1] * srLus + par[2]
        dens <- dtruncnorm(log(sr), b = log(srMax), mean = log(meanVal), sd = sqrt(par[3] / srHeight))
        return(-sum(log(dens)))
      }
      
      #likelihood function of multi-sequence model (modeling multiple loci together)
        #par(slope, intercept, variance, x)
      srMultiLikeGen <- function(par, srMax){
        lenCorMeanAll <- numeric(0)
        for(i in 1:length(motifLength)){
          mL <- motifLength[[i]]
          nML <- length(mL)
          lenCorSum <- rep(0, nML)
          for(j in 1:nML){
            lenCor <- mL[[j]] - round(par[4], 0)
            lenCorSum[j] <- sum(lenCor[which(lenCor > 0)])
          }
          srAlOneL <- srAlList[[i]]
          seqAlOneL <- seqAl[[i]]
          alNamesOneL <- sort(unique(seqAlOneL))
          seqCountOneL <- seqCount[[i]]
          lenCorMeanOneL <- rep(-100, length(srAlOneL))
          
          for(j in 1:length(alNamesOneL)){
            alPos <- which(seqAlOneL == alNamesOneL[j])
            lenCorSumOneAl <- lenCorSum[alPos]
            srAlPos <- which(srAlOneL == alNamesOneL[j])
            if(length(srAlPos) > 0){
              lenCorMeanOneL[srAlPos] <- sum(lenCorSumOneAl * seqCountOneL[alPos] / sum(seqCountOneL[alPos]))
            }
          }
          lenCorMeanAll <- c(lenCorMeanAll, lenCorMeanOneL)
        }
        meanVal <- par[1] * lenCorMeanAll + par[2]
        dens <- dtruncnorm(log(sr), b = log(srMax), mean = log(meanVal), sd = sqrt(par[3] / srHeight))
        return(-sum(log(dens)))
      }
      
      #likelihood function of multi-sequence model (locus specific)
        #par(slope, intercept, variance, x)
      srMultiLikeSp <- function(par, srMax){
        nML <- length(motifLength)
        lenCorSum <- rep(0, nML)
        for(i in 1:nML){
          lenCor <- motifLength[[i]] - round(par[4], 0)
          lenCorSum[i] <- sum(lenCor[which(lenCor > 0)])
        }
        lenCorMeanOneL <- rep(-100, length(srAl))
        alNames <- sort(unique(seqAl))
        for(i in 1:length(alNames)){
          alPos <- which(seqAl == alNames[i])
          lenCorSumOneAl <- lenCorSum[alPos]
          srAlPos <- which(srAl == alNames[i])
          if(length(srAlPos) > 0){
            lenCorMeanOneL[srAlPos] <-  sum(lenCorSumOneAl * seqCount[alPos] / sum(seqCount[alPos]))
          }
        }
        meanVal <- par[1] * lenCorMeanOneL + par[2]
        dens <- dtruncnorm(log(sr), b = log(srMax), mean = log(meanVal), sd = sqrt(par[3] / srHeight))
        return(-sum(log(dens)))
      }
      
      #likelihood function of uniform model
        #par(mean, variance)
      srUniformLike <- function(par, srMax){
        dens <- dtruncnorm(log(sr), b = log(srMax), mean = log(par[1]), sd = sqrt(par[2] / srHeight))
        return(-sum(log(dens)))
      }
      
      if(is.list(sr)){
        multiL <- TRUE
        srAlList <- srAl
        sr <- unlist(sr)
        srAl <- unlist(srAl)
        srHeight <- unlist(srHeight)
        srLus <- unlist(srLus)
      }else{
        multiL <- FALSE
      }
      
      srModel <- list()
      
      #Allele model
      modelPos <- which(modelName == "Allele")
      if(length(modelPos) == 1){
        srMleCondFinOneM <- srMleCondFinOneL[[modelPos]]
        initialVal <- srMleCondFinOneM[[1]]
        lowerVal <- srMleCondFinOneM[[2]]
        upperVal <- srMleCondFinOneM[[3]]
        srAlleleFit <- GenSA(par = initialVal, fn = srAlLike, srMax = srMax, lower = lowerVal, upper = upperVal)
        logLike <- - srAlleleFit[[1]]
        aic <- aicCalc(logLike, length(initialVal))
        modelInfo <- c(logLike, aic)
        names(modelInfo) <- c("likelihood", "AIC")
        estPar <- srAlleleFit[[2]]
        names(estPar) <- c("slope", "intercept", "variance")
        srModel[[modelPos]] <- list(modelInfo, estPar, modelName[modelPos])
      }
      
      #LUS model
      modelPos <- which(modelName == "LUS")
      if(length(modelPos) == 1){
        srMleCondFinOneM <- srMleCondFinOneL[[modelPos]]
        initialVal <- srMleCondFinOneM[[1]]
        lowerVal <- srMleCondFinOneM[[2]]
        upperVal <- srMleCondFinOneM[[3]]
        srLusFit <- GenSA(par = initialVal, fn = srLusLike, srMax = srMax, lower = lowerVal, upper = upperVal)
        logLike <- - srLusFit[[1]]
        aic <- aicCalc(logLike, length(initialVal))
        modelInfo <- c(logLike, aic)
        names(modelInfo) <- c("likelihood", "AIC")
        estPar <- srLusFit[[2]]
        names(estPar) <- c("slope", "intercept", "variance")
        srModel[[modelPos]] <- list(modelInfo, estPar, modelName[modelPos])
      }
      
      #Multi-seq model
      modelPos <- which(modelName == "Multi-seq")
      if(length(modelPos) == 1){
        srMleCondFinOneM <- srMleCondFinOneL[[modelPos]]
        initialVal <- srMleCondFinOneM[[1]]
        lowerVal <- srMleCondFinOneM[[2]]
        upperVal <- srMleCondFinOneM[[3]]
        if(multiL){
          srMultiFit <- GenSA(par = initialVal, fn = srMultiLikeGen, srMax = srMax, lower = lowerVal, upper = upperVal)
        }else{
          srMultiFit <- GenSA(par = initialVal, fn = srMultiLikeSp, srMax = srMax, lower = lowerVal, upper = upperVal)
        }
        logLike <- - srMultiFit[[1]]
        aic <- aicCalc(logLike, length(initialVal))
        modelInfo <- c(logLike, aic)
        names(modelInfo) <- c("likelihood", "AIC")
        estPar <- srMultiFit[[2]]
        estPar[4] <- round(estPar[4], 0)
        names(estPar) <- c("slope", "intercept", "variance", "parameter x")
        srModel[[modelPos]] <- list(modelInfo, estPar, modelName[modelPos])
      }
      
      #Uniform model
      modelPos <- which(modelName == "Uniform")
      if(length(modelPos) == 1){
        srMleCondFinOneM <- srMleCondFinOneL[[modelPos]]
        initialVal <- srMleCondFinOneM[[1]]
        lowerVal <- srMleCondFinOneM[[2]]
        upperVal <- srMleCondFinOneM[[3]]
        srUniformFit <- GenSA(par = initialVal, fn = srUniformLike, srMax = srMax, lower = lowerVal, upper = upperVal)
        logLike <- - srUniformFit[[1]]
        aic <- aicCalc(logLike, length(initialVal))
        modelInfo <- c(logLike, aic)
        names(modelInfo) <- c("likelihood", "AIC")
        estPar <- srUniformFit[[2]]
        names(estPar) <- c("mean", "variance")
        srModel[[modelPos]] <- list(modelInfo, estPar, modelName[modelPos])
      }
      
      names(srModel) <- modelName
      return(srModel)
    }
    
    #Calculate corrected allele numbers
    caParentCalc <- function(minUS, motifLength, seqAl, seqCount, alPar, lusOneL){
      nLus <- length(lusOneL)
      nAlPar <- length(alPar)
      caPar <- rep(0, nAlPar)
      alNames <- as.numeric(names(lusOneL))
      for(i in 1:nLus){
        seqAlPos <- which(seqAl == alNames[i])
        nSeqal <- length(seqAlPos)
        lenCorSumOneAl <- rep(0, nSeqal)
        for(j in 1:nSeqal){
          motifLengthOneAl <- motifLength[[seqAlPos[j]]]
          lenCor <- motifLengthOneAl - minUS
          lenCorSumOneAl[j] <- sum(lenCor[which(lenCor > 0)])
        }
        caParPre <- sum(lenCorSumOneAl * seqCount[seqAlPos] / sum(seqCount[seqAlPos]))
        caPar[which(alPar == alNames[i])] <- caParPre
      }
      return(caPar)
    }
    
    #Estimate parameters
    parEstimate <- function(parLocus, parDye, mt, repLength, 
                            bsrDeal, fsrDeal, dsrDeal, m2srDeal, 
                            aeModelName, hbModelName, bsrModelName, fsrModelName, dsrModelName, m2srModelName, 
                            bsrMultiModelName, fsrMultiModelName, dsrMultiModelName, m2srMultiModelName, 
                            aeMin, hbMin, bsrMax, fsrMax, dsrMax, m2srMax, 
                            aeCandFin, hbCandFin, bsrCandFin, fsrCandFin, dsrCandFin, m2srCandFin, 
                            aeMleCond, hbMleCond, bsrMleCond, fsrMleCond, dsrMleCond, m2srMleCond){
      nL <- length(parLocus)
      
      aeModelNameList <- modelNameListMake(aeModelName, aeCandFin)
      hbModelNameList <- modelNameListMake(hbModelName, hbCandFin)
      bsrModelNameList <- modelNameListMake(bsrModelName, bsrCandFin, bsrDeal, bsrMultiModelName)
      fsrModelNameList <- modelNameListMake(fsrModelName, fsrCandFin, fsrDeal, fsrMultiModelName)
      dsrModelNameList <- modelNameListMake(dsrModelName, dsrCandFin, dsrDeal, dsrMultiModelName)
      m2srModelNameList <- modelNameListMake(m2srModelName, m2srCandFin, m2srDeal, m2srMultiModelName)
      
      aeMleCondFin <- mleCondFinMake(aeMleCond, aeModelNameList)
      hbMleCondFin <- mleCondFinMake(hbMleCond, hbModelNameList)
      bsrMleCondFin <- mleCondFinMake(bsrMleCond, bsrModelNameList)
      fsrMleCondFin <- mleCondFinMake(fsrMleCond, fsrModelNameList)
      dsrMleCondFin <- mleCondFinMake(dsrMleCond, dsrModelNameList)
      m2srMleCondFin <- mleCondFinMake(m2srMleCond, m2srModelNameList)
      
      calData <- read.csv(tclvalue(calibFp), header = TRUE)
      calData <- as.matrix(calData)
      calData <- gsub(" ", "", calData, fixed = TRUE)
      gtAns <- read.csv(tclvalue(gtAnsFp), header = TRUE)
      gtAns <- as.matrix(gtAns)
      gtAns <- gsub(" ", "", gtAns, fixed = TRUE)
      
      srConsider <- matrix(0, nL, 4)
      srConsider[, 1] <- bsrDeal != "Not consider"
      srConsider[, 2] <- fsrDeal != "Not consider"
      srConsider[, 3] <- dsrDeal != "Not consider"
      srConsider[, 4] <- m2srDeal != "Not consider"
      
      srLSp <- matrix(0, nL, 4)
      srLSp[, 1] <- bsrDeal == "Locus specific"
      srLSp[, 2] <- fsrDeal == "Locus specific"
      srLSp[, 3] <- dsrDeal == "Locus specific"
      srLSp[, 4] <- m2srDeal == "Locus specific"
      
      calGtAnsData2 <- calDataArrange(calData, gtAns, parLocus, mt, srConsider)
      calData2 <- calGtAnsData2[[1]]
      gtAns2 <- calGtAnsData2[[2]]
      
      aeDataResult <- aeDataGet(calData2)
      aeData <- aeDataResult[[1]]
      aeHeightData <- aeDataResult[[2]]
      
      hbDataResult <- hbDataGet(calData2, gtAns2, srConsider)
      hbData <- hbDataResult[[1]]
      hbHeightData <- hbDataResult[[2]]
      
      srDataResult <- srDataGet(calData2, gtAns2, srConsider)
      bsrData <- srDataResult[[1]]
      fsrData <- srDataResult[[2]]
      dsrData <- srDataResult[[3]]
      m2srData <- srDataResult[[4]]
      srAlleleData <- srDataResult[[5]]
      srHeightData <- srDataResult[[6]]
      
      seqData <- read.csv(tclvalue(seqFp), header = TRUE)
      seqData <- as.matrix(seqData)
      seqAnalysisResult <- seqAnalysis(seqData, parLocus)
      lusData <- seqAnalysisResult[[1]]
      motifLengthList <- seqAnalysisResult[[2]]
      seqAlList <- seqAnalysisResult[[3]]
      seqCountList <- seqAnalysisResult[[4]]
      rm(calData, calData2, gtAns, gtAns2, calGtAnsData2, aeDataResult, hbDataResult, srDataResult, seqAnalysisResult)
      gc()
      
      cat("Estimation of parameters for locus-specific amplification efficiency was started.", "\n")
      aeUseData <- aeModel <- mapply(rep, 1:nL, 0)
      for(i in 1:nL){
        aePos <- which(aeData[, i] != 0)
        ae <- aeData[aePos, i]
        aeHeight <- aeHeightData[aePos, i]
        aeUseOneL <- cbind(ae, aeHeight)
        colnames(aeUseOneL) <- c("AE", "Height")
        aeUseData[[i]] <- aeUseOneL
        aeModel[[i]] <- aeModeling(aeModelNameList[[i]], ae, aeHeight, aeMin, aeMleCondFin[[i]])
        aeModel[[i]][[length(aeModel[[i]]) + 1]] <- "Locus specific"
      }
      names(aeUseData) <- names(aeModel) <- parLocus
      cat("Estimation of parameters for locus-specific amplification efficiency was finished.", "\n")
      
      cat("Estimation of parameters for heterozygote balance was started.", "\n")
      hbUseData <- hbModel <- mapply(rep, 1:nL, 0)
      for(i in 1:nL){
        hbPos <- which(hbData[, i] != 0)
        hb <- hbData[hbPos, i]
        hbHeight <- hbHeightData[hbPos, i]
        hbUseOneL <- cbind(hb, hbHeight)
        colnames(hbUseOneL) <- c("Hb", "Height")
        hbUseData[[i]] <- hbUseOneL
        hbModel[[i]] <- hbModeling(hbModelNameList[[i]], hb, hbHeight, hbMin, hbMleCondFin[[i]])
        hbModel[[i]][[length(hbModel[[i]]) + 1]] <- "Locus specific"
      }
      names(hbUseData) <- names(hbModel) <- parLocus
      cat("Estimation of parameters for heterozygote balance was finished.", "\n")
      
      cat("Estimation of parameters for back stutter ratio was started.", "\n")
      options(warn = -1)
      bsrGenPos <- which(apply(rbind(srConsider[, 1], !srLSp[, 1]), 2, all) == TRUE)
      options(warn = 0)
      if(length(bsrGenPos) > 0){
        bsrGenData <- srGenMake(bsrGenPos, bsrData, srAlleleData, srHeightData, lusData, motifLengthList, seqAlList, seqCountList, bsrMax)
        bsrGenModel <- srModeling(bsrModelNameList[[bsrGenPos[1]]], bsrGenData[[1]], bsrGenData[[2]], bsrGenData[[3]], bsrGenData[[4]], bsrGenData[[5]], bsrGenData[[6]], bsrGenData[[7]], bsrMax, bsrMleCondFin[[bsrGenPos[1]]])
      }else{
        bsrGenData <- integer(0)
      }
      bsrUseData <- bsrModel <- mapply(rep, 1:nL, 0)
      for(i in 1:nL){
        if(is.element(i, bsrGenPos)){
          bsrUseData[[i]] <- srUseMake(bsrData[, i], srAlleleData[, i], srHeightData[, i], lusData[[i]], bsrMax)
          bsrModel[[i]] <- bsrGenModel
          bsrModel[[i]][[length(bsrModel[[i]]) + 1]] <- "Multiple loci together"
          mulPos <- which(bsrModelNameList[[i]] == "Multi-seq")
          if(length(mulPos) == 1){
            minUS <- bsrModel[[i]][[mulPos]][[2]][4]
            bsrUseData[[i]][[length(bsrUseData[[i]]) + 1]] <- caParentCalc(minUS, motifLengthList[[i]], seqAlList[[i]], seqCountList[[i]], bsrUseData[[i]][[3]], lusData[[i]])
          }else{
            bsrUseData[[i]][[length(bsrUseData[[i]]) + 1]] <- integer(0)
          }
          names(bsrUseData[[i]])[length(bsrUseData[[i]])] <- "CA of parent allele"
        }else if(srConsider[i, 1]){
          bsrUseData[[i]] <- bsrUseOneL <- srUseMake(bsrData[, i], srAlleleData[, i], srHeightData[, i], lusData[[i]], bsrMax)
          bsrModel[[i]] <- srModeling(bsrModelNameList[[i]], bsrUseOneL[[1]], bsrUseOneL[[2]], bsrUseOneL[[3]], bsrUseOneL[[4]], motifLengthList[[i]], seqAlList[[i]], seqCountList[[i]], bsrMax, bsrMleCondFin[[i]])
          bsrModel[[i]][[length(bsrModel[[i]]) + 1]] <- "Locus specific"
          mulPos <- which(bsrModelNameList[[i]] == "Multi-seq")
          if(length(mulPos) == 1){
            minUS <- bsrModel[[i]][[mulPos]][[2]][4]
            bsrUseData[[i]][[length(bsrUseData[[i]]) + 1]] <- caParentCalc(minUS, motifLengthList[[i]], seqAlList[[i]], seqCountList[[i]], bsrUseData[[i]][[3]], lusData[[i]])
          }else{
            bsrUseData[[i]][[length(bsrUseData[[i]]) + 1]] <- integer(0)
          }
          names(bsrUseData[[i]])[length(bsrUseData[[i]])] <- "CA of parent allele"
        }
      }
      names(bsrUseData) <- names(bsrModel) <- parLocus
      cat("Estimation of parameters for back stutter ratio was finished.", "\n")
      
      cat("Estimation of parameters for forward stutter ratio was started.", "\n")
      options(warn = -1)
      fsrGenPos <- which(apply(rbind(srConsider[, 2], !srLSp[, 2]), 2, all) == TRUE)
      options(warn = 0)
      if(length(fsrGenPos) > 0){
        fsrGenData <- srGenMake(fsrGenPos, fsrData, srAlleleData, srHeightData, lusData, motifLengthList, seqAlList, seqCountList, fsrMax)
        fsrGenModel <- srModeling(fsrModelNameList[[fsrGenPos[1]]], fsrGenData[[1]], fsrGenData[[2]], fsrGenData[[3]], fsrGenData[[4]], fsrGenData[[5]], fsrGenData[[6]], fsrGenData[[7]], fsrMax, fsrMleCondFin[[fsrGenPos[1]]])
      }else{
        fsrGenData <- integer(0)
      }
      fsrUseData <- fsrModel <- mapply(rep, 1:nL, 0)
      for(i in 1:nL){
        if(is.element(i, fsrGenPos)){
          fsrUseData[[i]] <- srUseMake(fsrData[, i], srAlleleData[, i], srHeightData[, i], lusData[[i]], fsrMax)
          fsrModel[[i]] <- fsrGenModel
          fsrModel[[i]][[length(fsrModel[[i]]) + 1]] <- "Multiple loci together"
          mulPos <- which(fsrModelNameList[[i]] == "Multi-seq")
          if(length(mulPos) == 1){
            minUS <- fsrModel[[i]][[mulPos]][[2]][4]
            fsrUseData[[i]][[length(fsrUseData[[i]]) + 1]] <- caParentCalc(minUS, motifLengthList[[i]], seqAlList[[i]], seqCountList[[i]], fsrUseData[[i]][[3]], lusData[[i]])
          }else{
            fsrUseData[[i]][[length(fsrUseData[[i]]) + 1]] <- integer(0)
          }
          names(fsrUseData[[i]])[length(fsrUseData[[i]])] <- "CA of parent allele"
        }else if(srConsider[i, 2]){
          fsrUseData[[i]] <- fsrUseOneL <- srUseMake(fsrData[, i], srAlleleData[, i], srHeightData[, i], lusData[[i]], fsrMax)
          fsrModel[[i]] <- srModeling(fsrModelNameList[[i]], fsrUseOneL[[1]], fsrUseOneL[[2]], fsrUseOneL[[3]], fsrUseOneL[[4]], motifLengthList[[i]], seqAlList[[i]], seqCountList[[i]], fsrMax, fsrMleCondFin[[i]])
          fsrModel[[i]][[length(fsrModel[[i]]) + 1]] <- "Locus specific"
          mulPos <- which(fsrModelNameList[[i]] == "Multi-seq")
          if(length(mulPos) == 1){
            minUS <- fsrModel[[i]][[mulPos]][[2]][4]
            fsrUseData[[i]][[length(fsrUseData[[i]]) + 1]] <- caParentCalc(minUS, motifLengthList[[i]], seqAlList[[i]], seqCountList[[i]], fsrUseData[[i]][[3]], lusData[[i]])
          }else{
            fsrUseData[[i]][[length(fsrUseData[[i]]) + 1]] <- integer(0)
          }
          names(fsrUseData[[i]])[length(fsrUseData[[i]])] <- "CA of parent allele"
        }
      }
      names(fsrUseData) <- names(fsrModel) <- parLocus
      cat("Estimation of parameters for forward stutter ratio was finished.", "\n")
      
      cat("Estimation of parameters for double-back stutter ratio was started.", "\n")
      options(warn = -1)
      dsrGenPos <- which(apply(rbind(srConsider[, 3], !srLSp[, 3]), 2, all) == TRUE)
      options(warn = 0)
      if(length(dsrGenPos) > 0){
        dsrGenData <- srGenMake(dsrGenPos, dsrData, srAlleleData, srHeightData, lusData, motifLengthList, seqAlList, seqCountList, dsrMax)
        dsrGenModel <- srModeling(dsrModelNameList[[dsrGenPos[1]]], dsrGenData[[1]], dsrGenData[[2]], dsrGenData[[3]], dsrGenData[[4]], dsrGenData[[5]], dsrGenData[[6]], dsrGenData[[7]], dsrMax, dsrMleCondFin[[dsrGenPos[1]]])
      }else{
        dsrGenData <- integer(0)
      }
      dsrUseData <- dsrModel <- mapply(rep, 1:nL, 0)
      for(i in 1:nL){
        if(is.element(i, dsrGenPos)){
          dsrUseData[[i]] <- srUseMake(dsrData[, i], srAlleleData[, i], srHeightData[, i], lusData[[i]], dsrMax)
          dsrModel[[i]] <- dsrGenModel
          dsrModel[[i]][[length(dsrModel[[i]]) + 1]] <- "Multiple loci together"
          mulPos <- which(dsrModelNameList[[i]] == "Multi-seq")
          if(length(mulPos) == 1){
            minUS <- dsrModel[[i]][[mulPos]][[2]][4]
            dsrUseData[[i]][[length(dsrUseData[[i]]) + 1]] <- caParentCalc(minUS, motifLengthList[[i]], seqAlList[[i]], seqCountList[[i]], dsrUseData[[i]][[3]], lusData[[i]])
          }else{
            dsrUseData[[i]][[length(dsrUseData[[i]]) + 1]] <- integer(0)
          }
          names(dsrUseData[[i]])[length(dsrUseData[[i]])] <- "CA of parent allele"
        }else if(srConsider[i, 3]){
          dsrUseData[[i]] <- dsrUseOneL <- srUseMake(dsrData[, i], srAlleleData[, i], srHeightData[, i], lusData[[i]], dsrMax)
          dsrModel[[i]] <- srModeling(dsrModelNameList[[i]], dsrUseOneL[[1]], dsrUseOneL[[2]], dsrUseOneL[[3]], dsrUseOneL[[4]], motifLengthList[[i]], seqAlList[[i]], seqCountList[[i]], dsrMax, dsrMleCondFin[[i]])
          dsrModel[[i]][[length(dsrModel[[i]]) + 1]] <- "Locus specific"
          mulPos <- which(dsrModelNameList[[i]] == "Multi-seq")
          if(length(mulPos) == 1){
            minUS <- dsrModel[[i]][[mulPos]][[2]][4]
            dsrUseData[[i]][[length(dsrUseData[[i]]) + 1]] <- caParentCalc(minUS, motifLengthList[[i]], seqAlList[[i]], seqCountList[[i]], dsrUseData[[i]][[3]], lusData[[i]])
          }else{
            dsrUseData[[i]][[length(dsrUseData[[i]]) + 1]] <- integer(0)
          }
          names(dsrUseData[[i]])[length(dsrUseData[[i]])] <- "CA of parent allele"
        }
      }
      names(dsrUseData) <- names(dsrModel) <- parLocus
      cat("Estimation of parameters for double-back stutter ratio was finished.", "\n")
      
      cat("Estimation of parameters for minus 2-nt stutter ratio was started.", "\n")
      options(warn = -1)
      m2srGenPos <- which(apply(rbind(srConsider[, 4], !srLSp[, 4]), 2, all) == TRUE)
      options(warn = 0)
      if(length(m2srGenPos) > 0){
        m2srGenData <- srGenMake(m2srGenPos, m2srData, srAlleleData, srHeightData, lusData, motifLengthList, seqAlList, seqCountList, m2srMax)
        m2srGenModel <- srModeling(m2srModelNameList[[m2srGenPos[1]]], m2srGenData[[1]], m2srGenData[[2]], m2srGenData[[3]], m2srGenData[[4]], m2srGenData[[5]], m2srGenData[[6]], m2srGenData[[7]], m2srMax, m2srMleCondFin[[m2srGenPos[1]]])
      }else{
        m2srGenData <- integer(0)
      }
      m2srUseData <- m2srModel <- mapply(rep, 1:nL, 0)
      for(i in 1:nL){
        if(is.element(i, m2srGenPos)){
          m2srUseData[[i]] <- srUseMake(m2srData[, i], srAlleleData[, i], srHeightData[, i], lusData[[i]], m2srMax)
          m2srModel[[i]] <- m2srGenModel
          m2srModel[[i]][[length(m2srModel[[i]]) + 1]] <- "Multiple loci together"
          m2srUseData[[i]][[length(m2srUseData[[i]]) + 1]] <- integer(0)
          names(m2srUseData[[i]])[length(m2srUseData[[i]])] <- "CA of parent allele"
        }else if(srConsider[i, 4]){
          m2srUseData[[i]] <- m2srUseOneL <- srUseMake(m2srData[, i], srAlleleData[, i], srHeightData[, i], lusData[[i]], m2srMax)
          m2srModel[[i]] <- srModeling(m2srModelNameList[[i]], m2srUseOneL[[1]], m2srUseOneL[[2]], m2srUseOneL[[3]], m2srUseOneL[[4]], motifLengthList[[i]], seqAlList[[i]], seqCountList[[i]], m2srMax, m2srMleCondFin[[i]])
          m2srModel[[i]][[length(m2srModel[[i]]) + 1]] <- "Locus specific"
          m2srUseData[[i]][[length(m2srUseData[[i]]) + 1]] <- integer(0)
          names(m2srUseData[[i]])[length(m2srUseData[[i]])] <- "CA of parent allele"
        }
      }
      names(m2srUseData) <- names(m2srModel) <- parLocus
      cat("Estimation of parameters for minus 2-nt stutter ratio was finished.", "\n")
      cat("Estimation of all parameters was finished.", "\n")
      
      tclvalue(paramEstFin) <- "1"
      tab3ParMake(aeUseData, aeModel, hbUseData, hbModel, 
                  bsrGenData, bsrUseData, bsrModel, 
                  fsrGenData, fsrUseData, fsrModel, 
                  dsrGenData, dsrUseData, dsrModel, 
                  m2srGenData, m2srUseData, m2srModel, 
                  aeMin, hbMin, bsrMax, fsrMax, dsrMax, m2srMax, 
                  aeMleCondFin, hbMleCondFin, bsrMleCondFin, fsrMleCondFin, dsrMleCondFin, m2srMleCondFin, 
                  lusData, motifLengthList, seqAlList, seqCountList, 
                  srConsider, srLSp, parDye, repLength)
      tk2notetab.select(tabsPar, "Result")
    }
    
    
    ##################################
    # After estimation of parameters #
    ##################################
    
    #Choose the best model
    chooseAdoptModel <- function(parModel, method, posMOneL){
      nM <- length(parModel) - 1
      if(nM > 1){
        if(posMOneL == 0){
          modelInfo <- rep(0, nM)
          infoPos <- which(names(parModel[[1]][[1]]) == method)
          if(length(infoPos) == 0){
            stop(message = "Inappropriate method was selected!") 
          }else{
            for(i in 1:nM){
              modelInfo[i] <- parModel[[i]][[1]][infoPos]
            }
            if(method == "likelihood"){
              bestModel <- parModel[[which(modelInfo == max(modelInfo))]]
            }else if(method == "AIC"){
              bestModel <- parModel[[which(modelInfo == min(modelInfo))]]
            }
          }
        }else{
          bestModel <- parModel[[posMOneL]]
          bestModel[[3]] <- paste(bestModel[[3]], " \u2020", sep = "")
        }
      }else if(nM == 1){
        bestModel <- parModel[[1]]
      }else{
        bestModel <- integer(0)
      }
      return(bestModel)
    }
    
    #Output parameters
    parOutput <- function(aeBestList, hbBestList, bsrBestList, fsrBestList, dsrBestList, m2srBestList, srConsider, srLSp, parDye, repLength){
      saveAs <- tkgetSaveFile(filetypes = "{{CSV Files} {.csv}}")
      if(tclvalue(saveAs) != ""){
        if(substr(tclvalue(saveAs), nchar(tclvalue(saveAs)) - 3, nchar(tclvalue(saveAs))) == ".csv"){
          reportName <- tclvalue(saveAs)
        }else{
          reportName <- paste(tclvalue(saveAs), ".csv", sep = "")
        }
        parLocus <- names(aeBestList)
        nL <- length(parLocus)
        mcParFormat <- matrix("", 6 * nL, 10)
        colnames(mcParFormat) <- c("Factor", "Marker", "Repeat_length", "Dye", "Model", "Locus_specific", "Parameter1", "Parameter2", "Parameter3", "Parameter4")
        mcParFormat[1:nL, 1] <- "AE"
        mcParFormat[(nL + 1):(2 * nL), 1] <- "Hb"
        mcParFormat[(2 * nL + 1):(3 * nL), 1] <- "BSR"
        mcParFormat[(3 * nL + 1):(4 * nL), 1] <- "FSR"
        mcParFormat[(4 * nL + 1):(5 * nL), 1] <- "DSR"
        mcParFormat[(5 * nL + 1):(6 * nL), 1] <- "M2SR"
        mcParFormat[, 2] <- rep(parLocus, 6)
        mcParFormat[, 3] <- rep(repLength, 6)
        mcParFormat[, 4] <- rep(parDye, 6)
        for(i in 1:nL){
          aeBestModel <- aeBestList[[i]]
          hbBestModel <- hbBestList[[i]]
          bsrBestModel <- bsrBestList[[i]]
          fsrBestModel <- fsrBestList[[i]]
          dsrBestModel <- dsrBestList[[i]]
          m2srBestModel <- m2srBestList[[i]]
          
          if(length(grep(" \u2020", aeBestModel[[3]])) == 0){
            mcParFormat[i, 5] <- aeBestModel[[3]]
          }else{
            mcParFormat[i, 5] <- gsub(" \u2020", "", aeBestModel[[3]])
          }
          mcParFormat[i, 6] <- TRUE
          mcParFormat[i, 7:(6 + length(aeBestModel[[2]]))] <- aeBestModel[[2]]
          
          if(length(grep(" \u2020", hbBestModel[[3]])) == 0){
            mcParFormat[nL + i, 5] <- hbBestModel[[3]]
          }else{
            mcParFormat[nL + i, 5] <- gsub(" \u2020", "", hbBestModel[[3]])
          }
          mcParFormat[nL + i, 6] <- TRUE
          mcParFormat[nL + i, 7:(6 + length(hbBestModel[[2]]))] <- hbBestModel[[2]]
          
          if(srConsider[i, 1] == 1){
            if(length(grep(" \u2020", bsrBestModel[[3]])) == 0){
              mcParFormat[2 * nL + i, 5] <- bsrBestModel[[3]]
            }else{
              mcParFormat[2 * nL + i, 5] <- gsub(" \u2020", "", bsrBestModel[[3]])
            }
            mcParFormat[2 * nL + i, 6] <- srLSp[i, 1] == 1
            mcParFormat[2 * nL + i, 7:(6 + length(bsrBestModel[[2]]))] <- bsrBestModel[[2]]
          }else{
            mcParFormat[2 * nL + i, 5] <- "No"
          }
          if(srConsider[i, 2] == 1){
            if(length(grep(" \u2020", fsrBestModel[[3]])) == 0){
              mcParFormat[3 * nL + i, 5] <- fsrBestModel[[3]]
            }else{
              mcParFormat[3 * nL + i, 5] <- gsub(" \u2020", "", fsrBestModel[[3]])
            }
            mcParFormat[3 * nL + i, 6] <- srLSp[i, 2] == 1
            mcParFormat[3 * nL + i, 7:(6 + length(fsrBestModel[[2]]))] <- fsrBestModel[[2]]
          }else{
            mcParFormat[3 * nL + i, 5] <- "No"
          }
          if(srConsider[i, 3] == 1){
            if(length(grep(" \u2020", dsrBestModel[[3]])) == 0){
              mcParFormat[4 * nL + i, 5] <- dsrBestModel[[3]]
            }else{
              mcParFormat[4 * nL + i, 5] <- gsub(" \u2020", "", dsrBestModel[[3]])
            }
            mcParFormat[4 * nL + i, 6] <- srLSp[i, 3] == 1
            mcParFormat[4 * nL + i, 7:(6 + length(dsrBestModel[[2]]))] <- dsrBestModel[[2]]
          }else{
            mcParFormat[4 * nL + i, 5] <- "No"
          }
          if(srConsider[i, 4] == 1){
            if(length(grep(" \u2020", m2srBestModel[[3]])) == 0){
              mcParFormat[5 * nL + i, 5] <- m2srBestModel[[3]]
            }else{
              mcParFormat[5 * nL + i, 5] <- gsub(" \u2020", "", m2srBestModel[[3]])
            }
            mcParFormat[5 * nL + i, 6] <- srLSp[i, 4] == 1
            mcParFormat[5 * nL + i, 7:(6 + length(m2srBestModel[[2]]))] <- m2srBestModel[[2]]
          }else{
            mcParFormat[5 * nL + i, 5] <- "No"
          }
        }
        write.csv(mcParFormat, reportName, row.names = FALSE, col.names = TRUE)
      }
    }
    
    #Output information of allele repeat correction
    alCorOutput <- function(lusData, motifLengthList, seqAlList, seqCountList, bsrBestList, fsrBestList, dsrBestList){
      #Calculate weighted average values of corrected allele numbers
      wCACalc <- function(parX, motifLength, seqAl, seqCount){
        uniqAl <- sort(unique(seqAl))
        nU <- length(uniqAl)
        wCA <- rep(0, nU)
        countML <- 0
        for(i in 1:nU){
          alPos <- which(seqAl == uniqAl[i])
          seqCountOneA <- seqCount[alPos]
          nCA <- length(alPos)
          CAs <- rep(0, nCA)
          for(j in 1:nCA){
            countML <- countML + 1
            caPre <- motifLength[[countML]] - parX
            caPre[which(caPre < 0)] <- 0
            CAs[j] <- sum(caPre)
          }
          wCA[i] <- sum(seqCountOneA * CAs) / sum(seqCountOneA)
        }
        return(wCA)
      }
      
      saveAs <- tkgetSaveFile(filetypes = "{{CSV Files} {.csv}}")
      if(tclvalue(saveAs) != ""){
        if(substr(tclvalue(saveAs), nchar(tclvalue(saveAs)) - 3, nchar(tclvalue(saveAs))) == ".csv"){
          alCorName <- tclvalue(saveAs)
        }else{
          alCorName <- paste(tclvalue(saveAs), ".csv", sep = "")
        }
        parLocus <- names(bsrBestList)
        nL <- length(parLocus)
        nAl <- sapply(lusData, length)
        alCorData <- matrix("", sum(nAl), 6)
        colnames(alCorData) <- c("Marker", "Allele", "LUS", "CA_BSR", "CA_FSR", "CA_DSR")
        count <- 0
        for(i in 1:nL){
          rowPos <- (count + 1):(count + nAl[i])
          count <- count + nAl[i]
          lusOneL <- lusData[[i]]
          bsrBestOneL <- bsrBestList[[i]]
          fsrBestOneL <- fsrBestList[[i]]
          dsrBestOneL <- dsrBestList[[i]]
          motifLength <- motifLengthList[[i]]
          seqAl <- seqAlList[[i]] 
          seqCount <- seqCountList[[i]]
          
          alCorData[rowPos, 1] <- parLocus[i]
          alCorData[rowPos, 2] <- names(lusOneL)
          alCorData[rowPos, 3] <- lusOneL
          
          if(length(bsrBestOneL) != 0){
            if(length(grep(" \u2020", bsrBestOneL[[3]])) == 0){
              bsrBestName <- bsrBestOneL[[3]]
            }else{
              bsrBestName <- gsub(" \u2020", "", bsrBestOneL[[3]])
            }
            if(bsrBestName == "Multi-seq"){
              parX <- bsrBestOneL[[2]][4]
              nSeq <- length(motifLength)
              caOneL <- rep(0, nSeq)
              alCorData[rowPos, 4] <- wCACalc(parX, motifLength, seqAl, seqCount)
            }
          }

          if(length(fsrBestOneL) != 0){
            if(length(grep(" \u2020", fsrBestOneL[[3]])) == 0){
              fsrBestName <- fsrBestOneL[[3]]
            }else{
              fsrBestName <- gsub(" \u2020", "", fsrBestOneL[[3]])
            }
            if(fsrBestName == "Multi-seq"){
              parX <- fsrBestOneL[[2]][4]
              nSeq <- length(motifLength)
              caOneL <- rep(0, nSeq)
              alCorData[rowPos, 5] <- wCACalc(parX, motifLength, seqAl, seqCount)
            }
          }

          if(length(dsrBestOneL) != 0){
            if(length(grep(" \u2020", dsrBestOneL[[3]])) == 0){
              dsrBestName <- dsrBestOneL[[3]]
            }else{
              dsrBestName <- gsub(" \u2020", "", dsrBestOneL[[3]])
            }
            if(dsrBestName == "Multi-seq"){
              parX <- dsrBestOneL[[2]][4]
              nSeq <- length(motifLength)
              caOneL <- rep(0, nSeq)
              alCorData[rowPos, 6] <- wCACalc(parX, motifLength, seqAl, seqCount)
            }
          }
        }
        write.csv(alCorData, alCorName, row.names = FALSE, col.names = TRUE)
      }
    }
    
    #Delete current results
    delResult <- function(parLocus, parDye){
      tkdestroy(fTab2Par)
      fTab2Par <<- tkframe(tab2Par)
      tkgrid(tkbutton(fTab2Par, text = "    Delete current results    ", cursor = "hand2", command = function() tab2ParMake(parLocus, parDye)), padx = 5, pady = 5)
      tkgrid(fTab2Par)
    }
    
    #Make the 'Result' tab
    tab3ParMake <- function(aeUseData = NULL, aeModel = NULL, hbUseData = NULL, hbModel = NULL, 
                            bsrGenData = NULL, bsrUseData = NULL, bsrModel = NULL, 
                            fsrGenData = NULL, fsrUseData = NULL, fsrModel = NULL, 
                            dsrGenData = NULL, dsrUseData = NULL, dsrModel = NULL, 
                            m2srGenData = NULL, m2srUseData = NULL, m2srModel = NULL, 
                            aeMin = NULL, hbMin = NULL, bsrMax = NULL, fsrMax = NULL, dsrMax = NULL, m2srMax = NULL, 
                            aeMleCondFin = NULL, hbMleCondFin = NULL, bsrMleCondFin = NULL, fsrMleCondFin = NULL, dsrMleCondFin = NULL, m2srMleCondFin = NULL, 
                            lusData = NULL, motifLengthList = NULL, seqAlList = NULL, seqCountList = NULL, 
                            srConsider = NULL, srLSp = NULL, parDye = NULL, repLength = NULL
    ){
      tkdestroy(fTab3Par)
      fTab3Par <<- tkframe(tab3Par)
      if(tclvalue(calibName) == ""){
        tkgrid(tklabel(fTab3Par, text = "        Load experimental data!        "), pady = 10, sticky = "w")
        tkgrid(fTab3Par)
      }
      if(tclvalue(gtAnsName) == ""){
        tkgrid(tklabel(fTab3Par, text = "        Load actual genotypes!        "), pady = 10, sticky = "w")
        tkgrid(fTab3Par)
      }
      if(tclvalue(seqName) == ""){
        tkgrid(tklabel(fTab3Par, text = "        Load sequence data!        "), pady = 10, sticky = "w")
        tkgrid(fTab3Par)
      }
      if(all(c(tclvalue(calibName), tclvalue(gtAnsName), tclvalue(seqName)) != "")){
        if(tclvalue(fileCk2Fin) == "0"){
          tkgrid(tklabel(fTab3Par, text = "        Push 'Next' button in the Files tab.        "), pady = 10, sticky = "w")
          tkgrid(fTab3Par)
        }else if(tclvalue(paramEstFin) == "0"){
          tkgrid(tklabel(fTab3Par, text = "        Estimate parameters!        "), pady = 10, sticky = "w")
          tkgrid(fTab3Par)
        }else{
          #Detail of results
          resultDetail <- function(parModel, parUseData, factName, mleCondFin, minMax){
            #Make a graph of the variance of AE or Hb
            graphVar <- function(parModelOneL, parUseOneL, factName, factMin){
              estPar <- parModelOneL[[1]][[2]]
              if(factName == "AE"){
                yLab <- "locus-specific amplification efficiency"
              }else if(factName == "Hb"){
                yLab <- "heterozygote balance"
              }
              h <- parUseOneL[, 2]
              f <- parUseOneL[, 1]
              xMin <- 0
              xMax <- round(max(h) + 1, 0)
              yMin <- min(f)
              yMax <- max(f)
              plot(h, f, xlim = c(xMin, xMax), ylim = c(yMin, yMax), xlab = "Average heterozygous peak height (RFU)", ylab = yLab)
              myu <- estPar[1]
              sigma <- estPar[2]
              Q0.95 <- qtruncnorm(0.95, a = log(factMin), b = log(1 / factMin), mean = myu, sd = sqrt(sigma / (100:xMax)))
              Q0.05 <- qtruncnorm(0.05, a = log(factMin), b = log(1 / factMin), mean = myu, sd = sqrt(sigma / (100:xMax)))
              par(new = TRUE)
              plot(100:xMax, exp(Q0.95), type = "l", xlim = c(xMin, xMax), ylim = c(yMin, yMax), 
                   xlab = "", ylab = "", axes = FALSE)
              par(new = TRUE)
              plot(100:xMax, exp(Q0.05), type = "l", xlim = c(xMin, xMax), ylim = c(yMin, yMax), 
                   xlab = "", ylab = "", axes = FALSE)
            }
            
            #Make a graph of the regression line of SR
            srGraphMean <- function(parModelOneL, parUseOneL, nameModel, selectModel, factName){
              modelPos <- which(nameModel == selectModel)
              estPar <- parModelOneL[[modelPos]][[2]]
              if(factName == "BSR"){
                yLab = "back stutter ratio"
              }else if(factName == "FSR"){
                yLab = "forward stutter ratio"
              }else if(factName == "DSR"){
                yLab = "double-back stutter ratio"
              }else if(factName == "M2SR"){
                yLab = "minus 2-nt stutter ratio"
              }
              if(selectModel == "Allele"){
                al <- parUseOneL[[3]]
                sr <- parUseOneL[[1]]
                xMin <- min(al)
                xMax <- max(al)
                yMin <- min(sr)
                yMax <- max(sr)
                plot(al, sr, xlim = c(xMin, xMax), ylim = c(yMin, yMax), 
                     xlab = "Allele", ylab = yLab)
                slope <- estPar[1]
                intercept <- estPar[2]
                lineX <- seq(round(xMin - 1, 1), round(xMax + 1, 1), 0.1)
                lineY <- slope * lineX + intercept
                par(new = TRUE)
                plot(lineX, lineY, type = "l", xlim = c(xMin, xMax), ylim = c(yMin, yMax), 
                     xlab = "", ylab = "", axes = FALSE)
              }else if(selectModel == "LUS"){
                lus <- parUseOneL[[4]]
                sr <- parUseOneL[[1]]
                xMin <- min(lus)
                xMax <- max(lus)
                yMin <- min(sr)
                yMax <- max(sr)
                plot(lus, sr, xlim = c(xMin, xMax), ylim = c(yMin, yMax), 
                     xlab = "LUS", ylab = yLab)
                slope <- estPar[1]
                intercept <- estPar[2]
                lineX <- seq(round(xMin - 1, 1), round(xMax + 1, 1), 0.1)
                lineY <- slope * lineX + intercept
                par(new = TRUE)
                plot(lineX, lineY, type = "l", xlim = c(xMin, xMax), ylim = c(yMin, yMax), 
                     xlab = "", ylab = "", axes = FALSE)
              }else if(selectModel == "Multi-seq"){
                ca <- parUseOneL[[5]]
                sr <- parUseOneL[[1]]
                xMin <- min(ca)
                xMax <- max(ca)
                yMin <- min(sr)
                yMax <- max(sr)
                plot(ca, sr, xlim = c(xMin, xMax), ylim = c(yMin, yMax), 
                     xlab = "Corrected allele number", ylab = yLab)
                slope <- estPar[1]
                intercept <- estPar[2]
                lineX <- seq(round(xMin - 1, 1), round(xMax + 1, 1), 0.1)
                lineY <- slope * lineX + intercept
                par(new = TRUE)
                plot(lineX, lineY, type = "l", xlim = c(xMin, xMax), ylim = c(yMin, yMax), 
                     xlab = "", ylab = "", axes = FALSE)
              }else if(selectModel == "Uniform"){
                plot(parUseOneL[[3]], parUseOneL[[1]], xlab = "Allele", ylab = yLab)
                abline(h = estPar[1])
              }
            }
            
            #Make a graph of the variance of SR
            srGraphVar <- function(parModelOneL, parUseOneL, nameModel, selectModel, factName, srMax){
              modelPos <- which(nameModel == selectModel)
              sr <- parUseOneL[[1]]
              h <- parUseOneL[[2]]
              estPar <- parModelOneL[[modelPos]][[2]]
              if(factName == "BSR"){
                yLab = "back stutter ratio"
              }else if(factName == "FSR"){
                yLab = "forward stutter ratio"
              }else if(factName == "DSR"){
                yLab = "double-back stutter ratio"
              }else if(factName == "M2SR"){
                yLab = "minus 2-nt stutter ratio"
              }
              if(selectModel != "Uniform"){
                if(selectModel == "Allele"){
                  al <- parUseOneL[[3]]
                }else if(selectModel == "LUS"){
                  al <- parUseOneL[[4]]
                }else if(selectModel == "Multi-seq"){
                  al <- parUseOneL[[5]]
                }
                meanAl <- mean(al)
                slope <- estPar[1]
                intercept <- estPar[2]
                sigma <- estPar[3]
                myu <- slope * meanAl + intercept
                srMean <- slope * al + intercept
                srCor <- srMean - myu
                sr <- sr - srCor
                yLab <- paste(yLab, " (corrected)", sep = "")
              }else{
                myu <- estPar[1]
                sigma <- estPar[2]
              }
              xMin <- 0
              xMax <- round(max(h) + 1, 0)
              yMin <- min(sr)
              yMax <- max(sr)
              plot(h, sr, xlim = c(xMin, xMax), ylim = c(yMin, yMax), xlab = "Total allelic product (RFU)", ylab = yLab)
              Q0.95 <- qtruncnorm(0.95, b = log(srMax), mean = log(myu), sd = sqrt(sigma / (100:xMax)))
              Q0.05 <- qtruncnorm(0.05, b = log(srMax), mean = log(myu), sd = sqrt(sigma / (100:xMax)))
              par(new = TRUE)
              plot(100:xMax, exp(Q0.95), type = "l", xlim = c(xMin, xMax), ylim = c(yMin, yMax), 
                   xlab = "", ylab = "", axes = FALSE)
              par(new = TRUE)
              plot(100:xMax, exp(Q0.05), type = "l", xlim = c(xMin, xMax), ylim = c(yMin, yMax), 
                   xlab = "", ylab = "", axes = FALSE)
            }
            
            #Change the adopted model
            modelChange <- function(nameModel, posL, factName){
              #Adopt the selected model
              adoptModel <- function(selectModel){
                modelPos <- which(nameModel == selectModel)
                changeOk <- TRUE
                if(factName == "AE"){
                  aeAdoptModelPos[posL] <<- modelPos
                }else if(factName == "Hb"){
                  hbAdoptModelPos[posL] <<- modelPos
                }else if(factName == "BSR"){
                  if(srLSp[posL, 1] == 1){
                    bsrAdoptModelPos[posL] <<- modelPos
                  }else{
                    imputOk <- tclvalue(tkmessageBox(message = paste(factName, "of the selected locus was modeled considering multiple loci together. The selected model will be also reflected in these loci. Do you want to continue?", sep = ""), type = "okcancel", icon = "warning"))
                    if(imputOk == "ok"){
                      bsrAdoptModelPos[srLSp[, 1] != 1] <<- modelPos
                    }else{
                      changeOk <- FALSE
                    }
                  }
                }else if(factName == "FSR"){
                  if(srLSp[posL, 2] == 1){
                    fsrAdoptModelPos[posL] <<- modelPos
                  }else{
                    imputOk <- tclvalue(tkmessageBox(message = paste(factName, "of the selected locus was modeled considering multiple loci together. The selected model will be also reflected in these loci. Do you want to continue?", sep = ""), type = "okcancel", icon = "warning"))
                    if(imputOk == "ok"){
                      fsrAdoptModelPos[srLSp[, 2] != 1] <<- modelPos
                    }else{
                      changeOk <- FALSE
                    }
                  }
                }else if(factName == "DSR"){
                  if(srLSp[posL, 3] == 1){
                    dsrAdoptModelPos[posL] <<- modelPos
                  }else{
                    imputOk <- tclvalue(tkmessageBox(message = paste(factName, "of the selected locus was modeled considering multiple loci together. The selected model will be also reflected in these loci. Do you want to continue?", sep = ""), type = "okcancel", icon = "warning"))
                    if(imputOk == "ok"){
                      dsrAdoptModelPos[srLSp[, 3] != 1] <<- modelPos
                    }else{
                      changeOk <- FALSE
                    }
                  }
                }else if(factName == "M2SR"){
                  if(srLSp[posL, 3] == 1){
                    m2srAdoptModelPos[posL] <<- modelPos
                  }else{
                    imputOk <- tclvalue(tkmessageBox(message = paste(factName, "of the selected locus was modeled considering multiple loci together. The selected model will be also reflected in these loci. Do you want to continue?", sep = ""), type = "okcancel", icon = "warning"))
                    if(imputOk == "ok"){
                      m2srAdoptModelPos[srLSp[, 4] != 1] <<- modelPos
                    }else{
                      changeOk <- FALSE
                    }
                  }
                }
                if(changeOk){
                  tkdestroy(tfChange)
                  tkdestroy(tfParResult)
                  summaryMake()
                }
              }
              
              tfChange <- tktoplevel()
              tkwm.title(tfChange, "Change the adopted model")
              tkgrid(tklabel(tfChange, text = "Select a model : "), row = 0, column = 0, padx = 5, pady = 5)
              selectM <- tclVar(nameModel[1])
              tkgrid(ttkcombobox(tfChange, values = nameModel, textvariable = selectM, state = "readonly"), row = 0, column = 1, padx = 5, pady = 5)
              tkgrid(tkbutton(tfChange, text = "    Adopt    ", cursor = "hand2", command = function() adoptModel(tclvalue(selectM))), row = 0, column = 2, padx = 5, pady = 5)
            }
            
            #Reestimate parameters
            reestParam <- function(parUseOneL, nameModel, posL, factName, minMax, mleCondOneL){
              reestRun <- function(mleCOneM, posM, selectModel){
                tkdestroy(tfReest)
                tkdestroy(tfParResult)
                tkdestroy(fTab3Par)
                fTab3Par <<- tkframe(tab3Par)
                tkgrid(tklabel(fTab3Par, text = "        Wait!        "), pady = 10, sticky = "w")
                tkgrid(fTab3Par)
                if(factName == "AE"){
                  cat("Reestimation of parameters was started.", "\n")
                  aeMleCondFin[[posL]][[posM]] <<- mleCondFinMake(list(mleCOneM), list(selectModel))[[1]][[1]]
                  aeModel[[posL]][[posM]] <<- aeModeling(selectModel, parUseOneL[, 1], parUseOneL[, 2], minMax, list(aeMleCondFin[[posL]][[posM]]))[[1]]
                }else if(factName == "Hb"){
                  cat("Reestimation of parameters was started.", "\n")
                  hbMleCondFin[[posL]][[posM]] <<- mleCondFinMake(list(mleCOneM), list(selectModel))[[1]][[1]]
                  hbModel[[posL]][[posM]] <<- hbModeling(selectModel, parUseOneL[, 1], parUseOneL[, 2], minMax, list(hbMleCondFin[[posL]][[posM]]))[[1]]
                }else if(factName == "BSR"){
                  bsrMleCondFin[[posL]][[posM]] <<- mleCondFinMake(list(mleCOneM), list(selectModel))[[1]][[1]]
                  if(srLSp[posL, 1] == 0){
                    imputOk <- tclvalue(tkmessageBox(message = paste(factName, "of the selected locus was modeled considering multiple loci together. Re-estimation of parameters will be also performed by considering multiple loci together. Do you want to continue?", sep = ""), type = "okcancel", icon = "warning"))
                    if(imputOk == "ok"){
                      cat("Reestimation of parameters was started.", "\n")
                      options(warn = -1)
                      bsrGenPos <- which(apply(rbind(srConsider[, 1], !srLSp[, 1]), 2, all) == TRUE)
                      options(warn = 0)
                      bsrGenModel <- srModeling(selectModel, bsrGenData[[1]], bsrGenData[[2]], bsrGenData[[3]], bsrGenData[[4]], bsrGenData[[5]], bsrGenData[[6]], bsrGenData[[7]], minMax, bsrMleCondFin[[posL]][[posM]])
                      for(i in 1:length(bsrGenPos)){
                        bsrModel[[bsrGenPos[i]]][[posM]] <<- bsrGenModel[[1]]
                      }
                    }
                  }else{
                    cat("Reestimation of parameters was started.", "\n")
                    bsrModel[[posL]][[posM]] <<- srModeling(selectModel, parUseOneL[[1]], parUseOneL[[2]], parUseOneL[[3]], parUseOneL[[4]], motifLengthList[[posL]], seqAlList[[posL]], seqCountList[[posL]], minMax, list(bsrMleCondFin[[posL]][[posM]]))[[1]]
                  }
                }else if(factName == "FSR"){
                  fsrMleCondFin[[posL]][[posM]] <<- mleCondFinMake(list(mleCOneM), list(selectModel))[[1]][[1]]
                  if(srLSp[posL, 2] == 0){
                    imputOk <- tclvalue(tkmessageBox(message = paste(factName, "of the selected locus was modeled considering multiple loci together. Re-estimation of parameters will be also performed by considering multiple loci together. Do you want to continue?", sep = ""), type = "okcancel", icon = "warning"))
                    if(imputOk == "ok"){
                      cat("Reestimation of parameters was started.", "\n")
                      options(warn = -1)
                      fsrGenPos <- which(apply(rbind(srConsider[, 2], !srLSp[, 2]), 2, all) == TRUE)
                      options(warn = 0)
                      fsrGenModel <- srModeling(selectModel, fsrGenData[[1]], fsrGenData[[2]], fsrGenData[[3]], fsrGenData[[4]], fsrGenData[[5]], fsrGenData[[6]], fsrGenData[[7]], minMax, list(fsrMleCondFin[[posL]][[posM]]))
                      for(i in 1:length(fsrGenPos)){
                        fsrModel[[fsrGenPos[i]]][[posM]] <<- fsrGenModel[[1]]
                      }
                    }
                  }else{
                    cat("Reestimation of parameters was started.", "\n")
                    fsrModel[[posL]][[posM]] <<- srModeling(selectModel, parUseOneL[[1]], parUseOneL[[2]], parUseOneL[[3]], parUseOneL[[4]], motifLengthList[[posL]], seqAlList[[posL]], seqCountList[[posL]], minMax, list(fsrMleCondFin[[posL]][[posM]]))[[1]]
                  }
                }else if(factName == "DSR"){
                  dsrMleCondFin[[posL]][[posM]] <<- mleCondFinMake(list(mleCOneM), list(selectModel))[[1]][[1]]
                  if(srLSp[posL, 3] == 0){
                    imputOk <- tclvalue(tkmessageBox(message = paste(factName, "of the selected locus was modeled considering multiple loci together. Re-estimation of parameters will be also performed by considering multiple loci together. Do you want to continue?", sep = ""), type = "okcancel", icon = "warning"))
                    if(imputOk == "ok"){
                      cat("Reestimation of parameters was started.", "\n")
                      options(warn = -1)
                      dsrGenPos <- which(apply(rbind(srConsider[, 3], !srLSp[, 3]), 2, all) == TRUE)
                      options(warn = 0)
                      dsrGenModel <- srModeling(selectModel, dsrGenData[[1]], dsrGenData[[2]], dsrGenData[[3]], dsrGenData[[4]], dsrGenData[[5]], dsrGenData[[6]], dsrGenData[[7]], minMax, list(dsrMleCondFin[[posL]][[posM]]))
                      for(i in 1:length(dsrGenPos)){
                        dsrModel[[dsrGenPos[i]]][[posM]] <<- dsrGenModel[[1]]
                      }
                    }
                  }else{
                    cat("Reestimation of parameters was started.", "\n")
                    dsrModel[[posL]][[posM]] <<- srModeling(selectModel, parUseOneL[[1]], parUseOneL[[2]], parUseOneL[[3]], parUseOneL[[4]], motifLengthList[[posL]], seqAlList[[posL]], seqCountList[[posL]], minMax, list(dsrMleCondFin[[posL]][[posM]]))[[1]]
                  }
                }else if(factName == "M2SR"){
                  m2srMleCondFin[[posL]][[posM]] <<- mleCondFinMake(list(mleCOneM), list(selectModel))[[1]][[1]]
                  if(srLSp[posL, 4] == 0){
                    imputOk <- tclvalue(tkmessageBox(message = paste(factName, "of the selected locus was modeled considering multiple loci together. Re-estimation of parameters will be also performed by considering multiple loci together. Do you want to continue?", sep = ""), type = "okcancel", icon = "warning"))
                    if(imputOk == "ok"){
                      cat("Reestimation of parameters was started.", "\n")
                      options(warn = -1)
                      m2srGenPos <- which(apply(rbind(srConsider[, 4], !srLSp[, 4]), 2, all) == TRUE)
                      options(warn = 0)
                      m2srGenModel <- srModeling(selectModel, m2srGenData[[1]], m2srGenData[[2]], m2srGenData[[3]], m2srGenData[[4]], m2srGenData[[5]], m2srGenData[[6]], m2srGenData[[7]], minMax, list(m2srMleCondFin[[posL]][[posM]]))
                      for(i in 1:length(dsrGenPos)){
                        m2srModel[[m2srGenPos[i]]][[posM]] <<- m2srGenModel[[1]]
                      }
                    }
                  }else{
                    cat("Reestimation of parameters was started.", "\n")
                    m2srModel[[posL]][[posM]] <<- srModeling(selectModel, parUseOneL[[1]], parUseOneL[[2]], parUseOneL[[3]], parUseOneL[[4]], motifLengthList[[posL]], seqAlList[[posL]], seqCountList[[posL]], minMax, list(m2srMleCondFin[[posL]][[posM]]))[[1]]
                  }
                }
                summaryMake()
              }
              
              condFinTcl <- function(mleCF, selectModel){
                mleCOneM <- mleC <- list()
                parName <- names(mleCF[[1]])
                for(i in 1:length(mleCF)){
                  mleCFOne <- mleCF[[i]]
                  mleCOne <- list()
                  for(j in 1:length(mleCFOne)){
                    mleCOne[[j]] <- tclVar(mleCFOne[j])
                  }
                  names(mleCOne) <- parName
                  mleC[[i]] <- mleCOne
                }
                mleCOneM[[1]] <- mleC
                names(mleCOneM) <- selectModel
                return(mleCOneM)
              }
              
              resetCond <- function(selectModel){
                posM <- which(nameModel == selectModel)
                mleCF <- mleCondOneL[[posM]]
                mleCOneM <- condFinTcl(mleCF, selectModel)
                mleInitial <- mleCOneM[[1]][[1]]
                mleLower <- mleCOneM[[1]][[2]]
                mleUpper <- mleCOneM[[1]][[3]]
                nPar <- length(mleInitial)
                parName <- names(mleInitial)
                
                tkdestroy(fReest_2)
                fReest_2 <<- tkframe(tfReest, relief = "groove", borderwidth = 2)
                tkgrid(tklabel(fReest_2, text = "Model"), row = 0, column = 0, padx = 5, pady = 5)
                tkgrid(tklabel(fReest_2, text = "Parameter"), row = 0, column = 1, padx = 5, pady = 5)
                tkgrid(tklabel(fReest_2, text = "Initial value"), row = 0, column = 2, padx = 5, pady = 5)
                tkgrid(tklabel(fReest_2, text = "Lower value"), row = 0, column = 3, padx = 5, pady = 5)
                tkgrid(tklabel(fReest_2, text = "Upper value"), row = 0, column = 4, padx = 5, pady = 5)
                tkgrid(tklabel(fReest_2, text = selectModel), row = 1, column = 0, padx = 5, pady = 5)
                for(j in 1:nPar){
                  tkgrid(tklabel(fReest_2, text = parName[j]), row = j, column = 1, padx = 5, pady = 5)
                  tkgrid(tkentry(fReest_2, textvariable = mleInitial[[j]], width = 10, highlightthickness = 1, relief = "solid", justify = "center", background = "white"), row = j, column = 2, padx = 5, pady = 5)
                  tkgrid(tkentry(fReest_2, textvariable = mleLower[[j]], width = 10, highlightthickness = 1, relief = "solid", justify = "center", background = "white"), row = j, column = 3, padx = 5, pady = 5)
                  tkgrid(tkentry(fReest_2, textvariable = mleUpper[[j]], width = 10, highlightthickness = 1, relief = "solid", justify = "center", background = "white"), row = j, column = 4, padx = 5, pady = 5)
                }
                tkgrid(tkbutton(fReest_2, text = "    Estimate    ", cursor = "hand2", command = function() reestRun(mleCOneM, posM, selectModel)), row = j + 1, column = 0, padx = 5, pady = 5)
                tkgrid(fReest_2, padx = 5, pady = 5, sticky = "w")
              }
              
              selectM <- tclVar(nameModel[1])
              
              tfReest <- tktoplevel()
              tkwm.title(tfReest, "Reestimate parameters")
              fReest_1 <- tkframe(tfReest)
              fReest_2 <- tkframe(tfReest)
              tkgrid(tklabel(fReest_1, text = "Select a model : "), row = 0, column = 0, padx = 5, pady = 5)
              tkgrid(ttkcombobox(fReest_1, values = nameModel, textvariable = selectM, state = "readonly"), row = 0, column = 1, padx = 5, pady = 5)
              tkgrid(tkbutton(fReest_1, text = "    Set conditions    ", cursor = "hand2", command = function() resetCond(tclvalue(selectM))), row = 0, column = 2, padx = 5, pady = 5)
              tkgrid(fReest_1, sticky = "w")
              tkgrid(fReest_2, sticky = "w")
            }
            
            #Show results of one locus in detail
            showDetail <- function(){
              tkdestroy(fDetail_2)
              fDetail_2 <<- tkframe(tfParResult, relief = "groove", borderwidth = 2)
              tkdestroy(fDetail_3)
              fDetail_3 <<- tkframe(tfParResult, relief = "groove", borderwidth = 2)
              tkdestroy(fDetail_4)
              fDetail_4 <<- tkframe(tfParResult, relief = "groove", borderwidth = 2)
              tkdestroy(fDetail_5)
              fDetail_5 <<- tkframe(tfParResult)
              
              posL <- which(parLocus == tclvalue(selectL))
              parModelOneL <- parModel[[posL]]
              parUseOneL <- parUseData[[posL]]
              if(factName == "AE"){
                posMOneL <- aeAdoptModelPos[posL]
              }else if(factName == "Hb"){
                posMOneL <- hbAdoptModelPos[posL]
              }else if(factName == "BSR"){
                posMOneL <- bsrAdoptModelPos[posL]
              }else if(factName == "FSR"){
                posMOneL <- fsrAdoptModelPos[posL]
              }else if(factName == "DSR"){
                posMOneL <- dsrAdoptModelPos[posL]
              }else if(factName == "M2SR"){
                posMOneL <- m2srAdoptModelPos[posL]
              }
              parBestModel <- chooseAdoptModel(parModelOneL, "AIC", posMOneL)
              nM <- length(parModelOneL) - 1
              
              tkgrid(tklabel(fDetail_2, text = "Information", font = "Helvetica 10 bold"), row = 0, column = 0, padx = 5, pady = 5, sticky = "w")
              tkgrid(tklabel(fDetail_2, text = "Adopted model : "), row = 1, column = 0, padx = 5, pady = 5, sticky = "w")
              tkgrid(tklabel(fDetail_2, text = "Method of modeling : "), row = 2, column = 0, padx = 5, pady = 5, sticky = "w")
              if(length(parModelOneL) == 0){
                tkgrid(tklabel(fDetail_2, text = ""), row = 1, column = 1, padx = 5, pady = 5, sticky = "w")
                tkgrid(tklabel(fDetail_2, text = "Not consider"), row = 2, column = 1, padx = 5, pady = 5, sticky = "w")
                tkgrid(fDetail_2, padx = 5, pady = 5, sticky = "w")
                tkgrid(fDetail_3, padx = 5, pady = 5, sticky = "w")
                tkgrid(fDetail_4, padx = 5, pady = 5, sticky = "w")
              }else{
                if(length(grep("\u2020", parBestModel[[3]])) == 1){
                  adoptLabel <- sub("\u2020", "(selected by user)", parBestModel[[3]])
                }else{
                  adoptLabel <- parBestModel[[3]]
                }
                tkgrid(tklabel(fDetail_2, text = adoptLabel), row = 1, column = 1, padx = 5, pady = 5, sticky = "w")
                tkgrid(tklabel(fDetail_2, text = parModelOneL[[nM + 1]]), row = 2, column = 1, padx = 5, pady = 5, sticky = "w")
                tkgrid(fDetail_2, padx = 5, pady = 5, sticky = "w")
                
                tkgrid(tklabel(fDetail_3, text = "Comparison of each model", font = "Helvetica 10 bold"), row = 0, column = 0, padx = 5, pady = 5, sticky = "w")
                tkgrid(tklabel(fDetail_3, text = "Log10 (likelihood)"), row = 2, column = 0, padx = 5, pady = 5)
                tkgrid(tklabel(fDetail_3, text = "Number of parameters"), row = 3, column = 0, padx = 5, pady = 5)
                tkgrid(tklabel(fDetail_3, text = "AIC"), row = 4, column = 0, padx = 5, pady = 5)
                namePars <- character(0)
                nameModel <- rep("", nM)
                for(i in 1:nM){
                  pm <- parModelOneL[[i]]
                  estPar <- pm[[2]]
                  namP <- names(estPar)
                  nep <- length(estPar)
                  nameModel[i] <- pm[[3]]
                  tkgrid(tklabel(fDetail_3, text = nameModel[i]), row = 1, column = i, padx = 5, pady = 5)
                  tkgrid(tklabel(fDetail_3, text = sprintf("%.3f", pm[[1]][1])), row = 2, column = i, padx = 5, pady = 5)
                  tkgrid(tklabel(fDetail_3, text = nep), row = 3, column = i, padx = 5, pady = 5)
                  tkgrid(tklabel(fDetail_3, text = sprintf("%.3f", pm[[1]][2])), row = 4, column = i, padx = 5, pady = 5)
                  for(j in 1:nep){
                    namPos <- which(namePars == namP[j])
                    if(length(namPos) == 0){
                      namePars <- c(namePars, namP[j])
                      rowPos <- 4 + length(namePars)
                      tkgrid(tklabel(fDetail_3, text = namP[j]), row = rowPos, column = 0, padx = 5, pady = 5)
                    }else{
                      rowPos <- 4 + namPos
                    }
                    if(is.element(namP[j], c("slope", "intercept"))){
                      val <- sprintf("%.3e", estPar[j])
                    }else if(is.element(namP[j], c("parameter x"))){
                      val <- round(estPar[j], 0)
                    }else{
                      val <- sprintf("%.3f", estPar[j])
                    }
                    tkgrid(tklabel(fDetail_3, text = val), row = rowPos, column = i, padx = 5, pady = 5)
                  }
                }
                tkgrid(fDetail_3, padx = 5, pady = 5, sticky = "w")
                
                tkgrid(tklabel(fDetail_4, text = "Graph", font = "Helvetica 10 bold"), row = 0, column = 0, padx = 5, pady = 5, sticky = "w")
                selectM <- tclVar(nameModel[1])
                tkgrid(tklabel(fDetail_4, text = "Model : "), row = 1, column = 0, padx = 5, pady = 5, sticky = "w")
                tkgrid(ttkcombobox(fDetail_4, values = nameModel, textvariable = selectM, state = "readonly"), row = 1, column = 1, padx = 5, pady = 5, sticky = "w")
                if(is.element(factName, c("AE", "Hb"))){
                  tkgrid(tkbutton(fDetail_4, text = "    Variance    ", cursor = "hand2", command = function() graphVar(parModelOneL, parUseOneL, factName, minMax)), row = 2, column = 0, padx = 5, pady = 5, sticky = "w")
                }else{
                  tkgrid(tkbutton(fDetail_4, text = "    Mean    ", cursor = "hand2", command = function() srGraphMean(parModelOneL, parUseOneL, nameModel, tclvalue(selectM), factName)), row = 2, column = 0, padx = 5, pady = 5, sticky = "w")
                  tkgrid(tkbutton(fDetail_4, text = "    Variance    ", cursor = "hand2", command = function() srGraphVar(parModelOneL, parUseOneL, nameModel, tclvalue(selectM), factName, minMax)), row = 2, column = 1, padx = 5, pady = 5, sticky = "w")
                }
                tkgrid(fDetail_4, padx = 5, pady = 5, sticky = "w")
                
                tkgrid(tkbutton(fDetail_5, text = "    Change the adopted model    ", cursor = "hand2", command = function() modelChange(nameModel, posL, factName)), row = 0, column = 0, padx = 5, pady = 5, sticky = "w")
                tkgrid(tkbutton(fDetail_5, text = "    Reestimate parameters    ", cursor = "hand2", command = function() reestParam(parUseOneL, nameModel, posL, factName, minMax, mleCondFin[[posL]])), row = 0, column = 1, padx = 5, pady = 5, sticky = "w")
                tkgrid(fDetail_5, padx = 5, pady = 5, sticky = "w")
              }
            }
            
            tfParResult <- tktoplevel()
            tkwm.title(tfParResult, "Results of parameter estimation")
            
            fDetail_1 <- tkframe(tfParResult)
            fDetail_2 <- tkframe(tfParResult)
            fDetail_3 <- tkframe(tfParResult)
            fDetail_4 <- tkframe(tfParResult)
            fDetail_5 <- tkframe(tfParResult)
            parLocus <- names(parModel)
            selectL <- tclVar(parLocus[1])
            tkgrid(tklabel(fDetail_1, text = "Locus : "), row = 0, column = 0, padx = 5, pady = 5)
            tkgrid(ttkcombobox(fDetail_1, values = parLocus, textvariable = selectL, state = "readonly"), row = 0, column = 1, padx = 5, pady = 5)
            tkgrid(tkbutton(fDetail_1, text = "    Show    ", cursor = "hand2", command = function() showDetail()), row = 0, column = 2, padx = 5, pady = 5)
            tkgrid(fDetail_1, padx = 10, sticky = "w")
            tkgrid(fDetail_2, padx = 10, sticky = "w")
            tkgrid(fDetail_3, padx = 10, sticky = "w")
            tkgrid(fDetail_4, padx = 10, sticky = "w")
            tkgrid(fDetail_5, padx = 10, sticky = "w")
          }
          
          #Make a summary of the results
          summaryMake <- function(){
            tkdestroy(fTab3Par)
            fTab3Par <<- scrollMake(tab3Par)
            tkpack(fTab3Par, expand = TRUE, fill = "both")
            fTab3Par_scr <- scrollGet(fTab3Par)
            
            tkgrid(tklabel(fTab3Par_scr, text = "Adopted models", font = "Helvetica 10 bold"), row = 0, column = 0, columnspan = 4, padx = 5, pady = 10, sticky = "w")
            tkgrid(tklabel(fTab3Par_scr, text = "Locus", font = "Helvetica 10 bold"), row = 1, column = 0, padx = 5)
            tkgrid(tklabel(fTab3Par_scr, text = "     "), row = 1, column = 1, padx = 5)
            tkgrid(tklabel(fTab3Par_scr, text = "Locus-specific", font = "Helvetica 10 bold"), row = 1, column = 2, padx = 5)
            tkgrid(tklabel(fTab3Par_scr, text = "amplification efficiency", font = "Helvetica 10 bold"), row = 2, column = 2, padx = 5)
            tkgrid(tklabel(fTab3Par_scr, text = "     "), row = 1, column = 3, padx = 5)
            tkgrid(tklabel(fTab3Par_scr, text = "Heterozygote balance", font = "Helvetica 10 bold"), row = 1, column = 4, padx = 5)
            tkgrid(tklabel(fTab3Par_scr, text = "     "), row = 1, column = 5, padx = 5)
            tkgrid(tklabel(fTab3Par_scr, text = "Back stutter ratio", font = "Helvetica 10 bold"), row = 1, column = 6, padx = 5)
            tkgrid(tklabel(fTab3Par_scr, text = "     "), row = 1, column = 7, padx = 5)
            tkgrid(tklabel(fTab3Par_scr, text = "Forward stutter ratio", font = "Helvetica 10 bold"), row = 1, column = 8, padx = 5)
            tkgrid(tklabel(fTab3Par_scr, text = "     "), row = 1, column = 9, padx = 5)
            tkgrid(tklabel(fTab3Par_scr, text = "Double-back stutter ratio", font = "Helvetica 10 bold"), row = 1, column = 10, padx = 5)
            tkgrid(tklabel(fTab3Par_scr, text = "     "), row = 1, column = 11, padx = 5)
            tkgrid(tklabel(fTab3Par_scr, text = "Minus 2-nt stutter ratio", font = "Helvetica 10 bold"), row = 1, column = 12, padx = 5)
            
            parLocus <- names(aeModel)
            nL <- length(parLocus)
            aeBestList <- hbBestList <- bsrBestList <- fsrBestList <- dsrBestList <- m2srBestList <- list()
            daggerJudge <- FALSE
            for(i in 1:nL){
              tkgrid(tklabel(fTab3Par_scr, text = parLocus[i]), row = i + 2, column = 0, padx = 5, pady = 5)
              
              aeUseOneL <- aeUseData[[i]]
              aeModelOneL <- aeModel[[i]]
              aeBestList[[i]] <- aeBestModel <- chooseAdoptModel(aeModelOneL, "AIC", aeAdoptModelPos[i])
              if(length(aeBestModel) != 0){
                bestModel <- aeBestModel[[3]]
                tkgrid(tklabel(fTab3Par_scr, text = bestModel), row = i + 2, column = 2, padx = 5, pady = 5)
                if(length(grep("\u2020", bestModel)) == 1){
                  daggerJudge <- TRUE
                }
              }
              
              hbUseOneL <- hbUseData[[i]]
              hbModelOneL <- hbModel[[i]]
              hbBestList[[i]] <- hbBestModel <- chooseAdoptModel(hbModelOneL, "AIC", hbAdoptModelPos[i])
              if(length(hbBestModel) != 0){
                bestModel <- hbBestModel[[3]]
                tkgrid(tklabel(fTab3Par_scr, text = bestModel), row = i + 2, column = 4, padx = 5, pady = 5)
                if(length(grep("\u2020", bestModel)) == 1){
                  daggerJudge <- TRUE
                }
              }
              
              bsrUseOneL <- bsrUseData[[i]]
              bsrModelOneL <- bsrModel[[i]]
              bsrBestList[[i]] <- bsrBestModel <- chooseAdoptModel(bsrModelOneL, "AIC", bsrAdoptModelPos[i])
              if(length(bsrBestModel) != 0){
                bestModel <- bsrBestModel[[3]]
                tkgrid(tklabel(fTab3Par_scr, text = bestModel), row = i + 2, column = 6, padx = 5, pady = 5)
                if(length(grep("\u2020", bestModel)) == 1){
                  daggerJudge <- TRUE
                }
              }
              
              fsrUseOneL <- fsrUseData[[i]]
              fsrModelOneL <- fsrModel[[i]]
              fsrBestList[[i]] <- fsrBestModel <- chooseAdoptModel(fsrModelOneL, "AIC", fsrAdoptModelPos[i])
              if(length(fsrBestModel) != 0){
                bestModel <- fsrBestModel[[3]]
                tkgrid(tklabel(fTab3Par_scr, text = bestModel), row = i + 2, column = 8, padx = 5, pady = 5)
                if(length(grep("\u2020", bestModel)) == 1){
                  daggerJudge <- TRUE
                }
              }
              
              dsrUseOneL <- dsrUseData[[i]]
              dsrModelOneL <- dsrModel[[i]]
              dsrBestList[[i]] <- dsrBestModel <- chooseAdoptModel(dsrModelOneL, "AIC", dsrAdoptModelPos[i])
              if(length(dsrBestModel) != 0){
                bestModel <- dsrBestModel[[3]]
                tkgrid(tklabel(fTab3Par_scr, text = bestModel), row = i + 2, column = 10, padx = 5, pady = 5)
                if(length(grep("\u2020", bestModel)) == 1){
                  daggerJudge <- TRUE
                }
              }
              
              m2srUseOneL <- m2srUseData[[i]]
              m2srModelOneL <- m2srModel[[i]]
              m2srBestList[[i]] <- m2srBestModel <- chooseAdoptModel(m2srModelOneL, "AIC", m2srAdoptModelPos[i])
              if(length(m2srBestModel)){
                bestModel <- m2srBestModel[[3]]
                tkgrid(tklabel(fTab3Par_scr, text =  bestModel), row = i + 2, column = 12, padx = 5, pady = 5)
                if(length(grep("\u2020", bestModel)) == 1){
                  daggerJudge <- TRUE
                }
              }
            }
            names(aeBestList) <- names(hbBestList) <- names(bsrBestList) <- names(fsrBestList) <- names(dsrBestList) <- names(m2srBestList) <- parLocus
            
            tkgrid(tkbutton(fTab3Par_scr, text = "    Detail    ", cursor = "hand2", command = function() resultDetail(aeModel, aeUseData, "AE", aeMleCondFin, aeMin)), row = i + 3, column = 2, padx = 5, pady = 10)
            tkgrid(tkbutton(fTab3Par_scr, text = "    Detail    ", cursor = "hand2", command = function() resultDetail(hbModel, hbUseData, "Hb", hbMleCondFin, hbMin)), row = i + 3, column = 4, padx = 5, pady = 10)
            tkgrid(tkbutton(fTab3Par_scr, text = "    Detail    ", cursor = "hand2", command = function() resultDetail(bsrModel, bsrUseData, "BSR", bsrMleCondFin, bsrMax)), row = i + 3, column = 6, padx = 5, pady = 10)
            tkgrid(tkbutton(fTab3Par_scr, text = "    Detail    ", cursor = "hand2", command = function() resultDetail(fsrModel, fsrUseData, "FSR", fsrMleCondFin, fsrMax)), row = i + 3, column = 8, padx = 5, pady = 10)
            tkgrid(tkbutton(fTab3Par_scr, text = "    Detail    ", cursor = "hand2", command = function() resultDetail(dsrModel, dsrUseData, "DSR", dsrMleCondFin, dsrMax)), row = i + 3, column = 10, padx = 5, pady = 10)
            tkgrid(tkbutton(fTab3Par_scr, text = "    Detail    ", cursor = "hand2", command = function() resultDetail(m2srModel, m2srUseData, "M2SR", m2srMleCondFin, m2srMax)), row = i + 3, column = 12, padx = 5, pady = 10)
            tkgrid(tkbutton(fTab3Par_scr, text = "    Output parameters    ", cursor = "hand2", command = function() parOutput(aeBestList, hbBestList, bsrBestList, fsrBestList, dsrBestList, m2srBestList, srConsider, srLSp, parDye, repLength)), row = i + 4, column = 2, columnspan = 2, padx = 5, pady = 10)
            tkgrid(tkbutton(fTab3Par_scr, text = "    Output allele repeat correction    ", cursor = "hand2", command = function() alCorOutput(lusData, motifLengthList, seqAlList, seqCountList,bsrBestList, fsrBestList, dsrBestList)), row = i + 4, column = 4, columnspan = 4, padx = 5, pady = 10)
            if(daggerJudge){
              tkgrid(tklabel(fTab3Par_scr, text = paste("(", "\u2020", " selected by user)", sep = "")), row = i + 4, column = 12, padx = 5, pady = 10, sticky = "w")
            }
          }
          
          aeAdoptModelPos <<- hbAdoptModelPos <<- bsrAdoptModelPos <<- fsrAdoptModelPos <<- dsrAdoptModelPos <<- m2srAdoptModelPos <<- rep(0, length(aeModel))
          summaryMake()
        }
      }
    }
    
    
    #######################################
    # Information of generally used locus #
    #######################################
    
    #Autosomal STR markers
    strMar <- c("D3S1358", "vWA", "D16S539", "CSF1PO", "TPOX", "D8S1179", "D21S11", "D18S51", "D2S441", "D19S433", "TH01", "FGA", "D22S1045", "D5S818", "D13S317", "D7S820", "SE33", "D10S1248", "D1S1656", "D12S391", "D2S1338", "D6S1043", "Penta D", "Penta E")
    nStr <- length(strMar)
    
    #Sex chromosomal markers
    sexChrMar <- c("AMEL", "Amelogenin", "Yindel", "YIndel", "YInDel", "DYS391")
    
    #STR repeat length
    repLengthInfo <- c(4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 3, 4, 4, 4, 4, 4, 4, 4, 4, 4, 5, 5)
    
    
    ###################
    # Default setting #
    ###################
    
    #Make 'tcl' objects
    calibName <- tclVar("")
    calibFp <- tclVar("")
    gtAnsName <- tclVar("")
    gtAnsFp <- tclVar("")
    seqName <- tclVar("")
    seqFp <- tclVar("")
    fileCk2Fin <- tclVar("0")
    paramEstFin <- tclVar("0")
    
    #Default of the min. threshold
    mtDefault <- 30
    
    #Default of parameter setting in each locus
    parDefault <- matrix("", nStr, 13)
    colnames(parDefault) <- c("Locus", "Repeat length", "Min. threshold", "AE model", "Hb model", "BSR method", "BSR model", "FSR method", "FSR model", "DSR method", "DSR model", "M2SR method", "M2SR model")
    parDefault[, 1] <- strMar
    parDefault[, 2] <- repLengthInfo
    parDefault[, 3] <- mtDefault
    parDefault[, 4] <- rep("Log-normal", nStr)
    parDefault[, 5] <- rep("Log-normal", nStr)
    parDefault[, 6] <- rep("Locus specific", nStr)
    parDefault[, 7] <- rep("Best", nStr)
    parDefault[, 8] <- c(rep("Multiple loci together", 12), "Locus specific", rep("Multiple loci together", 11))
    parDefault[, 9] <- c(rep("", 12), "Best", rep("", 11))
    parDefault[, 10] <- rep("Multiple loci together", nStr)
    parDefault[, 11] <- rep("", nStr)
    parDefault[, 12] <- c(rep("Not consider", 16), "Locus specific", "Not consider", "Locus specific", rep("Not consider", 5))
    parDefault[, 13] <- c(rep("", 16), "Uniform", "", "Uniform", rep("", 5))
    
    #Default of parameter setting (for new loci)
    parGeneral <- rep("", 13)
    names(parGeneral) <- colnames(parDefault)
    parGeneral[1] <- "General"
    parGeneral[2] <- 4
    parGeneral[3] <- mtDefault
    parGeneral[4] <- "Log-normal"
    parGeneral[5] <- "Log-normal"
    parGeneral[6] <- "Locus specific"
    parGeneral[7] <- "Best"
    parGeneral[8] <- "Multiple loci together"
    parGeneral[9] <- ""
    parGeneral[10] <- "Multiple loci together"
    parGeneral[11] <- ""
    parGeneral[12] <- "Not consider"
    parGeneral[13] <- ""
    
    #Candidate models of each factor
    aeCand <- hbCand <- bsrCand <- fsrCand <- dsrCand <- m2srCand <- list()
    for(i in 1:nStr){
      aeCand[[i]] <- "Log-normal"
      hbCand[[i]] <- "Log-normal"
      bsrCand[[i]] <- c("Best", "Allele", "LUS", "Multi-seq")
      fsrCand[[i]] <- c("Best", "Allele", "LUS", "Multi-seq", "Uniform")
      dsrCand[[i]] <- c("Best", "Allele", "LUS", "Multi-seq", "Uniform")
      m2srCand[[i]] <- "Uniform"
    }
    names(aeCand) <- names(hbCand) <- names(bsrCand) <- names(fsrCand) <- names(dsrCand) <- names(m2srCand) <- strMar
    
    #Candidate models of each factor for new loci
    aeCandGen <- "Log-normal"
    hbCandGen <- "Log-normal"
    bsrCandGen <- c("Best", "Allele", "LUS", "Multi-seq")
    fsrCandGen <- c("Best", "Allele", "LUS", "Multi-seq", "Uniform")
    dsrCandGen <- c("Best", "Allele", "LUS", "Multi-seq", "Uniform")
    m2srCandGen <- "Uniform"
    
    #Minimum or maximum values of each factor
    aeMinVar <- tclVar(0.058)
    hbMinVar <- tclVar(0.046)
    bsrMaxVar <- tclVar(0.7)
    fsrMaxVar <- tclVar(0.35)
    dsrMaxVar <- tclVar(0.13)
    m2srMaxVar <- tclVar(0.12)
    
    
    ####################
    # Make a top frame #
    ####################
    
    tfPar <- tktoplevel()
    tkwm.title(tfPar, "parameter estimation")
    tabsPar <- tk2notebook(tfPar, tabs = c("Files", "Setting", "Result"))
    tkpack(tabsPar, fill = "both", expand = 1)
    tab1Par <- tk2notetab(tabsPar, "Files")
    tab2Par <- tk2notetab(tabsPar, "Setting")
    tab3Par <- tk2notetab(tabsPar, "Result")
    
    fTab1Par <- tkframe(tab1Par)
    tab1ParMake()
    fTab2Par <- tkframe(tab2Par)
    tab2ParMake()
    fTab3Par <- tkframe(tab3Par)
    tab3ParMake()
  }
  
  
  #################################
  # 'Estimation of parameter' tab #
  #################################
  
  #Make 'Estimation of parameters' tab
  tab5Make <- function(){
    tkdestroy(frameTab5)
    frameTab5 <<- tkframe(tab5)
    tkgrid(tkbutton(frameTab5, text = "    Open a new window   ", cursor = "hand2", command = function() parEstWindow()), padx = 5, pady = 5)
    tkgrid(frameTab5)
  }

  
  ##########################
  # Make a toplevel window #
  ##########################
  
  tf <- tktoplevel()
  tkwm.title(tf, softVer)
  topMenu <- tkmenu(tf)
  tkconfigure(tf, menu = topMenu)
  tabs <- tk2notebook(tf, tabs = c("Files", "Calculation", "Probabilistic genotyping", "Likelihood ratio", "Estimation of parameters"))
  tkpack(tabs, fill = "both", expand = 1)
  tab1 <- tk2notetab(tabs, "Files")
  tab2 <- tk2notetab(tabs, "Calculation")
  tab3 <- tk2notetab(tabs, "Probabilistic genotyping")
  tab4 <- tk2notetab(tabs, "Likelihood ratio")
  tab5 <- tk2notetab(tabs, "Estimation of parameters")
  frameTab1 <- tkframe(tab1)
  tab1Make()
  frameTab2 <- tkframe(tab2)
  tab2Make()
  frameTab3 <- tkframe(tab3)
  tab3Make()
  frameTab4 <- tkframe(tab4)
  tab4Make()
  frameTab5 <- tkframe(tab5)
  tab5Make()
}


