---
title: "Kongoh version 3.2.2 User Manual"
author: "Sho Manabe (manabesh@hirakata.kmu.ac.jp)"
date: "2022-10-21"
header-includes:
  \usepackage{ulem}
output:
  pdf_document:
    latex_engine: xelatex 
    number_sections: false
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

\tableofcontents
\clearpage

\setlength{\baselineskip}{18pt}

\setcounter{page}{3}

\section{Information on Kongoh}

Kongoh (named after the Japanese word “mixture”) is open-source software for DNA evidence interpretation based on a quantitative continuous model\cite{Manabe2017,Manabe2022}. The software is a graphical user interface and has been released as the R-package \textit{Kongoh} (>= ver. 3.2.0). 

Kongoh can perform mixture deconvolution and the assignment of the likelihood ratio for crime stain profiles typed by the AmpFlSTR\textsuperscript{\textregistered} Identifiler\textsuperscript{\textregistered} Plus PCR Amplification Kit (Thermo Fisher Scientific, Waltham, MA) and GlobalFiler\textsuperscript{TM} PCR Amplification Kit (Thermo Fisher Scientific). Profiles typed by kits other than Identifiler Plus and GlobalFiler can be analyzed if experimental data are prepared from the kit. A peak located at the position of the back stutter, forward stutter, double-back stutter, and minus 2-nt stutter need not be designated as an allele or stutter because the derivation of the peak in the stutter position can be determined probabilistically. Hence, the stutter filters of all loci can be removed. Moreover, Kongoh considers allelic drop-out, which is the event of a peak under the analytical threshold. By contrast, drop-in is not considered; however, spontaneous drop-in peaks can be explained based on additional unknown contributors.

\newpage

\section{Version changes}

\textbf{Note}

\textbf{Users are recommended to switch to version 3.2.2 because of the following updates.}

\subsection{Changes in ver. 3.2.2}

\subsubsection{Important updates}

\begin{itemize}

  \item A calculation error was fixed. In the previous version of Kongoh, the error occurred in the case that one of the expected forward stutter ratios (FSRs) or the expected double-back stutter ratios (DSRs) were not greater than zero under the model for these stutters. In this case, the new version of Kongoh does not regard the observed peak at the stutter position as only the stutter product.
  
  \item A calculation error was fixed. In the previous version of Kongoh, the error occurred when there were some locus names with spaces (e.g., Penta D and Penta E) in the input files.
  
  \item An error about setting analytical thresholds in the "Set analysis methods" was fixed.

\end{itemize}

\subsubsection{Minor changes}

\begin{itemize}

  \item The default file "GF\_29cycles\_3500xL\_24sec.csv" is updated. The update does not affect the result of the mixture deconvolution and the likelihood ratio.

\end{itemize}

\subsection{Changes in ver. 3.2.1}

\subsubsection{New features}

\begin{itemize}

  \item The likelihood ratio is displayed on both normal and logarithmic scales in the "Result" of the "Likelihood ratio" tab.

  \item Users can export likelihoods and LRs of all assumed hypotheses.

\end{itemize}

\subsubsection{Minor changes}

\begin{itemize}

  \item The term "Likelihood" has been changed to "Log10 (Likelihood)" in the "Result" of the "Deconvolution" tab.
  
  \item The following minor issue was fixed. The unexpected error message "Error: The total number of contributors is smaller than the number of known contributors." was displayed when the assumed number of contributors was smaller than the number of reference profiles inputted.

\end{itemize}

\subsection{Changes in ver. 3.2.0}

\subsubsection{New features}

\begin{itemize}

  \item Kongoh can be used as the R-package \textit{Kongoh}.

  \item UI design has been changed.

  \item Project data can be saved and loaded.
  
  \item A new feature "mixture deconvolution" has been implemented. In this feature, probabilistic genotyping can be performed without reference profiles.
  
  \item The expected peak height of drop-out alleles (Q) is regarded as that of the most frequent allele in Q regardless of known and unknown contributors.

  \item A new feature "Set analysis methods" has been implemented.
  
  \begin{itemize}
  
    \item Users can set analysis methods per typing kit in the software.
  
    \item Analytical thresholds can be set per locus.

    \item The "mixture ratio filter" is added to exclude unrealistic genotype combinations.
  
    \item The files for Monte Carlo parameters and the information on allele repeat correction can be specified when setting analysis methods. Therefore, users no longer need to load these files in the "Files" tab.
    
  \end{itemize}

  \item The file format for Monte Carlo parameters is changed as follows.
  
  \begin{itemize}

      \item Names of parameters in each model are explicitly declared in the column.
    
      \item The scale of "Mean" parameters for AE and Hb is changed from a logarithmic scale to a linear scale.
    
      \item See example files for GlobalFiler (GF\_29cycles\_3500xL\_24sec.csv) and Identifiler Plus (IDP\_28cycles\_3130xl\_10sec.csv) which are located at extdata > parameters in the package \textit{Kongoh}.
  
  \end{itemize}
  
  \item A new feature "Edit repeat correction" has been implemented. Users can edit information on allele repeat correction used for models of stutter ratios.
  
  \item The default files of allele repeat correction are updated. See files for GlobalFiler (GF\_repeat\_correction.csv) and Identifiler Plus (IDP\_repeat\_correction.csv) which are located at extdata > repeat\_correction in the package \textit{Kongoh}.

  \item When the input allele frequencies are represented as allele counts, the input frequencies can be viewed in both the format of allele counts and the format of allele probabilities estimated by Dirichlet distribution.

  \item Estimated gamma distributions are replaced with the expected peak heights of gamma distributions in the graph of probabilistic genotyping.
  
  \item There are some improvements in the feature "Estimate parameters".
  
  \begin{itemize}
  
    \item Project data can be saved and loaded.
    
    \item The information on "Sample File" and "Dye" in the file of experimental data is no longer needed. See the example file "Experimental-data\_GF\_Example.csv" which is located at extdata > example in the package \textit{Kongoh}.

    \item Parameters of min AE, min Hb, max BSR, max FSR, max DSR, and max M2SR can be determined based on the experimental data.
    
    \item Functions for checking input files have been updated to detect the difference in the locus set and the sample names between each file.
    
    \item The graph for stutter ratios has been improved when the option "Multiple loci together" is selected for the method of modeling.
  
  \end{itemize}
  
  \item A new feature "Set information on typing kits" has been implemented. Users can create or edit information on the locus set, alleles, and sizes of each allele.
  
\end{itemize}

\subsubsection{Minor changes}

\begin{itemize}

  \item The term "mixture ratio" has been changed to "mixture proportion".
  
  \item The term "Weight" has been changed to "Weight (0-1 scale)" in the result of probabilistic genotyping.

  \item An error message has been improved when there is a lack of information about allele repeat correction for the representative of the unobserved alleles (Q).

\end{itemize}

\newpage

\section{Tutorial}

\subsection{Getting started}

\begin{enumerate}

  \item Ensure that R (>= 4.2.0) is installed. It is available from the R Development Core Team website (http://www.R-project.org).

  \item Begin an R session.
  
  \item Execute the following command in R to install required packages.

\end{enumerate}
  
```
install.packages(c("tcltk2", "gtools", "truncnorm", "GenSA"))
```

\vskip\baselineskip

\begin{enumerate}

\setcounter{enumi}{3} % <- \begin{enumerate}

  \item Go to https://github.com/manabe0322/Kongoh/releases.
  
  \item Download "Kongoh\_3.2.2.zip".
  
  \begin{figure}[htbp]
  
    \centering
  
    \includegraphics[height=12cm]{figures/figure_install1.jpg}
    
  \end{figure}
  
  \newpage
  
  \item Install "Kongoh\_3.2.2.zip" from "Install package(s) from local files...".
  
  \begin{figure}[htbp]
  
    \centering
  
    \includegraphics[height=6cm]{figures/figure_install2.jpg}
    
  \end{figure}
  
  \vskip\baselineskip

  \item Execute the following commands in R to start GUI.
  
\end{enumerate}

```
library(Kongoh)
Kongoh()
```

\newpage

\subsection{Input files}

\begin{enumerate}

  \item Load a file of the crime stain profile.
  
  \begin{figure}[htbp]
  
    \centering
  
    \includegraphics[height=6cm]{figures/screen1.jpg}
    
  \end{figure}

  \textbf{Note}

  \begin{itemize}

    \item The example file "CSP\_GF\_Example.csv" is provided. Go to extdata > example in the package \textit{Kongoh}.

    \item This file can be exported from the GeneMapper\textsuperscript{\textregistered} ID-X software.

    \item This file must include information regarding the “Sample Name”, “Marker”, “Allele”, “Size”, and “Height” as shown in Figure 1.

    \item Back stutters, forward stutters, double-back stutters, and minus 2-nt stutters are not necessarily removed manually\footnote{Whether these stutters should be removed depends on the setting of the Monte Carlo parameters.}.

    \item Pull-up peaks and noises must be removed manually.
  
  \end{itemize}
  
  \newpage
  
  \begin{figure}[htbp]
  
    \centering
  
    \includegraphics[height=9cm]{figures/figure1.jpg}
    
    \caption{Format of the crime stain profile}
    
  \end{figure}
  
  \newpage

  \item Load a file of the reference profiles.
  
  \begin{figure}[htbp]
    
    \centering
  
    \includegraphics[height=6cm]{figures/screen2.jpg}
    
  \end{figure}

  \textbf{Note}

  \begin{itemize}

    \item The example file "Ref\_GF\_Example.csv" is provided. Go to extdata > example in the package \textit{Kongoh}.

    \item This file must include information regarding the “Marker” and the name of each profile (e.g., victim and suspect), as shown in Figure 2.

    \item Two alleles in homozygotes must be entered in each column.
    
  \end{itemize}
  
  \begin{figure}[htbp]
  
    \centering
  
    \includegraphics[height=6cm]{figures/figure2.jpg}
    
    \caption{Format of the reference profiles}
    
  \end{figure}
  
  \newpage

  \item Load a file of the allele frequencies.
  
  \begin{figure}[htbp]
    
    \centering
  
    \includegraphics[height=6cm]{figures/screen3.jpg}
    
  \end{figure}

  \textbf{Note}

  \begin{itemize}

    \item Users can input two formats of allele frequencies: the allele counts (Figure 3) and the allele probabilities (Figure 4).

    \item If the allele frequencies are represented as the allele counts, the probabilities of each allele are determined by the expected values based on the Dirichlet distribution.

    \item If the allele frequencies are represented as the probabilities, these values are directly used for the probabilities of each allele. When a crime stain profile or the reference profiles contain alleles that are not listed in the input file, then the minimum allele frequency is used as the frequency of these alleles.

    \item The allele frequencies for the Identifiler\cite{Yoshida2005} and the GlobalFiler\cite{Fujii2015} in the Japanese population are provided. Go to extdata > example in the package \textit{Kongoh}. The allele frequencies for the Identifiler are represented as the probabilities. The allele frequencies for the GlobalFiler are represented as the allele counts. 
  
  \end{itemize}
  
  \begin{figure}[htbp]
    
    \centering
  
    \includegraphics[height=9cm]{figures/figure3.jpg}
    
    \caption{Format of the allele frequencies, which are represented as the allele counts}
    
  \end{figure}
  
  \begin{figure}[htbp]
    
    \centering
  
    \includegraphics[height=9cm]{figures/figure4.jpg}
    
    \caption{Format of the allele frequencies, which are represented as the allele probabilities}
    
  \end{figure}

\end{enumerate}

\newpage

\subsection{Deconvolution}

\begin{enumerate}

  \item Click the "Deconvolution" button in the "Files" tab.
  
  \begin{figure}[htbp]
    
    \centering
  
    \includegraphics[height=6cm]{figures/screen4.jpg}
    
  \end{figure}
  
  \vskip\baselineskip

  \item Set the range of the assumed number of contributors.
  
  \begin{figure}[htbp]
    
    \centering
  
    \includegraphics[height=6cm]{figures/screen_D1.jpg}
    
  \end{figure}
  
  \textbf{Note}
  
  The upper limit of the assumed number of contributors is four.
  
  \newpage

  \item Select an analysis method\footnote{Analysis methods can be set from Tools > Set analysis methods. See the section "Set analysis methods".}.
  
  \begin{figure}[htbp]
    
    \centering
  
    \includegraphics[height=6cm]{figures/screen_D2.jpg}
    
  \end{figure}
  
  \vskip\baselineskip

  \item Click the "Analyze" button.
  
  \begin{figure}[htbp]
    
    \centering
  
    \includegraphics[height=6cm]{figures/screen_D3.jpg}
    
  \end{figure}
  
  \newpage
  
  Then, the "Progress Bar" window will appear. After completing the analysis, the "Result" tab will appear automatically.
  
  \begin{figure}[htbp]
    
    \centering
  
    \includegraphics[height=12cm]{figures/progress1.jpg}
    
  \end{figure}
  
  \newpage

  \item Review the result of probabilistic genotyping.
  
  \begin{itemize}

    \item The weight values of each genotype combination under the selected conditions are displayed.
    
    \begin{figure}[htbp]
    
      \centering
  
      \includegraphics[height=12cm]{figures/screen_D4.jpg}
    
    \end{figure}
    
    \newpage
    
    \item The observed peak heights and the expected peak heights of gamma distributions under the selected genotype combination can be compared by clicking the "Graph" button. An example of the graph is shown in Figure 5.
    
    \begin{figure}[htbp]
    
      \centering
  
      \includegraphics[height=12cm]{figures/screen_D5.jpg}
    
    \end{figure}
    
    \newpage
    
    \begin{figure}[htbp]
    
      \centering
  
      \includegraphics[height=12cm]{figures/figure5.jpg}
      
      \caption{An example of the comparison between the observed peak heights (blue) and the expected peak heights (red)}
    
    \end{figure}
    
    \newpage
    
    \item Displayed data can be exported by clicking the "Export displayed data" button.
    
    \begin{figure}[htbp]
    
      \centering
  
      \includegraphics[height=12cm]{figures/screen_D6.jpg}
    
    \end{figure}
    
    \newpage
    
    \item The results of all conditions can be exported by clicking the "Export all" button.
    
    \begin{figure}[htbp]
    
      \centering
  
      \includegraphics[height=12cm]{figures/screen_D7.jpg}
    
    \end{figure}
    
  \end{itemize}
  
  \newpage

  \item Review the result of the likelihoods and the estimated parameters.
  
  \begin{itemize}
  
    \item The report of probabilistic genotyping can be exported by clicking the "Report" button.
    
    \begin{figure}[htbp]
    
      \centering
  
      \includegraphics[height=12cm]{figures/screen_D8.jpg}
    
    \end{figure}
  
  \end{itemize}

\end{enumerate}

\newpage

\subsection{Likelihood ratio}

\begin{enumerate}

  \item Click the "Likelihood ratio" button in the "Files" tab.
  
  \begin{figure}[htbp]
    
    \centering
  
    \includegraphics[height=6cm]{figures/screen5.jpg}
    
  \end{figure}
  
  \vskip\baselineskip

  \item Set both the prosecutor (Hp) and defense (Hd) hypotheses. Check the individuals to include them as contributors in each hypothesis. 
  
  \begin{figure}[htbp]
    
    \centering
  
    \includegraphics[height=6cm]{figures/screen_L1.jpg}
    
  \end{figure}
  
  \newpage
  
  Fig. 6 shows an example of setting the hypotheses:
  
  \begin{itemize}

    \item Hp: victim + suspect (+ unknown contributor(s))
    
    \item Hd: victim (+ unknown contributor(s))
  
  \end{itemize}
  
  \begin{figure}[htbp]
    
    \centering
  
    \includegraphics[height=6cm]{figures/screen_L2.jpg}
    
    \caption{An example of setting the hypotheses}
    
  \end{figure}
  
  \vskip\baselineskip
  
  \item Set the range of the assumed number of contributors. The likelihoods of all set numbers are calculated for both Hp and Hd.
  
  \begin{figure}[htbp]
    
    \centering
  
    \includegraphics[height=6cm]{figures/screen_L3.jpg}
    
  \end{figure}
  
  \textbf{Note}
  
  The upper limit of the assumed number of contributors is four.
  
  \newpage

  \item Select an analysis method\footnote{Analysis methods can be set from Tools > Set analysis methods. See the section "Set analysis methods".}.
  
  \begin{figure}[htbp]
    
    \centering
  
    \includegraphics[height=6cm]{figures/screen_L4.jpg}
    
  \end{figure}
  
  \vskip\baselineskip

  \item The likelihoods of all assumed hypotheses can be calculated by checking the "Calculate the likelihood of all assumed hypotheses".
  
  \begin{figure}[htbp]
    
    \centering
  
    \includegraphics[height=6cm]{figures/screen_L5.jpg}
    
  \end{figure}
  
  \newpage
  
  For example, when the reference profiles of the victim and the suspect are inputted and the range of the assumed number of contributors is 2 to 3, all assumed hypotheses are as follows:
  
  \begin{itemize}

    \item two unknown contributors
    
    \item victim + one unknown contributor
    
    \item suspect + one unknown contributor
    
    \item victim + suspect
    
    \item three unknown contributors
    
    \item victim + two unknown contributors
    
    \item suspect + two unknown contributors
    
    \item victim + suspect + one unknown contributor
    
  \end{itemize}
  
  \vskip\baselineskip

  \item Click the "Analyze" button.
  
  \begin{figure}[htbp]
    
    \centering
  
    \includegraphics[height=6cm]{figures/screen_L6.jpg}
    
  \end{figure}
  
  \newpage
  
  Then, the "Progress Bar" window will appear. After completing the analysis, the "Result" tab will appear automatically.
  
  \begin{figure}[htbp]
    
    \centering
  
    \includegraphics[height=12cm]{figures/progress2.jpg}
    
  \end{figure}
  
  \newpage

  \item Review the result of probabilistic genotyping.
  
  \begin{itemize}

    \item The weight values of each genotype combination under the selected conditions are displayed.
    
    \begin{figure}[htbp]
    
      \centering
  
      \includegraphics[height=12cm]{figures/screen_L7.jpg}
    
    \end{figure}
    
    \newpage
    
    \item The observed peak heights and the expected peak heights of gamma distributions under the selected genotype combination can be compared by clicking the "Graph" button.
    
    \begin{figure}[htbp]
    
      \centering
  
      \includegraphics[height=12cm]{figures/screen_L8.jpg}
    
    \end{figure}
    
    \newpage
    
    \item Displayed data can be exported by clicking the "Export displayed data" button.
    
    \begin{figure}[htbp]
    
      \centering
  
      \includegraphics[height=12cm]{figures/screen_L9.jpg}
    
    \end{figure}
    
    \newpage
    
    \item The results of all conditions can be exported by clicking the "Export all" button.
    
    \begin{figure}[htbp]
    
      \centering
  
      \includegraphics[height=12cm]{figures/screen_L10.jpg}
    
    \end{figure}
    
  \end{itemize}
  
  \newpage

  \item Review the result of the likelihood ratio.
  
  \begin{itemize}

    \item The likelihood ratio, the estimated mixture proportion, and the estimated degradation parameter under the selected hypotheses are displayed.
    
    \begin{figure}[htbp]
    
      \centering
  
      \includegraphics[height=12cm]{figures/screen_L11.jpg}
    
    \end{figure}
    
    \newpage
    
    \item The report of the likelihood ratio for displayed data can be exported by clicking the "Report (displayed data)" button.
    
    \begin{figure}[htbp]
    
      \centering
  
      \includegraphics[height=12cm]{figures/screen_L12.jpg}
    
    \end{figure}
    
    \newpage
    
    \item Likelihoods and LRs of all assumed hypotheses can be exported by clicking the "Export all likelihoods" button.
    
    \begin{figure}[htbp]
    
      \centering
  
      \includegraphics[height=12cm]{figures/screen_L13.jpg}
    
    \end{figure}
    
  \end{itemize}

\end{enumerate}

\newpage

\section{Set analysis methods}

Since ver. 3.2.0, users must create an analysis method to determine the conditions for performing deconvolution and assigning likelihood ratios. This section describes how to create and edit an analysis method.

\subsection{Create an analysis method}

\begin{enumerate}

  \item Go to Tools > Set analysis methods.
  
  \begin{figure}[htbp]
    
    \centering
  
    \includegraphics[height=6cm]{figures/screen_setAN1.jpg}
    
  \end{figure}
  
  \vskip\baselineskip
  
  \item Click the "New" button to create an analysis method.
  
  \begin{figure}[htbp]
    
    \centering
  
    \includegraphics[height=4cm]{figures/screen_setAN2.jpg}
    
  \end{figure}
  
  \vskip\baselineskip
  
  \item Enter or select the settings shown in the figures on the following pages.
  
  \newpage
  
  \uuline{"Conditions" tab}
  
  \begin{figure}[htbp]
    
    \centering
  
    \includegraphics[height=12cm]{figures/screen_setAN3.jpg}
    
  \end{figure}
  
  \begin{description}
  
  \setlength{\leftskip}{0.5cm}
  
    \item[Parameters for Monte Carlo simulation]\mbox{}\\
    Select a file of the parameters for Monte Carlo simulation. The file can be generated in the function of “Estimate parameters”\footnote{See the section "Estimate parameters".}. The default files for Identifiler Plus (IDP\_28cycles\_3130xl\_10sec.csv)\footnote{Default parameters of Identifiler Plus were estimated using 392 single-source profiles, which were obtained from publicly available datasets in the Project Research Openness for Validation with Empirical Data (PROVEDIt: https://lftdi.camden.rutgers.edu/provedit/files/). These profiles were derived from non-degraded DNA of 50 individuals. The DNA amount of each sample was 0.0078–0.73 ng. The DNA samples were amplified at 28 cycles, and PCR products were analyzed on an Applied Biosystems\textsuperscript{TM} 3130xl Genetic Analyzer (Thermo Fisher Scientific) with injection parameters set at 3 kV for 10 s.} and GlobalFiler (GF\_29cycles\_3500xL\_24sec.csv)\footnote{Default parameters of GlobalFiler were estimated using 300 single-source profiles. These profiles were derived from pristine DNA of 50 individuals. The extracted DNA was diluted to 0.1, 0.05, 0.025, 0.0125, 0.00625, and 0.003125 ng/μL. A total of 300 diluted DNA solutions were amplified in 29 cycles using the GlobalFiler kit. Subsequently, PCR products were analyzed on an Applied Biosystems\textsuperscript{TM} 3500xL Genetic Analyzer (Thermo Fisher Scientific) with injection parameters set at 1.2 kV for 24 s in the Data Collection Software v3.1 (Thermo Fisher Scientific). See Manabe et al., 2021\cite{Manabe2021}.} are provided in the package \textit{Kongoh}.
    
    \newpage
    
    \item[Allele repeat correction]\mbox{}\\
    Select a file of the allele repeat correction. The file can be generated in the function of “Estimate parameters”. The default files for Identifiler Plus (IDP\_repeat\_correction.csv) and GlobalFiler (GF\_repeat\_correction.csv) are provided in the package \textit{Kongoh}.
    
    \item[Method of estimating allele frequencies]\mbox{}\\
    If the allele frequencies of input files are represented as the allele counts, select "Dirichlet". If the allele frequencies of input files are represented as the allele probabilities, select "No correction".
    
    \item[Minimum allele frequency]\mbox{}\\
    The minimum allele frequency is applied to the frequency of the unobserved alleles when the "Method of estimating allele frequencies" is set to "No correction".
    
    \item[Theta]\mbox{}\\
    The theta value is used to consider the subpopulation effect.
    
    \item[Number of Monte Carlo simulation]\mbox{}\\
    The default value is 1,000, which is considered sufficient for simulations based on developmental validation. If the number of simulations is increased, then the result becomes more robust; however, the runtime increases as well.
  
  \end{description}
  
  \newpage
  
  \uuline{"Analytical threshold" tab}
  
  \begin{figure}[htbp]
    
    \centering
  
    \includegraphics[height=12cm]{figures/screen_setAN4.jpg}
    
  \end{figure}
  
  \vskip\baselineskip
  
  Select a kit, then enter the analytical thresholds of each locus.
  
  \newpage
  
  \uuline{"Mixture proportion" tab}
  
  \begin{figure}[htbp]
    
    \centering
  
    \includegraphics[height=11cm]{figures/screen_setAN5.jpg}
    
  \end{figure}
  
  \vskip\baselineskip
  
  Users can manually add or delete the mixture proportion per contributor. $ MP_n $ is the DNA proportion of the contributor $ n (n = 1, 2, ..., N) $ in a crime stain profile, where $ N $ is the number of contributors. In Kongoh, $ MP_n $ must satisfy the following conditions:
  
\vskip\baselineskip

$ \sum^{N}_{n=1}MP_n = 1 $, 

$ 0 < MP_1 \leq MP_2 \leq ... \leq MP_{N-1} \leq MP_N (N \geq 2)$.

\vskip\baselineskip

In the "Mixture proportion" tab, candidate values for $ MP_n (n = 1, 2, ..., N-1) $ can be set. $ MP_N $ is calculated as $ 1 - \sum^{N-1}_{n=1}MP_n $ in Kongoh.

\vskip\baselineskip

\textbf{Note}

In addition to the sets of $ MP_n $ calculated by the above mentioned method, a set of the same amount of DNA in each contributor is automatically considered in Kongoh (e.g., $ MP_1 = 0.33 $ , $ MP_2 = 0.33 $, $ MP_3 = 0.34 $ for $ N = 3 $).
  
  \newpage
  
  \uuline{"Degradation" tab}
  
  \begin{figure}[htbp]
    
    \centering
  
    \includegraphics[height=11cm]{figures/screen_setAN6.jpg}
    
  \end{figure}
  
  \vskip\baselineskip
  
  Users can manually add or delete the degradation parameters ($ d $). $ d = 0 $ means no degradation. The smaller the $ d $ value, the greater the degree of degradation.
  
  \vskip\baselineskip
  
  \textbf{Note}
  
  The default range of $ d $ is $ [-0.05, 0] $, which is discretized into 20 intervals of width 0.0025. When $ d = -0.05 $, DNA has significantly degraded. For example, when an allelic peak height located at the 100 base is 4,000 RFU, the peak height of the other allele located at the 225 base (which is approximately the middle detection point in the Identifiler Plus system) is expected to be only 8 RFU (i.e., typically less than the analytical threshold). 
  
  \newpage
  
  \uuline{"Exclude unrealistic genotype combinations" tab}
  
  \begin{figure}[htbp]
    
    \centering
  
    \includegraphics[height=12cm]{figures/screen_setAN7.jpg}
    
  \end{figure}
  
  \vskip\baselineskip
  
  \begin{itemize}
    
  \setlength{\leftskip}{0.5cm}
    
    \item \textbf{Stochastic threshold}: The mixture ratio filter and the heterozygote balance filter do not work if one of the peaks is below the threshold.
      
    \item \textbf{Mixture ratio filter}: A candidate genotype combination of minor and major contributors is inappropriate when the peak height of the minor ($ h_{minor} $) is much larger than that of the major ($ h_{major} $). If $ h_{minor} / h_{major} $ is greater than the mixture ratio filter, the genotype combination is excluded. $ h_{minor} $ and $ h_{major} $ are exactly the mean peak heights of the unique alleles in each.
      
    \item \textbf{Heterozygote balance filter}: If the peak height ratio of the two arbitrary peaks is smaller than the set value, the two peaks are not regarded as the heterozygote derived only from a single contributor.
    
    \item \textbf{Back stutter filter}: If the back stutter ratio is greater than the set value, the peak in the position of the back stutter is not regarded as the product derived only from the back stutter.
    
    \item \textbf{Forward stutter filter}: If the forward stutter ratio is greater than the set value, the peak in the position of the forward stutter is not regarded as the product derived only from the forward stutter.
    
    \item \textbf{Double-back stutter filter}: If the double-back stutter ratio is greater than the set value, the peak in the position of the double-back stutter is not regarded as the product derived only from the double-back stutter.
    
    \item \textbf{Minus 2-nt stutter filter}: If the minus 2-nt stutter ratio is greater than the set value, the peak in the position of the minus 2-nt stutter is not regarded as the product derived only from the minus 2-nt stutter.
    
  \end{itemize}
  
  \vskip\baselineskip
  
  \item Click the "Save as" button after finishing all the settings.
  
  \begin{figure}[htbp]
    
    \centering
  
    \includegraphics[height=12cm]{figures/screen_setAN8.jpg}
    
  \end{figure}

\end{enumerate}

\newpage

\subsection{Edit an analysis method}

\begin{enumerate}

  \item Go to Tools > Set analysis methods.
  
  \begin{figure}[htbp]
    
    \centering
  
    \includegraphics[height=6cm]{figures/screen_setAN1.jpg}
    
  \end{figure}
  
  \vskip\baselineskip
  
  \item Select an analysis method.
  
  \begin{figure}[htbp]
    
    \centering
  
    \includegraphics[height=4cm]{figures/screen_setAN9.jpg}
    
  \end{figure}
  
  \vskip\baselineskip
  
  \item Click the "Edit" button to edit the selected analysis method.
  
  \begin{figure}[htbp]
    
    \centering
  
    \includegraphics[height=4cm]{figures/screen_setAN10.jpg}
    
  \end{figure}
  
  \vskip\baselineskip
  
  \item Enter or select the settings just like creating an analysis method.
  
  \newpage
  
  \item Click the "Save" button after finishing all the settings.
  
  \begin{figure}[htbp]
    
    \centering
  
    \includegraphics[height=12cm]{figures/screen_setAN11.jpg}
    
  \end{figure}

\end{enumerate}

\newpage

\section{Estimate parameters}

In Kongoh, probability distribution models for locus-specific amplification efficiency (AE), heterozygote balance (Hb), back stutter ratio (BSR), forward stutter ratio (FSR), double-back stutter ratio (DSR), and minus 2-nt stutter ratio (M2SR) can be estimated using experimental data prepared by users. The maximum likelihood estimation (MLE) for each parameter of the probability distribution models (e.g., mean, variance, etc.) is performed by the generalized simulated annealing, which is implemented into the R package "GenSA". In this section, the procedure for the estimation of parameters is described.

\subsection{Input files}

\begin{enumerate}
  
  \item Go to Tools > Estimate parameters.
  
  \begin{figure}[htbp]
    
    \centering
  
    \includegraphics[height=6cm]{figures/screen_MCpar1.jpg}
    
  \end{figure}
  
  \vskip\baselineskip
  
  \item Load a file of the experimental data.
  
  \begin{figure}[htbp]
    
    \centering
  
    \includegraphics[height=5cm]{figures/screen_MCpar2.jpg}
    
  \end{figure}
  
  \newpage
  
  \textbf{Note}
  
  The experimental data should satisfy the following conditions:

  \begin{itemize}

    \item The data are composed of single-source profiles.

    \item The profiles are derived from pristine (non-degraded) DNA samples.
    
    \item DNA samples should be prepared to the maximum extent, including samples of various amounts of DNA and various individuals.
    
    \item All DNA samples are analyzed based on one protocol.
    
    \item No filters are to be used to remove stutter peaks when analyzing sample electrophoretic data in software programs such as the GeneMapper\textsuperscript{\textregistered} ID-X software.
    
    \item It is desirable to remove pull-up peaks and noises manually, but not required\footnote{Kongoh can automatically determine allele and stutter peaks based on the genotypes of each experimental data, as well as remove peaks located within ±1 base\cite{Fujii2018} of an allele peak(s) in different color channels to exclude the effect of the spectral pull-up.}.
    
  \end{itemize}
  
  \vskip\baselineskip
  
  \textbf{Note}
  
  The example file "Experimental-data\_GF\_Example.csv" is provided. Go to extdata > example in the package \textit{Kongoh}.
  
  \begin{itemize}
    
    \item The file of the experimental data can be exported from the GeneMapper\textsuperscript{\textregistered} ID-X software.

    \item The file of the experimental data must include information regarding the “Sample Name”, “Marker”, “Allele”, “Size”, and “Height”.
  
  \end{itemize}
  
  \vskip\baselineskip
  
  \item Load a file of the genotype data.
  
  \begin{figure}[htbp]
    
    \centering
  
    \includegraphics[height=5cm]{figures/screen_MCpar3.jpg}
    
  \end{figure}
  
  \newpage
  
  \textbf{Note}
  
  The example file "Genotype-data\_GF\_Example.csv" is provided. Go to extdata > example in the package \textit{Kongoh}.
  
  \begin{itemize}

    \item This file must include information regarding the “Sample Name”, “Marker”, “Allele 1”, and “Allele 2”.

    \item Information regarding the “Sample Name” in the “Genotype Data” file must be the same as that in the “Experimental Data” file.
    
    \item Two alleles in homozygotes must be entered in each column (except for Yindel and DYS391).
    
  \end{itemize}
  
  \vskip\baselineskip
  
  \item Load a file of the sequence data.
  
  \begin{figure}[htbp]
    
    \centering
  
    \includegraphics[height=5cm]{figures/screen_MCpar4.jpg}
    
  \end{figure}
  
  \vskip\baselineskip
  
  \textbf{Note}
  
  The example file "Sequence-data\_GF\_Example.csv" is provided\footnote{The example data is derived from sequence-based U.S. population data for SE33\cite{Borsuk2018} and for loci other than SE33\cite{Gettings2018}.}. Go to extdata > example in the package \textit{Kongoh}.
  
  \begin{itemize}
  
    \item This file must include information regarding the “Marker”, “Allele”, “Repeat Region”, and “Count”. Information on the "Flanking Region" is optional.
    
    \item “Count” implies the number of observations in sequence-based population data.
  
  \end{itemize}
  
  \newpage
  
  \item Click the "Next" button. Then the "Setting" tab will be automatically opened.
  
  \begin{figure}[htbp]
    
    \centering
  
    \includegraphics[height=5cm]{figures/screen_MCpar5.jpg}
    
  \end{figure}
  
  \newpage
  
\end{enumerate}
  
\subsection{Set conditions}

\begin{enumerate}

  \item Set conditions shown in the figures on the following pages.
  
  \uuline{"General" tab}
  
  \begin{figure}[htbp]
    
    \centering
  
    \includegraphics[height=12cm]{figures/screen_MCpar_Setting_Gen.jpg}
    
  \end{figure}
  
  \begin{description}
  
  \setlength{\leftskip}{0.5cm}
  
    \item[Min. threshold]\mbox{}\\
    Peaks below the thresholds are not used to estimate the parameters.
    
    \item[Kit]\mbox{}\\
    The kit name is automatically selected based on the input files. Users should select the appropriate kit name when more than one kit with the same locus set is registered in the feature "Edit information on typing kits\footnote{See the section "Edit information on typing kits" in the user manual.}" (e.g., Identifiler and Identifiler Plus).
  
  \end{description}
  
  \newpage
  
  \uuline{"Amplification efficiency (AE)" tab}
  
  \begin{figure}[htbp]
    
    \centering
  
    \includegraphics[height=12cm]{figures/screen_MCpar_Setting_AE.jpg}
    
  \end{figure}
  
  \begin{description}
  
  \setlength{\leftskip}{0.5cm}
  
    \item[Model]\mbox{}\\
    The model for AE is fixed to the “Log-normal” distribution in the current version of Kongoh.
    
    \item[Min. AE]\mbox{}\\
    Min. AE is the lower limit of the truncated log-normal distribution for AE. The upper limit is automatically set to 1 / Min. AE. If "Based on experimental data" is checked, Min. AE is set to the minimum of the experimental AE values. If "Based on experimental data" is unchecked, users can enter an arbitrary value for Min. AE.
    
    \item[Conditions of MLE]\mbox{}\\
    Users can set initial, lower, and upper values of each parameter in the maximum likelihood estimation (MLE). These values do not need to be changed in the initial analysis of MLE. If the estimated probability distribution does not fit to the experimental data, users may try to change these values.
  
  \end{description}
  
  \newpage
  
  \uuline{"Heterozygote balance (Hb)" tab}
  
  \begin{figure}[htbp]
    
    \centering
  
    \includegraphics[height=12cm]{figures/screen_MCpar_Setting_Hb.jpg}
    
  \end{figure}
  
  \begin{description}
  
  \setlength{\leftskip}{0.5cm}
  
    \item[Model]\mbox{}\\
    The model for Hb is fixed to the “Log-normal” distribution in the current version of Kongoh.
    
    \item[Min. Hb]\mbox{}\\
    Min. Hb is the lower limit of the truncated log-normal distribution for Hb. The upper limit is automatically set to 1 / Min. Hb. If "Based on experimental data" is checked, Min. Hb is set to the minimum of the experimental Hb values. If "Based on experimental data" is unchecked, users can enter an arbitrary value for Min. Hb.
    
    \item[Conditions of MLE]\mbox{}\\
    Users can set initial, lower, and upper values of each parameter in the maximum likelihood estimation (MLE). These values do not need to be changed in the initial analysis of MLE. If the estimated probability distribution does not fit to the experimental data, users may try to change these values.
  
  \end{description}
  
  \newpage
  
  \uuline{"Back stutter ratio (BSR)" tab}
  
  \begin{figure}[htbp]
    
    \centering
  
    \includegraphics[height=12cm]{figures/screen_MCpar_Setting_BSR.jpg}
    
  \end{figure}
  
  \begin{description}
  
  \setlength{\leftskip}{0.5cm}
  
    \item[Model]\mbox{}\\
    There are four options for the model for BSR.
    
    \begin{itemize}
    
    \setlength{\leftskip}{1.0cm}
    
      \item \textbf{Best}: The best model is selected from Allele, LUS, and Multi-seq models based on the AIC (Akaike information criterion) values of each model.
      
      \item \textbf{Allele}: The BSRs are positively correlated with the allele numbers\cite{Walsh1996}.
      
      \item \textbf{LUS}: The BSRs are positively correlated with the longest uninterrupted stretch (LUS) values\cite{Brookes2012,Bright2013}.
      
      \item \textbf{Multi-seq}: The BSRs are positively correlated with the corrected allele numbers, which consider not only the LUS, but also other uninterrupted stretches\cite{Taylor2016}.
    
    \end{itemize}
    
    \item[Method]\mbox{}\\
    There are three options for the methods of modeling.
    
    \begin{itemize}
    
    \setlength{\leftskip}{1.0cm}
    
      \item \textbf{Locus specific}: The experimental data of the locus is only used for estimating the probability distribution for this locus.
      
      \item \textbf{Multiple loci together}: The experimental data of all loci with the option "Multiple loci together" is used for estimating the probability distribution for these loci.
      
      \item \textbf{Not consider}: The probability distribution model is not considered in the locus.
    
    \end{itemize}
    
    \item[Model when considering multiple loci together]\mbox{}\\
    Users can select models from four options (i.e., Best, Allele, LUS, and Multi-seq when considering multiple loci together.
    
    \item[Max. BSR]\mbox{}\\
    Max. BSR is the upper limit of the truncated log-normal distribution for BSR. If "Based on experimental data" is checked, Max. BSR is set to the maximum of the experimental BSRs in all loci. If "Based on experimental data" is unchecked, users can enter an arbitrary value for Max. BSR.
    
    \item[Conditions of MLE]\mbox{}\\
    Users can set initial, lower, and upper values of each parameter in the maximum likelihood estimation (MLE). These values do not need to be changed in the initial analysis of MLE. If the estimated probability distribution does not fit to the experimental data, users may try to change these values.
  
  \end{description}
  
  \newpage
  
  \uuline{"Forward stutter ratio (FSR)" tab}
  
  \begin{figure}[htbp]
    
    \centering
  
    \includegraphics[height=12cm]{figures/screen_MCpar_Setting_FSR.jpg}
    
  \end{figure}
  
  \begin{description}
  
  \setlength{\leftskip}{0.5cm}
  
    \item[Model]\mbox{}\\
    There are five options for the model for FSR.
    
    \begin{itemize}
    
    \setlength{\leftskip}{1.0cm}
    
      \item \textbf{Best}: The best model is selected from Allele, LUS, Multi-seq, and Uniform models based on the AIC values of each model.
      
      \item \textbf{Allele}: The FSRs are positively correlated with the allele numbers.
      
      \item \textbf{LUS}: The FSRs are positively correlated with the longest uninterrupted stretch (LUS) values.
      
      \item \textbf{Multi-seq}: The FSRs are positively correlated with the corrected allele numbers, which consider not only the LUS, but also other uninterrupted stretches.
      
      \item \textbf{Uniform}: The mean value of the FSRs is uniform regardless of the allele number.
    
    \end{itemize}
    
    \item[Method]\mbox{}\\
    There are three options for the methods of modeling.
    
    \begin{itemize}
    
    \setlength{\leftskip}{1.0cm}
    
      \item \textbf{Locus specific}: The experimental data of the locus is only used for estimating the probability distribution for this locus.
      
      \item \textbf{Multiple loci together}: The experimental data of all loci with the option "Multiple loci together" is used for estimating the probability distribution for these loci.
      
      \item \textbf{Not consider}: The probability distribution model is not considered in the locus.
    
    \end{itemize}
    
    \item[Model when considering multiple loci together]\mbox{}\\
    Users can select models from five options (i.e., Best, Allele, LUS, Multi-seq, and Uniform when considering multiple loci together.
    
    \item[Max. FSR]\mbox{}\\
    Max. FSR is the upper limit of the truncated log-normal distribution for FSR. If "Based on experimental data" is checked, Max. FSR is set to the maximum of the experimental FSRs in all loci. If "Based on experimental data" is unchecked, users can enter an arbitrary value for Max. FSR.
    
    \item[Conditions of MLE]\mbox{}\\
    Users can set initial, lower, and upper values of each parameter in the maximum likelihood estimation (MLE). These values do not need to be changed in the initial analysis of MLE. If the estimated probability distribution does not fit to the experimental data, users may try to change these values.
  
  \end{description}
  
  \newpage
  
  \uuline{"Double-back stutter ratio (DSR)" tab}
  
  \begin{figure}[htbp]
    
    \centering
  
    \includegraphics[height=12cm]{figures/screen_MCpar_Setting_DSR.jpg}
    
  \end{figure}
  
  \begin{description}
  
  \setlength{\leftskip}{0.5cm}
  
    \item[Model]\mbox{}\\
    There are five options for the model for DSR.
    
    \begin{itemize}
    
    \setlength{\leftskip}{1.0cm}
    
      \item \textbf{Best}: The best model is selected from Allele, LUS, Multi-seq, and Uniform models based on the AIC values of each model.
      
      \item \textbf{Allele}: The DSRs are positively correlated with the allele numbers.
      
      \item \textbf{LUS}: The DSRs are positively correlated with the longest uninterrupted stretch (LUS) values.
      
      \item \textbf{Multi-seq}: The DSRs are positively correlated with the corrected allele numbers, which consider not only the LUS, but also other uninterrupted stretches.
      
      \item \textbf{Uniform}: The mean value of the DSRs is uniform regardless of the allele number.
    
    \end{itemize}
    
    \item[Method]\mbox{}\\
    There are three options for the methods of modeling.
    
    \begin{itemize}
    
    \setlength{\leftskip}{1.0cm}
    
      \item \textbf{Locus specific}: The experimental data of the locus is only used for estimating the probability distribution for this locus.
      
      \item \textbf{Multiple loci together}: The experimental data of all loci with the option "Multiple loci together" is used for estimating the probability distribution for these loci.
      
      \item \textbf{Not consider}: The probability distribution model is not considered in the locus.
    
    \end{itemize}
    
    \item[Model when considering multiple loci together]\mbox{}\\
    Users can select models from five options (i.e., Best, Allele, LUS, Multi-seq, and Uniform when considering multiple loci together.
  
    \item[Max. DSR]\mbox{}\\
    Max. DSR is the upper limit of the truncated log-normal distribution for DSR. If "Based on experimental data" is checked, Max. DSR is set to the maximum of the experimental DSRs in all loci. If "Based on experimental data" is unchecked, users can enter an arbitrary value for Max. DSR.
    
    \item[Conditions of MLE]\mbox{}\\
    Users can set initial, lower, and upper values of each parameter in the maximum likelihood estimation (MLE). These values do not need to be changed in the initial analysis of MLE. If the estimated probability distribution does not fit to the experimental data, users may try to change these values.
  
  \end{description}
  
  \newpage
  
  \uuline{"Minus 2-nt stutter ratio (M2SR)" tab}
  
  \begin{figure}[htbp]
    
    \centering
  
    \includegraphics[height=12cm]{figures/screen_MCpar_Setting_M2SR.jpg}
    
  \end{figure}
  
  \begin{description}
  
  \setlength{\leftskip}{0.5cm}
  
    \item[Model]\mbox{}\\
    The model for M2SR is fixed to the “Uniform” distribution in the current version of Kongoh.
    
    \item[Method]\mbox{}\\
    There are three options for the methods of modeling.
    
    \begin{itemize}
    
    \setlength{\leftskip}{1.0cm}
    
      \item \textbf{Locus specific}: The experimental data of the locus is only used for estimating the probability distribution for this locus.
      
      \item \textbf{Multiple loci together}: The experimental data of all loci with the option "Multiple loci together" is used for estimating the probability distribution for these loci.
      
      \item \textbf{Not consider}: The probability distribution model is not considered in the locus.
    
    \end{itemize}
    
    \item[Model when considering multiple loci together]\mbox{}\\
    The model for M2SR is fixed to the “Uniform” distribution in the current version of Kongoh when considering multiple loci together.
    
    \item[Max. M2SR]\mbox{}\\
    Max. M2SR is the upper limit of the truncated log-normal distribution for M2SR. If "Based on experimental data" is checked, Max. M2SR is set to the maximum of the experimental M2SRs in all loci. If "Based on experimental data" is unchecked, users can enter an arbitrary value for Max. M2SR.
    
    \item[Conditions of MLE]\mbox{}\\
    Users can set initial, lower, and upper values of each parameter in the maximum likelihood estimation (MLE). These values do not need to be changed in the initial analysis of MLE. If the estimated probability distribution does not fit to the experimental data, users may try to change these values.
  
  \end{description}
  
  \vskip\baselineskip
  
  \item Click the "Estimate parameters" button after setting all conditions.
  
  \begin{figure}[htbp]
    
    \centering
  
    \includegraphics[height=12cm]{figures/screen_MCpar6.jpg}
    
  \end{figure}
  
  \textbf{Note}
  
  It takes a few hours to estimate parameters. After the parameter estimation is completed, the “Result” tab appears automatically.

\end{enumerate}

\newpage

\subsection{Review the results}

\begin{enumerate}
  
  \item Review the results of estimated parameters.
  
  \begin{itemize}
  
  \setlength{\leftskip}{0.5cm}
  
    \item Adopted models and estimated parameters in each locus are displayed on the left side of the screen.
    
    \begin{figure}[htbp]
    
      \centering
  
      \includegraphics[height=12cm]{figures/screen_MCpar7.jpg}
    
    \end{figure}
    
    \newpage
    
    \item Detailed results of one locus can be displayed by selecting a locus and clicking the "Show detail" button.
    
    \begin{figure}[htbp]
    
      \centering
  
      \includegraphics[height=12cm]{figures/screen_MCpar8.jpg}
    
    \end{figure}
    
    \newpage
    
    \item AIC (Akaike information criterion) values\footnote{A model featuring a low AIC is considered a superior fit for the experimental SR data as compared with a model featuring a high AIC.} of each model are displayed in the frame of "Comparison of each model".
    
    \begin{figure}[htbp]
    
      \centering
  
      \includegraphics[height=12cm]{figures/screen_MCpar9.jpg}
    
    \end{figure}
    
    \newpage
    
    \item Users can verify whether the estimated parameters are fitted to the experimental data. The graph for "5\% and 95\% quantiles" can be described in terms of the AE and Hb. The graphs for "Mean" and "5\% and 95\% quantiles" can be described in terms of the BSR, FSR, DSR, and M2SR. For example, the graphs for "Mean" and "5\% and 95\% quantiles" of BSR can be described as shown in Figures 6 and 7, respectively.
    
    \begin{figure}[htbp]
    
      \centering
  
      \includegraphics[height=12cm]{figures/screen_MCpar10.jpg}
    
    \end{figure}
    
    \newpage
    
    \begin{figure}[htbp]
    
      \centering
  
      \includegraphics[height=12cm]{figures/figure6.jpg}
      
      \caption{An example of the graph for "Mean" of back stutter ratios. The solid line indicates a regression line of mean values.}
    
    \end{figure}
    
    \newpage
    
    \begin{figure}[htbp]
    
      \centering
  
      \includegraphics[height=12cm]{figures/figure7.jpg}
      
      \caption{An example of the graph for "5\% and 95\% quantiles" of back stutter ratios. The relationship between back stutter ratios and total allelic products is shown. The upper and lower solid lines indicate 95\% and 5\% quantiles of back stutter ratios, respectively. Back stutter ratios are corrected by eliminating the difference in mean values in each allele.}
    
    \end{figure}
    
    \newpage
    
    \item Users can change the adopted model for BSR, FSR, and DSR by clicking the "Change the adopted model" button.
    
    \begin{figure}[htbp]
    
      \centering
  
      \includegraphics[height=12cm]{figures/screen_MCpar11.jpg}
    
    \end{figure}
    
    \vskip\baselineskip
    
    Then, a new window will appear. Select the desired model, and click the “Adopt” button.
    
    \begin{figure}[htbp]
    
      \centering
  
      \includegraphics[height=12cm]{figures/screen_MCpar12.jpg}
    
    \end{figure}
    
    \newpage
    
    The user-selected model will be reflected and the "User-selected" column of the target locus is changed to "Yes".
    
    \begin{figure}[htbp]
    
      \centering
  
      \includegraphics[height=12cm]{figures/screen_MCpar13.jpg}
    
    \end{figure}
    
    \newpage
    
    \item Users can re-estimate the parameters by clicking the "Re-estimate parameters" button.
    
    \begin{figure}[htbp]
    
      \centering
  
      \includegraphics[height=12cm]{figures/screen_MCpar14.jpg}
    
    \end{figure}
    
    Then, a new window will appear. Select a model and set initial, lower, and upper values of each parameter in the maximum likelihood estimation (MLE). After that, click the "Estimate" button to start the re-estimation.
    
    \begin{figure}[htbp]
    
      \centering
  
      \includegraphics[height=5cm]{figures/screen_MCpar15.jpg}
    
    \end{figure}
  
  \end{itemize}
  
  \newpage
  
  \item After reviewing all the results of estimated parameters, click the "Export" button.
  
  \begin{figure}[htbp]
    
    \centering
  
    \includegraphics[height=12cm]{figures/screen_MCpar16.jpg}
    
  \end{figure}
  
  \newpage
  
  \item Enter names for "Parameters for Monte Carlo simulation" and "Allele repeat correction".
  
  \begin{figure}[htbp]
    
    \centering
  
    \includegraphics[height=5cm]{figures/screen_MCpar17.jpg}
    
  \end{figure}
  
  \item Click the "Save" button.
  
  \begin{figure}[htbp]
    
    \centering
  
    \includegraphics[height=5cm]{figures/screen_MCpar18.jpg}
    
  \end{figure}
  
  \textbf{Note}
  
  Two exported files can be directly used for setting analysis methods of "Parameters for Monte Carlo simulation" and "Allele repeat correction".
  
  \newpage
  
  \item Click the "Save project" button to save the current project.
  
  \begin{figure}[htbp]
    
    \centering
  
    \includegraphics[height=12cm]{figures/screen_MCpar19.jpg}
    
  \end{figure}
  
  \vskip\baselineskip
  
  \textbf{Note}
  
  Users can load the saved project by clicking the "Load project" button in the "Files" tab.
  
  \begin{figure}[htbp]
    
    \centering
  
    \includegraphics[height=5cm]{figures/screen_MCpar20.jpg}
    
  \end{figure}
  
\end{enumerate}

\newpage

\section{Set information on typing kits}

Users can create or edit information on typing kits. The information is essential to perform deconvolution and the assignment of likelihood ratios.

\subsection{Create information on a typing kit}

\begin{enumerate}

  \item Go to Tools > Set information on typing kits.
  
  \begin{figure}[htbp]
    
    \centering
  
    \includegraphics[height=6cm]{figures/screen_setKit1.jpg}
    
  \end{figure}
  
  \vskip\baselineskip
  
  \item Click the "New" button.
  
  \begin{figure}[htbp]
    
    \centering
  
    \includegraphics[height=5cm]{figures/screen_setKit2.jpg}
    
  \end{figure}
  
  \newpage
  
  \item Enter a kit name.
  
  \begin{figure}[htbp]
    
    \centering
  
    \includegraphics[height=12cm]{figures/screen_setKit3.jpg}
    
  \end{figure}
  
  \newpage
  
  \item Click the "Add a locus" button.
  
  \begin{figure}[htbp]
    
    \centering
  
    \includegraphics[height=12cm]{figures/screen_setKit4.jpg}
    
  \end{figure}
  
  \vskip\baselineskip
  
  Then, a new window will be automatically opened.
  
  \begin{figure}[htbp]
    
    \centering
  
    \includegraphics[height=4cm]{figures/screen_setKit5.jpg}
    
  \end{figure}
  
  \newpage
  
  \item Enter a locus name, the dye information, and the length of the repeat unit\footnote{Kongoh is only applicable to the STR typing kit; therefore, the length of the repeat unit must be set in all loci other than sex chromosomal markers.}.
  
  \begin{figure}[htbp]
    
    \centering
  
    \includegraphics[height=4cm]{figures/screen_setKit6.jpg}
    
  \end{figure}
  
  \vskip\baselineskip
  
  \item If the locus is a sex chromosomal marker, Check the "Sex chromosomal marker".
  
  \begin{figure}[htbp]
    
    \centering
  
    \includegraphics[height=4cm]{figures/screen_setKit7.jpg}
    
  \end{figure}
  
  \vskip\baselineskip
  
  \item Click the "Add" button. Then, the information on the locus will be displayed.
  
  \begin{figure}[htbp]
    
    \centering
  
    \includegraphics[height=4cm]{figures/screen_setKit8.jpg}
    
  \end{figure}
  
  \newpage
  
  \item Select a locus and click the "Set sizes" button to set the sizes of each allele\footnote{The information on the sizes is used to estimate the size of a drop-out allele in the case of locus drop-out when performing deconvolution and assigning likelihood ratios.}.
  
  \begin{figure}[htbp]
    
    \centering
  
    \includegraphics[height=12cm]{figures/screen_setKit9.jpg}
    
  \end{figure}
  
  \newpage
  
  Then, a new window will be automatically opened.
  
  \begin{figure}[htbp]
    
    \centering
  
    \includegraphics[height=12cm]{figures/screen_setKit10.jpg}
    
  \end{figure}
  
  \newpage
  
  \item Click the "Add an allele" button.
  
  \begin{figure}[htbp]
    
    \centering
  
    \includegraphics[height=12cm]{figures/screen_setKit11.jpg}
    
  \end{figure}
  
  \vskip\baselineskip
  
  \item Enter an allele name, and the size.
  
  \begin{figure}[htbp]
    
    \centering
  
    \includegraphics[height=3cm]{figures/screen_setKit12.jpg}
    
  \end{figure}
  
  \newpage
  
  \item Click the "Add" button.
  
  \begin{figure}[htbp]
    
    \centering
  
    \includegraphics[height=3cm]{figures/screen_setKit13.jpg}
    
  \end{figure}
  
  \vskip\baselineskip
  
  Then, the information on the allele and the size will be displayed.
  
  \begin{figure}[htbp]
    
    \centering
  
    \includegraphics[height=12cm]{figures/screen_setKit14.jpg}
    
  \end{figure}
  
  \newpage
  
  \item Repeat the addition of alleles and the sizes as many as possible.
  
  \vskip\baselineskip
  
  \textbf{Note}
  
  Users can edit an allele and the size from the "Edit an allele" button. Users can also delete an allele and the size from the "Delete an allele" button. Alleles can be sorted from the "Sort" button.
  
  \vskip\baselineskip
  
  \item Click the "Finish" button.
  
  \begin{figure}[htbp]
    
    \centering
  
    \includegraphics[height=12cm]{figures/screen_setKit15.jpg}
    
  \end{figure}
  
  \newpage
  
  Then, the "Set sizes" column will be changed to "yes"\footnote{Users must set sizes of all loci.}.
  
  \begin{figure}[htbp]
    
    \centering
  
    \includegraphics[height=12cm]{figures/screen_setKit16.jpg}
    
  \end{figure}
  
  \newpage
  
  \item Click the "Save" button after setting all markers.
  
  \begin{figure}[htbp]
    
    \centering
  
    \includegraphics[height=12cm]{figures/screen_setKit17.jpg}
    
  \end{figure}

\end{enumerate}

\newpage

\subsection{Edit information on a typing kit}

\begin{enumerate}

  \item Go to Tools > Set information on typing kits.
  
  \begin{figure}[htbp]
    
    \centering
  
    \includegraphics[height=6cm]{figures/screen_setKit1.jpg}
    
  \end{figure}
  
  \vskip\baselineskip
  
  \item Select a kit and click the "Edit" button.
  
  \begin{figure}[htbp]
    
    \centering
  
    \includegraphics[height=5cm]{figures/screen_setKit18.jpg}
    
  \end{figure}
  
  \vskip\baselineskip
  
  \item Edit the information on the selected kit just like creating information on a typing kit.
  
  \newpage
  
  \item Click the "Save" button.
  
  \begin{figure}[htbp]
    
    \centering
  
    \includegraphics[height=12cm]{figures/screen_setKit17.jpg}
    
  \end{figure}

\end{enumerate}

\newpage

\section{Edit repeat correction}

Users can edit LUS, CA\_BSR, CA\_FSR, and CA\_DSR. LUS is the abbreviation of the longest uninterrupted stretch\cite{Bright2013}, which is used in the LUS model for stutter ratios. CA\_BSR, CA\_FSR, and CA\_DSR are corrected allele numbers used in the multi-sequence model for back stutter ratios, forward stutter ratios, and double-back stutter ratios, respectively.

\begin{enumerate}

  \item Go to Tools > Edit repeat correction.
  
  \begin{figure}[htbp]
    
    \centering
  
    \includegraphics[height=6cm]{figures/screen_setRep1.jpg}
    
  \end{figure}
  
  \vskip\baselineskip
  
  \item Select a dataset of allele repeat correction and click the "Edit" button.
  
  \begin{figure}[htbp]
    
    \centering
  
    \includegraphics[height=4cm]{figures/screen_setRep2.jpg}
    
  \end{figure}
  
  \newpage
  
  \item Select a locus.
  
  \begin{figure}[htbp]
    
    \centering
  
    \includegraphics[height=8.5cm]{figures/screen_setRep3.jpg}
    
  \end{figure}
  
  \vskip\baselineskip
  
  \item Add or edit LUS, CA\_BSR, CA\_FSR, or CA\_DSR from the "Add an allele" button or the "Edit an allele" buttons, respectively.
  
  \begin{figure}[htbp]
    
    \centering
  
    \includegraphics[height=8.5cm]{figures/screen_setRep4.jpg}
    
  \end{figure}
  
  \newpage
  
  \item Enter the name of an allele.
  
  \begin{figure}[htbp]
    
    \centering
  
    \includegraphics[height=4cm]{figures/screen_setRep5.jpg}
    
  \end{figure}
  
  \vskip\baselineskip
  
  \item Enter values of LUS, CA\_BSR, CA\_FSR, or CA\_DSR as needed.
  
  \begin{figure}[htbp]
    
    \centering
  
    \includegraphics[height=4cm]{figures/screen_setRep6.jpg}
    
  \end{figure}
  
  \vskip\baselineskip
  
  \item Click the "Save" button.
  
  \begin{figure}[htbp]
    
    \centering
  
    \includegraphics[height=4cm]{figures/screen_setRep7.jpg}
    
  \end{figure}
  
  \newpage
  
  \item Repeat the edit of LUS, CA\_BSR, CA\_FSR, and CA\_DSR.
  
  \vskip\baselineskip
  
  \textbf{Note}
  
  Users can delete an allele from the "Delete an allele" button.
  
  \vskip\baselineskip
  
  \item Click the ”Finish” button after finishing all the settings.
  
  \begin{figure}[htbp]
    
    \centering
  
    \includegraphics[height=12cm]{figures/screen_setRep8.jpg}
    
  \end{figure}
  
\end{enumerate}

\newpage

\bibliographystyle{elsarticle-num}
\bibliography{references.bib}
\addcontentsline{toc}{section}{References}
